--- a/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
@@ -20,36 +_,63 @@
     }
 
     @Override
-    public ItemStack execute(BlockSource p_302460_, ItemStack p_123376_) {
-        Direction direction = p_302460_.state().getValue(DispenserBlock.FACING);
-        ServerLevel serverlevel = p_302460_.level();
-        Vec3 vec3 = p_302460_.center();
-        double d0 = 0.5625 + this.type.getWidth() / 2.0;
-        double d1 = vec3.x() + direction.getStepX() * d0;
+    public ItemStack execute(BlockSource blockSource, ItemStack item) {
+        Direction direction = blockSource.state().getValue(DispenserBlock.FACING);
+        ServerLevel serverLevel = blockSource.level();
+        Vec3 vec3 = blockSource.center();
+        double d = 0.5625 + this.type.getWidth() / 2.0;
+        double d1 = vec3.x() + direction.getStepX() * d;
         double d2 = vec3.y() + direction.getStepY() * 1.125F;
-        double d3 = vec3.z() + direction.getStepZ() * d0;
-        BlockPos blockpos = p_302460_.pos().relative(direction);
+        double d3 = vec3.z() + direction.getStepZ() * d;
+        BlockPos blockPos = blockSource.pos().relative(direction);
+        AbstractBoat abstractBoat = this.type.create(serverLevel, EntitySpawnReason.DISPENSER);
         double d4;
-        if (serverlevel.getFluidState(blockpos).is(FluidTags.WATER)) {
+        if (canBoatInFluid(abstractBoat, serverLevel.getFluidState(blockPos))) {
             d4 = 1.0;
         } else {
-            if (!serverlevel.getBlockState(blockpos).isAir() || !serverlevel.getFluidState(blockpos.below()).is(FluidTags.WATER)) {
-                return this.defaultDispenseItemBehavior.dispense(p_302460_, p_123376_);
+            if (!serverLevel.getBlockState(blockPos).isAir() || !canBoatInFluid(abstractBoat, serverLevel.getFluidState(blockPos.below()))) {
+                return this.defaultDispenseItemBehavior.dispense(blockSource, item);
             }
 
             d4 = 0.0;
         }
 
-        AbstractBoat abstractboat = this.type.create(serverlevel, EntitySpawnReason.DISPENSER);
-        if (abstractboat != null) {
-            abstractboat.setInitialPos(d1, d2 + d4, d3);
-            EntityType.<AbstractBoat>createDefaultStackConfig(serverlevel, p_123376_, null).accept(abstractboat);
-            abstractboat.setYRot(direction.toYRot());
-            serverlevel.addFreshEntity(abstractboat);
-            p_123376_.shrink(1);
-        }
-
-        return p_123376_;
+        // CraftBukkit start
+        ItemStack singleItemStack = item.copyWithCount(1);
+        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(serverLevel, blockSource.pos());
+        org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(singleItemStack);
+
+        org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d1, d2 + d4, d3));
+        serverLevel.getCraftServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            return item;
+        }
+
+        boolean shrink = true;
+        if (!event.getItem().equals(craftItem)) {
+            shrink = false;
+            // Chain to handler for new item
+            ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior dispenseBehavior = DispenserBlock.getDispenseBehavior(blockSource, eventStack);
+            if (dispenseBehavior != DispenseItemBehavior.NOOP && dispenseBehavior != this) {
+                dispenseBehavior.dispense(blockSource, eventStack);
+                return item;
+            }
+        }
+        // CraftBukkit end
+        if (abstractBoat != null) {
+            abstractBoat.setInitialPos(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ()); // CraftBukkit
+            EntityType.<AbstractBoat>createDefaultStackConfig(serverLevel, item, null).accept(abstractBoat);
+            abstractBoat.setYRot(direction.toYRot());
+            if (serverLevel.addFreshEntity(abstractBoat) && shrink) item.shrink(1); // Paper - if entity add was successful and supposed to shrink
+        }
+
+        return item;
+    }
+
+    private static boolean canBoatInFluid(@org.jetbrains.annotations.Nullable AbstractBoat boat, net.minecraft.world.level.material.FluidState fluid) {
+        return boat != null ? boat.canBoatInFluid(fluid) : fluid.is(FluidTags.WATER);
     }
 
     @Override
