--- a/net/minecraft/core/dispenser/MinecartDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/MinecartDispenseItemBehavior.java
@@ -34,7 +_,7 @@
         BlockState blockstate = serverlevel.getBlockState(blockpos);
         double d3;
         if (blockstate.is(BlockTags.RAILS)) {
-            if (getRailShape(blockstate).isSlope()) {
+            if (getRailShape(blockstate, serverlevel, blockpos).isSlope()) {
                 d3 = 0.6;
             } else {
                 d3 = 0.1;
@@ -49,7 +_,7 @@
                 return this.defaultDispenseItemBehavior.dispense(p_374045_, p_374580_);
             }
 
-            if (direction != Direction.DOWN && getRailShape(blockstate1).isSlope()) {
+            if (direction != Direction.DOWN && getRailShape(blockstate1, serverlevel, blockpos.below()).isSlope()) {
                 d3 = -0.4;
             } else {
                 d3 = -0.9;
@@ -57,19 +_,42 @@
         }
 
         Vec3 vec31 = new Vec3(d0, d1 + d3, d2);
-        AbstractMinecart abstractminecart = AbstractMinecart.createMinecart(
-            serverlevel, vec31.x, vec31.y, vec31.z, this.entityType, EntitySpawnReason.DISPENSER, p_374580_, null
-        );
-        if (abstractminecart != null) {
-            serverlevel.addFreshEntity(abstractminecart);
-            p_374580_.shrink(1);
+        ItemStack itemstack1 = p_374580_.copyWithCount(1); // Paper - shrink below and single item in event
+        org.bukkit.block.Block block2 = org.bukkit.craftbukkit.block.CraftBlock.at(serverlevel, p_374045_.pos());
+        org.bukkit.craftbukkit.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+        org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block2, craftItem.clone(), org.bukkit.craftbukkit.util.CraftVector.toBukkit(vec31));
+        serverlevel.getCraftServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            return p_374580_;
+        }
+
+        boolean shrink = true;
+        if (!event.getItem().equals(craftItem)) {
+            shrink = false;
+            // Chain to handler for new item
+            ItemStack eventStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior dispenseItemBehavior = DispenserBlock.getDispenseBehavior(p_374045_, eventStack);
+            if (dispenseItemBehavior != DispenseItemBehavior.NOOP && dispenseItemBehavior != this) {
+                dispenseItemBehavior.dispense(p_374045_, eventStack);
+                return p_374580_;
+            }
+        }
+
+        itemstack1 = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+        AbstractMinecart abstractMinecart = AbstractMinecart.createMinecart(serverlevel, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), this.entityType, EntitySpawnReason.DISPENSER, itemstack1, null);
+
+        if (abstractMinecart != null) {
+            if (serverlevel.addFreshEntity(abstractMinecart) && shrink) p_374580_.shrink(1); // Paper - if entity add was successful and supposed to shrink
+            // CraftBukkit end
         }
 
         return p_374580_;
     }
 
-    private static RailShape getRailShape(BlockState p_374571_) {
-        return p_374571_.getBlock() instanceof BaseRailBlock baserailblock ? p_374571_.getValue(baserailblock.getShapeProperty()) : RailShape.NORTH_SOUTH;
+    private static RailShape getRailShape(BlockState p_374571_, ServerLevel level, BlockPos pos) {
+        return p_374571_.getBlock() instanceof BaseRailBlock baserailblock ? baserailblock.getRailDirection(p_374571_, level, pos, null) : RailShape.NORTH_SOUTH;
     }
 
     @Override
