--- a/net/minecraft/server/packs/PathPackResources.java
+++ b/net/minecraft/server/packs/PathPackResources.java
@@ -113,6 +_,12 @@
         try (DirectoryStream<Path> directorystream = Files.newDirectoryStream(path)) {
             for (Path path1 : directorystream) {
                 String s = path1.getFileName().toString();
+                // Paper start - Improve logging and errors
+                if (!Files.isDirectory(path1)) {
+                    LOGGER.error("Invalid directory entry: {} in {}.", s, this.root, new java.nio.file.NotDirectoryException(s));
+                    continue;
+                }
+                // Paper end - Improve logging and errors
                 if (ResourceLocation.isValidNamespace(s)) {
                     set.add(s);
                 } else {
