--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -69,228 +_,330 @@
         reason = "System.out needed before bootstrap"
     )
     @DontObfuscate
-    public static void main(String[] p_129699_) {
+    public static void main(final OptionSet optionSet) { // CraftBukkit - replaces main(String[] args)
+        io.papermc.paper.util.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
         SharedConstants.tryDetectVersion();
-        OptionParser optionparser = new OptionParser();
-        OptionSpec<Void> optionspec = optionparser.accepts("nogui");
-        OptionSpec<Void> optionspec1 = optionparser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
-        OptionSpec<Void> optionspec2 = optionparser.accepts("demo");
-        OptionSpec<Void> optionspec3 = optionparser.accepts("bonusChest");
-        OptionSpec<Void> optionspec4 = optionparser.accepts("forceUpgrade");
-        OptionSpec<Void> optionspec5 = optionparser.accepts("eraseCache");
-        OptionSpec<Void> optionspec6 = optionparser.accepts("recreateRegionFiles");
-        OptionSpec<Void> optionspec7 = optionparser.accepts("safeMode", "Loads level with vanilla datapack only");
-        OptionSpec<Void> optionspec8 = optionparser.accepts("help").forHelp();
-        OptionSpec<String> optionspec9 = optionparser.accepts("universe").withRequiredArg().defaultsTo(".");
-        OptionSpec<String> optionspec10 = optionparser.accepts("world").withRequiredArg();
-        OptionSpec<Integer> optionspec11 = optionparser.accepts("port").withRequiredArg().ofType(Integer.class).defaultsTo(-1);
-        OptionSpec<String> optionspec12 = optionparser.accepts("serverId").withRequiredArg();
-        OptionSpec<Void> optionspec13 = optionparser.accepts("jfrProfile");
-        OptionSpec<Path> optionspec14 = optionparser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
-        OptionSpec<String> optionspec15 = optionparser.nonOptions();
+        /* CraftBukkit start - Replace everything
+        OptionParser optionParser = new OptionParser();
+        OptionSpec<Void> optionSpec = optionParser.accepts("nogui");
+        OptionSpec<Void> optionSpec1 = optionParser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
+        OptionSpec<Void> optionSpec2 = optionParser.accepts("demo");
+        OptionSpec<Void> optionSpec3 = optionParser.accepts("bonusChest");
+        OptionSpec<Void> optionSpec4 = optionParser.accepts("forceUpgrade");
+        OptionSpec<Void> optionSpec5 = optionParser.accepts("eraseCache");
+        OptionSpec<Void> optionSpec6 = optionParser.accepts("recreateRegionFiles");
+        OptionSpec<Void> optionSpec7 = optionParser.accepts("safeMode", "Loads level with vanilla datapack only");
+        OptionSpec<Void> optionSpec8 = optionParser.accepts("help").forHelp();
+        OptionSpec<String> optionSpec9 = optionParser.accepts("universe").withRequiredArg().defaultsTo(".");
+        OptionSpec<String> optionSpec10 = optionParser.accepts("world").withRequiredArg();
+        OptionSpec<Integer> optionSpec11 = optionParser.accepts("port").withRequiredArg().ofType(Integer.class).defaultsTo(-1);
+        OptionSpec<String> optionSpec12 = optionParser.accepts("serverId").withRequiredArg();
+        OptionSpec<Void> optionSpec13 = optionParser.accepts("jfrProfile");
+        OptionSpec<Path> optionSpec14 = optionParser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
+        OptionSpec<String> optionSpec15 = optionParser.nonOptions();
 
         try {
-            OptionSet optionset = optionparser.parse(p_129699_);
-            if (optionset.has(optionspec8)) {
-                optionparser.printHelpOn(System.err);
+            OptionSet optionSet = optionParser.parse(args);
+            if (optionSet.has(optionSpec8)) {
+                optionParser.printHelpOn(System.err);
                 return;
             }
+            */ // CraftBukkit end
+        try {
 
-            Path path = optionset.valueOf(optionspec14);
+            Path path = (Path) optionSet.valueOf("pidFile"); // CraftBukkit
             if (path != null) {
                 writePidFile(path);
             }
 
             CrashReport.preload();
-            if (optionset.has(optionspec13)) {
+            if (optionSet.has("jfrProfile")) { // CraftBukkit
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            io.papermc.paper.plugin.PluginInitializerManager.load(optionSet); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
             Util.startTimerHackThread();
             Path path1 = Paths.get("server.properties");
-            DedicatedServerSettings dedicatedserversettings = new DedicatedServerSettings(path1);
-            dedicatedserversettings.forceSave();
-            RegionFileVersion.configure(dedicatedserversettings.getProperties().regionFileComression);
+            if (!optionSet.has("initSettings")) net.neoforged.neoforge.server.loading.ServerModLoader.load(); // Load mods before we load almost anything else anymore. Single spot now. Only loads if they haven't passed the initserver param
+            DedicatedServerSettings dedicatedServerSettings = new DedicatedServerSettings(optionSet); // CraftBukkit - CLI argument support
+            dedicatedServerSettings.forceSave();
+            RegionFileVersion.configure(dedicatedServerSettings.getProperties().regionFileComression);
             Path path2 = Paths.get("eula.txt");
             Eula eula = new Eula(path2);
-            if (optionset.has(optionspec1)) {
+            // Paper start - load config files early for access below if needed
+            org.bukkit.configuration.file.YamlConfiguration bukkitConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) optionSet.valueOf("bukkit-settings"));
+            org.bukkit.configuration.file.YamlConfiguration spigotConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) optionSet.valueOf("spigot-settings"));
+            // Paper end - load config files early for access below if needed
+            if (optionSet.has("initSettings")) { // CraftBukkit
+                // CraftBukkit start - SPIGOT-5761: Create bukkit.yml and commands.yml if not present
+                File configFile = (File) optionSet.valueOf("bukkit-settings");
+                org.bukkit.configuration.file.YamlConfiguration configuration = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(configFile);
+                configuration.options().copyDefaults(true);
+                configuration.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/bukkit.yml"), java.nio.charset.StandardCharsets.UTF_8)));
+                configuration.save(configFile);
+
+                File commandFile = (File) optionSet.valueOf("commands-settings");
+                org.bukkit.configuration.file.YamlConfiguration commandsConfiguration = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(commandFile);
+                commandsConfiguration.options().copyDefaults(true);
+                commandsConfiguration.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/commands.yml"), java.nio.charset.StandardCharsets.UTF_8)));
+                commandsConfiguration.save(commandFile);
+                // CraftBukkit end
                 LOGGER.info("Initialized '{}' and '{}'", path1.toAbsolutePath(), path2.toAbsolutePath());
                 return;
             }
 
-            if (!eula.hasAgreedToEULA()) {
+            // Spigot start
+            boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
+            if (eulaAgreed) {
+                LOGGER.error("You have used the Spigot command line EULA agreement flag.");
+                LOGGER.error("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
+                LOGGER.error("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
+            }
+            if (!eula.hasAgreedToEULA() && !eulaAgreed) {
+                // Spigot end
                 LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
                 return;
             }
 
-            File file1 = new File(optionset.valueOf(optionspec9));
-            Services services = Services.create(new YggdrasilAuthenticationService(Proxy.NO_PROXY), file1);
-            String s = Optional.ofNullable(optionset.valueOf(optionspec10)).orElse(dedicatedserversettings.getProperties().levelName);
-            LevelStorageSource levelstoragesource = LevelStorageSource.createDefault(file1.toPath());
-            LevelStorageSource.LevelStorageAccess levelstoragesource$levelstorageaccess = levelstoragesource.validateAndCreateAccess(s);
-            Dynamic<?> dynamic;
-            if (levelstoragesource$levelstorageaccess.hasWorldData()) {
-                LevelSummary levelsummary;
+            // Paper start - Detect headless JRE
+            String awtException = io.papermc.paper.util.ServerEnvironment.awtDependencyCheck();
+            if (awtException != null) {
+                LOGGER.error("You are using a headless JRE distribution.");
+                LOGGER.error("This distribution is missing certain graphic libraries that the Minecraft server needs to function.");
+                LOGGER.error("For instructions on how to install the non-headless JRE, see https://docs.papermc.io/misc/java-install");
+                LOGGER.error("");
+                LOGGER.error(awtException);
+                return;
+            }
+            // Paper end - Detect headless JRE
+
+            org.spigotmc.SpigotConfig.disabledAdvancements = spigotConfiguration.getStringList("advancements.disabled"); // Paper - fix SPIGOT-5885, must be set early in init
+
+            // Paper start - fix SPIGOT-5824
+            File file;
+            File userCacheFile = new File(Services.USERID_CACHE_FILE);
+            if (optionSet.has("universe")) {
+                file = (File) optionSet.valueOf("universe"); // CraftBukkit
+                userCacheFile = new File(file, Services.USERID_CACHE_FILE);
+            } else {
+                file = new File(bukkitConfiguration.getString("settings.world-container", "."));
+            }
+            // Paper end - fix SPIGOT-5824
+            Services services = Services.create(new com.destroystokyo.paper.profile.PaperAuthenticationService(Proxy.NO_PROXY), file, userCacheFile, optionSet); // Paper - pass OptionSet to load paper config files; override authentication service; fix world-container
+            // CraftBukkit start
+            String string = Optional.ofNullable((String) optionSet.valueOf("world")).orElse(dedicatedServerSettings.getProperties().levelName);
+            if (string == null || string.isEmpty() || new File(file, string).getAbsolutePath().equals(new File(string).getAbsolutePath())) { // Neo: validate world directory
+                LOGGER.error("Invalid world directory specified, must not be null, empty or the same directory as your universe! " + string);
+                return;
+            }
+            LevelStorageSource levelStorageSource = LevelStorageSource.createDefault(file.toPath());
+            LevelStorageSource.LevelStorageAccess levelStorageAccess = levelStorageSource.validateAndCreateAccess(string, LevelStem.OVERWORLD);
+            // CraftBukkit end
+            Dynamic<?> dataTag;
+            if (levelStorageAccess.hasWorldData()) {
+                LevelSummary summary;
                 try {
-                    dynamic = levelstoragesource$levelstorageaccess.getDataTag();
-                    levelsummary = levelstoragesource$levelstorageaccess.getSummary(dynamic);
-                } catch (NbtException | ReportedNbtException | IOException ioexception1) {
-                    LevelStorageSource.LevelDirectory levelstoragesource$leveldirectory = levelstoragesource$levelstorageaccess.getLevelDirectory();
-                    LOGGER.warn("Failed to load world data from {}", levelstoragesource$leveldirectory.dataFile(), ioexception1);
+                    dataTag = levelStorageAccess.getDataTag();
+                    summary = levelStorageAccess.getSummary(dataTag);
+                    levelStorageAccess.readAdditionalLevelSaveData(false); // Neo: mod hook for additional level save data
+                } catch (NbtException | ReportedNbtException | IOException var41) {
+                    LevelStorageSource.LevelDirectory levelDirectory = levelStorageAccess.getLevelDirectory();
+                    LOGGER.warn("Failed to load world data from {}", levelDirectory.dataFile(), var41);
                     LOGGER.info("Attempting to use fallback");
 
                     try {
-                        dynamic = levelstoragesource$levelstorageaccess.getDataTagFallback();
-                        levelsummary = levelstoragesource$levelstorageaccess.getSummary(dynamic);
-                    } catch (NbtException | ReportedNbtException | IOException ioexception) {
-                        LOGGER.error("Failed to load world data from {}", levelstoragesource$leveldirectory.oldDataFile(), ioexception);
+                        dataTag = levelStorageAccess.getDataTagFallback();
+                        summary = levelStorageAccess.getSummary(dataTag);
+                        levelStorageAccess.readAdditionalLevelSaveData(true); // Neo: mod hook for additional level save data
+                    } catch (NbtException | ReportedNbtException | IOException var40) {
+                        LOGGER.error("Failed to load world data from {}", levelDirectory.oldDataFile(), var40);
                         LOGGER.error(
                             "Failed to load world data from {} and {}. World files may be corrupted. Shutting down.",
-                            levelstoragesource$leveldirectory.dataFile(),
-                            levelstoragesource$leveldirectory.oldDataFile()
+                            levelDirectory.dataFile(),
+                            levelDirectory.oldDataFile()
                         );
                         return;
                     }
 
-                    levelstoragesource$levelstorageaccess.restoreLevelDataFromOld();
+                    levelStorageAccess.restoreLevelDataFromOld();
                 }
 
-                if (levelsummary.requiresManualConversion()) {
+                if (summary.requiresManualConversion()) {
                     LOGGER.info("This world must be opened in an older version (like 1.6.4) to be safely converted");
                     return;
                 }
 
-                if (!levelsummary.isCompatible()) {
+                if (!summary.isCompatible()) {
                     LOGGER.info("This world was created by an incompatible version.");
                     return;
                 }
             } else {
-                dynamic = null;
+                dataTag = null;
             }
 
-            Dynamic<?> dynamic1 = dynamic;
-            boolean flag = optionset.has(optionspec7);
-            if (flag) {
+            Dynamic<?> dynamic = dataTag;
+            boolean hasOptionSpec = optionSet.has("safeMode"); // CraftBukkit
+            if (hasOptionSpec) {
                 LOGGER.warn("Safe mode active, only vanilla datapack will be loaded");
             }
 
-            PackRepository packrepository = ServerPacksSource.createPackRepository(levelstoragesource$levelstorageaccess);
+            PackRepository packRepository = ServerPacksSource.createPackRepository(levelStorageAccess);
+            // CraftBukkit start
+            File bukkitDataPackFolder = new File(levelStorageAccess.getLevelPath(net.minecraft.world.level.storage.LevelResource.DATAPACK_DIR).toFile(), "bukkit");
+            if (!bukkitDataPackFolder.exists()) {
+                bukkitDataPackFolder.mkdirs();
+            }
+            File mcMeta = new File(bukkitDataPackFolder, "pack.mcmeta");
+            try {
+                com.google.common.io.Files.asCharSink(mcMeta, java.nio.charset.StandardCharsets.UTF_8).write("""
+                    {
+                        "pack": {
+                            "description": "Data pack for resources provided by Bukkit plugins",
+                            "pack_format": %d
+                        }
+                    }
+                    """.formatted(SharedConstants.getCurrentVersion().packVersion(net.minecraft.server.packs.PackType.SERVER_DATA))
+                );
+            } catch (java.io.IOException ex) {
+                throw new RuntimeException("Could not initialize Bukkit datapack", ex);
+            }
+            java.util.concurrent.atomic.AtomicReference<WorldLoader.DataLoadContext> worldLoader = new java.util.concurrent.atomic.AtomicReference<>();
+            // CraftBukkit end
 
-            WorldStem worldstem;
+            WorldStem worldStem;
             try {
-                WorldLoader.InitConfig worldloader$initconfig = loadOrCreateConfig(dedicatedserversettings.getProperties(), dynamic1, flag, packrepository);
-                worldstem = Util.<WorldStem>blockUntilDone(
-                        p_372648_ -> WorldLoader.load(
-                            worldloader$initconfig,
-                            p_359487_ -> {
-                                Registry<LevelStem> registry = p_359487_.datapackDimensions().lookupOrThrow(Registries.LEVEL_STEM);
-                                if (dynamic1 != null) {
-                                    LevelDataAndDimensions leveldataanddimensions = LevelStorageSource.getLevelDataAndDimensions(
-                                        dynamic1, p_359487_.dataConfiguration(), registry, p_359487_.datapackWorldgen()
+                WorldLoader.InitConfig initConfig = loadOrCreateConfig(dedicatedServerSettings.getProperties(), dynamic, hasOptionSpec, packRepository);
+                worldStem = Util.<WorldStem>blockUntilDone(
+                        executor -> WorldLoader.load(
+                            initConfig,
+                            context -> {
+                                worldLoader.set(context); // CraftBukkit
+                                Registry<LevelStem> registry = context.datapackDimensions().lookupOrThrow(Registries.LEVEL_STEM);
+                                if (dynamic != null) {
+                                    LevelDataAndDimensions levelDataAndDimensions = LevelStorageSource.getLevelDataAndDimensions(
+                                        dynamic, context.dataConfiguration(), registry, context.datapackWorldgen()
                                     );
                                     return new WorldLoader.DataLoadOutput<>(
-                                        leveldataanddimensions.worldData(), leveldataanddimensions.dimensions().dimensionsRegistryAccess()
+                                        levelDataAndDimensions.worldData(), levelDataAndDimensions.dimensions().dimensionsRegistryAccess()
                                     );
                                 } else {
                                     LOGGER.info("No existing world data, creating new world");
-                                    LevelSettings levelsettings;
-                                    WorldOptions worldoptions;
-                                    WorldDimensions worlddimensions;
-                                    if (optionset.has(optionspec2)) {
-                                        levelsettings = MinecraftServer.DEMO_SETTINGS;
-                                        worldoptions = WorldOptions.DEMO_OPTIONS;
-                                        worlddimensions = WorldPresets.createNormalWorldDimensions(p_359487_.datapackWorldgen());
+                                    LevelSettings levelSettings;
+                                    WorldOptions worldOptions;
+                                    WorldDimensions worldDimensions;
+                                    if (optionSet.has("demo")) { // CraftBukkit
+                                        levelSettings = MinecraftServer.DEMO_SETTINGS;
+                                        worldOptions = WorldOptions.DEMO_OPTIONS;
+                                        worldDimensions = WorldPresets.createNormalWorldDimensions(context.datapackWorldgen());
                                     } else {
-                                        DedicatedServerProperties dedicatedserverproperties = dedicatedserversettings.getProperties();
-                                        levelsettings = new LevelSettings(
-                                            dedicatedserverproperties.levelName,
-                                            dedicatedserverproperties.gamemode,
-                                            dedicatedserverproperties.hardcore,
-                                            dedicatedserverproperties.difficulty,
+                                        DedicatedServerProperties properties = dedicatedServerSettings.getProperties();
+                                        levelSettings = new LevelSettings(
+                                            properties.levelName,
+                                            properties.gamemode,
+                                            properties.hardcore,
+                                            properties.difficulty,
                                             false,
-                                            new GameRules(p_359487_.dataConfiguration().enabledFeatures()),
-                                            p_359487_.dataConfiguration()
+                                            new GameRules(context.dataConfiguration().enabledFeatures()),
+                                            context.dataConfiguration()
                                         );
-                                        worldoptions = optionset.has(optionspec3)
-                                            ? dedicatedserverproperties.worldOptions.withBonusChest(true)
-                                            : dedicatedserverproperties.worldOptions;
-                                        worlddimensions = dedicatedserverproperties.createDimensions(p_359487_.datapackWorldgen());
+                                        worldOptions = optionSet.has("bonusChest") ? properties.worldOptions.withBonusChest(true) : properties.worldOptions; // CraftBukkit
+                                        worldDimensions = properties.createDimensions(context.datapackWorldgen());
                                     }
 
-                                    WorldDimensions.Complete worlddimensions$complete = worlddimensions.bake(registry);
-                                    Lifecycle lifecycle = worlddimensions$complete.lifecycle().add(p_359487_.datapackWorldgen().allRegistriesLifecycle());
+                                    // Neo: Do a write-read-cycle to inject modded dimensions on first start of a dedicated server into its generated world dimensions list.
+                                    var registryOps = net.minecraft.resources.RegistryOps.create(net.minecraft.nbt.NbtOps.INSTANCE, context.datapackWorldgen());
+                                    worldDimensions = WorldDimensions.CODEC.encoder().encodeStart(registryOps, worldDimensions).flatMap((writtenPayloadWithModdedDimensions) -> WorldDimensions.CODEC.decoder().parse(registryOps, writtenPayloadWithModdedDimensions)).resultOrPartial(LOGGER::error).orElse(worldDimensions);
+                                    WorldDimensions.Complete worldDimensionsComplete = worldDimensions.bake(registry);
+                                    Lifecycle lifecycle = worldDimensionsComplete.lifecycle().add(context.datapackWorldgen().allRegistriesLifecycle());
                                     return new WorldLoader.DataLoadOutput<>(
-                                        new PrimaryLevelData(levelsettings, worldoptions, worlddimensions$complete.specialWorldProperty(), lifecycle),
-                                        worlddimensions$complete.dimensionsRegistryAccess()
+                                        new PrimaryLevelData(levelSettings, worldOptions, worldDimensionsComplete.specialWorldProperty(), lifecycle),
+                                        worldDimensionsComplete.dimensionsRegistryAccess()
                                     );
                                 }
                             },
                             WorldStem::new,
                             Util.backgroundExecutor(),
-                            p_372648_
+                            executor
                         )
                     )
                     .get();
-            } catch (Exception exception) {
+            } catch (Exception var39) {
                 LOGGER.warn(
                     "Failed to load datapacks, can't proceed with server load. You can either fix your datapacks or reset to vanilla with --safeMode",
-                    (Throwable)exception
+                    (Throwable)var39
                 );
                 return;
             }
 
-            RegistryAccess.Frozen registryaccess$frozen = worldstem.registries().compositeAccess();
-            WorldData worlddata = worldstem.worldData();
-            boolean flag1 = optionset.has(optionspec6);
-            if (optionset.has(optionspec4) || flag1) {
-                forceUpgrade(
-                    levelstoragesource$levelstorageaccess,
-                    worlddata,
-                    DataFixers.getDataFixer(),
-                    optionset.has(optionspec5),
-                    () -> true,
-                    registryaccess$frozen,
-                    flag1
-                );
+            /*
+            RegistryAccess.Frozen frozen = worldStem.registries().compositeAccess();
+            WorldData worldData = worldStem.worldData();
+            boolean hasOptionSpec1 = optionSet.has(optionSpec6);
+            if (optionSet.has(optionSpec4) || hasOptionSpec1) {
+                forceUpgrade(levelStorageAccess, worldData, DataFixers.getDataFixer(), optionSet.has(optionSpec5), () -> true, frozen, hasOptionSpec1);
             }
 
-            levelstoragesource$levelstorageaccess.saveDataTag(registryaccess$frozen, worlddata);
-            final DedicatedServer dedicatedserver = MinecraftServer.spin(
-                p_423201_ -> {
-                    DedicatedServer dedicatedserver1 = new DedicatedServer(
-                        p_423201_,
-                        levelstoragesource$levelstorageaccess,
-                        packrepository,
-                        worldstem,
-                        dedicatedserversettings,
+            levelStorageAccess.saveDataTag(frozen, worldData);
+            */
+            Class.forName(net.minecraft.world.entity.npc.VillagerTrades.class.getName()); // Paper - load this sync so it won't fail later async
+            final DedicatedServer dedicatedServer = MinecraftServer.spin(
+                thread1 -> {
+                    DedicatedServer dedicatedServer1 = new DedicatedServer(
+                        // CraftBukkit start
+                        optionSet,
+                        worldLoader.get(),
+                        thread1,
+                        levelStorageAccess,
+                        packRepository,
+                        worldStem,
+                        dedicatedServerSettings,
                         DataFixers.getDataFixer(),
                         services,
                         LoggerChunkProgressListener::createFromGameruleRadius
                     );
-                    dedicatedserver1.setPort(optionset.valueOf(optionspec11));
-                    dedicatedserver1.setDemo(optionset.has(optionspec2));
-                    dedicatedserver1.setId(optionset.valueOf(optionspec12));
-                    boolean flag2 = !optionset.has(optionspec) && !optionset.valuesOf(optionspec15).contains("nogui");
-                    if (flag2 && !GraphicsEnvironment.isHeadless()) {
-                        dedicatedserver1.showGui();
-                    }
+                    /*
+                    dedicatedServer1.setPort(optionSet.valueOf(optionSpec11));
+                     */
+                    // Paper start
+                    if (optionSet.has("serverId")) {
+                        dedicatedServer1.setId((String) optionSet.valueOf("serverId"));
+                    }
+                    dedicatedServer1.setDemo(optionSet.has("demo"));
+                    // Paper end
+                    /*
+                    dedicatedServer1.setId(optionSet.valueOf(optionSpec12));
+                     */
+                    boolean flag = !optionSet.has("nogui") && !optionSet.nonOptionArguments().contains("nogui");
+                    if (flag && !GraphicsEnvironment.isHeadless()) {
+                        dedicatedServer1.showGui();
+                    }
+                    // Paper start
+                    if (optionSet.has("port")) {
+                        int port = (Integer) optionSet.valueOf("port");
+                        if (port > 0) {
+                            dedicatedServer1.setPort(port);
+                        }
+                    }
+                    // Paper end
 
                     GameTestTicker.SINGLETON.startTicking();
-                    return dedicatedserver1;
+                    return dedicatedServer1;
                 }
             );
+            /* CraftBukkit start
             Thread thread = new Thread("Server Shutdown Thread") {
                 @Override
                 public void run() {
-                    dedicatedserver.halt(true);
+                    dedicatedServer.halt(true);
                 }
             };
             thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
             Runtime.getRuntime().addShutdownHook(thread);
-        } catch (Exception exception1) {
-            LOGGER.error(LogUtils.FATAL_MARKER, "Failed to start the minecraft server", (Throwable)exception1);
+            */ // CraftBukkit end
+        } catch (Exception var42) {
+            LOGGER.error(LogUtils.FATAL_MARKER, "Failed to start the minecraft server", (Throwable)var42);
         }
     }
 
@@ -330,7 +_,7 @@
         RegistryAccess p_323503_,
         boolean p_321704_
     ) {
-        LOGGER.info("Forcing world upgrade!");
+        LOGGER.info("Forcing world upgrade! {}", p_195489_.getLevelId()); // CraftBukkit
 
         try (WorldUpgrader worldupgrader = new WorldUpgrader(p_195489_, p_195490_, p_401284_, p_323503_, p_195491_, p_321704_)) {
             Component component = null;
