--- a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -72,6 +_,7 @@
                             return false;
                         } else {
                             mutablelong.setValue(p_390573_ + 20L + p_390571_.getRandom().nextInt(20));
+                            if (p_390571_.paperConfig().entities.behavior.stuckEntityPoiRetryDelay.enabled() && p_390572_.getNavigation().isStuck()) mutablelong.add(p_390571_.paperConfig().entities.behavior.stuckEntityPoiRetryDelay.intValue()); // Paper - Next stuck check delay config
                             PoiManager poimanager = p_390571_.getPoiManager();
                             long2objectmap.long2ObjectEntrySet().removeIf(p_22338_ -> !p_22338_.getValue().isStillValid(p_390573_));
                             Predicate<BlockPos> predicate = p_258266_ -> {
@@ -85,12 +_,16 @@
                                     return true;
                                 }
                             };
-                            Set<Pair<Holder<PoiType>, BlockPos>> set = poimanager.findAllClosestFirstWithType(
-                                    p_390608_, predicate, p_390572_.blockPosition(), 48, PoiManager.Occupancy.HAS_SPACE
-                                )
-                                .limit(5L)
-                                .filter(p_390563_ -> p_390593_.test(p_390571_, p_390563_.getSecond()))
-                                .collect(Collectors.toSet());
+                            // Paper start - optimise POI access
+                            final java.util.List<Pair<Holder<PoiType>, BlockPos>> poiposes = new java.util.ArrayList<>();
+                            io.papermc.paper.util.PoiAccess.findNearestPoiPositions(poimanager, p_390608_, predicate, p_390572_.blockPosition(), p_390571_.purpurConfig.villagerAcquirePoiSearchRadius, p_390571_.purpurConfig.villagerAcquirePoiSearchRadius*p_390571_.purpurConfig.villagerAcquirePoiSearchRadius, PoiManager.Occupancy.HAS_SPACE, false, 5, poiposes); // Purpur - Configurable villager search radius
+                            final Set<Pair<Holder<PoiType>, BlockPos>> set = new java.util.HashSet<>(poiposes.size());
+                            for (final Pair<Holder<PoiType>, BlockPos> poiPose : poiposes) {
+                                if (p_390593_.test(p_390571_, poiPose.getSecond())) {
+                                    set.add(poiPose);
+                                }
+                            }
+                            // Paper end - optimise POI access
                             Path path = findPathToPois(p_390572_, set);
                             if (path != null && path.canReach()) {
                                 BlockPos blockpos = path.getTarget();
