--- a/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
+++ b/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
@@ -28,9 +_,18 @@
                             return false;
                         } else {
                             LivingEntity livingentity = p_419440_.get(p_419445_);
-                            if (p_416204_.closerThan(livingentity, p_259321_.getMaxValue() + 1) && !p_416204_.closerThan(livingentity, p_259321_.getMinValue())
-                                )
-                             {
+                            if (p_416204_.closerThan(livingentity, p_259321_.getMaxValue() + 1) && !p_416204_.closerThan(livingentity, p_259321_.getMinValue())) {
+                                // CraftBukkit start
+                                org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(p_416204_, livingentity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER);
+                                if (event.isCancelled()) {
+                                    return false;
+                                }
+                                if (event.getTarget() == null) {
+                                    p_419445_.erase();
+                                    return true;
+                                }
+                                livingentity = ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle();
+                                // CraftBukkit end
                                 WalkTarget walktarget = new WalkTarget(
                                     new EntityTracker(livingentity, p_420043_, p_420043_), p_259190_.apply(p_416204_), p_259321_.getMinValue() - 1
                                 );
