--- a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
@@ -72,8 +_,20 @@
     protected void start(ServerLevel p_147736_, PathfinderMob p_147737_, long p_147738_) {
         Brain<?> brain = p_147737_.getBrain();
         brain.getMemory(MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES)
-            .flatMap(p_375645_ -> p_375645_.findClosest(p_375650_ -> this.ramTargeting.test(p_147736_, p_147737_, p_375650_)))
-            .ifPresent(p_147778_ -> this.chooseRamPosition(p_147737_, p_147778_));
+            .flatMap(
+                nearestVisibleLivingEntities -> nearestVisibleLivingEntities.findClosest(livingEntity -> this.ramTargeting.test(p_147736_, p_147737_, livingEntity))
+            )
+            // CraftBukkit start
+            .map((livingEntity) -> {
+                org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(p_147737_, livingEntity, (livingEntity instanceof net.minecraft.server.level.ServerPlayer) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+                if (event.isCancelled() || event.getTarget() == null) {
+                    return null;
+                }
+                livingEntity = ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle();
+                return livingEntity;
+            })
+            // CraftBukkit end
+            .ifPresent(entity1 -> this.chooseRamPosition(p_147737_, entity1));
     }
 
     protected void stop(ServerLevel p_147762_, E p_147763_, long p_147764_) {
