--- a/net/minecraft/world/entity/ai/behavior/ShowTradesToPlayer.java
+++ b/net/minecraft/world/entity/ai/behavior/ShowTradesToPlayer.java
@@ -44,6 +_,7 @@
     }
 
     public boolean canStillUse(ServerLevel p_24109_, Villager p_24110_, long p_24111_) {
+        if (!p_24110_.level().purpurConfig.villagerDisplayTradeItem) return false; // Purpur - Option for villager display trade item
         return this.checkExtraStartConditions(p_24109_, p_24110_)
             && this.lookTime > 0
             && p_24110_.getBrain().getMemory(MemoryModuleType.INTERACTION_TARGET).isPresent();
