--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -30,7 +_,7 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private final ServerRecipeBook.DisplayResolver displayResolver;
     @VisibleForTesting
-    protected final Set<ResourceKey<Recipe<?>>> known = Sets.newIdentityHashSet();
+    public final Set<ResourceKey<Recipe<?>>> known = Sets.newIdentityHashSet(); // Youer - fix: widen visibility to match Paper (used by CraftPlayer)
     @VisibleForTesting
     protected final Set<ResourceKey<Recipe<?>>> highlight = Sets.newIdentityHashSet();
 
@@ -59,24 +_,28 @@
         this.highlight.add(p_379380_);
     }
 
-    public int addRecipes(Collection<RecipeHolder<?>> p_12792_, ServerPlayer p_12793_) {
+    public int addRecipes(Collection<RecipeHolder<?>> recipes, ServerPlayer player) {
         List<ClientboundRecipeBookAddPacket.Entry> list = new ArrayList<>();
 
-        for (RecipeHolder<?> recipeholder : p_12792_) {
-            ResourceKey<Recipe<?>> resourcekey = recipeholder.id();
-            if (!this.known.contains(resourcekey) && !recipeholder.value().isSpecial()) {
-                this.add(resourcekey);
-                this.addHighlight(resourcekey);
+        for (RecipeHolder<?> recipeHolder : recipes) {
+            ResourceKey<Recipe<?>> resourceKey = recipeHolder.id();
+            if (!this.known.contains(resourceKey) && !recipeHolder.value().isSpecial()) {
+                // Paper start - PlayerRecipeDiscoverEvent event
+                final org.bukkit.event.player.PlayerRecipeDiscoverEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerRecipeListUpdateEvent(player, recipeHolder);
+                if (event.isCancelled()) continue;
+                // Paper end - PlayerRecipeDiscoverEvent event
+                this.add(resourceKey);
+                this.addHighlight(resourceKey);
                 this.displayResolver
                     .displaysForRecipe(
-                        resourcekey, p_379913_ -> list.add(new ClientboundRecipeBookAddPacket.Entry(p_379913_, recipeholder.value().showNotification(), true))
+                        resourceKey, entry -> list.add(new ClientboundRecipeBookAddPacket.Entry(entry, event.shouldShowNotification(), true)) // Paper - set notification from the event
                     );
-                CriteriaTriggers.RECIPE_UNLOCKED.trigger(p_12793_, recipeholder);
+                CriteriaTriggers.RECIPE_UNLOCKED.trigger(player, recipeHolder);
             }
         }
 
-        if (!list.isEmpty()) {
-            p_12793_.connection.send(new ClientboundRecipeBookAddPacket(list, false));
+        if (!list.isEmpty() && player.connection != null) { // SPIGOT-4478 during PlayerLoginEvent
+            player.connection.send(new ClientboundRecipeBookAddPacket(list, false));
         }
 
         return list.size();
@@ -93,7 +_,7 @@
             }
         }
 
-        if (!list.isEmpty()) {
+        if (!list.isEmpty() && p_12808_.connection != null) { // SPIGOT-4478 during PlayerLoginEvent
             p_12808_.connection.send(new ClientboundRecipeBookRemovePacket(list));
         }
 
@@ -103,6 +_,7 @@
     private void loadRecipes(List<ResourceKey<Recipe<?>>> p_404832_, Consumer<ResourceKey<Recipe<?>>> p_12799_, Predicate<ResourceKey<Recipe<?>>> p_379754_) {
         for (ResourceKey<Recipe<?>> resourcekey : p_404832_) {
             if (!p_379754_.test(resourcekey)) {
+                if (!org.purpurmc.purpur.PurpurConfig.loggerSuppressUnrecognizedRecipeErrors) // Purpur - Logger settings (suppressing pointless logs)
                 LOGGER.error("Tried to load unrecognized recipe: {} removed now.", resourcekey);
             } else {
                 p_12799_.accept(resourcekey);
