--- a/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
+++ b/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
@@ -57,7 +_,7 @@
         return this;
     }
 
-    public boolean test(ServerLevel p_376690_, @Nullable LivingEntity p_26886_, LivingEntity p_26887_) {
+    public boolean test(ServerLevel level, @Nullable LivingEntity p_26886_, LivingEntity p_26887_) {
         if (p_26886_ == p_26887_) {
             return false;
         } else if (!p_26887_.canBeSeenByAnyone()) {
@@ -94,6 +_,6 @@
 
     @FunctionalInterface
     public interface Selector {
-        boolean test(LivingEntity p_376107_, ServerLevel p_376747_);
+        boolean test(LivingEntity entity, ServerLevel level);
     }
 }
