--- a/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -60,8 +_,10 @@
         this.eatAnimationTick = Math.max(0, this.eatAnimationTick - 1);
         if (this.eatAnimationTick == this.adjustedTickDelay(4)) {
             BlockPos blockpos = this.mob.blockPosition();
-            if (IS_EDIBLE.test(this.level.getBlockState(blockpos))) {
-                if (getServerLevel(this.level).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            final BlockState blockstate = this.level.getBlockState(blockpos); // Paper - fix wrong block state
+            if (IS_EDIBLE.test(blockstate)) { // Paper - fix wrong block state
+                Boolean mobGriefingOverride = getServerLevel(this.level).purpurConfig.sheepMobGriefingOverride; // Purpur - Add mobGriefing override to everything affected
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockpos, blockstate.getFluidState().createLegacyBlock(), !(mobGriefingOverride != null ? mobGriefingOverride : net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.level), this.mob)))) { // CraftBukkit // Paper - fix wrong block state
                     this.level.destroyBlock(blockpos, false);
                 }
 
@@ -69,7 +_,8 @@
             } else {
                 BlockPos blockpos1 = blockpos.below();
                 if (this.level.getBlockState(blockpos1).is(Blocks.GRASS_BLOCK)) {
-                    if (getServerLevel(this.level).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                    Boolean mobGriefingOverride = getServerLevel(this.level).purpurConfig.sheepMobGriefingOverride; // Purpur - Add mobGriefing override to everything affected
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockpos1, Blocks.DIRT.defaultBlockState(), !(mobGriefingOverride != null ? mobGriefingOverride : net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.level), this.mob)))) { // CraftBukkit // Paper - Fix wrong block state
                         this.level.levelEvent(2001, blockpos1, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
                         this.level.setBlock(blockpos1, Blocks.DIRT.defaultBlockState(), 2);
                     }
