--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -25,7 +_,7 @@
     private double pRotX;
     private double pRotY;
     @Nullable
-    protected Player player;
+    protected LivingEntity player; // CraftBukkit
     private int calmDown;
     private boolean isRunning;
     private final Predicate<ItemStack> items;
@@ -58,12 +_,21 @@
         } else {
             this.player = getServerLevel(this.mob)
                 .getNearestPlayer(this.targetingConditions.range(this.mob.getAttributeValue(Attributes.TEMPT_RANGE)), this.mob);
+            // CraftBukkit start
+            if (this.player != null) {
+                org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(this.mob, this.player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TEMPT);
+                if (event.isCancelled()) {
+                    return false;
+                }
+                this.player = (event.getTarget() == null) ? null : ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle();
+            }
+            // CraftBukkit end
             return this.player != null;
         }
     }
 
-    private boolean shouldFollow(LivingEntity p_148139_) {
-        return this.items.test(p_148139_.getMainHandItem()) || this.items.test(p_148139_.getOffhandItem());
+    private boolean shouldFollow(LivingEntity entity) {
+        return (this.items.test(entity.getMainHandItem()) || this.items.test(entity.getOffhandItem())) && (!(this.mob instanceof net.minecraft.world.entity.npc.Villager villager) || !villager.isSleeping()); // Purpur - Villagers follow emerald blocks
     }
 
     @Override
@@ -124,8 +_,8 @@
         this.mob.getNavigation().stop();
     }
 
-    protected void navigateTowards(Player p_416293_) {
-        this.mob.getNavigation().moveTo(p_416293_, this.speedModifier);
+    protected void navigateTowards(LivingEntity player) { // Paper
+        this.mob.getNavigation().moveTo(player, this.speedModifier);
     }
 
     public boolean isRunning() {
@@ -143,8 +_,8 @@
         }
 
         @Override
-        protected void navigateTowards(Player p_415789_) {
-            Vec3 vec3 = p_415789_.getEyePosition().subtract(this.mob.position()).scale(this.mob.getRandom().nextDouble()).add(this.mob.position());
+        protected void navigateTowards(LivingEntity player) { // Paper
+            Vec3 vec3 = player.getEyePosition().subtract(this.mob.position()).scale(this.mob.getRandom().nextDouble()).add(this.mob.position());
             this.mob.getMoveControl().setWantedPosition(vec3.x, vec3.y, vec3.z, this.speedModifier);
         }
     }
