--- a/net/minecraft/world/item/component/ItemContainerContents.java
+++ b/net/minecraft/world/item/component/ItemContainerContents.java
@@ -29,6 +_,7 @@
         .xmap(ItemContainerContents::fromSlots, ItemContainerContents::asSlots);
     public static final StreamCodec<RegistryFriendlyByteBuf, ItemContainerContents> STREAM_CODEC = ItemStack.OPTIONAL_STREAM_CODEC
         .apply(ByteBufCodecs.list(256))
+        .apply(ByteBufCodecs::increaseDepth) // Paper - Track codec depth
         .map(ItemContainerContents::new, p_331691_ -> p_331691_.items);
     public final NonNullList<ItemStack> items;
     private final int hashCode;
@@ -161,6 +_,35 @@
 
         if (j - i > 0) {
             p_399797_.accept(Component.translatable("item.container.more_items", j - i).withStyle(ChatFormatting.ITALIC));
+        }
+    }
+
+    /**
+     * Neo:
+     * {@return the number of slots in this container}
+     */
+    public int getSlots() {
+        return this.items.size();
+    }
+
+    /**
+     * Neo: Gets a copy of the stack at a particular slot.
+     *
+     * @param slot The slot to check. Must be within [0, {@link #getSlots()}]
+     * @return A copy of the stack in that slot
+     * @throws UnsupportedOperationException if the provided slot index is out-of-bounds.
+     */
+    public ItemStack getStackInSlot(int slot) {
+        validateSlotIndex(slot);
+        return this.items.get(slot).copy();
+    }
+
+    /**
+     * Neo: Throws {@link UnsupportedOperationException} if the provided slot index is invalid.
+     */
+    private void validateSlotIndex(int slot) {
+        if (slot < 0 || slot >= getSlots()) {
+            throw new UnsupportedOperationException("Slot " + slot + " not in valid range - [0," + getSlots() + ")");
         }
     }
 
