--- a/net/minecraft/world/item/component/OminousBottleAmplifier.java
+++ b/net/minecraft/world/item/component/OminousBottleAmplifier.java
@@ -29,8 +_,15 @@
 
     @Override
     public void onConsume(Level p_366547_, LivingEntity p_366849_, ItemStack p_366530_, Consumable p_366459_) {
-        p_366849_.addEffect(new MobEffectInstance(MobEffects.BAD_OMEN, 120000, this.value, false, false, true));
-    }
+        p_366849_.addEffect(new MobEffectInstance(MobEffects.BAD_OMEN, 120000, this.value, false, false, true)); // Paper - properly resend entities - diff on change for below
+    }
+
+    // Paper start - properly resend entities - collect packets for bundle
+    @Override
+    public void cancelUsingItem(net.minecraft.server.level.ServerPlayer player, ItemStack stack, List<net.minecraft.network.protocol.Packet<? super net.minecraft.network.protocol.game.ClientGamePacketListener>> collectedPackets) {
+        collectedPackets.add(new net.minecraft.network.protocol.game.ClientboundRemoveMobEffectPacket(player.getId(), MobEffects.BAD_OMEN));
+    }
+    // Paper end - properly resend entities - collect packets for bundle
 
     @Override
     public void addToTooltip(Item.TooltipContext p_366592_, Consumer<Component> p_366659_, TooltipFlag p_366416_, DataComponentGetter p_399937_) {
