--- a/net/minecraft/world/item/component/ResolvableProfile.java
+++ b/net/minecraft/world/item/component/ResolvableProfile.java
@@ -22,9 +_,10 @@
         p_337950_ -> p_337950_.group(
                 ExtraCodecs.PLAYER_NAME.optionalFieldOf("name").forGetter(ResolvableProfile::name),
                 UUIDUtil.CODEC.optionalFieldOf("id").forGetter(ResolvableProfile::id),
+                UUIDUtil.STRING_CODEC.lenientOptionalFieldOf("Id").forGetter($ -> Optional.empty()), // Paper
                 ExtraCodecs.PROPERTY_MAP.optionalFieldOf("properties", new PropertyMap()).forGetter(ResolvableProfile::properties)
             )
-            .apply(p_337950_, ResolvableProfile::new)
+            .apply(p_337950_, (p_332623_, uuid, uuid2, propertyMap) -> new ResolvableProfile(p_332623_, uuid2.or(() -> uuid), propertyMap)) // Paper
     );
     public static final Codec<ResolvableProfile> CODEC = Codec.withAlternative(
         FULL_CODEC, ExtraCodecs.PLAYER_NAME, p_332623_ -> new ResolvableProfile(Optional.of(p_332623_), Optional.empty(), new PropertyMap())
@@ -54,7 +_,7 @@
         } else {
             Optional<GameProfile> optional;
             if (this.id.isPresent()) {
-                optional = SkullBlockEntity.fetchGameProfile(this.id.get()).getNow(null);
+                optional = SkullBlockEntity.fetchGameProfile(this.id.get(), this.name.orElse(null)).getNow(null); // Paper - player profile events - pass name
             } else {
                 optional = SkullBlockEntity.fetchGameProfile(this.name.orElseThrow()).getNow(null);
             }
@@ -68,7 +_,7 @@
             return CompletableFuture.completedFuture(this);
         } else {
             return this.id.isPresent()
-                ? SkullBlockEntity.fetchGameProfile(this.id.get()).thenApply(this::createProfile)
+                ? SkullBlockEntity.fetchGameProfile(this.id.get(), this.name.orElse(null)).thenApply(this::createProfile)  // Paper - player profile events - pass name
                 : SkullBlockEntity.fetchGameProfile(this.name.orElseThrow()).thenApply(this::createProfile);
         }
     }
