--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -124,7 +_,31 @@
         this.setPathfindingMalus(PathType.LAVA, 8.0F);
         this.setPathfindingMalus(PathType.DAMAGE_FIRE, 0.0F);
         this.setPathfindingMalus(PathType.DANGER_FIRE, 0.0F);
-    }
+        this.moveControl = new org.purpurmc.purpur.controller.MoveControllerWASD(this, 0.5F); // Purpur - Ridables
+    }
+
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.wardenRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.wardenRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.wardenControllable;
+    }
+
+    @Override
+    protected void registerGoals() {
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
+        this.targetSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
+    }
+    // Purpur end - Ridables
 
     @Override
     public Packet<ClientGamePacketListener> getAddEntityPacket(ServerEntity p_352154_) {
@@ -281,6 +_,7 @@
     protected void customServerAiStep(ServerLevel p_376677_) {
         ProfilerFiller profilerfiller = Profiler.get();
         profilerfiller.push("wardenBrain");
+        if ((getRider() == null || !this.isControllable())) // Purpur - only use brain if no rider
         this.getBrain().tick(p_376677_, this);
         profilerfiller.pop();
         super.customServerAiStep(p_376677_);
@@ -389,6 +_,7 @@
 
     @Contract("null->false")
     public boolean canTargetEntity(@Nullable Entity p_219386_) {
+        if (getRider() != null && isControllable()) return false; // Purpur - Ridables
         return p_219386_ instanceof LivingEntity livingentity
             && this.level() == p_219386_.level()
             && EntitySelector.NO_CREATIVE_OR_SPECTATOR.test(p_219386_)
@@ -402,7 +_,7 @@
 
     public static void applyDarknessAround(ServerLevel p_219376_, Vec3 p_219377_, @Nullable Entity p_219378_, int p_219379_) {
         MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.DARKNESS, 260, 0, false, false);
-        MobEffectUtil.addEffectToPlayersAround(p_219376_, p_219378_, p_219377_, p_219379_, mobeffectinstance, 200);
+        MobEffectUtil.addEffectToPlayersAround(p_219376_, p_219378_, p_219377_, p_219379_, mobeffectinstance, 200, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WARDEN); // CraftBukkit - Add EntityPotionEffectEvent.Cause
     }
 
     @Override
@@ -444,16 +_,25 @@
     }
 
     @VisibleForTesting
-    public void increaseAngerAt(@Nullable Entity p_219388_, int p_219389_, boolean p_219390_) {
-        if (!this.isNoAi() && this.canTargetEntity(p_219388_)) {
+    public void increaseAngerAt(@Nullable Entity entity, int offset, boolean playListeningSound) {
+        if (!this.isNoAi() && this.canTargetEntity(entity)) {
+            // Paper start - Add WardenAngerChangeEvent
+            int activeAnger = this.angerManagement.getActiveAnger(entity);
+            io.papermc.paper.event.entity.WardenAngerChangeEvent event = new io.papermc.paper.event.entity.WardenAngerChangeEvent((org.bukkit.entity.Warden) this.getBukkitEntity(), entity.getBukkitEntity(), activeAnger, Math.min(150, activeAnger + offset));
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
+            offset = event.getNewAnger() - activeAnger;
+            // Paper end - Add WardenAngerChangeEvent
             WardenAi.setDigCooldown(this);
             boolean flag = !(this.getTarget() instanceof Player);
-            int i = this.angerManagement.increaseAnger(p_219388_, p_219389_);
-            if (p_219388_ instanceof Player && flag && AngerLevel.byAnger(i).isAngry()) {
+            int i = this.angerManagement.increaseAnger(entity, offset);
+            if (entity instanceof Player && flag && AngerLevel.byAnger(i).isAngry()) {
                 this.getBrain().eraseMemory(MemoryModuleType.ATTACK_TARGET);
             }
 
-            if (p_219390_) {
+            if (playListeningSound) {
                 this.playListeningSound();
             }
         }
