--- a/net/minecraft/world/level/entity/EntityLookup.java
+++ b/net/minecraft/world/level/entity/EntityLookup.java
@@ -29,13 +_,21 @@
         return Iterables.unmodifiableIterable(this.byId.values());
     }
 
-    public void add(T p_156815_) {
-        UUID uuid = p_156815_.getUUID();
+    public void add(T entity) {
+        UUID uuid = entity.getUUID();
         if (this.byUuid.containsKey(uuid)) {
-            LOGGER.warn("Duplicate entity UUID {}: {}", uuid, p_156815_);
+            LOGGER.warn("Duplicate entity UUID {}: {}", uuid, entity);
+            // Paper start - extra debug info
+            if (entity instanceof net.minecraft.world.entity.Entity) {
+                final T old = this.byUuid.get(entity.getUUID());
+                if (old instanceof net.minecraft.world.entity.Entity oldCast && oldCast.getId() != entity.getId() && oldCast.valid) {
+                    LOGGER.error("Overwrote an existing entity {} with {}", oldCast, entity);
+                }
+            }
+            // Paper end - extra debug info
         } else {
-            this.byUuid.put(uuid, p_156815_);
-            this.byId.put(p_156815_.getId(), p_156815_);
+            this.byUuid.put(uuid, entity);
+            this.byId.put(entity.getId(), entity);
         }
     }
 
