--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -53,6 +_,7 @@
     public int maxNearbyEntities = 6;
     public int requiredPlayerRange = 16;
     public int spawnRange = 4;
+    private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public void setEntityId(EntityType<?> p_253682_, @Nullable Level p_254041_, RandomSource p_254221_, BlockPos p_254050_) {
         this.getOrCreateNextSpawnData(p_254041_, p_254221_, p_254050_)
@@ -84,13 +_,25 @@
     }
 
     public void serverTick(ServerLevel p_151312_, BlockPos p_151313_) {
+        // Youer start - remove the spawner outright when spawners are disabled
+        if (com.mohistmc.youer.YouerConfig.custom_disabled_spawner) {
+            p_151312_.setBlock(p_151313_, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(), 3);
+            return;
+        }
+        // Youer end
+        if (spawnCount <= 0 || maxNearbyEntities <= 0) return; // Paper - Ignore impossible spawn tick
+        // Paper start - Configurable mob spawner tick rate
+        if (spawnDelay > 0 && --tickDelay > 0) return;
+        tickDelay = p_151312_.paperConfig().tickRates.mobSpawner;
+        if (tickDelay == -1) { return; } // If disabled
+        // Paper end - Configurable mob spawner tick rate
         if (this.isNearPlayer(p_151312_, p_151313_)) {
-            if (this.spawnDelay == -1) {
+            if (this.spawnDelay < -tickDelay) { // Paper - Configurable mob spawner tick rate
                 this.delay(p_151312_, p_151313_);
             }
 
             if (this.spawnDelay > 0) {
-                this.spawnDelay--;
+                this.spawnDelay -= tickDelay; // Paper - Configurable mob spawner tick rate
             } else {
                 boolean flag = false;
                 RandomSource randomsource = p_151312_.getRandom();
@@ -128,6 +_,21 @@
                                 continue;
                             }
 
+                            // Paper start - PreCreatureSpawnEvent
+                            com.destroystokyo.paper.event.entity.PreSpawnerSpawnEvent event = new com.destroystokyo.paper.event.entity.PreSpawnerSpawnEvent(
+                                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(vec3, p_151312_.getWorld()),
+                                org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(optional.get()),
+                                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(blockpos, p_151312_)
+                            );
+                            if (!event.callEvent()) {
+                                flag = true;
+                                if (event.shouldAbortSpawn()) {
+                                    break;
+                                }
+                                continue;
+                            }
+                            // Paper end - PreCreatureSpawnEvent
+
                             Entity entity = EntityType.loadEntityRecursive(valueinput, p_151312_, EntitySpawnReason.SPAWNER, p_404552_ -> {
                                 p_404552_.snapTo(vec3.x, vec3.y, vec3.z, p_404552_.getYRot(), p_404552_.getXRot());
                                 return p_404552_;
@@ -158,21 +_,32 @@
 
                             entity.snapTo(entity.getX(), entity.getY(), entity.getZ(), randomsource.nextFloat() * 360.0F, 0.0F);
                             if (entity instanceof Mob mob) {
-                                if (spawndata.getCustomSpawnRules().isEmpty() && !mob.checkSpawnRules(p_151312_, EntitySpawnReason.SPAWNER)
-                                    || !mob.checkSpawnObstruction(p_151312_)) {
+                                if (!net.neoforged.neoforge.event.EventHooks.checkSpawnPositionSpawner(mob, p_151312_, EntitySpawnReason.SPAWNER, spawndata, this)) {
                                     continue;
                                 }
 
                                 boolean flag1 = spawndata.getEntityToSpawn().size() == 1 && spawndata.getEntityToSpawn().getString("id").isPresent();
-                                if (flag1) {
-                                    ((Mob)entity)
-                                        .finalizeSpawn(p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.SPAWNER, null);
-                                }
+                                // Neo: Patch in FinalizeSpawn for spawners so it may be fired unconditionally, instead of only when vanilla would normally call it.
+                                // The local flag1 is the conditions under which the spawner will normally call Mob#finalizeSpawn.
+                                net.neoforged.neoforge.event.EventHooks.finalizeMobSpawnSpawner(mob, p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.SPAWNER, null, this, flag1);
 
                                 spawndata.getEquipment().ifPresent(mob::equip);
+                                // Spigot start
+                                if (mob.level().spigotConfig.nerfSpawnerMobs) {
+                                    mob.aware = false;
+                                }
+                                // Spigot end
                             }
 
-                            if (!p_151312_.tryAddFreshEntityWithPassengers(entity)) {
+                            // Paper start
+                            entity.spawnedViaMobSpawner = true;
+                            entity.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER;
+                            flag = true;
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callSpawnerSpawnEvent(entity, p_151313_).isCancelled()) {
+                                continue;
+                            }
+                            if (!p_151312_.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER)) {
+                                // Paper end
                                 this.delay(p_151312_, p_151313_);
                                 return;
                             }
@@ -183,7 +_,7 @@
                                 ((Mob)entity).spawnAnim();
                             }
 
-                            flag = true;
+                            //flag = true; // Paper - moved up above cancellable event
                         }
                     }
                 }
@@ -210,12 +_,14 @@
     }
 
     public void load(@Nullable Level p_151329_, BlockPos p_151330_, ValueInput p_421598_) {
-        this.spawnDelay = p_421598_.getShortOr("Delay", (short)20);
+        this.spawnDelay = p_421598_.getIntOr("Paper.Delay", p_421598_.getShortOr("Delay", (short) 20)); // Paper - use int if set
         p_421598_.read("SpawnData", SpawnData.CODEC).ifPresent(p_400944_ -> this.setNextSpawnData(p_151329_, p_151330_, p_400944_));
         this.spawnPotentials = p_421598_.read("SpawnPotentials", SpawnData.LIST_CODEC)
             .orElseGet(() -> WeightedList.of(this.nextSpawnData != null ? this.nextSpawnData : new SpawnData()));
-        this.minSpawnDelay = p_421598_.getIntOr("MinSpawnDelay", 200);
-        this.maxSpawnDelay = p_421598_.getIntOr("MaxSpawnDelay", 800);
+        // Paper start - use int if set
+        this.minSpawnDelay = p_421598_.getIntOr("Paper.MinSpawnDelay", p_421598_.getIntOr("MinSpawnDelay", 200));
+        this.maxSpawnDelay = p_421598_.getIntOr("Paper.MaxSpawnDelay", p_421598_.getIntOr("MaxSpawnDelay", 800));
+        // Paper end - use int if set
         this.spawnCount = p_421598_.getIntOr("SpawnCount", 4);
         this.maxNearbyEntities = p_421598_.getIntOr("MaxNearbyEntities", 6);
         this.requiredPlayerRange = p_421598_.getIntOr("RequiredPlayerRange", 16);
@@ -224,9 +_,19 @@
     }
 
     public void save(ValueOutput p_422002_) {
-        p_422002_.putShort("Delay", (short)this.spawnDelay);
-        p_422002_.putShort("MinSpawnDelay", (short)this.minSpawnDelay);
-        p_422002_.putShort("MaxSpawnDelay", (short)this.maxSpawnDelay);
+        // Paper start
+        if (this.spawnDelay > Short.MAX_VALUE) {
+            p_422002_.putInt("Paper.Delay", this.spawnDelay);
+        }
+        p_422002_.putShort("Delay", (short) Math.min(Short.MAX_VALUE, this.spawnDelay));
+
+        if (this.minSpawnDelay > Short.MAX_VALUE || this.maxSpawnDelay > Short.MAX_VALUE) {
+            p_422002_.putInt("Paper.MinSpawnDelay", this.minSpawnDelay);
+            p_422002_.putInt("Paper.MaxSpawnDelay", this.maxSpawnDelay);
+        }
+        p_422002_.putShort("MinSpawnDelay", (short) Math.min(Short.MAX_VALUE, this.minSpawnDelay));
+        p_422002_.putShort("MaxSpawnDelay", (short) Math.min(Short.MAX_VALUE, this.maxSpawnDelay));
+        // Paper end
         p_422002_.putShort("SpawnCount", (short)this.spawnCount);
         p_422002_.putShort("MaxNearbyEntities", (short)this.maxNearbyEntities);
         p_422002_.putShort("RequiredPlayerRange", (short)this.requiredPlayerRange);
@@ -284,5 +_,12 @@
 
     public double getoSpin() {
         return this.oSpin;
+    }
+
+    @Override
+    @org.jetbrains.annotations.Nullable
+    public com.mojang.datafixers.util.Either<net.minecraft.world.level.block.entity.BlockEntity, Entity> getOwner() {
+        // The vanilla anonymous classes have proper overrides, but we return null here for compatibility.
+        return null;
     }
 }
