--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -33,10 +_,21 @@
 
     BlockState getBlockState(BlockPos p_45571_);
 
+    // Paper start - if loaded util
+    @Nullable BlockState getBlockStateIfLoaded(BlockPos pos);
+
+    default @Nullable net.minecraft.world.level.block.Block getBlockIfLoaded(BlockPos pos) {
+        BlockState type = this.getBlockStateIfLoaded(pos);
+        return type == null ? null : type.getBlock();
+    }
+
+    @Nullable FluidState getFluidIfLoaded(BlockPos pos);
+    // Paper end
+
     FluidState getFluidState(BlockPos p_45569_);
 
     default int getLightEmission(BlockPos p_45572_) {
-        return this.getBlockState(p_45572_).getLightEmission();
+        return this.getBlockState(p_45572_).getLightEmission(this, p_45572_);
     }
 
     default Stream<BlockState> getBlockStates(AABB p_45557_) {
@@ -64,22 +_,49 @@
         );
     }
 
+    // CraftBukkit start - moved block handling into separate method for use by Block#rayTrace
+    default BlockHitResult clip(ClipContext traverseContext, BlockPos traversePos) {
+        // Paper start - Add predicate for blocks when raytracing
+        return clip(traverseContext, traversePos, null);
+    }
+
+    default BlockHitResult clip(ClipContext traverseContext, BlockPos traversePos, @Nullable java.util.function.Predicate<? super org.bukkit.block.Block> canCollide) {
+        // Paper end - Add predicate for blocks when raytracing
+        // Paper start - Prevent raytrace from loading chunks
+        BlockState blockState = this.getBlockStateIfLoaded(traversePos);
+        if (blockState == null) {
+            // copied the last function parameter (listed below)
+            Vec3 vec3d = traverseContext.getFrom().subtract(traverseContext.getTo());
+
+            return BlockHitResult.miss(traverseContext.getTo(), Direction.getApproximateNearest(vec3d.x, vec3d.y, vec3d.z), BlockPos.containing(traverseContext.getTo()));
+        }
+        // Paper end - Prevent raytrace from loading chunks
+        if (blockState.isAir() || (canCollide != null && this instanceof LevelAccessor levelAccessor && !canCollide.test(org.bukkit.craftbukkit.block.CraftBlock.at(levelAccessor, traversePos)))) return null; // Paper - Perf: optimise air cases & check canCollide predicate
+            FluidState fluidState = blockState.getFluidState(); // Paper - Perf: don't need to go to world state again
+            Vec3 from = traverseContext.getFrom();
+            Vec3 to = traverseContext.getTo();
+            VoxelShape blockShape = traverseContext.getBlockShape(blockState, this, traversePos);
+            BlockHitResult blockHitResult = this.clipWithInteractionOverride(from, to, traversePos, blockShape, blockState);
+            VoxelShape fluidShape = traverseContext.getFluidShape(fluidState, this, traversePos);
+            BlockHitResult blockHitResult1 = fluidShape.clip(from, to, traversePos);
+            double d = blockHitResult == null ? Double.MAX_VALUE : traverseContext.getFrom().distanceToSqr(blockHitResult.getLocation());
+            double d1 = blockHitResult1 == null ? Double.MAX_VALUE : traverseContext.getFrom().distanceToSqr(blockHitResult1.getLocation());
+            return d <= d1 ? blockHitResult : blockHitResult1;
+    }
+    // CraftBukkit end
+
     default BlockHitResult clip(ClipContext p_45548_) {
-        return traverseBlocks(p_45548_.getFrom(), p_45548_.getTo(), p_45548_, (p_151359_, p_151360_) -> {
-            BlockState blockstate = this.getBlockState(p_151360_);
-            FluidState fluidstate = this.getFluidState(p_151360_);
-            Vec3 vec3 = p_151359_.getFrom();
-            Vec3 vec31 = p_151359_.getTo();
-            VoxelShape voxelshape = p_151359_.getBlockShape(blockstate, this, p_151360_);
-            BlockHitResult blockhitresult = this.clipWithInteractionOverride(vec3, vec31, p_151360_, voxelshape, blockstate);
-            VoxelShape voxelshape1 = p_151359_.getFluidShape(fluidstate, this, p_151360_);
-            BlockHitResult blockhitresult1 = voxelshape1.clip(vec3, vec31, p_151360_);
-            double d0 = blockhitresult == null ? Double.MAX_VALUE : p_151359_.getFrom().distanceToSqr(blockhitresult.getLocation());
-            double d1 = blockhitresult1 == null ? Double.MAX_VALUE : p_151359_.getFrom().distanceToSqr(blockhitresult1.getLocation());
-            return d0 <= d1 ? blockhitresult : blockhitresult1;
-        }, p_360110_ -> {
-            Vec3 vec3 = p_360110_.getFrom().subtract(p_360110_.getTo());
-            return BlockHitResult.miss(p_360110_.getTo(), Direction.getApproximateNearest(vec3.x, vec3.y, vec3.z), BlockPos.containing(p_360110_.getTo()));
+        // Paper start - Add predicate for blocks when raytracing
+        return clip(p_45548_, (java.util.function.Predicate<org.bukkit.block.Block>) null);
+    }
+
+    default BlockHitResult clip(ClipContext context, @Nullable java.util.function.Predicate<? super org.bukkit.block.Block> canCollide) {
+        // Paper end - Add predicate for blocks when raytracing
+        return (BlockHitResult) BlockGetter.traverseBlocks(context.getFrom(), context.getTo(), context, (raytrace1, blockposition) -> {
+            return this.clip(raytrace1, blockposition, canCollide); // CraftBukkit - moved into separate method // Paper - Add predicate for blocks when raytracing
+        }, failContext -> {
+            Vec3 vec3 = failContext.getFrom().subtract(failContext.getTo());
+            return BlockHitResult.miss(failContext.getTo(), Direction.getApproximateNearest(vec3.x, vec3.y, vec3.z), BlockPos.containing(failContext.getTo()));
         });
     }
 
