--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -432,7 +_,7 @@
         p_422367_.read("Trusted", TRUSTED_LIST_CODEC).orElse(List.of()).forEach(this::addTrustedEntity);
         this.setSleeping(p_422367_.getBooleanOr("Sleeping", false));
         this.setVariant(p_422367_.read("Type", Fox.Variant.CODEC).orElse(Fox.Variant.DEFAULT));
-        this.setSitting(p_422367_.getBooleanOr("Sitting", false));
+        this.setSitting(p_422367_.getBooleanOr("Sitting", false), false); // Paper - Add EntityToggleSitEvent
         this.setIsCrouching(p_422367_.getBooleanOr("Crouching", false));
         if (this.level() instanceof ServerLevel) {
             this.setTargetGoals();
@@ -449,7 +_,13 @@
     }
 
     public void setSitting(boolean p_28611_) {
-        this.setFlag(1, p_28611_);
+        // Paper start - Add EntityToggleSitEvent
+        this.setSitting(p_28611_, true);
+    }
+    public void setSitting(boolean sitting, boolean fireEvent) {
+        if (fireEvent && !new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), sitting).callEvent()) return;
+        // Paper end - Add EntityToggleSitEvent
+        this.setFlag(1, sitting);
     }
 
     public boolean isFaceplanted() {
@@ -508,19 +_,20 @@
             itementity.setPickUpDelay(40);
             itementity.setThrower(this);
             this.playSound(SoundEvents.FOX_SPIT, 1.0F, 1.0F);
-            this.level().addFreshEntity(itementity);
+            this.spawnAtLocation((net.minecraft.server.level.ServerLevel) this.level(), itementity); // Paper - Call EntityDropItemEvent
         }
     }
 
     private void dropItemStack(ItemStack p_28606_) {
         ItemEntity itementity = new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), p_28606_);
-        this.level().addFreshEntity(itementity);
+        this.spawnAtLocation((net.minecraft.server.level.ServerLevel) this.level(), itementity); // Paper - Call EntityDropItemEvent
     }
 
     @Override
     protected void pickUpItem(ServerLevel p_376527_, ItemEntity p_28514_) {
         ItemStack itemstack = p_28514_.getItem();
-        if (this.canHoldItem(itemstack)) {
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(this, p_28514_, itemstack.getCount() - 1, !this.canHoldItem(itemstack)).isCancelled()) { // CraftBukkit - call EntityPickupItemEvent
+            itemstack = p_28514_.getItem(); // CraftBukkit - update itemstack after event
             int i = itemstack.getCount();
             if (i > 1) {
                 this.dropItemStack(itemstack.split(i - 1));
@@ -531,7 +_,7 @@
             this.setItemSlot(EquipmentSlot.MAINHAND, itemstack.split(1));
             this.setGuaranteedDrop(EquipmentSlot.MAINHAND);
             this.take(p_28514_, itemstack.getCount());
-            p_28514_.discard();
+            p_28514_.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             this.ticksSinceEaten = 0;
         }
     }
@@ -622,12 +_,12 @@
     }
 
     @Override
-    public void setTarget(@Nullable LivingEntity p_28574_) {
-        if (this.isDefending() && p_28574_ == null) {
+    public boolean setTarget(@Nullable LivingEntity livingEntity, @Nullable org.bukkit.event.entity.EntityTargetEvent.TargetReason reason) { // CraftBukkit
+        if (this.isDefending() && livingEntity == null) {
             this.setDefending(false);
         }
 
-        super.setTarget(p_28574_);
+        return super.setTarget(livingEntity, reason); // CraftBukkit
     }
 
     void wakeUp() {
@@ -691,15 +_,25 @@
         return this.getTrustedEntities().anyMatch(p_393132_ -> p_393132_.matches(p_394151_));
     }
 
+    // Paper start - handle the bitten item separately like vanilla
     @Override
-    protected void dropAllDeathLoot(ServerLevel p_348640_, DamageSource p_28536_) {
-        ItemStack itemstack = this.getItemBySlot(EquipmentSlot.MAINHAND);
-        if (!itemstack.isEmpty()) {
-            this.spawnAtLocation(p_348640_, itemstack);
+    protected org.bukkit.event.entity.EntityDeathEvent dropAllDeathLoot(ServerLevel level, DamageSource damageSource) {
+        ItemStack itemBySlot = this.getItemBySlot(EquipmentSlot.MAINHAND);
+        boolean releaseMouth = false;
+        if (!itemBySlot.isEmpty() && level.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) { // Fix MC-153010
+            this.spawnAtLocation(level, itemBySlot);
+            releaseMouth = true;
+        }
+
+        org.bukkit.event.entity.EntityDeathEvent deathEvent = super.dropAllDeathLoot(level, damageSource);
+        // Below is code to drop
+        if (deathEvent == null || deathEvent.isCancelled()) return deathEvent;
+
+        if (releaseMouth) {
+            // Paper end
             this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
         }
-
-        super.dropAllDeathLoot(p_348640_, p_28536_);
+        return deathEvent; // Paper
     }
 
     public static boolean isPathClear(Fox p_28472_, LivingEntity p_28473_) {
@@ -856,39 +_,61 @@
 
         @Override
         protected void breed() {
-            ServerLevel serverlevel = this.level;
-            Fox fox = (Fox)this.animal.getBreedOffspring(serverlevel, this.partner);
+            ServerLevel serverLevel = this.level;
+            Fox fox = (Fox)this.animal.getBreedOffspring(serverLevel, this.partner);
+            final net.neoforged.neoforge.event.entity.living.BabyEntitySpawnEvent event = new net.neoforged.neoforge.event.entity.living.BabyEntitySpawnEvent(animal, partner, fox);
+            final boolean cancelled = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event).isCanceled();
+            fox = (Fox) event.getChild();
+            if (cancelled) {
+                //Reset the "inLove" state for the animals
+                this.animal.setAge(6000);
+                this.partner.setAge(6000);
+                this.animal.resetLove();
+                this.partner.resetLove();
+                return;
+            }
             if (fox != null) {
-                ServerPlayer serverplayer = this.animal.getLoveCause();
-                ServerPlayer serverplayer1 = this.partner.getLoveCause();
-                ServerPlayer serverplayer2 = serverplayer;
-                if (serverplayer != null) {
-                    fox.addTrustedEntity(serverplayer);
+                ServerPlayer loveCause = this.animal.getLoveCause();
+                ServerPlayer loveCause1 = this.partner.getLoveCause();
+                ServerPlayer serverPlayer = loveCause;
+                if (loveCause != null) {
+                    fox.addTrustedEntity(loveCause);
                 } else {
-                    serverplayer2 = serverplayer1;
-                }
-
-                if (serverplayer1 != null && serverplayer != serverplayer1) {
-                    fox.addTrustedEntity(serverplayer1);
-                }
-
-                if (serverplayer2 != null) {
-                    serverplayer2.awardStat(Stats.ANIMALS_BRED);
-                    CriteriaTriggers.BRED_ANIMALS.trigger(serverplayer2, this.animal, this.partner, fox);
-                }
-
-                this.animal.setAge(6000);
-                this.partner.setAge(6000);
-                this.animal.resetLove();
-                this.partner.resetLove();
+                    serverPlayer = loveCause1;
+                }
+
+                if (loveCause1 != null && loveCause != loveCause1) {
+                    fox.addTrustedEntity(loveCause1);
+                }
+
+                // CraftBukkit start - call EntityBreedEvent
                 fox.setAge(-24000);
                 fox.snapTo(this.animal.getX(), this.animal.getY(), this.animal.getZ(), 0.0F, 0.0F);
-                serverlevel.addFreshEntityWithPassengers(fox);
+                int experience = this.animal.getRandom().nextInt(7) + 1;
+                org.bukkit.event.entity.EntityBreedEvent entityBreedEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreedEvent(fox, this.animal, this.partner, loveCause, this.animal.breedItem, experience);
+                if (entityBreedEvent.isCancelled()) {
+                    this.animal.resetLove();
+                    this.partner.resetLove();
+                    return;
+                }
+                experience = entityBreedEvent.getExperience();
+                // CraftBukkit end - call EntityBreedEvent
+
+                if (serverPlayer != null) {
+                    serverPlayer.awardStat(Stats.ANIMALS_BRED);
+                    CriteriaTriggers.BRED_ANIMALS.trigger(serverPlayer, this.animal, this.partner, fox);
+                }
+
+                this.animal.setAge(6000);
+                this.partner.setAge(6000);
+                this.animal.resetLove();
+                this.partner.resetLove();
+                serverLevel.addFreshEntityWithPassengers(fox, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
                 this.level.broadcastEntityEvent(this.animal, (byte)18);
-                if (serverlevel.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
+                if (experience > 0 && serverLevel.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) { // Paper - call EntityBreedEvent
                     this.level
                         .addFreshEntity(
-                            new ExperienceOrb(this.level, this.animal.getX(), this.animal.getY(), this.animal.getZ(), this.animal.getRandom().nextInt(7) + 1)
+                            new ExperienceOrb(this.level, this.animal.position(), net.minecraft.world.phys.Vec3.ZERO, experience, org.bukkit.entity.ExperienceOrb.SpawnReason.BREED, loveCause, fox) // Paper - call EntityBreedEvent, add spawn context
                         );
                 }
             }
@@ -935,7 +_,7 @@
         }
 
         protected void onReachedTarget() {
-            if (getServerLevel(Fox.this.level()).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            if (net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(Fox.this.level()), Fox.this)) {
                 BlockState blockstate = Fox.this.level().getBlockState(this.blockPos);
                 if (blockstate.is(Blocks.SWEET_BERRY_BUSH)) {
                     this.pickSweetBerries(blockstate);
@@ -952,6 +_,7 @@
         private void pickSweetBerries(BlockState p_148929_) {
             int i = p_148929_.getValue(SweetBerryBushBlock.AGE);
             p_148929_.setValue(SweetBerryBushBlock.AGE, 1);
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(Fox.this, this.blockPos, p_148929_.setValue(SweetBerryBushBlock.AGE, 1))) return; // CraftBukkit - call EntityChangeBlockEvent
             int j = 1 + Fox.this.level().random.nextInt(2) + (i == 3 ? 1 : 0);
             ItemStack itemstack = Fox.this.getItemBySlot(EquipmentSlot.MAINHAND);
             if (itemstack.isEmpty()) {
@@ -994,7 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25 || Fox.this.isInLava();
+            return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25D || Fox.this.isInLava() || Fox.this.isInFluidType((fluidType, height) -> Fox.this.canSwimInFluidType(fluidType) && height > 0.25D);
         }
     }
 
