--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -142,6 +_,73 @@
         this.getNavigation().setRequiredPathLength(32.0F);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.foxRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.foxRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.foxControllable;
+    }
+
+    @Override
+    public float getJumpPower() {
+        return getRider() != null && this.isControllable() ? 0.5F : super.getJumpPower();
+    }
+
+    @Override
+    public void onMount(Player rider) {
+        super.onMount(rider);
+        setCanPickUpLoot(false);
+        clearStates();
+        setIsPouncing(false);
+        spitOutItem(getItemBySlot(EquipmentSlot.MAINHAND));
+        setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
+    }
+
+    @Override
+    public void onDismount(Player rider) {
+        super.onDismount(rider);
+        setCanPickUpLoot(true);
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.foxMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.foxScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Make entity breeding times configurable
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level().purpurConfig.foxBreedingTicks;
+    }
+    // Purpur end - Make entity breeding times configurable
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.foxTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.foxAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
     @Override
     protected void defineSynchedData(SynchedEntityData.Builder p_326335_) {
         super.defineSynchedData(p_326335_);
@@ -161,6 +_,7 @@
             this, AbstractFish.class, 20, false, false, (p_28600_, p_376887_) -> p_28600_ instanceof AbstractSchoolingFish
         );
         this.goalSelector.addGoal(0, new Fox.FoxFloatGoal());
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(0, new ClimbOnTopOfPowderSnowGoal(this, this.level()));
         this.goalSelector.addGoal(1, new Fox.FaceplantGoal());
         this.goalSelector.addGoal(2, new Fox.FoxPanicGoal(2.2));
@@ -187,6 +_,7 @@
         this.goalSelector.addGoal(11, new Fox.FoxSearchForItemsGoal());
         this.goalSelector.addGoal(12, new Fox.FoxLookAtPlayerGoal(this, Player.class, 24.0F));
         this.goalSelector.addGoal(13, new Fox.PerchAndSearchGoal());
+        this.targetSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.targetSelector
             .addGoal(
                 3,
@@ -348,6 +_,11 @@
     }
 
     private void setTargetGoals() {
+        // Purpur start - Tulips change fox type - do not add duplicate goals
+        this.targetSelector.removeGoal(this.landTargetGoal);
+        this.targetSelector.removeGoal(this.turtleEggTargetGoal);
+        this.targetSelector.removeGoal(this.fishTargetGoal);
+        // Purpur end - Tulips change fox type
         if (this.getVariant() == Fox.Variant.RED) {
             this.targetSelector.addGoal(4, this.landTargetGoal);
             this.targetSelector.addGoal(4, this.turtleEggTargetGoal);
@@ -375,6 +_,7 @@
 
     private void setVariant(Fox.Variant p_364650_) {
         this.entityData.set(DATA_TYPE_ID, p_364650_.getId());
+        this.setTargetGoals(); // Purpur - Tulips change fox type - fix API bug not updating pathfinders on type change
     }
 
     @Nullable
@@ -432,7 +_,7 @@
         p_422367_.read("Trusted", TRUSTED_LIST_CODEC).orElse(List.of()).forEach(this::addTrustedEntity);
         this.setSleeping(p_422367_.getBooleanOr("Sleeping", false));
         this.setVariant(p_422367_.read("Type", Fox.Variant.CODEC).orElse(Fox.Variant.DEFAULT));
-        this.setSitting(p_422367_.getBooleanOr("Sitting", false));
+        this.setSitting(p_422367_.getBooleanOr("Sitting", false), false); // Paper - Add EntityToggleSitEvent
         this.setIsCrouching(p_422367_.getBooleanOr("Crouching", false));
         if (this.level() instanceof ServerLevel) {
             this.setTargetGoals();
@@ -449,7 +_,13 @@
     }
 
     public void setSitting(boolean p_28611_) {
-        this.setFlag(1, p_28611_);
+        // Paper start - Add EntityToggleSitEvent
+        this.setSitting(p_28611_, true);
+    }
+    public void setSitting(boolean sitting, boolean fireEvent) {
+        if (fireEvent && !new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), sitting).callEvent()) return;
+        // Paper end - Add EntityToggleSitEvent
+        this.setFlag(1, sitting);
     }
 
     public boolean isFaceplanted() {
@@ -508,19 +_,20 @@
             itementity.setPickUpDelay(40);
             itementity.setThrower(this);
             this.playSound(SoundEvents.FOX_SPIT, 1.0F, 1.0F);
-            this.level().addFreshEntity(itementity);
+            this.spawnAtLocation((net.minecraft.server.level.ServerLevel) this.level(), itementity); // Paper - Call EntityDropItemEvent
         }
     }
 
     private void dropItemStack(ItemStack p_28606_) {
         ItemEntity itementity = new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), p_28606_);
-        this.level().addFreshEntity(itementity);
+        this.spawnAtLocation((net.minecraft.server.level.ServerLevel) this.level(), itementity); // Paper - Call EntityDropItemEvent
     }
 
     @Override
     protected void pickUpItem(ServerLevel p_376527_, ItemEntity p_28514_) {
         ItemStack itemstack = p_28514_.getItem();
-        if (this.canHoldItem(itemstack)) {
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(this, p_28514_, itemstack.getCount() - 1, !this.canHoldItem(itemstack)).isCancelled()) { // CraftBukkit - call EntityPickupItemEvent
+            itemstack = p_28514_.getItem(); // CraftBukkit - update itemstack after event
             int i = itemstack.getCount();
             if (i > 1) {
                 this.dropItemStack(itemstack.split(i - 1));
@@ -531,7 +_,7 @@
             this.setItemSlot(EquipmentSlot.MAINHAND, itemstack.split(1));
             this.setGuaranteedDrop(EquipmentSlot.MAINHAND);
             this.take(p_28514_, itemstack.getCount());
-            p_28514_.discard();
+            p_28514_.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             this.ticksSinceEaten = 0;
         }
     }
@@ -622,12 +_,12 @@
     }
 
     @Override
-    public void setTarget(@Nullable LivingEntity p_28574_) {
-        if (this.isDefending() && p_28574_ == null) {
+    public boolean setTarget(@Nullable LivingEntity livingEntity, @Nullable org.bukkit.event.entity.EntityTargetEvent.TargetReason reason) { // CraftBukkit
+        if (this.isDefending() && livingEntity == null) {
             this.setDefending(false);
         }
 
-        super.setTarget(p_28574_);
+        return super.setTarget(livingEntity, reason); // CraftBukkit
     }
 
     void wakeUp() {
@@ -691,15 +_,55 @@
         return this.getTrustedEntities().anyMatch(p_393132_ -> p_393132_.matches(p_394151_));
     }
 
-    @Override
-    protected void dropAllDeathLoot(ServerLevel p_348640_, DamageSource p_28536_) {
-        ItemStack itemstack = this.getItemBySlot(EquipmentSlot.MAINHAND);
-        if (!itemstack.isEmpty()) {
-            this.spawnAtLocation(p_348640_, itemstack);
+    // Paper start - handle the bitten item separately like vanilla
+    @Override
+    protected boolean shouldSkipLoot(EquipmentSlot slot) {
+        return slot == EquipmentSlot.MAINHAND;
+    }
+    // Paper end
+
+    // Purpur start - Tulips change fox type
+    @Override
+    public net.minecraft.world.InteractionResult mobInteract(Player player, net.minecraft.world.InteractionHand hand) {
+        if (level().purpurConfig.foxTypeChangesWithTulips) {
+            ItemStack itemstack = player.getItemInHand(hand);
+            if (getVariant() == Variant.RED && itemstack.getItem() == Items.WHITE_TULIP) {
+                setVariant(Variant.SNOW);
+                if (!player.getAbilities().instabuild) {
+                    itemstack.shrink(1);
+                }
+                return net.minecraft.world.InteractionResult.SUCCESS;
+            } else if (getVariant() == Variant.SNOW && itemstack.getItem() == Items.ORANGE_TULIP) {
+                setVariant(Variant.RED);
+                if (!player.getAbilities().instabuild) {
+                    itemstack.shrink(1);
+                }
+                return net.minecraft.world.InteractionResult.SUCCESS;
+            }
+        }
+        return super.mobInteract(player, hand);
+    }
+    // Purpur end - Tulips change fox type
+
+    @Override
+    // Paper start - Cancellable death event
+    protected org.bukkit.event.entity.EntityDeathEvent dropAllDeathLoot(ServerLevel level, DamageSource damageSource) {
+        ItemStack itemBySlot = this.getItemBySlot(EquipmentSlot.MAINHAND);
+        boolean releaseMouth = false;
+        if (!itemBySlot.isEmpty() && level.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) { // Fix MC-153010
+            this.spawnAtLocation(level, itemBySlot);
+            releaseMouth = true;
+        }
+
+        org.bukkit.event.entity.EntityDeathEvent deathEvent = super.dropAllDeathLoot(level, damageSource);
+        // Below is code to drop
+        if (deathEvent == null || deathEvent.isCancelled()) return deathEvent;
+
+        if (releaseMouth) {
+            // Paper end
             this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
         }
-
-        super.dropAllDeathLoot(p_348640_, p_28536_);
+        return deathEvent; // Paper
     }
 
     public static boolean isPathClear(Fox p_28472_, LivingEntity p_28473_) {
@@ -856,39 +_,63 @@
 
         @Override
         protected void breed() {
-            ServerLevel serverlevel = this.level;
-            Fox fox = (Fox)this.animal.getBreedOffspring(serverlevel, this.partner);
-            if (fox != null) {
-                ServerPlayer serverplayer = this.animal.getLoveCause();
-                ServerPlayer serverplayer1 = this.partner.getLoveCause();
-                ServerPlayer serverplayer2 = serverplayer;
-                if (serverplayer != null) {
-                    fox.addTrustedEntity(serverplayer);
-                } else {
-                    serverplayer2 = serverplayer1;
-                }
-
-                if (serverplayer1 != null && serverplayer != serverplayer1) {
-                    fox.addTrustedEntity(serverplayer1);
-                }
-
-                if (serverplayer2 != null) {
-                    serverplayer2.awardStat(Stats.ANIMALS_BRED);
-                    CriteriaTriggers.BRED_ANIMALS.trigger(serverplayer2, this.animal, this.partner, fox);
-                }
-
+            ServerLevel serverLevel = this.level;
+            Fox fox = (Fox)this.animal.getBreedOffspring(serverLevel, this.partner);
+            final net.neoforged.neoforge.event.entity.living.BabyEntitySpawnEvent event = new net.neoforged.neoforge.event.entity.living.BabyEntitySpawnEvent(animal, partner, fox);
+            final boolean cancelled = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event).isCanceled();
+            fox = (Fox) event.getChild();
+            if (cancelled) {
+                //Reset the "inLove" state for the animals
                 this.animal.setAge(6000);
                 this.partner.setAge(6000);
                 this.animal.resetLove();
                 this.partner.resetLove();
+                return;
+            }
+            if (fox != null) {
+                ServerPlayer loveCause = this.animal.getLoveCause();
+                ServerPlayer loveCause1 = this.partner.getLoveCause();
+                ServerPlayer serverPlayer = loveCause;
+                if (loveCause != null) {
+                    fox.addTrustedEntity(loveCause);
+                } else {
+                    serverPlayer = loveCause1;
+                }
+
+                if (loveCause1 != null && loveCause != loveCause1) {
+                    fox.addTrustedEntity(loveCause1);
+                }
+
+                // CraftBukkit start - call EntityBreedEvent
                 fox.setAge(-24000);
                 fox.snapTo(this.animal.getX(), this.animal.getY(), this.animal.getZ(), 0.0F, 0.0F);
-                serverlevel.addFreshEntityWithPassengers(fox);
+                int experience = this.animal.getRandom().nextInt(7) + 1;
+                org.bukkit.event.entity.EntityBreedEvent entityBreedEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreedEvent(fox, this.animal, this.partner, loveCause, this.animal.breedItem, experience);
+                if (entityBreedEvent.isCancelled()) {
+                    this.animal.resetLove();
+                    this.partner.resetLove();
+                    return;
+                }
+                experience = entityBreedEvent.getExperience();
+                // CraftBukkit end - call EntityBreedEvent
+
+                if (serverPlayer != null) {
+                    serverPlayer.awardStat(Stats.ANIMALS_BRED);
+                    CriteriaTriggers.BRED_ANIMALS.trigger(serverPlayer, this.animal, this.partner, fox);
+                }
+
+                // Purpur start - Make entity breeding times configurable
+                this.animal.setAge(this.animal.getPurpurBreedTime());
+                this.partner.setAge(this.partner.getPurpurBreedTime());
+                // Purpur end - Make entity breeding times configurable
+                this.animal.resetLove();
+                this.partner.resetLove();
+                serverLevel.addFreshEntityWithPassengers(fox, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
                 this.level.broadcastEntityEvent(this.animal, (byte)18);
-                if (serverlevel.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
+                if (experience > 0 && serverLevel.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) { // Paper - call EntityBreedEvent
                     this.level
                         .addFreshEntity(
-                            new ExperienceOrb(this.level, this.animal.getX(), this.animal.getY(), this.animal.getZ(), this.animal.getRandom().nextInt(7) + 1)
+                            new ExperienceOrb(this.level, this.animal.position(), net.minecraft.world.phys.Vec3.ZERO, experience, org.bukkit.entity.ExperienceOrb.SpawnReason.BREED, loveCause, fox) // Paper - call EntityBreedEvent, add spawn context
                         );
                 }
             }
@@ -935,7 +_,8 @@
         }
 
         protected void onReachedTarget() {
-            if (getServerLevel(Fox.this.level()).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            Boolean mobGriefingOverride = getServerLevel(Fox.this.level()).purpurConfig.foxMobGriefingOverride; // Purpur - Add mobGriefing override to everything affected
+            if (mobGriefingOverride != null ? mobGriefingOverride : net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(Fox.this.level()), Fox.this)) {
                 BlockState blockstate = Fox.this.level().getBlockState(this.blockPos);
                 if (blockstate.is(Blocks.SWEET_BERRY_BUSH)) {
                     this.pickSweetBerries(blockstate);
@@ -952,6 +_,7 @@
         private void pickSweetBerries(BlockState p_148929_) {
             int i = p_148929_.getValue(SweetBerryBushBlock.AGE);
             p_148929_.setValue(SweetBerryBushBlock.AGE, 1);
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(Fox.this, this.blockPos, p_148929_.setValue(SweetBerryBushBlock.AGE, 1))) return; // CraftBukkit - call EntityChangeBlockEvent
             int j = 1 + Fox.this.level().random.nextInt(2) + (i == 3 ? 1 : 0);
             ItemStack itemstack = Fox.this.getItemBySlot(EquipmentSlot.MAINHAND);
             if (itemstack.isEmpty()) {
@@ -994,7 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25 || Fox.this.isInLava();
+            return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25D || Fox.this.isInLava() || Fox.this.isInFluidType((fluidType, height) -> Fox.this.canSwimInFluidType(fluidType) && height > 0.25D);
         }
     }
 
@@ -1048,15 +_,15 @@
         }
     }
 
-    public class FoxLookControl extends LookControl {
+    public class FoxLookControl extends org.purpurmc.purpur.controller.LookControllerWASD { // Purpur - Ridables
         public FoxLookControl() {
             super(Fox.this);
         }
 
         @Override
-        public void tick() {
+        public void vanillaTick() { // Purpur - Ridables
             if (!Fox.this.isSleeping()) {
-                super.tick();
+                super.vanillaTick(); // Purpur - Ridables
             }
         }
 
@@ -1092,15 +_,15 @@
         }
     }
 
-    class FoxMoveControl extends MoveControl {
+    class FoxMoveControl extends org.purpurmc.purpur.controller.MoveControllerWASD { // Purpur - Ridables
         public FoxMoveControl() {
             super(Fox.this);
         }
 
         @Override
-        public void tick() {
+        public void vanillaTick() { // Purpur - Ridables
             if (Fox.this.canMove()) {
-                super.tick();
+                super.vanillaTick(); // Purpur - Ridables
             }
         }
     }
