--- a/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
+++ b/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
@@ -24,7 +_,7 @@
             this.saveWithoutId(tagvalueoutput);
             tagvalueoutput.putString("id", this.getEncodeId());
             if (p_29896_.setEntityOnShoulder(tagvalueoutput.buildResult())) {
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
                 return true;
             }
         }
