--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -56,6 +_,80 @@
     public final Player player;
     private final EntityEquipment equipment;
     private int timesChanged;
+    // Paper start - add fields and methods
+    public static final EquipmentSlot[] EQUIPMENT_SLOTS_SORTED_BY_INDEX = EQUIPMENT_SLOT_MAPPING.int2ObjectEntrySet()
+        .stream()
+        .sorted(java.util.Comparator.comparingInt(it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry::getIntKey))
+        .map(java.util.Map.Entry::getValue).toArray(EquipmentSlot[]::new);
+    public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+    private int maxStack = MAX_STACK;
+
+    @Override
+    public java.util.List<ItemStack> getContents() {
+        java.util.List<ItemStack> combined = new java.util.ArrayList<>(this.items.size() + EQUIPMENT_SLOT_MAPPING.size());
+        combined.addAll(this.items);
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
+            ItemStack itemStack = this.equipment.get(equipmentSlot);
+            combined.add(itemStack); // Include empty items
+        };
+        return combined;
+    }
+
+    public java.util.List<ItemStack> getArmorContents() {
+        java.util.List<ItemStack> items = new java.util.ArrayList<>(4);
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
+            if (equipmentSlot.getType() == EquipmentSlot.Type.HUMANOID_ARMOR) {
+                items.add(this.equipment.get(equipmentSlot));
+            }
+        }
+        return items;
+    }
+
+    public java.util.List<ItemStack> getExtraContent() {
+        java.util.List<ItemStack> items = new java.util.ArrayList<>();
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
+            if (equipmentSlot.getType() != EquipmentSlot.Type.HUMANOID_ARMOR) { // Non humanoid armor is considered extra
+                items.add(this.equipment.get(equipmentSlot));
+            }
+        }
+        return items;
+    }
+
+    @Override
+    public void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        this.transaction.add(player);
+    }
+
+    @Override
+    public void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        this.transaction.remove(player);
+    }
+
+    @Override
+    public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return this.transaction;
+    }
+
+    @Override
+    public org.bukkit.inventory.InventoryHolder getOwner() {
+        return this.player.getBukkitEntity();
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return this.maxStack;
+    }
+
+    @Override
+    public void setMaxStackSize(int size) {
+        this.maxStack = size;
+    }
+
+    @Override
+    public org.bukkit.Location getLocation() {
+        return this.player.getBukkitEntity().getLocation();
+    }
+    // Paper end - add fields and methods
 
     public Inventory(Player p_35983_, EntityEquipment p_401828_) {
         this.player = p_35983_;
@@ -92,10 +_,39 @@
 
     private boolean hasRemainingSpaceForItem(ItemStack p_36015_, ItemStack p_36016_) {
         return !p_36015_.isEmpty()
-            && ItemStack.isSameItemSameComponents(p_36015_, p_36016_)
             && p_36015_.isStackable()
-            && p_36015_.getCount() < this.getMaxStackSize(p_36015_);
-    }
+            && p_36015_.getCount() < this.getMaxStackSize(p_36015_)
+            && ItemStack.isSameItemSameComponents(p_36015_, p_36016_); // Paper - check if itemstack is stackable first
+    }
+
+    // CraftBukkit start - Watch method above! :D
+    public int canHold(ItemStack itemStack) {
+        int remains = itemStack.getCount();
+        for (int slot = 0; slot < this.items.size(); ++slot) {
+            ItemStack itemInSlot = this.getItem(slot);
+            if (itemInSlot.isEmpty()) {
+                return itemStack.getCount();
+            }
+
+            if (this.hasRemainingSpaceForItem(itemInSlot, itemStack)) {
+                remains -= (itemInSlot.getMaxStackSize() < this.getMaxStackSize() ? itemInSlot.getMaxStackSize() : this.getMaxStackSize()) - itemInSlot.getCount();
+            }
+            if (remains <= 0) {
+                return itemStack.getCount();
+            }
+        }
+
+        ItemStack itemInOffhand = this.equipment.get(EquipmentSlot.OFFHAND);
+        if (this.hasRemainingSpaceForItem(itemInOffhand, itemStack)) {
+            remains -= (itemInOffhand.getMaxStackSize() < this.getMaxStackSize() ? itemInOffhand.getMaxStackSize() : this.getMaxStackSize()) - itemInOffhand.getCount();
+        }
+        if (remains <= 0) {
+            return itemStack.getCount();
+        }
+
+        return itemStack.getCount() - remains;
+    }
+    // CraftBukkit end
 
     public int getFreeSlot() {
         for (int i = 0; i < this.items.size(); i++) {
@@ -107,23 +_,27 @@
         return -1;
     }
 
-    public void addAndPickItem(ItemStack p_383009_) {
-        this.setSelectedSlot(this.getSuitableHotbarSlot());
+    // Paper start - Add PlayerPickItemEvent
+    public void addAndPickItem(ItemStack stack, final int targetSlot) {
+        this.setSelectedSlot(targetSlot);
+        // Paper end - Add PlayerPickItemEvent
         if (!this.items.get(this.selected).isEmpty()) {
-            int i = this.getFreeSlot();
-            if (i != -1) {
-                this.items.set(i, this.items.get(this.selected));
+            int freeSlot = this.getFreeSlot();
+            if (freeSlot != -1) {
+                this.items.set(freeSlot, this.items.get(this.selected));
             }
         }
 
-        this.items.set(this.selected, p_383009_);
+        this.items.set(this.selected, stack);
     }
 
-    public void pickSlot(int p_36039_) {
-        this.setSelectedSlot(this.getSuitableHotbarSlot());
-        ItemStack itemstack = this.items.get(this.selected);
-        this.items.set(this.selected, this.items.get(p_36039_));
-        this.items.set(p_36039_, itemstack);
+    // Paper start - Add PlayerPickItemEvent
+    public void pickSlot(int index, final int targetSlot) {
+        this.setSelectedSlot(targetSlot);
+    // Paper end - Add PlayerPickItemEvent
+        ItemStack itemStack = this.items.get(this.selected);
+        this.items.set(this.selected, this.items.get(index));
+        this.items.set(index, itemStack);
     }
 
     public static boolean isHotbarSlot(int p_36046_) {
@@ -144,13 +_,13 @@
         return !p_365004_.isDamaged() && !p_365004_.isEnchanted() && !p_365004_.has(DataComponents.CUSTOM_NAME);
     }
 
-    public int findSlotMatchingCraftingIngredient(Holder<Item> p_363103_, ItemStack p_390604_) {
+    public int findSlotMatchingCraftingIngredient(io.papermc.paper.inventory.recipe.ItemOrExact item, ItemStack stack) { // Paper - Improve exact choice recipe ingredients
         for (int i = 0; i < this.items.size(); i++) {
-            ItemStack itemstack = this.items.get(i);
-            if (!itemstack.isEmpty()
-                && itemstack.is(p_363103_)
-                && isUsableForCrafting(itemstack)
-                && (p_390604_.isEmpty() || ItemStack.isSameItemSameComponents(p_390604_, itemstack))) {
+            ItemStack itemStack = this.items.get(i);
+            if (!itemStack.isEmpty()
+                && item.matches(itemStack) // Paper - Improve exact choice recipe ingredients
+                && (!(item instanceof io.papermc.paper.inventory.recipe.ItemOrExact.Item) || Inventory.isUsableForCrafting(itemStack)) // Paper - Improve exact choice recipe ingredients
+                && (stack.isEmpty() || ItemStack.isSameItemSameComponents(stack, itemStack))) {
                 return i;
             }
         }
@@ -168,7 +_,7 @@
 
         for (int k = 0; k < 9; k++) {
             int l = (this.selected + k) % 9;
-            if (!this.items.get(l).isEnchanted()) {
+            if (!this.items.get(l).isNotReplaceableByPickAction(this.player, l)) {
                 return l;
             }
         }
@@ -289,6 +_,8 @@
             } catch (Throwable throwable) {
                 CrashReport crashreport = CrashReport.forThrowable(throwable, "Adding item to inventory");
                 CrashReportCategory crashreportcategory = crashreport.addCategory("Item being added");
+                crashreportcategory.setDetail("Registry Name", () -> String.valueOf(net.minecraft.core.registries.BuiltInRegistries.ITEM.getKey(p_36042_.getItem())));
+                crashreportcategory.setDetail("Item Class", () -> p_36042_.getItem().getClass().getName());
                 crashreportcategory.setDetail("Item ID", Item.getId(p_36042_.getItem()));
                 crashreportcategory.setDetail("Item data", p_36042_.getDamageValue());
                 crashreportcategory.setDetail("Item name", () -> p_36042_.getHoverName().getString());
