--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1085,6 +_,7 @@
     private final boolean canSpawnFarFromPlayer;
     private final int clientTrackingRange;
     private final int updateInterval;
+    public int ttl = -1; // Pufferfish - entity TTL
     private final String descriptionId;
     @Nullable
     private Component description;
@@ -1093,6 +_,11 @@
     private final float spawnDimensionsScale;
     private final FeatureFlagSet requiredFeatures;
 
+    private final java.util.function.Predicate<EntityType<?>> trackDeltasSupplier;
+    private final java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier;
+    private final java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier;
+    private final boolean onlyOpCanSetNbt;
+
     private static <T extends Entity> EntityType<T> register(ResourceKey<EntityType<?>> p_368669_, EntityType.Builder<T> p_368714_) {
         return Registry.register(BuiltInRegistries.ENTITY_TYPE, p_368669_, p_368714_.build(p_368669_));
     }
@@ -1105,6 +_,16 @@
         return register(vanillaEntityId(p_20635_), p_20636_);
     }
 
+    // Purpur start - PlayerSetSpawnerTypeWithEggEvent
+    public static EntityType<?> getFromBukkitType(org.bukkit.entity.EntityType bukkitType) {
+        return getFromKey(ResourceLocation.parse(bukkitType.getKey().toString()));
+    }
+
+    public static EntityType<?> getFromKey(ResourceLocation location) {
+        return BuiltInRegistries.ENTITY_TYPE.getValue(location);
+    }
+    // Purpur end - PlayerSetSpawnerTypeWithEggEvent
+
     public static ResourceLocation getKey(EntityType<?> p_20614_) {
         return BuiltInRegistries.ENTITY_TYPE.getKey(p_20614_);
     }
@@ -1129,6 +_,29 @@
         Optional<ResourceKey<LootTable>> p_368696_,
         FeatureFlagSet p_273518_
     ) {
+        this(p_273268_, p_272918_, p_273417_, p_273389_, p_273556_, p_272654_, p_273631_, p_272946_, p_338404_, p_272895_, p_273451_, p_368582_, p_368696_, p_273518_, EntityType::defaultTrackDeltasSupplier, EntityType::defaultTrackingRangeSupplier, EntityType::defaultUpdateIntervalSupplier, false);
+    }
+
+    public EntityType(
+        EntityType.EntityFactory<T> p_273268_,
+        MobCategory p_272918_,
+        boolean p_273417_,
+        boolean p_273389_,
+        boolean p_273556_,
+        boolean p_272654_,
+        ImmutableSet<Block> p_273631_,
+        EntityDimensions p_272946_,
+        float p_338404_,
+        int p_272895_,
+        int p_273451_,
+        String p_368582_,
+        Optional<ResourceKey<LootTable>> p_368696_,
+        FeatureFlagSet p_273518_,
+        final java.util.function.Predicate<EntityType<?>> trackDeltasSupplier,
+        final java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier,
+        final java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier,
+        boolean onlyOpCanSetNbt
+    ) {
         this.factory = p_273268_;
         this.category = p_272918_;
         this.canSpawnFarFromPlayer = p_272654_;
@@ -1143,26 +_,46 @@
         this.descriptionId = p_368582_;
         this.lootTable = p_368696_;
         this.requiredFeatures = p_273518_;
-    }
-
-    @Nullable
-    public T spawn(
-        ServerLevel p_20593_,
-        @Nullable ItemStack p_20594_,
-        @Nullable LivingEntity p_393729_,
-        BlockPos p_20596_,
-        EntitySpawnReason p_361023_,
-        boolean p_20598_,
-        boolean p_20599_
-    ) {
+        this.trackDeltasSupplier = trackDeltasSupplier;
+        this.trackingRangeSupplier = trackingRangeSupplier;
+        this.updateIntervalSupplier = updateIntervalSupplier;
+        this.onlyOpCanSetNbt = onlyOpCanSetNbt;
+    }
+
+    @Nullable
+    public T spawn(
+        ServerLevel level,
+        @Nullable ItemStack spawnedFrom,
+        @Nullable LivingEntity owner,
+        BlockPos pos,
+        EntitySpawnReason reason,
+        boolean shouldOffsetY,
+        boolean shouldOffsetYMore
+    ) {
+        // CraftBukkit start
+        return this.spawn(level, spawnedFrom, owner, pos, reason, shouldOffsetY, shouldOffsetYMore, reason == EntitySpawnReason.DISPENSER ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG); // Paper - use correct spawn reason for dispenser spawn eggs
+    }
+
+    @Nullable
+    public T spawn(
+        ServerLevel level,
+        @Nullable ItemStack spawnedFrom,
+        @Nullable LivingEntity owner,
+        BlockPos pos,
+        EntitySpawnReason reason,
+        boolean shouldOffsetY,
+        boolean shouldOffsetYMore,
+        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason createSpawnReason
+    ) {
+        // CraftBukkit end
         Consumer<T> consumer;
-        if (p_20594_ != null) {
-            consumer = createDefaultStackConfig(p_20593_, p_20594_, p_393729_);
+        if (spawnedFrom != null) {
+            consumer = createDefaultStackConfig(level, spawnedFrom, owner);
         } else {
             consumer = p_263563_ -> {};
         }
 
-        return this.spawn(p_20593_, consumer, p_20596_, p_361023_, p_20598_, p_20599_);
+        return this.spawn(level, consumer, pos, reason, shouldOffsetY, shouldOffsetYMore, createSpawnReason); // CraftBukkit
     }
 
     public static <T extends Entity> Consumer<T> createDefaultStackConfig(Level p_366618_, ItemStack p_263568_, @Nullable LivingEntity p_393794_) {
@@ -1183,27 +_,62 @@
         Consumer<T> p_263579_, Level p_366735_, ItemStack p_263582_, @Nullable LivingEntity p_394266_
     ) {
         CustomData customdata = p_263582_.getOrDefault(DataComponents.ENTITY_DATA, CustomData.EMPTY);
-        return !customdata.isEmpty() ? p_263579_.andThen(p_392975_ -> updateCustomEntityTag(p_366735_, p_394266_, p_392975_, customdata)) : p_263579_;
+        // CraftBukkit start - SPIGOT-5665
+        return !customdata.isEmpty() ? p_263579_.andThen(entity -> {
+            try {
+                updateCustomEntityTag(p_366735_, p_394266_, entity, customdata);
+            } catch (Throwable t) {
+                EntityType.LOGGER.warn("Error loading spawn egg NBT", t);
+            }
+        }) : p_263579_;
+        // CraftBukkit end
     }
 
     @Nullable
     public T spawn(ServerLevel p_262704_, BlockPos p_262672_, EntitySpawnReason p_363697_) {
-        return this.spawn(p_262704_, null, p_262672_, p_363697_, false, false);
-    }
-
-    @Nullable
-    public T spawn(
-        ServerLevel p_262634_, @Nullable Consumer<T> p_365118_, BlockPos p_262707_, EntitySpawnReason p_362115_, boolean p_361504_, boolean p_362006_
-    ) {
-        T t = this.create(p_262634_, p_365118_, p_262707_, p_362115_, p_361504_, p_362006_);
-        if (t != null) {
-            p_262634_.addFreshEntityWithPassengers(t);
-            if (t instanceof Mob mob) {
+        // CraftBukkit start
+        return this.spawn(p_262704_, p_262672_, p_363697_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    @Nullable
+    public T spawn(ServerLevel level, BlockPos pos, EntitySpawnReason reason, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason creatureSpawnReason) {
+        return this.spawn(level, null, pos, reason, false, false, creatureSpawnReason);
+        // CraftBukkit End
+    }
+
+    @Nullable
+    public T spawn(ServerLevel level, @Nullable Consumer<T> consumer, BlockPos pos, EntitySpawnReason reason, boolean shouldOffsetY, boolean shouldOffsetYMore) {
+        // CraftBukkit start
+        return this.spawn(level, consumer, pos, reason, shouldOffsetY, shouldOffsetYMore, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    @Nullable
+    public T spawn(ServerLevel level, @Nullable Consumer<T> consumer, BlockPos pos, EntitySpawnReason reason, boolean shouldOffsetY, boolean shouldOffsetYMore, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason creatureSpawnReason) {
+        // CraftBukkit end
+        // Paper start - PreCreatureSpawnEvent
+        com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
+            org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
+            org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(this),
+            creatureSpawnReason
+        );
+        if (!event.callEvent()) {
+            return null;
+        }
+        // Paper end - PreCreatureSpawnEvent
+        T entity = this.create(level, consumer, pos, reason, shouldOffsetY, shouldOffsetYMore);
+        if (entity != null) {
+            // CraftBukkit start
+            level.addFreshEntityWithPassengers(entity, creatureSpawnReason);
+            if (entity.isRemoved()) {
+                return null; // Don't return an entity when CreatureSpawnEvent is canceled
+            }
+            // CraftBukkit end
+            if (entity instanceof Mob mob) {
                 mob.playAmbientSound();
             }
         }
 
-        return t;
+        return entity;
     }
 
     @Nullable
@@ -1227,6 +_,13 @@
                 mob.yHeadRot = mob.getYRot();
                 mob.yBodyRot = mob.getYRot();
                 mob.finalizeSpawn(p_262637_, p_262637_.getCurrentDifficultyAt(mob.blockPosition()), p_360546_, null);
+                if (mob.isSpawnCancelled()) {
+                    // Neo: Discard mob, spawn was cancelled
+                    mob.discard();
+                    // return null, mob was killed, context should be lost
+                    // fixes llamas for wandering trader spawning if wandering trader was cancelled
+                    return null;
+                }
             }
 
             if (p_262629_ != null) {
@@ -1255,6 +_,15 @@
                 if (p_20621_.isClientSide
                     || !p_20623_.getType().onlyOpCanSetNbt()
                     || p_394374_ instanceof Player player && minecraftserver.getPlayerList().isOp(player.getGameProfile())) {
+                    // Paper start - filter out protected tags
+                    if (p_394374_ == null || !p_394374_.getBukkitEntity().hasPermission("minecraft.nbt.place")) {
+                        p_331752_ = p_331752_.update((compound) -> {
+                            for (net.minecraft.commands.arguments.NbtPathArgument.NbtPath tag : p_20621_.paperConfig().entities.spawning.filteredEntityTagNbtPaths) {
+                                tag.remove(compound);
+                            }
+                        });
+                    }
+                    // Paper end - filter out protected tags
                     p_331752_.loadInto(p_20623_);
                 }
             }
@@ -1281,6 +_,16 @@
         return this.category;
     }
 
+    // Purpur start - PlayerSetSpawnerTypeWithEggEvent
+    public String getName() {
+        return BuiltInRegistries.ENTITY_TYPE.getKey(this).getPath();
+    }
+
+    public String getTranslatedName() {
+        return getDescription().getString();
+    }
+    // Purpur end - PlayerSetSpawnerTypeWithEggEvent
+
     public String getDescriptionId() {
         return this.descriptionId;
     }
@@ -1326,10 +_,25 @@
     }
 
     public static Optional<Entity> create(ValueInput p_422014_, Level p_20644_, EntitySpawnReason p_360780_) {
+        // Paper start - Don't fire sync event during generation
+        return create(p_422014_, p_20644_, p_360780_, false);
+    }
+
+    public static Optional<Entity> create(ValueInput input, Level level, EntitySpawnReason spawnReason, boolean generation) {
+        // Paper end - Don't fire sync event during generation
         return Util.ifElse(
-            by(p_422014_).map(p_359692_ -> p_359692_.create(p_20644_, p_360780_)),
-            p_421348_ -> p_421348_.load(p_422014_),
-            () -> LOGGER.warn("Skipping Entity with id {}", p_422014_.getStringOr("id", "[invalid]"))
+            by(input).map(entityType -> entityType.create(level, spawnReason)),
+            // Paper start - Don't fire sync event during generation
+            entity -> {
+                if (generation) entity.generation = true; // Paper - Don't fire sync event during generation
+                entity.load(input);
+            },
+            // Paper end - Don't fire sync event during generation
+            // Purpur start - log skipped entity's position
+            () -> {LOGGER.warn("Skipping Entity with id {}", input.getStringOr("id", "[invalid]"));
+                EntityType.LOGGER.warn("Location: {} {}", level.getWorld().getName(), input.read("Pos", net.minecraft.world.phys.Vec3.CODEC).orElse(net.minecraft.world.phys.Vec3.ZERO));
+            }
+            // Purpur end - log skipped entity's position
         );
     }
 
@@ -1383,10 +_,21 @@
         }).orElse(null);
     }
 
-    public static Stream<Entity> loadEntitiesRecursive(ValueInput.ValueInputList p_422664_, Level p_147047_, EntitySpawnReason p_360400_) {
-        return p_422664_.stream().mapMulti((p_421351_, p_421352_) -> loadEntityRecursive(p_421351_, p_147047_, p_360400_, p_404267_ -> {
-            p_421352_.accept(p_404267_);
-            return p_404267_;
+    public static Stream<Entity> loadEntitiesRecursive(ValueInput.ValueInputList input, Level level, EntitySpawnReason spawnReason) {
+        final java.util.Map<EntityType<?>, Integer> loadedEntityCounts = new java.util.HashMap<>(); // Paper - Entity load/save limit per chunk
+        return input.stream().mapMulti((valueInput, consumer) -> loadEntityRecursive(valueInput, level, spawnReason, entity -> {
+            // Paper start - Entity load/save limit per chunk
+            final EntityType<?> entityType = entity.getType();
+            final int saveLimit = level.paperConfig().chunks.entityPerChunkSaveLimit.getOrDefault(entityType, -1);
+            if (saveLimit > -1) {
+                if (loadedEntityCounts.getOrDefault(entityType, 0) >= saveLimit) {
+                    return null;
+                }
+                loadedEntityCounts.merge(entityType, 1, Integer::sum);
+            }
+            // Paper end - Entity load/save limit per chunk
+            consumer.accept(entity);
+            return entity;
         }));
     }
 
@@ -1400,14 +_,23 @@
     }
 
     public int clientTrackingRange() {
+        return trackingRangeSupplier.applyAsInt(this);
+    }
+    private int defaultTrackingRangeSupplier() {
         return this.clientTrackingRange;
     }
 
     public int updateInterval() {
+        return updateIntervalSupplier.applyAsInt(this);
+    }
+    private int defaultUpdateIntervalSupplier() {
         return this.updateInterval;
     }
 
     public boolean trackDeltas() {
+        return this.trackDeltasSupplier.test(this);
+    }
+    private boolean defaultTrackDeltasSupplier() {
         return this != PLAYER
             && this != LLAMA_SPIT
             && this != WITHER
@@ -1460,9 +_,12 @@
     }
 
     public boolean onlyOpCanSetNbt() {
+        if (onlyOpCanSetNbt) return true;
         return OP_ONLY_CUSTOM_DATA.contains(this);
     }
 
+    public Stream<TagKey<EntityType<?>>> getTags() {return this.builtInRegistryHolder().tags();}
+
     public static class Builder<T extends Entity> {
         private final EntityType.EntityFactory<T> factory;
         private final MobCategory category;
@@ -1481,6 +_,10 @@
             ResourceKey.create(Registries.LOOT_TABLE, p_367917_.location().withPrefix("entities/"))
         );
         private final DependantName<EntityType<?>, String> descriptionId = p_367918_ -> Util.makeDescriptionId("entity", p_367918_.location());
+        private java.util.function.Predicate<EntityType<?>> velocityUpdateSupplier = EntityType::defaultTrackDeltasSupplier;
+        private java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier = EntityType::defaultTrackingRangeSupplier;
+        private java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier = EntityType::defaultUpdateIntervalSupplier;
+        private boolean onlyOpCanSetNbt = false;
 
         private Builder(EntityType.EntityFactory<T> p_20696_, MobCategory p_20697_) {
             this.factory = p_20696_;
@@ -1594,6 +_,26 @@
             return this;
         }
 
+        public EntityType.Builder<T> setUpdateInterval(int interval) {
+            this.updateIntervalSupplier = t->interval;
+            return this;
+        }
+
+        public EntityType.Builder<T> setTrackingRange(int range) {
+            this.trackingRangeSupplier = t->range;
+            return this;
+        }
+
+        public EntityType.Builder<T> setShouldReceiveVelocityUpdates(boolean value) {
+            this.velocityUpdateSupplier = t->value;
+            return this;
+        }
+
+        public EntityType.Builder<T> setOnlyOpCanSetNbt(boolean onlyOpCanSetNbt) {
+            this.onlyOpCanSetNbt = onlyOpCanSetNbt;
+            return this;
+        }
+
         public EntityType<T> build(ResourceKey<EntityType<?>> p_368626_) {
             if (this.serialize) {
                 Util.fetchChoiceType(References.ENTITY_TREE, p_368626_.location().toString());
@@ -1613,7 +_,11 @@
                 this.updateInterval,
                 this.descriptionId.get(p_368626_),
                 this.lootTable.get(p_368626_),
-                this.requiredFeatures
+                this.requiredFeatures,
+                this.velocityUpdateSupplier,
+                this.trackingRangeSupplier,
+                this.updateIntervalSupplier,
+                this.onlyOpCanSetNbt
             );
         }
     }
