--- a/net/minecraft/world/entity/GlowSquid.java
+++ b/net/minecraft/world/entity/GlowSquid.java
@@ -65,13 +_,13 @@
     }
 
     @Override
-    protected void addAdditionalSaveData(ValueOutput p_422305_) {
+    protected void addAdditionalSaveData(ValueOutput output) {
         super.addAdditionalSaveData(p_422305_);
         p_422305_.putInt("DarkTicksRemaining", this.getDarkTicksRemaining());
     }
 
     @Override
-    protected void readAdditionalSaveData(ValueInput p_422093_) {
+    protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(p_422093_);
         this.setDarkTicks(p_422093_.getIntOr("DarkTicksRemaining", 0));
     }
@@ -79,7 +_,7 @@
     @Override
     public void aiStep() {
         super.aiStep();
-        int i = this.getDarkTicksRemaining();
+        int darkTicksRemaining = this.getDarkTicksRemaining();
         if (i > 0) {
             this.setDarkTicks(i - 1);
         }
@@ -88,7 +_,7 @@
     }
 
     @Override
-    public boolean hurtServer(ServerLevel p_376490_, DamageSource p_376927_, float p_376723_) {
+    public boolean hurtServer(ServerLevel level, DamageSource damageSource, float p_376723_) {
         boolean flag = super.hurtServer(p_376490_, p_376927_, p_376723_);
         if (flag) {
             this.setDarkTicks(100);
@@ -97,8 +_,8 @@
         return flag;
     }
 
-    public void setDarkTicks(int p_147120_) {
-        this.entityData.set(DATA_DARK_TICKS_REMAINING, p_147120_);
+    public void setDarkTicks(int darkTicks) {
+        this.entityData.set(DATA_DARK_TICKS_REMAINING, darkTicks);
     }
 
     public int getDarkTicksRemaining() {
@@ -106,7 +_,7 @@
     }
 
     public static boolean checkGlowSquidSpawnRules(
-        EntityType<? extends LivingEntity> p_217018_, ServerLevelAccessor p_217019_, EntitySpawnReason p_365182_, BlockPos p_217021_, RandomSource p_217022_
+        EntityType<? extends LivingEntity> entityType, ServerLevelAccessor p_217019_, EntitySpawnReason p_365182_, BlockPos p_217021_, RandomSource p_217022_
     ) {
         return p_217021_.getY() <= p_217019_.getSeaLevel() - 33
             && p_217019_.getRawBrightness(p_217021_, 0) == 0
