--- a/net/minecraft/world/entity/InsideBlockEffectType.java
+++ b/net/minecraft/world/entity/InsideBlockEffectType.java
@@ -4,23 +_,36 @@
 import net.minecraft.world.level.block.BaseFireBlock;
 
 public enum InsideBlockEffectType {
-    FREEZE(p_405140_ -> {
-        p_405140_.setIsInPowderSnow(true);
-        if (p_405140_.canFreeze()) {
-            p_405140_.setTicksFrozen(Math.min(p_405140_.getTicksRequiredToFreeze(), p_405140_.getTicksFrozen() + 1));
+    FREEZE(entity -> {
+        entity.setIsInPowderSnow(true);
+        if (entity.canFreeze() && !entity.freezeLocked) { // Paper - Freeze Tick Lock API
+            entity.setTicksFrozen(Math.min(entity.getTicksRequiredToFreeze(), entity.getTicksFrozen() + 1));
         }
     }),
     FIRE_IGNITE(BaseFireBlock::fireIgnite),
-    LAVA_IGNITE(Entity::lavaIgnite),
+    LAVA_IGNITE((entity, pos) -> entity.lavaIgnite(pos)), // Paper - track lava contact
     EXTINGUISH(Entity::clearFire);
 
-    private final Consumer<Entity> effect;
-
-    private InsideBlockEffectType(Consumer<Entity> p_405275_) {
-        this.effect = p_405275_;
-    }
-
-    public Consumer<Entity> effect() {
+    private final Applier effect; // Paper - track position inside effect was triggered on
+
+    private InsideBlockEffectType(final Consumer<Entity> p_405275_) {
+    // Paper start - track position inside effect was triggered on
+        this((entity, block) -> p_405275_.accept(entity));
+    }
+    private InsideBlockEffectType(final Applier effect) {
+    // Paper end - track position inside effect was triggered on
+        this.effect = effect;
+    }
+
+    public Applier effect() { // Paper - track position inside effect was triggered on
         return this.effect;
     }
+
+    // Paper start - track position inside effect was triggered on
+    // Use over biconsumer for less fqn spamming.
+    @FunctionalInterface
+    public interface Applier {
+        void affect(final Entity entity, final net.minecraft.core.BlockPos blockPos);
+    }
+    // Paper end - track position inside effect was triggered on
 }
