--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -17,6 +_,8 @@
         super(p_21683_, p_21684_);
     }
 
+    public BlockPos movingTarget; public BlockPos getMovingTarget() { return movingTarget; } // Paper
+
     public float getWalkTargetValue(BlockPos p_21693_) {
         return this.getWalkTargetValue(p_21693_, this.level());
     }
@@ -67,7 +_,7 @@
     }
 
     @Override
-    public void whenLeashedTo(Entity p_418459_) {
+    public void whenLeashedTo(Entity p_352153_) {
         this.setHomeTo(p_418459_.blockPosition(), (int)this.leashElasticDistance() - 1);
         super.whenLeashedTo(p_418459_);
     }
