--- a/net/minecraft/gametest/framework/GameTestHelper.java
+++ b/net/minecraft/gametest/framework/GameTestHelper.java
@@ -294,6 +_,10 @@
                 return p_323908_;
             }
 
+            public void setAfk(final boolean afk) {} // Purpur - AFK API
+
+            public void resetLastActionTime() {} // Purpur - Ridables
+
             @Override
             public boolean isClientAuthoritative() {
                 return false;
