--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -823,95 +_,98 @@
 
         @Override
         public void postProcess(
-            WorldGenLevel p_229635_,
-            StructureManager p_229636_,
-            ChunkGenerator p_229637_,
-            RandomSource p_229638_,
-            BoundingBox p_229639_,
-            ChunkPos p_229640_,
-            BlockPos p_229641_
+            WorldGenLevel level,
+            StructureManager structureManager,
+            ChunkGenerator generator,
+            RandomSource random,
+            BoundingBox box,
+            ChunkPos chunkPos,
+            BlockPos pos
         ) {
-            this.generateBox(p_229635_, p_229639_, 0, 0, 0, 10, 7, 15, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateSmallDoor(p_229635_, p_229638_, p_229639_, StrongholdPieces.StrongholdPiece.SmallDoorType.GRATES, 4, 1, 0);
+            this.generateBox(level, box, 0, 0, 0, 10, 7, 15, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateSmallDoor(level, random, box, StrongholdPieces.StrongholdPiece.SmallDoorType.GRATES, 4, 1, 0);
             int i = 6;
-            this.generateBox(p_229635_, p_229639_, 1, 6, 1, 1, 6, 14, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 9, 6, 1, 9, 6, 14, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 2, 6, 1, 8, 6, 2, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 2, 6, 14, 8, 6, 14, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 1, 1, 1, 2, 1, 4, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 8, 1, 1, 9, 1, 4, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 1, 1, 1, 1, 1, 3, Blocks.LAVA.defaultBlockState(), Blocks.LAVA.defaultBlockState(), false);
-            this.generateBox(p_229635_, p_229639_, 9, 1, 1, 9, 1, 3, Blocks.LAVA.defaultBlockState(), Blocks.LAVA.defaultBlockState(), false);
-            this.generateBox(p_229635_, p_229639_, 3, 1, 8, 7, 1, 12, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 4, 1, 9, 6, 1, 11, Blocks.LAVA.defaultBlockState(), Blocks.LAVA.defaultBlockState(), false);
-            BlockState blockstate = Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.NORTH, true).setValue(IronBarsBlock.SOUTH, true);
-            BlockState blockstate1 = Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.WEST, true).setValue(IronBarsBlock.EAST, true);
+            this.generateBox(level, box, 1, 6, 1, 1, 6, 14, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 9, 6, 1, 9, 6, 14, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 2, 6, 1, 8, 6, 2, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 2, 6, 14, 8, 6, 14, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 1, 1, 1, 2, 1, 4, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 8, 1, 1, 9, 1, 4, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 1, 1, 1, 1, 1, 3, Blocks.LAVA.defaultBlockState(), Blocks.LAVA.defaultBlockState(), false);
+            this.generateBox(level, box, 9, 1, 1, 9, 1, 3, Blocks.LAVA.defaultBlockState(), Blocks.LAVA.defaultBlockState(), false);
+            this.generateBox(level, box, 3, 1, 8, 7, 1, 12, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 4, 1, 9, 6, 1, 11, Blocks.LAVA.defaultBlockState(), Blocks.LAVA.defaultBlockState(), false);
+            BlockState blockState = Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.NORTH, true).setValue(IronBarsBlock.SOUTH, true);
+            BlockState blockState1 = Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.WEST, true).setValue(IronBarsBlock.EAST, true);
 
-            for (int j = 3; j < 14; j += 2) {
-                this.generateBox(p_229635_, p_229639_, 0, 3, j, 0, 4, j, blockstate, blockstate, false);
-                this.generateBox(p_229635_, p_229639_, 10, 3, j, 10, 4, j, blockstate, blockstate, false);
+            for (int i1 = 3; i1 < 14; i1 += 2) {
+                this.generateBox(level, box, 0, 3, i1, 0, 4, i1, blockState, blockState, false);
+                this.generateBox(level, box, 10, 3, i1, 10, 4, i1, blockState, blockState, false);
             }
 
             for (int i1 = 2; i1 < 9; i1 += 2) {
-                this.generateBox(p_229635_, p_229639_, i1, 3, 15, i1, 4, 15, blockstate1, blockstate1, false);
-            }
-
-            BlockState blockstate5 = Blocks.STONE_BRICK_STAIRS.defaultBlockState().setValue(StairBlock.FACING, Direction.NORTH);
-            this.generateBox(p_229635_, p_229639_, 4, 1, 5, 6, 1, 7, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 4, 2, 6, 6, 2, 7, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-            this.generateBox(p_229635_, p_229639_, 4, 3, 7, 6, 3, 7, false, p_229638_, StrongholdPieces.SMOOTH_STONE_SELECTOR);
-
-            for (int k = 4; k <= 6; k++) {
-                this.placeBlock(p_229635_, blockstate5, k, 1, 4, p_229639_);
-                this.placeBlock(p_229635_, blockstate5, k, 2, 5, p_229639_);
-                this.placeBlock(p_229635_, blockstate5, k, 3, 6, p_229639_);
-            }
-
-            BlockState blockstate6 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.NORTH);
-            BlockState blockstate2 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.SOUTH);
-            BlockState blockstate3 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.EAST);
-            BlockState blockstate4 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.WEST);
+                this.generateBox(level, box, i1, 3, 15, i1, 4, 15, blockState1, blockState1, false);
+            }
+
+            BlockState blockState2 = Blocks.STONE_BRICK_STAIRS.defaultBlockState().setValue(StairBlock.FACING, Direction.NORTH);
+            this.generateBox(level, box, 4, 1, 5, 6, 1, 7, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 4, 2, 6, 6, 2, 7, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+            this.generateBox(level, box, 4, 3, 7, 6, 3, 7, false, random, StrongholdPieces.SMOOTH_STONE_SELECTOR);
+
+            for (int i2 = 4; i2 <= 6; i2++) {
+                this.placeBlock(level, blockState2, i2, 1, 4, box);
+                this.placeBlock(level, blockState2, i2, 2, 5, box);
+                this.placeBlock(level, blockState2, i2, 3, 6, box);
+            }
+
+            BlockState blockState3 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.NORTH);
+            BlockState blockState4 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.SOUTH);
+            BlockState blockState5 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.EAST);
+            BlockState blockState6 = Blocks.END_PORTAL_FRAME.defaultBlockState().setValue(EndPortalFrameBlock.FACING, Direction.WEST);
             boolean flag = true;
-            boolean[] aboolean = new boolean[12];
+            boolean[] flags = new boolean[12];
 
-            for (int l = 0; l < aboolean.length; l++) {
-                aboolean[l] = p_229638_.nextFloat() > 0.9F;
-                flag &= aboolean[l];
+            for (int i3 = 0; i3 < flags.length; i3++) {
+                flags[i3] = random.nextFloat() > 0.9F;
+                flag &= flags[i3];
             }
 
-            this.placeBlock(p_229635_, blockstate6.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[0]), 4, 3, 8, p_229639_);
-            this.placeBlock(p_229635_, blockstate6.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[1]), 5, 3, 8, p_229639_);
-            this.placeBlock(p_229635_, blockstate6.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[2]), 6, 3, 8, p_229639_);
-            this.placeBlock(p_229635_, blockstate2.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[3]), 4, 3, 12, p_229639_);
-            this.placeBlock(p_229635_, blockstate2.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[4]), 5, 3, 12, p_229639_);
-            this.placeBlock(p_229635_, blockstate2.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[5]), 6, 3, 12, p_229639_);
-            this.placeBlock(p_229635_, blockstate3.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[6]), 3, 3, 9, p_229639_);
-            this.placeBlock(p_229635_, blockstate3.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[7]), 3, 3, 10, p_229639_);
-            this.placeBlock(p_229635_, blockstate3.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[8]), 3, 3, 11, p_229639_);
-            this.placeBlock(p_229635_, blockstate4.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[9]), 7, 3, 9, p_229639_);
-            this.placeBlock(p_229635_, blockstate4.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[10]), 7, 3, 10, p_229639_);
-            this.placeBlock(p_229635_, blockstate4.setValue(EndPortalFrameBlock.HAS_EYE, aboolean[11]), 7, 3, 11, p_229639_);
+            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, flags[0]), 4, 3, 8, box);
+            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, flags[1]), 5, 3, 8, box);
+            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, flags[2]), 6, 3, 8, box);
+            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, flags[3]), 4, 3, 12, box);
+            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, flags[4]), 5, 3, 12, box);
+            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, flags[5]), 6, 3, 12, box);
+            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, flags[6]), 3, 3, 9, box);
+            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, flags[7]), 3, 3, 10, box);
+            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, flags[8]), 3, 3, 11, box);
+            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, flags[9]), 7, 3, 9, box);
+            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, flags[10]), 7, 3, 10, box);
+            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, flags[11]), 7, 3, 11, box);
             if (flag) {
-                BlockState blockstate7 = Blocks.END_PORTAL.defaultBlockState();
-                this.placeBlock(p_229635_, blockstate7, 4, 3, 9, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 5, 3, 9, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 6, 3, 9, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 4, 3, 10, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 5, 3, 10, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 6, 3, 10, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 4, 3, 11, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 5, 3, 11, p_229639_);
-                this.placeBlock(p_229635_, blockstate7, 6, 3, 11, p_229639_);
+                BlockState blockState7 = Blocks.END_PORTAL.defaultBlockState();
+                this.placeBlock(level, blockState7, 4, 3, 9, box);
+                this.placeBlock(level, blockState7, 5, 3, 9, box);
+                this.placeBlock(level, blockState7, 6, 3, 9, box);
+                this.placeBlock(level, blockState7, 4, 3, 10, box);
+                this.placeBlock(level, blockState7, 5, 3, 10, box);
+                this.placeBlock(level, blockState7, 6, 3, 10, box);
+                this.placeBlock(level, blockState7, 4, 3, 11, box);
+                this.placeBlock(level, blockState7, 5, 3, 11, box);
+                this.placeBlock(level, blockState7, 6, 3, 11, box);
             }
 
             if (!this.hasPlacedSpawner) {
-                BlockPos blockpos = this.getWorldPos(5, 3, 6);
-                if (p_229639_.isInside(blockpos)) {
+                BlockPos worldPos = this.getWorldPos(5, 3, 6);
+                if (box.isInside(worldPos)) {
                     this.hasPlacedSpawner = true;
-                    p_229635_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
-                    if (p_229635_.getBlockEntity(blockpos) instanceof SpawnerBlockEntity spawnerblockentity) {
-                        spawnerblockentity.setEntityId(EntityType.SILVERFISH, p_229638_);
-                    }
+                    // CraftBukkit start
+                    // level.setBlock(worldPos, Blocks.SPAWNER.defaultBlockState(), 2);
+                    // if (level.getBlockEntity(worldPos) instanceof SpawnerBlockEntity spawnerBlockEntity) {
+                    //     spawnerBlockEntity.setEntityId(EntityType.SILVERFISH, random);
+                    // }
+                    this.placeCraftSpawner(level, worldPos, org.bukkit.entity.EntityType.SILVERFISH, 2);
+                    // CraftBukkit end
                 }
             }
         }
