--- a/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -71,17 +_,25 @@
         }
     }
 
-    public void signalTo(Vec3 p_423653_) {
-        Vec3 vec3 = p_423653_.subtract(this.position());
-        double d0 = vec3.horizontalDistance();
-        if (d0 > 12.0) {
-            this.target = this.position().add(vec3.x / d0 * 12.0, 8.0, vec3.z / d0 * 12.0);
+    public void signalTo(Vec3 pos) {
+        // Paper start - Change EnderEye target without changing other things
+        this.signalTo(pos, true);
+    }
+
+    public void signalTo(Vec3 pos, boolean update) {
+        // Paper end - Change EnderEye target without changing other things
+        Vec3 vec3 = pos.subtract(this.position());
+        double d = vec3.horizontalDistance();
+        if (d > 12.0) {
+            this.target = this.position().add(vec3.x / d * 12.0, 8.0, vec3.z / d * 12.0);
         } else {
-            this.target = p_423653_;
+            this.target = pos;
         }
 
+        if (update) { // Paper - Change EnderEye target without changing other things
         this.life = 0;
         this.surviveAfterDeath = this.random.nextInt(5) > 0;
+        } // Paper - Change EnderEye target without changing other things
     }
 
     @Override
@@ -102,7 +_,7 @@
             this.life++;
             if (this.life > 80 && !this.level().isClientSide) {
                 this.playSound(SoundEvents.ENDER_EYE_DEATH, 1.0F, 1.0F);
-                this.discard();
+                this.discard(this.surviveAfterDeath ? org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP : org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 if (this.surviveAfterDeath) {
                     this.level().addFreshEntity(new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), this.getItem()));
                 } else {
