--- a/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
+++ b/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
@@ -26,7 +_,7 @@
     }
 
     @Override
-    protected Item getDefaultItem() {
+    public Item getDefaultItem() {
         return Items.EXPERIENCE_BOTTLE;
     }
 
@@ -36,19 +_,28 @@
     }
 
     @Override
-    protected void onHit(HitResult p_37521_) {
-        super.onHit(p_37521_);
-        if (this.level() instanceof ServerLevel serverlevel) {
-            serverlevel.levelEvent(2002, this.blockPosition(), -13083194);
-            int i = 3 + serverlevel.random.nextInt(5) + serverlevel.random.nextInt(5);
-            if (p_37521_ instanceof BlockHitResult blockhitresult) {
-                Vec3 vec3 = blockhitresult.getDirection().getUnitVec3();
-                ExperienceOrb.awardWithDirection(serverlevel, p_37521_.getLocation(), vec3, i);
+    protected void onHit(HitResult result) {
+        super.onHit(result);
+        if (this.level() instanceof ServerLevel serverLevel) {
+            // CraftBukkit - moved to after event
+            int i = 3 + serverLevel.random.nextInt(5) + serverLevel.random.nextInt(5);
+            // Paper start - exp bottle event
+            org.bukkit.event.entity.ExpBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExpBottleEvent(this, result, i);
+            i = event.getExperience();
+            // Paper end - exp bottle event
+            if (result instanceof BlockHitResult blockHitResult) {
+                Vec3 unitVec3 = blockHitResult.getDirection().getUnitVec3();
+                ExperienceOrb.awardWithDirection(serverLevel, result.getLocation(), unitVec3, i, org.bukkit.entity.ExperienceOrb.SpawnReason.EXP_BOTTLE, this.getOwner(), this); // Paper
             } else {
-                ExperienceOrb.awardWithDirection(serverlevel, p_37521_.getLocation(), this.getDeltaMovement().scale(-1.0), i);
-            }
+                ExperienceOrb.awardWithDirection(serverLevel, result.getLocation(), this.getDeltaMovement().scale(-1.0), i, org.bukkit.entity.ExperienceOrb.SpawnReason.EXP_BOTTLE, this.getOwner(), this); // Paper
+            }
+            // Paper start - exp bottle event
+            if (event.getShowEffect()) {
+                this.level().levelEvent(net.minecraft.world.level.block.LevelEvent.PARTICLES_SPELL_POTION_SPLASH, this.blockPosition(), net.minecraft.world.item.alchemy.PotionContents.BASE_POTION_COLOR);
+            }
+            // Paper end - exp bottle event
 
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
 }
