--- a/net/minecraft/world/level/redstone/ExperimentalRedstoneUtils.java
+++ b/net/minecraft/world/level/redstone/ExperimentalRedstoneUtils.java
@@ -17,6 +_,11 @@
             if (p_366872_ != null) {
                 orientation = orientation.withFront(p_366872_);
             }
+            // Paper start - Optimize redstone (Alternate Current) - use default front instead of random
+            else if (p_366754_.paperConfig().misc.redstoneImplementation == io.papermc.paper.configuration.WorldConfiguration.Misc.RedstoneImplementation.ALTERNATE_CURRENT) {
+                orientation = orientation.withFront(Direction.WEST);
+            }
+            // Paper end - Optimize redstone (Alternate Current)
 
             return orientation;
         } else {
