--- a/net/minecraft/world/inventory/AbstractCraftingMenu.java
+++ b/net/minecraft/world/inventory/AbstractCraftingMenu.java
@@ -12,8 +_,8 @@
 public abstract class AbstractCraftingMenu extends RecipeBookMenu {
     private final int width;
     private final int height;
-    protected final CraftingContainer craftSlots;
-    protected final ResultContainer resultSlots = new ResultContainer();
+    public final CraftingContainer craftSlots; // Youer - fix: widen visibility to match Paper (used by CraftServer)
+    public final ResultContainer resultSlots = new ResultContainer(); // Youer - fix: widen visibility to match Paper (used by CraftServer)
 
     public AbstractCraftingMenu(MenuType<?> p_362493_, int p_360673_, int p_364200_, int p_363034_) {
         super(p_362493_, p_360673_);
