--- a/net/minecraft/world/inventory/AnvilMenu.java
+++ b/net/minecraft/world/inventory/AnvilMenu.java
@@ -23,6 +_,12 @@
 import net.minecraft.world.level.block.state.BlockState;
 import org.slf4j.Logger;
 
+// Purpur start - Anvil API
+import net.minecraft.network.protocol.game.ClientboundContainerSetDataPacket;
+import net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket;
+import net.minecraft.server.level.ServerPlayer;
+// Purpur end - Anvil API
+
 public class AnvilMenu extends ItemCombinerMenu {
     public static final int INPUT_SLOT = 0;
     public static final int ADDITIONAL_SLOT = 1;
@@ -46,6 +_,12 @@
     private static final int ADDITIONAL_SLOT_X_PLACEMENT = 76;
     private static final int RESULT_SLOT_X_PLACEMENT = 134;
     private static final int SLOT_Y_PLACEMENT = 47;
+    // CraftBukkit start
+    public static final int DEFAULT_DENIED_COST = -1;
+    public int maximumRepairCost = 40;
+    private org.bukkit.craftbukkit.inventory.view.CraftAnvilView bukkitEntity;
+    // CraftBukkit end
+    public boolean bypassEnchantmentLevelRestriction = false; // Paper - bypass anvil level restrictions
 
     public AnvilMenu(int p_39005_, Inventory p_39006_) {
         this(p_39005_, p_39006_, ContainerLevelAccess.NULL);
@@ -71,11 +_,21 @@
 
     @Override
     protected boolean mayPickup(Player p_39023_, boolean p_39024_) {
-        return (p_39023_.hasInfiniteMaterials() || p_39023_.experienceLevel >= this.cost.get()) && this.cost.get() > 0;
+        return (p_39023_.hasInfiniteMaterials() || p_39023_.experienceLevel >= this.cost.get()) && this.cost.get() > AnvilMenu.DEFAULT_DENIED_COST && hasStack; // CraftBukkit - allow cost 0 like a free item
     }
 
     @Override
     protected void onTake(Player p_150474_, ItemStack p_150475_) {
+        // Neo: Fire the AnvilCraftEvent.Pre here, before any changes are made to the anvil or the contained items.
+        // If a user cancels the event, we presume they are handling any of the below logic themselves.
+        var event = net.neoforged.neoforge.common.CommonHooks.fireAnvilCraftPre(this, p_150474_, p_150475_, this.inputSlots.getItem(0), this.inputSlots.getItem(1));
+        if (event.isCanceled()) {
+            return;
+        }
+
+        ItemStack leftInputSnapshot = this.inputSlots.getItem(0).copy();
+        ItemStack rightInputSnapshot = this.inputSlots.getItem(1).copy();
+
         if (!p_150474_.hasInfiniteMaterials()) {
             p_150474_.giveExperienceLevels(-this.cost.get());
         }
@@ -92,7 +_,7 @@
             this.inputSlots.setItem(1, ItemStack.EMPTY);
         }
 
-        this.cost.set(0);
+        this.cost.set(AnvilMenu.DEFAULT_DENIED_COST); // CraftBukkit - use a variable for set a cost for denied item
         if (p_150474_ instanceof ServerPlayer serverplayer
             && !StringUtil.isBlank(this.itemName)
             && !this.inputSlots.getItem(0).getHoverName().getString().equals(this.itemName)) {
@@ -104,6 +_,16 @@
             BlockState blockstate = p_150479_.getBlockState(p_150480_);
             if (!p_150474_.hasInfiniteMaterials() && blockstate.is(BlockTags.ANVIL) && p_150474_.getRandom().nextFloat() < 0.12F) {
                 BlockState blockstate1 = AnvilBlock.damage(blockstate);
+                // Paper start - AnvilDamageEvent
+                com.destroystokyo.paper.event.block.AnvilDamagedEvent anvilDamagedEvent = new com.destroystokyo.paper.event.block.AnvilDamagedEvent(getBukkitView(), blockstate1 != null ? org.bukkit.craftbukkit.block.data.CraftBlockData.fromData(blockstate1) : null);
+                if (!anvilDamagedEvent.callEvent()) {
+                    return;
+                } else if (anvilDamagedEvent.getDamageState() == com.destroystokyo.paper.event.block.AnvilDamagedEvent.DamageState.BROKEN) {
+                    blockstate1 = null;
+                } else {
+                    blockstate1 = ((org.bukkit.craftbukkit.block.data.CraftBlockData) anvilDamagedEvent.getDamageState().getMaterial().createBlockData()).getState().setValue(AnvilBlock.FACING, blockstate.getValue(AnvilBlock.FACING));
+                }
+                // Paper end - AnvilDamageEvent
                 if (blockstate1 == null) {
                     p_150479_.removeBlock(p_150480_, false);
                     p_150479_.levelEvent(1029, p_150480_, 0);
@@ -115,10 +_,30 @@
                 p_150479_.levelEvent(1030, p_150480_, 0);
             }
         });
+
+        // Neo: Fire post after all logic has taken place. For ease of use, we record the original input stacks and pass them to this method.
+        // The modified stacks are available through the menu.
+        net.neoforged.neoforge.common.CommonHooks.fireAnvilCraftPost(this, p_150474_, p_150475_, leftInputSnapshot, rightInputSnapshot);
     }
 
     @Override
-    public void createResult() {
+    public final void createResult() {
+        // Neo: Override the real createResult() method to invoke the vanilla logic, and then call the event hook.
+        // Since the vanilla createResult() has multiple returns, we need to wrap it. The choices are this, or pack it in a lambda.
+        this.createResultInternal();
+        ItemStack leftInput = this.inputSlots.getItem(0);
+        ItemStack rightInput = this.inputSlots.getItem(1);
+        net.neoforged.neoforge.common.CommonHooks.onAnvilUpdate(this, leftInput, rightInput, resultSlots, itemName, this.player);
+    }
+
+    /**
+     * Neo: This is the original createResult() method, which is called by the above method ({@link #createResult()}.
+     * <p>
+     * If you would normally override {@link #createResult()}, you should override this method instead.
+     *
+     * @implNote This replacement exists so we can fire the AnvilUpdateEvent with the vanilla result, despite the multiple returns in the original method.
+     */
+    protected void createResultInternal() {
         ItemStack itemstack = this.inputSlots.getItem(0);
         this.onlyRenaming = false;
         this.cost.set(1);
@@ -136,8 +_,8 @@
                 if (itemstack1.isDamageableItem() && itemstack.isValidRepairItem(itemstack2)) {
                     int l2 = Math.min(itemstack1.getDamageValue(), itemstack1.getMaxDamage() / 4);
                     if (l2 <= 0) {
-                        this.resultSlots.setItem(0, ItemStack.EMPTY);
-                        this.cost.set(0);
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), ItemStack.EMPTY); // CraftBukkit
+                        this.cost.set(AnvilMenu.DEFAULT_DENIED_COST); // CraftBukkit - use a variable for set a cost for denied item
                         return;
                     }
 
@@ -152,8 +_,8 @@
                     this.repairItemCountCost = j3;
                 } else {
                     if (!flag && (!itemstack1.is(itemstack2.getItem()) || !itemstack1.isDamageableItem())) {
-                        this.resultSlots.setItem(0, ItemStack.EMPTY);
-                        this.cost.set(0);
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), ItemStack.EMPTY); // CraftBukkit
+                        this.cost.set(AnvilMenu.DEFAULT_DENIED_COST); // CraftBukkit - use a variable for set a cost for denied item
                         return;
                     }
 
@@ -183,8 +_,10 @@
                         int j2 = entry.getIntValue();
                         j2 = i2 == j2 ? j2 + 1 : Math.max(j2, i2);
                         Enchantment enchantment = holder.value();
-                        boolean flag1 = enchantment.canEnchant(itemstack);
-                        if (this.player.hasInfiniteMaterials() || itemstack.is(Items.ENCHANTED_BOOK)) {
+                        // Neo: Respect IItemExtension#supportsEnchantment - we also delegate the logic for Enchanted Books to this method.
+                        // Though we still allow creative players to combine any item with any enchantment in the anvil here.
+                        boolean flag1 = itemstack.supportsEnchantment(holder);
+                        if (this.player.getAbilities().instabuild) {
                             flag1 = true;
                         }
 
@@ -199,7 +_,7 @@
                             flag3 = true;
                         } else {
                             flag2 = true;
-                            if (j2 > enchantment.getMaxLevel()) {
+                            if (j2 > enchantment.getMaxLevel() && !this.bypassEnchantmentLevelRestriction) { // Paper - bypass anvil level restrictions
                                 j2 = enchantment.getMaxLevel();
                             }
 
@@ -217,8 +_,8 @@
                     }
 
                     if (flag3 && !flag2) {
-                        this.resultSlots.setItem(0, ItemStack.EMPTY);
-                        this.cost.set(0);
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), ItemStack.EMPTY); // CraftBukkit
+                        this.cost.set(AnvilMenu.DEFAULT_DENIED_COST); // CraftBukkit - use a variable for set a cost for denied item
                         return;
                     }
                 }
@@ -228,6 +_,54 @@
                 if (!this.itemName.equals(itemstack.getHoverName().getString())) {
                     k = 1;
                     i += k;
+                    // Purpur start - Allow anvil colors
+                    if (this.player != null) {
+                        org.bukkit.craftbukkit.entity.CraftHumanEntity player = this.player.getBukkitEntity();
+                        String name = this.itemName;
+                        boolean removeItalics = false;
+                        if (player.hasPermission("purpur.anvil.remove_italics")) {
+                            if (name.startsWith("&r")) {
+                                name = name.substring(2);
+                                removeItalics = true;
+                            } else if (name.startsWith("<r>")) {
+                                name = name.substring(3);
+                                removeItalics = true;
+                            } else if (name.startsWith("<reset>")) {
+                                name = name.substring(7);
+                                removeItalics = true;
+                            }
+                        }
+                        if (this.player.level().purpurConfig.anvilAllowColors) {
+                            if (player.hasPermission("purpur.anvil.color")) {
+                                java.util.regex.Matcher matcher = java.util.regex.Pattern.compile("(?i)&([0-9a-fr])").matcher(name);
+                                while (matcher.find()) {
+                                    String match = matcher.group(1);
+                                    name = name.replace("&" + match, "\u00a7" + match.toLowerCase(java.util.Locale.ROOT));
+                                }
+                                //name = name.replaceAll("(?i)&([0-9a-fr])", "\u00a7$1");
+                            }
+                            if (player.hasPermission("purpur.anvil.format")) {
+                                java.util.regex.Matcher matcher = java.util.regex.Pattern.compile("(?i)&([k-or])").matcher(name);
+                                while (matcher.find()) {
+                                    String match = matcher.group(1);
+                                    name = name.replace("&" + match, "\u00a7" + match.toLowerCase(java.util.Locale.ROOT));
+                                }
+                                //name = name.replaceAll("(?i)&([l-or])", "\u00a7$1");
+                            }
+                        }
+                        net.kyori.adventure.text.Component component;
+                        if (this.player.level().purpurConfig.anvilColorsUseMiniMessage && player.hasPermission("purpur.anvil.minimessage")) {
+                            component = net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(org.bukkit.ChatColor.stripColor(name));
+                        } else {
+                            component = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(name);
+                        }
+                        if (removeItalics) {
+                            component = component.decoration(net.kyori.adventure.text.format.TextDecoration.ITALIC, false);
+                        }
+                        itemStack.set(DataComponents.CUSTOM_NAME, io.papermc.paper.adventure.PaperAdventure.asVanilla(component));
+                    }
+                    else
+                    // Purpur end - Allow anvil colors
                     itemstack1.set(DataComponents.CUSTOM_NAME, Component.literal(this.itemName));
                 }
             } else if (itemstack.has(DataComponents.CUSTOM_NAME)) {
@@ -243,16 +_,25 @@
             }
 
             if (k == i && k > 0) {
-                if (this.cost.get() >= 40) {
-                    this.cost.set(39);
+                // CraftBukkit start
+                if (this.cost.get() >= this.maximumRepairCost) {
+                    this.cost.set(this.maximumRepairCost - 1);
+                // CraftBukkit end
                 }
 
                 this.onlyRenaming = true;
             }
 
-            if (this.cost.get() >= 40 && !this.player.hasInfiniteMaterials()) {
+            // Purpur start - Anvil API
+            if (this.bypassCost && this.cost.get() >= this.maximumRepairCost) {
+                this.cost.set(this.maximumRepairCost - 1);
+            }
+            // Purpur end - Anvil API
+
+            if (this.cost.get() >= this.maximumRepairCost && !this.player.hasInfiniteMaterials()) { // CraftBukkit
                 itemstack1 = ItemStack.EMPTY;
             }
+
 
             if (!itemstack1.isEmpty()) {
                 int i3 = itemstack1.getOrDefault(DataComponents.REPAIR_COST, 0);
@@ -277,7 +_,7 @@
     }
 
     public static int calculateIncreasedRepairCost(int p_39026_) {
-        return (int)Math.min(p_39026_ * 2L + 1L, 2147483647L);
+        return org.purpurmc.purpur.PurpurConfig.anvilCumulativeCost ? (int)Math.min(p_39026_ * 2L + 1L, 2147483647L) : 0; // Purpur - Make anvil cumulative cost configurable
     }
 
     public boolean setItemName(String p_288970_) {
@@ -294,6 +_,7 @@
             }
 
             this.createResult();
+            org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareResultEvent(this, RESULT_SLOT); // Paper - Add PrepareResultEvent
             return true;
         } else {
             return false;
@@ -309,4 +_,11 @@
     public int getCost() {
         return this.cost.get();
     }
+
+    /**
+     * Neo: Sets the cost. Negative values are clamped to zero.
+     */
+    public void setCost(int value) {
+        this.cost.set(Math.max(0, value));
+     }
 }
