--- a/net/minecraft/world/inventory/ContainerListener.java
+++ b/net/minecraft/world/inventory/ContainerListener.java
@@ -6,4 +_,10 @@
     void slotChanged(AbstractContainerMenu p_39315_, int p_39316_, ItemStack p_39317_);
 
     void dataChanged(AbstractContainerMenu p_150524_, int p_150525_, int p_150526_);
+
+    // Paper start - Add PlayerInventorySlotChangeEvent
+    default void slotChanged(AbstractContainerMenu containerToSend, int dataSlotIndex, ItemStack oldStack, ItemStack stack) {
+        slotChanged(containerToSend, dataSlotIndex, stack);
+    }
+    // Paper end - Add PlayerInventorySlotChangeEvent
 }
