--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -21,6 +_,21 @@
 import net.minecraft.world.phys.Vec3;
 
 public class GrindstoneMenu extends AbstractContainerMenu {
+    // CraftBukkit start
+    private @javax.annotation.Nullable org.bukkit.craftbukkit.inventory.CraftInventoryView view = null;
+    private final org.bukkit.entity.Player player;
+
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (this.view != null) {
+            return this.view;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventoryGrindstone inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryGrindstone(this.repairSlots, this.resultSlots);
+        this.view = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.player, inventory, this);
+        return this.view;
+    }
+    // CraftBukkit end
     public static final int MAX_NAME_LENGTH = 35;
     public static final int INPUT_SLOT = 0;
     public static final int ADDITIONAL_SLOT = 1;
@@ -29,15 +_,10 @@
     private static final int INV_SLOT_END = 30;
     private static final int USE_ROW_SLOT_START = 30;
     private static final int USE_ROW_SLOT_END = 39;
-    private final Container resultSlots = new ResultContainer();
-    final Container repairSlots = new SimpleContainer(2) {
-        @Override
-        public void setChanged() {
-            super.setChanged();
-            GrindstoneMenu.this.slotsChanged(this);
-        }
-    };
+    private final Container resultSlots; // Paper - Add missing InventoryHolders - move down
+    final Container repairSlots; // Paper - Add missing InventoryHolders - move down
     private final ContainerLevelAccess access;
+    private int xp = -1;
 
     public GrindstoneMenu(int p_39563_, Inventory p_39564_) {
         this(p_39563_, p_39564_, ContainerLevelAccess.NULL);
@@ -45,17 +_,33 @@
 
     public GrindstoneMenu(int p_39566_, Inventory p_39567_, final ContainerLevelAccess p_39568_) {
         super(MenuType.GRINDSTONE, p_39566_);
+        // Paper start - Add missing InventoryHolders
+        this.resultSlots = new ResultContainer(this.createBlockHolder(p_39568_)); // Paper - Add missing InventoryHolders
+        this.repairSlots = new SimpleContainer(this.createBlockHolder(p_39568_), 2) { // Paper - Add missing InventoryHolders
+            @Override
+            public void setChanged() {
+                super.setChanged();
+                GrindstoneMenu.this.slotsChanged(this);
+            }
+            // CraftBukkit start
+            @Override
+            public org.bukkit.Location getLocation() {
+                return access.getLocation();
+            }
+            // CraftBukkit end
+        };
+        // Paper end - Add missing InventoryHolders
         this.access = p_39568_;
         this.addSlot(new Slot(this.repairSlots, 0, 49, 19) {
             @Override
             public boolean mayPlace(ItemStack p_39607_) {
-                return p_39607_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39607_);
+                return p_39607_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39607_) || p_39607_.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.repairSlots, 1, 49, 40) {
             @Override
             public boolean mayPlace(ItemStack p_39616_) {
-                return p_39616_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39616_);
+                return p_39616_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39616_) || p_39616_.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.resultSlots, 2, 129, 34) {
@@ -66,6 +_,7 @@
 
             @Override
             public void onTake(Player p_150574_, ItemStack p_150575_) {
+                if (net.neoforged.neoforge.common.CommonHooks.onGrindstoneTake(GrindstoneMenu.this.repairSlots, p_39568_, p_150574_, this::getExperienceAmount)) return;
                 p_39568_.execute((p_39634_, p_39635_) -> {
                     if (p_39634_ instanceof ServerLevel) {
                         ExperienceOrb.award((ServerLevel)p_39634_, Vec3.atCenterOf(p_39635_), this.getExperienceAmount(p_39634_));
@@ -78,6 +_,7 @@
             }
 
             private int getExperienceAmount(Level p_39632_) {
+                if (xp > -1) return xp;
                 int i = 0;
                 i += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(0));
                 i += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(1));
@@ -96,7 +_,7 @@
                 for (Entry<Holder<Enchantment>> entry : itemenchantments.entrySet()) {
                     Holder<Enchantment> holder = entry.getKey();
                     int j = entry.getIntValue();
-                    if (!holder.is(EnchantmentTags.CURSE)) {
+                    if (!org.purpurmc.purpur.PurpurConfig.grindstoneIgnoredEnchants.contains(holder.value())) { // Purpur - Config for grindstones
                         i += holder.value().getMinCost(j);
                     }
                 }
@@ -105,6 +_,7 @@
             }
         });
         this.addStandardInventorySlots(p_39567_, 8, 84);
+        this.player = (org.bukkit.entity.Player) p_39567_.player.getBukkitEntity(); // CraftBukkit
     }
 
     @Override
@@ -116,6 +_,8 @@
     }
 
     private void createResult() {
+        this.xp = net.neoforged.neoforge.common.CommonHooks.onGrindstoneChange(this.repairSlots.getItem(0), this.repairSlots.getItem(1), this.resultSlots, -1);
+        if (this.xp == Integer.MIN_VALUE)
         this.resultSlots.setItem(0, this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1)));
         this.broadcastChanges();
     }
@@ -146,7 +_,7 @@
             int k = p_332686_.getMaxDamage() - p_332686_.getDamageValue();
             int l = j + k + i * 5 / 100;
             int i1 = 1;
-            if (!p_332723_.isDamageableItem()) {
+            if (!p_332723_.isDamageableItem() || !p_332723_.isCombineRepairable()) {
                 if (p_332723_.getMaxStackSize() < 2 || !ItemStack.matches(p_332723_, p_332686_)) {
                     return ItemStack.EMPTY;
                 }
@@ -158,6 +_,7 @@
             if (itemstack.isDamageableItem()) {
                 itemstack.set(DataComponents.MAX_DAMAGE, i);
                 itemstack.setDamageValue(Math.max(i - l, 0));
+                if (!p_332686_.isCombineRepairable()) itemstack.setDamageValue(p_332723_.getDamageValue());
             }
 
             this.mergeEnchantsFrom(itemstack, p_332686_);
@@ -178,6 +_,15 @@
         });
     }
 
+    // Purpur start - Config for grindstones
+    private java.util.List<net.minecraft.core.component.DataComponentType<?>> GRINDSTONE_REMOVE_ATTRIBUTES_REMOVAL_LIST = java.util.List.of(
+        DataComponents.CUSTOM_DATA,
+        DataComponents.ATTRIBUTE_MODIFIERS,
+        DataComponents.CUSTOM_MODEL_DATA,
+        DataComponents.DYED_COLOR,
+        DataComponents.SUSPICIOUS_STEW_EFFECTS
+    );
+    // Purpur end - Config for grindstones
     private ItemStack removeNonCursesFrom(ItemStack p_332709_) {
         ItemEnchantments itemenchantments = EnchantmentHelper.updateEnchantments(
             p_332709_, p_330066_ -> p_330066_.removeIf(p_344368_ -> !p_344368_.is(EnchantmentTags.CURSE))
@@ -193,6 +_,23 @@
         }
 
         p_332709_.set(DataComponents.REPAIR_COST, i);
+
+        // Purpur start - Config for grindstones
+        net.minecraft.core.component.DataComponentPatch.Builder builder = net.minecraft.core.component.DataComponentPatch.builder();
+        if (org.purpurmc.purpur.PurpurConfig.grindstoneRemoveAttributes) {
+            p_332709_.getComponents().forEach(typedDataComponent -> {
+                if (GRINDSTONE_REMOVE_ATTRIBUTES_REMOVAL_LIST.contains(typedDataComponent.type())) {
+                    builder.remove(typedDataComponent.type());
+                }
+            });
+        }
+        if (org.purpurmc.purpur.PurpurConfig.grindstoneRemoveDisplay) {
+            builder.remove(DataComponents.CUSTOM_NAME);
+            builder.remove(DataComponents.LORE);
+        }
+        p_332709_.applyComponents(builder.build());
+        // Purpur end - Config for grindstones
+
         return p_332709_;
     }
 
@@ -204,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player p_39572_) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, p_39572_, Blocks.GRINDSTONE);
     }
 
@@ -248,7 +_,9 @@
                 return ItemStack.EMPTY;
             }
 
+            this.activeQuickItem = itemstack; // Purpur - Grindstone API
             slot.onTake(p_39588_, itemstack1);
+            this.activeQuickItem = null; // Purpur - Grindstone API
         }
 
         return itemstack;
