--- a/net/minecraft/world/inventory/HorseInventoryMenu.java
+++ b/net/minecraft/world/inventory/HorseInventoryMenu.java
@@ -15,13 +_,27 @@
     private static final ResourceLocation LLAMA_ARMOR_SLOT_SPRITE = ResourceLocation.withDefaultNamespace("container/slot/llama_armor");
     private static final ResourceLocation ARMOR_SLOT_SPRITE = ResourceLocation.withDefaultNamespace("container/slot/horse_armor");
     private final Container horseContainer;
-    private final AbstractHorse horse;
+    public final AbstractHorse horse; // Youer - fix: widen visibility to match Paper (used by CraftHumanEntity)
     public static final int SLOT_SADDLE = 0;
     public static final int SLOT_BODY_ARMOR = 1;
     public static final int SLOT_HORSE_INVENTORY_START = 2;
+    // CraftBukkit start
+    private @javax.annotation.Nullable org.bukkit.craftbukkit.inventory.CraftInventoryView view;
+    private final Inventory inventory;
+
+    @Override
+    public org.bukkit.inventory.InventoryView getBukkitView() {
+        if (this.view != null) {
+            return this.view;
+        }
+
+        return this.view = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.inventory.player.getBukkitEntity(), this.horseContainer.getOwner().getInventory(), this);
+    }
+    // CraftBukkit end
 
     public HorseInventoryMenu(int p_39656_, Inventory p_39657_, Container p_39658_, final AbstractHorse p_39659_, int p_352384_) {
         super(null, p_39656_);
+        this.inventory = p_39657_; // CraftBukkit
         this.horseContainer = p_39658_;
         this.horse = p_39659_;
         p_39658_.startOpen(p_39657_.player);
