--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -179,6 +_,8 @@
         MemoryModuleType.MEETING_POINT,
         (p_219616_, p_219617_) -> p_219617_.is(PoiTypes.MEETING)
     );
+    private boolean isLobotomized = false; public boolean isLobotomized() { return this.isLobotomized; } // Purpur - Lobotomize stuck villagers
+    private int notLobotomizedCount = 0; // Purpur - Lobotomize stuck villagers
 
     public Villager(EntityType<? extends Villager> p_35381_, Level p_35382_) {
         this(p_35381_, p_35382_, VillagerType.PLAINS);
@@ -197,6 +_,103 @@
         this.setVillagerData(this.getVillagerData().withType(p_397796_).withProfession(p_35385_.registryAccess(), VillagerProfession.NONE));
     }
 
+    // Purpur start - Allow leashing villagers
+    @Override
+    public boolean canBeLeashed() {
+        return level().purpurConfig.villagerCanBeLeashed;
+    }
+    // Purpur end - Allow leashing villagers
+
+    // Purpur start - Lobotomize stuck villagers
+    private boolean checkLobotomized() {
+        int interval = this.level().purpurConfig.villagerLobotomizeCheckInterval;
+        boolean shouldCheckForTradeLocked = this.level().purpurConfig.villagerLobotomizeWaitUntilTradeLocked;
+        if (this.notLobotomizedCount > 3) {
+            // check half as often if not lobotomized for the last 3+ consecutive checks
+            interval *= 2;
+        }
+        if (this.level().getGameTime() % interval == 0) {
+            // offset Y for short blocks like dirt_path/farmland
+            this.isLobotomized = !(shouldCheckForTradeLocked && this.getVillagerXp() == 0) && !canTravelFrom(BlockPos.containing(this.position().x, this.getBoundingBox().minY + 0.0625D, this.position().z));
+
+            if (this.isLobotomized) {
+                this.notLobotomizedCount = 0;
+            } else {
+                this.notLobotomizedCount++;
+            }
+        }
+        return this.isLobotomized;
+    }
+
+    private boolean canTravelFrom(BlockPos pos) {
+        return canTravelTo(pos.east()) || canTravelTo(pos.west()) || canTravelTo(pos.north()) || canTravelTo(pos.south());
+    }
+
+    private boolean canTravelTo(BlockPos pos) {
+        net.minecraft.world.level.block.state.BlockState state = this.level().getBlockStateIfLoaded(pos);
+        if (state == null) {
+            // chunk not loaded
+            return false;
+        }
+        net.minecraft.world.level.block.Block bottom = state.getBlock();
+        if (bottom instanceof net.minecraft.world.level.block.FenceBlock ||
+            bottom instanceof net.minecraft.world.level.block.FenceGateBlock ||
+            bottom instanceof net.minecraft.world.level.block.WallBlock) {
+            // bottom block is too tall to get over
+            return false;
+        }
+        net.minecraft.world.level.block.Block top = level().getBlockState(pos.above()).getBlock();
+        // only if both blocks have no collision
+        return !bottom.hasCollision && !top.hasCollision;
+    }
+    // Purpur end - Lobotomize stuck villagers
+
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.villagerRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.villagerRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.villagerControllable;
+    }
+
+    @Override
+    protected void registerGoals() {
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this));
+        if (level().purpurConfig.villagerFollowEmeraldBlock) this.goalSelector.addGoal(3, new net.minecraft.world.entity.ai.goal.TemptGoal(this, 1.0D, TEMPT_ITEMS, false)); // Purpur - Villagers follow emerald blocks
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.villagerMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.villagerScale);
+        this.getAttribute(Attributes.TEMPT_RANGE).setBaseValue(this.level().purpurConfig.villagerTemptRange); // Purpur - Villagers follow emerald blocks
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.villagerTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.villagerAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
     @Override
     public Brain<Villager> getBrain() {
         return (Brain<Villager>)super.getBrain();
@@ -230,7 +_,7 @@
             p_35425_.setSchedule(Schedule.VILLAGER_DEFAULT);
             p_35425_.addActivityWithConditions(
                 Activity.WORK,
-                VillagerGoalPackages.getWorkPackage(holder, 0.5F),
+                VillagerGoalPackages.getWorkPackage(holder, 0.5F, this.level().purpurConfig.villagerClericsFarmWarts), // Purpur - Option for Villager Clerics to farm Nether Wart
                 ImmutableSet.of(Pair.of(MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_PRESENT))
             );
         }
@@ -262,19 +_,54 @@
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.5);
+        return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.5).add(Attributes.TEMPT_RANGE, 10.0D); // Purpur - Villagers follow emerald blocks
     }
 
     public boolean assignProfessionWhenSpawned() {
         return this.assignProfessionWhenSpawned;
     }
 
-    @Override
-    protected void customServerAiStep(ServerLevel p_376753_) {
-        ProfilerFiller profilerfiller = Profiler.get();
-        profilerfiller.push("villagerBrain");
-        this.getBrain().tick(p_376753_, this);
-        profilerfiller.pop();
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        // SPIGOT-3874, SPIGOT-3894, SPIGOT-3846, SPIGOT-5286 :(
+        if (this.getUnhappyCounter() > 0) {
+            this.setUnhappyCounter(this.getUnhappyCounter() - 1);
+        }
+        if (this.isEffectiveAi()) {
+            if (this.level().spigotConfig.tickInactiveVillagers) {
+                this.customServerAiStep(this.level().getMinecraftWorld());
+            } else {
+                this.customServerAiStep(this.level().getMinecraftWorld(), true);
+            }
+        }
+        this.maybeDecayGossip();
+        super.inactiveTick();
+    }
+    // Paper end - EAR 2
+
+    @Override
+    protected void customServerAiStep(ServerLevel level) {
+        // Paper start - EAR 2
+        this.customServerAiStep(level, false);
+    }
+    protected void customServerAiStep(ServerLevel level, boolean inactive) { // Purpur - Lobotomize stuck villagers - not final
+        // Paper end - EAR 2
+        ProfilerFiller profilerFiller = Profiler.get();
+        profilerFiller.push("villagerBrain");
+        // Purpur start - Lobotomize stuck villagers
+        if (this.level().purpurConfig.villagerLobotomizeEnabled) {
+            // treat as inactive if lobotomized
+            inactive = inactive || checkLobotomized();
+        } else {
+            this.isLobotomized = false;
+        }
+        // Purpur end - Lobotomize stuck villagers
+        if (!inactive && (getRider() == null || !this.isControllable())) { // Purpur - Ridables
+            this.getBrain().tick(level, this); // Paper - EAR 2
+        }
+        else if (this.isLobotomized && shouldRestock()) restock(); // Purpur - Lobotomize stuck villagers
+        profilerFiller.pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
         }
@@ -287,20 +_,20 @@
                     this.increaseProfessionLevelOnUpdate = false;
                 }
 
-                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0));
+                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.VILLAGER_TRADE); // CraftBukkit
             }
         }
 
         if (this.lastTradedPlayer != null) {
-            p_376753_.onReputationEvent(ReputationEventType.TRADE, this.lastTradedPlayer, this);
-            p_376753_.broadcastEntityEvent(this, (byte)14);
+            level.onReputationEvent(ReputationEventType.TRADE, this.lastTradedPlayer, this);
+            level.broadcastEntityEvent(this, (byte)14);
             this.lastTradedPlayer = null;
         }
 
-        if (!this.isNoAi() && this.random.nextInt(100) == 0) {
-            Raid raid = p_376753_.getRaidAt(this.blockPosition());
-            if (raid != null && raid.isActive() && !raid.isOver()) {
-                p_376753_.broadcastEntityEvent(this, (byte)42);
+        if (!inactive && !this.isNoAi() && this.random.nextInt(100) == 0) { // Paper - EAR 2
+            Raid raidAt = level.getRaidAt(this.blockPosition());
+            if (raidAt != null && raidAt.isActive() && !raidAt.isOver()) {
+                level.broadcastEntityEvent(this, (byte)42);
             }
         }
 
@@ -308,7 +_,8 @@
             this.stopTrading();
         }
 
-        super.customServerAiStep(p_376753_);
+        if (inactive) return; // Paper - EAR 2
+        super.customServerAiStep(level);
     }
 
     @Override
@@ -324,11 +_,11 @@
     @Override
     public InteractionResult mobInteract(Player p_35472_, InteractionHand p_35473_) {
         ItemStack itemstack = p_35472_.getItemInHand(p_35473_);
-        if (itemstack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping()) {
+        if (itemstack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping() || p_35472_.isSecondaryUseActive()) {
             return super.mobInteract(p_35472_, p_35473_);
         } else if (this.isBaby()) {
             this.setUnhappy();
-            return InteractionResult.SUCCESS;
+            return tryRide(p_35472_, p_35473_, InteractionResult.SUCCESS); // Purpur - Ridables
         } else {
             if (!this.level().isClientSide) {
                 boolean flag = this.getOffers().isEmpty();
@@ -341,9 +_,12 @@
                 }
 
                 if (flag) {
-                    return InteractionResult.CONSUME;
+                    return tryRide(p_35472_, p_35473_, InteractionResult.CONSUME); // Purpur - Ridables
                 }
 
+                if (level().purpurConfig.villagerRidable && itemstack.isEmpty()) return tryRide(p_35472_, p_35473_); // Purpur - Ridables
+
+                if (this.level().purpurConfig.villagerAllowTrading) // Purpur - Add config for villager trading
                 this.startTrading(p_35472_);
             }
 
@@ -395,8 +_,13 @@
     public void restock() {
         this.updateDemand();
 
-        for (MerchantOffer merchantoffer : this.getOffers()) {
-            merchantoffer.resetUses();
+        for (MerchantOffer merchantOffer : this.getOffers()) {
+            // CraftBukkit start
+            org.bukkit.event.entity.VillagerReplenishTradeEvent event = new org.bukkit.event.entity.VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), merchantOffer.asBukkit());
+            if (event.callEvent()) {
+                merchantOffer.resetUses();
+            }
+            // CraftBukkit end
         }
 
         this.resendOffersToTradingPlayer();
@@ -456,12 +_,17 @@
     private void catchUpDemand() {
         int i = 2 - this.numberOfRestocksToday;
         if (i > 0) {
-            for (MerchantOffer merchantoffer : this.getOffers()) {
-                merchantoffer.resetUses();
+            for (MerchantOffer merchantOffer : this.getOffers()) {
+                // CraftBukkit start
+                org.bukkit.event.entity.VillagerReplenishTradeEvent event = new org.bukkit.event.entity.VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), merchantOffer.asBukkit());
+                if (event.callEvent()) {
+                    merchantOffer.resetUses();
+                }
+                // CraftBukkit end
             }
         }
 
-        for (int j = 0; j < i; j++) {
+        for (int i1 = 0; i1 < i; i1++) {
             this.updateDemand();
         }
 
@@ -470,7 +_,7 @@
 
     public void updateDemand() {
         for (MerchantOffer merchantoffer : this.getOffers()) {
-            merchantoffer.updateDemand();
+            merchantoffer.updateDemand(this.level().purpurConfig.villagerMinimumDemand); // Purpur - Configurable minimum demand for trades
         }
     }
 
@@ -478,6 +_,7 @@
         int i = this.getPlayerReputation(p_35541_);
         if (i != 0) {
             for (MerchantOffer merchantoffer : this.getOffers()) {
+                if (merchantoffer.ignoreDiscounts) continue; // Paper - Add ignore discounts API
                 merchantoffer.addToSpecialPriceDiff(-Mth.floor(i * merchantoffer.getPriceMultiplier()));
             }
         }
@@ -487,6 +_,7 @@
             int k = mobeffectinstance.getAmplifier();
 
             for (MerchantOffer merchantoffer1 : this.getOffers()) {
+                if (merchantoffer1.ignoreDiscounts) continue; // Paper - Add ignore discounts API
                 double d0 = 0.3 + 0.0625 * k;
                 int j = (int)Math.floor(d0 * merchantoffer1.getBaseCostA().getCount());
                 merchantoffer1.addToSpecialPriceDiff(-Math.max(j, 1));
@@ -546,7 +_,7 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         if (this.isSleeping()) {
             return null;
         } else {
@@ -555,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_35498_) {
+    public SoundEvent getHurtSound(DamageSource p_35498_) {
         return SoundEvents.VILLAGER_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.VILLAGER_DEATH;
     }
 
@@ -595,7 +_,7 @@
         }
 
         if (p_35475_.shouldRewardExp()) {
-            this.level().addFreshEntity(new ExperienceOrb(this.level(), this.getX(), this.getY() + 0.5, this.getZ(), i));
+            this.level().addFreshEntity(new ExperienceOrb(this.level(), this.getX(), this.getY() + 0.5, this.getZ(), i, org.bukkit.entity.ExperienceOrb.SpawnReason.VILLAGER_TRADE, this.getTradingPlayer(), this)); // Paper
         }
     }
 
@@ -613,7 +_,7 @@
 
     @Override
     public void die(DamageSource p_35419_) {
-        LOGGER.info("Villager {} died, message: '{}'", this, p_35419_.getLocalizedDeathMessage(this).getString());
+        if (org.spigotmc.SpigotConfig.logVillagerDeaths) LOGGER.info("Villager {} died, message: '{}'", this, p_35419_.getLocalizedDeathMessage(this).getString()); // Spigot
         Entity entity = p_35419_.getEntity();
         if (entity != null) {
             this.tellWitnessesThatIWasMurdered(entity);
@@ -661,7 +_,7 @@
 
     @Override
     public boolean canBreed() {
-        return this.foodLevel + this.countFoodPointsInInventory() >= 12 && !this.isSleeping() && this.getAge() == 0;
+        return this.level().purpurConfig.villagerCanBreed && this.foodLevel + this.countFoodPointsInInventory() >= 12 && !this.isSleeping() && this.getAge() == 0; // Purpur - Configurable villager breeding
     }
 
     private boolean hungry() {
@@ -779,13 +_,21 @@
 
     @Override
     public void thunderHit(ServerLevel p_35409_, LightningBolt p_35410_) {
-        if (p_35409_.getDifficulty() != Difficulty.PEACEFUL) {
-            LOGGER.info("Villager {} was struck by lightning {}.", this, p_35410_);
-            Witch witch = this.convertTo(EntityType.WITCH, ConversionParams.single(this, false, false), p_427125_ -> {
-                p_427125_.finalizeSpawn(p_35409_, p_35409_.getCurrentDifficultyAt(p_427125_.blockPosition()), EntitySpawnReason.CONVERSION, null);
-                p_427125_.setPersistenceRequired();
+        if (p_35409_.getDifficulty() != Difficulty.PEACEFUL && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.WITCH, (timer) -> {})) {
+            // Paper - Add EntityZapEvent; move log down, event can cancel
+            Witch witch = this.convertTo(EntityType.WITCH, ConversionParams.single(this, false, false), witch1 -> {
+                // Paper start - Add EntityZapEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityZapEvent(this, p_35410_, witch1).isCancelled()) {
+                    return false;
+                }
+                if (org.spigotmc.SpigotConfig.logVillagerDeaths) LOGGER.info("Villager {} was struck by p_35410_ {}.", this, p_35410_); // Move down
+                // Paper end - Add EntityZapEvent
+                witch1.finalizeSpawn(p_35409_, p_35409_.getCurrentDifficultyAt(witch1.blockPosition()), EntitySpawnReason.CONVERSION, null);
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, witch1);
+                witch1.setPersistenceRequired();
                 this.releaseAllPois();
-            });
+                return true; // Paper start - Add EntityZapEvent
+            }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.LIGHTNING, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
             if (witch == null) {
                 super.thunderHit(p_35409_, p_35410_);
             }
@@ -802,7 +_,10 @@
     @Override
     public boolean wantsToPickUp(ServerLevel p_376823_, ItemStack p_35543_) {
         Item item = p_35543_.getItem();
-        return (p_35543_.is(ItemTags.VILLAGER_PICKS_UP) || this.getVillagerData().profession().value().requestedItems().contains(item))
+        // Neo: Allow Farmer Villagers to pick up SpecialPlantable items.
+        boolean isFarmerDesiredSeed = (p_35543_.getItem() instanceof net.neoforged.neoforge.common.SpecialPlantable specialPlantable && specialPlantable.villagerCanPlantItem(this))
+                && this.getVillagerData().profession().value().secondaryPoi().stream().anyMatch(secondaryPoi -> secondaryPoi.defaultBlockState().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS));
+        return (p_35543_.is(ItemTags.VILLAGER_PICKS_UP) || this.getVillagerData().profession().value().requestedItems().contains(item) || isFarmerDesiredSeed)
             && this.getInventory().canAddItem(p_35543_);
     }
 
@@ -820,30 +_,38 @@
     }
 
     public boolean hasFarmSeeds() {
-        return this.getInventory().hasAnyMatching(p_281096_ -> p_281096_.is(ItemTags.VILLAGER_PLANTABLE_SEEDS));
+        return this.getInventory().hasAnyMatching(stack -> this.level().purpurConfig.villagerClericsFarmWarts && this.getVillagerData().profession().is(VillagerProfession.CLERIC) ? stack.is(Items.NETHER_WART) : stack.is(ItemTags.VILLAGER_PLANTABLE_SEEDS)); // Purpur - Option for Villager Clerics to farm Nether Wart
     }
 
     @Override
     protected void updateTrades() {
-        VillagerData villagerdata = this.getVillagerData();
-        ResourceKey<VillagerProfession> resourcekey = villagerdata.profession().unwrapKey().orElse(null);
-        if (resourcekey != null) {
-            Int2ObjectMap<VillagerTrades.ItemListing[]> int2objectmap;
+        // Paper start - More vanilla friendly methods to update trades
+        this.updateTrades(TRADES_PER_LEVEL);
+    }
+
+    public boolean updateTrades(int amount) {
+        // Paper end - More vanilla friendly methods to update trades
+        VillagerData villagerData = this.getVillagerData();
+        ResourceKey<VillagerProfession> resourceKey = villagerData.profession().unwrapKey().orElse(null);
+        if (resourceKey != null) {
+            Int2ObjectMap<VillagerTrades.ItemListing[]> map1;
             if (this.level().enabledFeatures().contains(FeatureFlags.TRADE_REBALANCE)) {
-                Int2ObjectMap<VillagerTrades.ItemListing[]> int2objectmap1 = VillagerTrades.EXPERIMENTAL_TRADES.get(resourcekey);
-                int2objectmap = int2objectmap1 != null ? int2objectmap1 : VillagerTrades.TRADES.get(resourcekey);
+                Int2ObjectMap<VillagerTrades.ItemListing[]> map = VillagerTrades.EXPERIMENTAL_TRADES.get(resourceKey);
+                map1 = map != null ? map : VillagerTrades.TRADES.get(resourceKey);
             } else {
-                int2objectmap = VillagerTrades.TRADES.get(resourcekey);
+                map1 = VillagerTrades.TRADES.get(resourceKey);
             }
 
-            if (int2objectmap != null && !int2objectmap.isEmpty()) {
-                VillagerTrades.ItemListing[] avillagertrades$itemlisting = int2objectmap.get(villagerdata.level());
-                if (avillagertrades$itemlisting != null) {
-                    MerchantOffers merchantoffers = this.getOffers();
-                    this.addOffersFromItemListings(merchantoffers, avillagertrades$itemlisting, 2);
+            if (map1 != null && !map1.isEmpty()) {
+                VillagerTrades.ItemListing[] itemListings = map1.get(villagerData.level());
+                if (itemListings != null) {
+                    MerchantOffers offers = this.getOffers();
+                    this.addOffersFromItemListings(offers, itemListings, amount); // Paper - More vanilla friendly methods to update trades
+                    return true; // Paper - More vanilla friendly methods to update trades
                 }
             }
         }
+        return false; // Paper - More vanilla friendly methods to update trades
     }
 
     public void gossip(ServerLevel p_35412_, Villager p_35413_, long p_35414_) {
@@ -867,12 +_,13 @@
     }
 
     public void spawnGolemIfNeeded(ServerLevel p_35398_, long p_35399_, int p_35400_) {
+        if (p_35398_.purpurConfig.villagerSpawnIronGolemRadius > 0 && p_35398_.getEntitiesOfClass(net.minecraft.world.entity.animal.IronGolem.class, getBoundingBox().inflate(p_35398_.purpurConfig.villagerSpawnIronGolemRadius)).size() > p_35398_.purpurConfig.villagerSpawnIronGolemLimit) return; // Purpur - Implement configurable search radius for villagers to spawn iron golems
         if (this.wantsToSpawnGolem(p_35399_)) {
             AABB aabb = this.getBoundingBox().inflate(10.0, 10.0, 10.0);
             List<Villager> list = p_35398_.getEntitiesOfClass(Villager.class, aabb);
             List<Villager> list1 = list.stream().filter(p_186293_ -> p_186293_.wantsToSpawnGolem(p_35399_)).limit(5L).toList();
             if (list1.size() >= p_35400_) {
-                if (!SpawnUtil.trySpawnMob(
+                if (SpawnUtil.trySpawnMob( // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
                         EntityType.IRON_GOLEM,
                         EntitySpawnReason.MOB_SUMMONED,
                         p_35398_,
@@ -881,9 +_,11 @@
                         8,
                         6,
                         SpawnUtil.Strategy.LEGACY_IRON_GOLEM,
-                        false
+                        false,
+                        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE, // CraftBukkit,
+                        () -> {GolemSensor.golemDetected(this);} // Paper - Set Golem Last Seen to stop it from spawning another one
                     )
-                    .isEmpty()) {
+                    .isPresent()) { // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
                     list.forEach(GolemSensor::golemDetected);
                 }
             }
@@ -938,6 +_,12 @@
 
     @Override
     public void startSleeping(BlockPos p_35479_) {
+        // Purpur start - Option for beds to explode on villager sleep
+        if (level().purpurConfig.bedExplodeOnVillagerSleep && this.level().getBlockState(p_35479_).getBlock() instanceof net.minecraft.world.level.block.BedBlock) {
+            this.level().explode(null, (double) p_35479_.getX() + 0.5D, (double) p_35479_.getY() + 0.5D, (double) p_35479_.getZ() + 0.5D, (float) this.level().purpurConfig.bedExplosionPower, this.level().purpurConfig.bedExplosionFire, this.level().purpurConfig.bedExplosionEffect);
+            return;
+        }
+        // Purpur end - Option for beds to explode on villager sleep
         super.startSleeping(p_35479_);
         this.brain.setMemory(MemoryModuleType.LAST_SLEPT, this.level().getGameTime());
         this.brain.eraseMemory(MemoryModuleType.WALK_TARGET);
