--- a/net/minecraft/world/entity/monster/breeze/Breeze.java
+++ b/net/minecraft/world/entity/monster/breeze/Breeze.java
@@ -204,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.BREEZE_DEATH;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_312252_) {
+    public SoundEvent getHurtSound(DamageSource p_312252_) {
         return SoundEvents.BREEZE_HURT;
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return this.onGround() ? SoundEvents.BREEZE_IDLE_GROUND : SoundEvents.BREEZE_IDLE_AIR;
     }
 
@@ -235,6 +_,7 @@
     protected void customServerAiStep(ServerLevel p_376422_) {
         ProfilerFiller profilerfiller = Profiler.get();
         profilerfiller.push("breezeBrain");
+        if ((getRider() == null || !this.isControllable())) // Purpur - only use brain if no rider
         this.getBrain().tick(p_376422_, this);
         profilerfiller.popPush("breezeActivityUpdate");
         BreezeAi.updateActivity(this);
@@ -251,6 +_,7 @@
 
     @Override
     public boolean canAttackType(EntityType<?> p_312806_) {
+        if (this.getTarget() != null) return this.getTarget().getType() == p_312806_; // SPIGOT-7957: Allow attack if target from brain was set
         return p_312806_ == EntityType.PLAYER || p_312806_ == EntityType.IRON_GOLEM;
     }
 
