--- a/net/minecraft/world/level/chunk/ImposterProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ImposterProtoChunk.java
@@ -57,6 +_,13 @@
         return this.wrapped.getBlockState(p_62749_);
     }
 
+    // Paper start
+    @Override
+    public final BlockState getBlockState(final int x, final int y, final int z) {
+        return this.wrapped.getBlockStateFinal(x, y, z);
+    }
+    // Paper end
+
     @Override
     public FluidState getFluidState(BlockPos p_62736_) {
         return this.wrapped.getFluidState(p_62736_);
@@ -223,6 +_,10 @@
     }
 
     @Override
+    public void findBlocks(Predicate<BlockState> p_285465_, java.util.function.BiPredicate<BlockState, BlockPos> fineFilter, BiConsumer<BlockPos, BlockState> p_285030_) {
+        this.wrapped.findBlocks(p_285465_, fineFilter, p_285030_);
+    }
+
     public TickContainerAccess<Block> getBlockTicks() {
         return this.allowWrites ? this.wrapped.getBlockTicks() : BlackholeTickAccess.emptyContainer();
     }
@@ -290,5 +_,10 @@
     @Override
     public ChunkSkyLightSources getSkyLightSources() {
         return this.wrapped.getSkyLightSources();
+    }
+
+    @Override
+    public net.neoforged.neoforge.attachment.AttachmentHolder.AsField getAttachmentHolder() {
+        return wrapped.getAttachmentHolder();
     }
 }
