--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -88,16 +_,33 @@
         return new ChunkAccess.PackedTicks(this.blockTicks.pack(p_360967_), this.fluidTicks.pack(p_360967_));
     }
 
-    @Override
-    public BlockState getBlockState(BlockPos p_63264_) {
-        int i = p_63264_.getY();
-        if (this.isOutsideBuildHeight(i)) {
+    // Paper start - If loaded util
+    @Override
+    public final FluidState getFluidIfLoaded(BlockPos pos) {
+        return this.getFluidState(pos);
+    }
+
+    @Override
+    public final BlockState getBlockStateIfLoaded(BlockPos pos) {
+        return this.getBlockState(pos);
+    }
+    // Paper end
+
+    @Override
+    public BlockState getBlockState(BlockPos pos) {
+        // Paper start
+        return this.getBlockState(pos.getX(), pos.getY(), pos.getZ());
+    }
+
+    public BlockState getBlockState(final int x, final int y, final int z) {
+        // Paper end
+        if (this.isOutsideBuildHeight(y)) {
             return Blocks.VOID_AIR.defaultBlockState();
         } else {
-            LevelChunkSection levelchunksection = this.getSection(this.getSectionIndex(i));
-            return levelchunksection.hasOnlyAir()
-                ? Blocks.AIR.defaultBlockState()
-                : levelchunksection.getBlockState(p_63264_.getX() & 15, i & 15, p_63264_.getZ() & 15);
+            // Paper start
+            LevelChunkSection section = this.getSections()[this.getSectionIndex(y)];
+            return section.hasOnlyAir() ? Blocks.AIR.defaultBlockState() : section.getBlockState(x & 15, y & 15, z & 15);
+            // Paper end
         }
     }
 
@@ -139,8 +_,9 @@
                         this.lightEngine.updateSectionStatus(p_63217_, flag1);
                     }
 
-                    if (LightEngine.hasDifferentLightProperties(blockstate, p_63218_)) {
+                    if (LightEngine.hasDifferentLightProperties(this, p_63217_, blockstate, p_63218_)) {
                         this.skyLightSources.update(this, i1, j, k1);
+                        // Paper - rewrite chunk system
                         this.lightEngine.checkBlock(p_63217_);
                     }
                 }
