--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -61,7 +_,7 @@
     }
 
     @Override
-    protected float getSoundVolume() {
+    public float getSoundVolume() {
         return 0.1F;
     }
 
@@ -77,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_27451_) {
+    public SoundEvent getHurtSound(DamageSource p_27451_) {
         return SoundEvents.BAT_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.BAT_DEATH;
     }
 
     @Override
-    public boolean isPushable() {
+    public boolean isCollidable(boolean ignoreClimbing) { // Paper - Climbing should not bypass cramming gamerule
         return false;
     }
 
@@ -141,13 +_,13 @@
                     this.yHeadRot = this.random.nextInt(360);
                 }
 
-                if (p_376388_.getNearestPlayer(BAT_RESTING_TARGETING, this) != null) {
+                if (p_376388_.getNearestPlayer(BAT_RESTING_TARGETING, this) != null && org.bukkit.craftbukkit.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
                     this.setResting(false);
                     if (!flag) {
                         p_376388_.levelEvent(null, 1025, blockpos, 0);
                     }
                 }
-            } else {
+            } else if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
                 this.setResting(false);
                 if (!flag) {
                     p_376388_.levelEvent(null, 1025, blockpos, 0);
@@ -176,7 +_,7 @@
             float f1 = Mth.wrapDegrees(f - this.getYRot());
             this.zza = 0.5F;
             this.setYRot(this.getYRot() + f1);
-            if (this.random.nextInt(100) == 0 && p_376388_.getBlockState(blockpos1).isRedstoneConductor(p_376388_, blockpos1)) {
+            if (this.random.nextInt(100) == 0 && p_376388_.getBlockState(blockpos1).isRedstoneConductor(p_376388_, blockpos1) && org.bukkit.craftbukkit.event.CraftEventFactory.handleBatToggleSleepEvent(this, false)) { // CraftBukkit - Call BatToggleSleepEvent
                 this.setResting(true);
             }
         }
@@ -201,7 +_,7 @@
         if (this.isInvulnerableTo(p_376275_, p_376205_)) {
             return false;
         } else {
-            if (this.isResting()) {
+            if (this.isResting() && org.bukkit.craftbukkit.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
                 this.setResting(false);
             }
 
@@ -229,7 +_,7 @@
         } else {
             int i = p_218100_.getMaxLocalRawBrightness(p_218102_);
             int j = 4;
-            if (isHalloween()) {
+            if (Bat.isHalloweenSeason(p_218100_.getMinecraftWorld())) { // Purpur - Halloween options and optimizations
                 j = 7;
             } else if (p_218103_.nextBoolean()) {
                 return false;
@@ -245,6 +_,7 @@
         }
     }
 
+    public static boolean isHalloweenSeason(Level level) { return level.purpurConfig.forceHalloweenSeason || isHalloween(); } // Purpur - Halloween options and optimizations
     private static boolean isHalloween() {
         LocalDate localdate = LocalDate.now();
         int i = localdate.get(ChronoField.DAY_OF_MONTH);
