--- a/net/minecraft/world/level/block/CakeBlock.java
+++ b/net/minecraft/world/level/block/CakeBlock.java
@@ -57,6 +_,12 @@
     ) {
         Item item = p_316238_.getItem();
         if (p_316238_.is(ItemTags.CANDLES) && p_316837_.getValue(BITES) == 0 && Block.byItem(item) instanceof CandleBlock candleblock) {
+            // Paper start - call change block event
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(p_316853_, p_316227_, CandleCakeBlock.byCandle(candleblock))) {
+                p_316853_.containerMenu.forceHeldSlot(p_316422_); // update inv because candle could decrease
+                return InteractionResult.TRY_WITH_EMPTY_HAND;
+            }
+            // Paper end - call change block event
             p_316238_.consume(1, p_316853_);
             p_316766_.playSound(null, p_316227_, SoundEvents.CAKE_ADD_CANDLE, SoundSource.BLOCKS, 1.0F, 1.0F);
             p_316766_.setBlockAndUpdate(p_316227_, CandleCakeBlock.byCandle(candleblock));
@@ -83,19 +_,38 @@
         return eat(p_316406_, p_316218_, p_316481_, p_316212_);
     }
 
-    protected static InteractionResult eat(LevelAccessor p_51186_, BlockPos p_51187_, BlockState p_51188_, Player p_51189_) {
-        if (!p_51189_.canEat(false)) {
+    protected static InteractionResult eat(LevelAccessor level, BlockPos pos, BlockState state, Player player) {
+        if (!player.canEat(false)) {
             return InteractionResult.PASS;
         } else {
-            p_51189_.awardStat(Stats.EAT_CAKE_SLICE);
-            p_51189_.getFoodData().eat(2, 0.1F);
-            int i = p_51188_.getValue(BITES);
-            p_51186_.gameEvent(p_51189_, GameEvent.EAT, p_51187_);
-            if (i < 6) {
-                p_51186_.setBlock(p_51187_, p_51188_.setValue(BITES, i + 1), 3);
+            // Paper start - call change block event
+            int bitesValue = state.getValue(CakeBlock.BITES);
+            final BlockState newState = bitesValue < MAX_BITES ? state.setValue(CakeBlock.BITES, bitesValue + 1) : level.getFluidState(pos).createLegacyBlock();
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, newState)) {
+                ((net.minecraft.server.level.ServerPlayer) player).getBukkitEntity().sendHealthUpdate();
+                return InteractionResult.PASS; // return a non-consume result to cake blocks don't drop their candles
+            }
+            // Paper end - call change block event
+            player.awardStat(Stats.EAT_CAKE_SLICE);
+            // CraftBukkit start
+            // player.getFoodData().eat(2, 0.1F);
+            int oldFoodLevel = player.getFoodData().foodLevel;
+
+            org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFoodLevelChangeEvent(player, 2 + oldFoodLevel);
+
+            if (!event.isCancelled()) {
+                player.getFoodData().eat(event.getFoodLevel() - oldFoodLevel, 0.1F);
+            }
+
+            ((net.minecraft.server.level.ServerPlayer) player).getBukkitEntity().sendHealthUpdate();
+            // CraftBukkit end
+            // Paper - move up
+            level.gameEvent(player, GameEvent.EAT, pos);
+            if (bitesValue < 6) {
+                level.setBlock(pos, state.setValue(BITES, bitesValue + 1), 3);
             } else {
-                p_51186_.removeBlock(p_51187_, false);
-                p_51186_.gameEvent(p_51189_, GameEvent.BLOCK_DESTROY, p_51187_);
+                level.removeBlock(pos, false);
+                level.gameEvent(player, GameEvent.BLOCK_DESTROY, pos);
             }
 
             return InteractionResult.SUCCESS;
