--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -47,6 +_,8 @@
     public static final int MIN_LEVEL = 0;
     public static final int MAX_LEVEL = 7;
     public static final IntegerProperty LEVEL = BlockStateProperties.LEVEL_COMPOSTER;
+    /** @deprecated Neo: Use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#COMPOSTABLES compostable} data map instead */
+    @Deprecated
     public static final Object2FloatMap<ItemLike> COMPOSTABLES = new Object2FloatOpenHashMap<>();
     private static final int HOLE_WIDTH = 12;
     private static final VoxelShape[] SHAPES = Util.make(
@@ -242,6 +_,15 @@
         if (p_51978_.getValue(LEVEL) == 7) {
             p_51979_.scheduleTick(p_51980_, p_51978_.getBlock(), 20);
         }
+        // Neo: Invalidate composter capabilities when a composter is added
+        if (!p_51981_.is(this)) p_51979_.invalidateCapabilities(p_51980_);
+    }
+
+    @Override
+    protected void affectNeighborsAfterRemoval(BlockState p_394424_, ServerLevel p_394241_, BlockPos p_393520_, boolean p_394545_) {
+        super.affectNeighborsAfterRemoval(p_394424_, p_394241_, p_393520_, p_394545_);
+        // Neo: Invalidate composter capabilities when a composter is removed
+        if (!p_394241_.getBlockState(p_393520_).is(p_394424_.getBlock())) p_394241_.invalidateCapabilities(p_393520_);
     }
 
     @Override
@@ -249,13 +_,33 @@
         ItemStack p_316332_, BlockState p_316118_, Level p_316624_, BlockPos p_316660_, Player p_316715_, InteractionHand p_316472_, BlockHitResult p_316606_
     ) {
         int i = p_316118_.getValue(LEVEL);
-        if (i < 8 && COMPOSTABLES.containsKey(p_316332_.getItem())) {
-            if (i < 7 && !p_316624_.isClientSide) {
-                BlockState blockstate = addItem(p_316715_, p_316118_, p_316624_, p_316660_, p_316332_);
-                p_316624_.levelEvent(1500, p_316660_, p_316118_ != blockstate ? 1 : 0);
-                p_316715_.awardStat(Stats.ITEM_USED.get(p_316332_.getItem()));
-                p_316332_.consume(1, p_316715_);
-            }
+        if (i < 8 && getValue(p_316332_) > 0) {
+            // Purpur start - Sneak to bulk process composter
+            BlockState blockstate = process(i, p_316715_, p_316118_, p_316624_, p_316660_, p_316332_);
+            if (blockstate == null) {
+                return InteractionResult.PASS;
+            }
+
+            if (p_316624_.purpurConfig.composterBulkProcess && p_316715_.isShiftKeyDown() && blockstate != p_316118_) {
+                BlockState blockstate1;
+                int j;
+                int k;
+                int l;
+                int i1;
+                do {
+                    blockstate1 = blockstate;
+                    j = p_316332_.getCount();
+                    l = blockstate1.getValue(LEVEL);
+                    blockstate = process(l, p_316715_, blockstate1, p_316624_, p_316660_, p_316332_);
+                    if (blockstate == null) {
+                        return InteractionResult.PASS;
+                    }
+
+                    k = p_316332_.getCount();
+                    i1 = blockstate.getValue(LEVEL);
+                } while (k > 0 && (k != j || i1 != l || blockstate != blockstate1));
+            }
+            // Purpur end - Sneak to bulk process composter
 
             return InteractionResult.SUCCESS;
         } else {
@@ -263,6 +_,22 @@
         }
     }
 
+    // Purpur start - Sneak to bulk process composter
+    @Nullable
+    private static BlockState process(int levelValue, Player player, BlockState state, Level level, BlockPos pos, ItemStack stack) {
+        if (levelValue < 7 && !level.isClientSide) {
+            BlockState blockstate = addItem(player, state, level, pos, stack);
+            if (blockstate == null) return null; // Paper - Add CompostItemEvent and EntityCompostItemEvent
+            level.levelEvent(1500, pos, state != blockstate ? 1 : 0);
+            player.awardStat(Stats.ITEM_USED.get(stack.getItem()));
+            stack.consume(1, player);
+            return blockstate;
+        }
+
+        return state;
+    }
+    // Purpur end - Sneak to bulk process composter
+
     @Override
     protected InteractionResult useWithoutItem(BlockState p_316361_, Level p_316271_, BlockPos p_316647_, Player p_316633_, BlockHitResult p_316555_) {
         int i = p_316361_.getValue(LEVEL);
@@ -276,8 +_,9 @@
 
     public static BlockState insertItem(Entity p_270919_, BlockState p_270087_, ServerLevel p_270284_, ItemStack p_270253_, BlockPos p_270678_) {
         int i = p_270087_.getValue(LEVEL);
-        if (i < 7 && COMPOSTABLES.containsKey(p_270253_.getItem())) {
+        if (i < 7 && getValue(p_270253_) > 0) {
             BlockState blockstate = addItem(p_270919_, p_270087_, p_270284_, p_270678_, p_270253_);
+            if (blockstate == null) return p_270087_; // Paper - Add CompostItemEvent and EntityCompostItemEvent
             p_270253_.shrink(1);
             return blockstate;
         } else {
@@ -286,6 +_,14 @@
     }
 
     public static BlockState extractProduce(Entity p_270467_, BlockState p_51999_, Level p_52000_, BlockPos p_52001_) {
+        // CraftBukkit start
+        if (p_270467_ != null && !(p_270467_ instanceof Player)) {
+            BlockState emptyState = ComposterBlock.empty(p_270467_, p_51999_, org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE, p_52001_);
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(p_270467_, p_52001_, emptyState)) {
+                return p_51999_;
+            }
+        }
+        // CraftBukkit end
         if (!p_52000_.isClientSide) {
             Vec3 vec3 = Vec3.atLowerCornerWithOffset(p_52001_, 0.5, 1.01, 0.5).offsetRandom(p_52000_.random, 0.7F);
             ItemEntity itementity = new ItemEntity(p_52000_, vec3.x(), vec3.y(), vec3.z(), new ItemStack(Items.BONE_MEAL));
@@ -305,14 +_,40 @@
         return blockstate;
     }
 
+    @Nullable // Paper
     static BlockState addItem(@Nullable Entity p_270464_, BlockState p_270603_, LevelAccessor p_270151_, BlockPos p_270547_, ItemStack p_270354_) {
+        // CraftBukkit start
+        return ComposterBlock.addItem(p_270464_, p_270603_, p_270151_, p_270547_, p_270354_, p_270151_.getRandom().nextDouble());
+    }
+
+    @Nullable // Paper - make it nullable
+    static BlockState addItem(@Nullable Entity p_270464_, BlockState p_270603_, LevelAccessor p_270151_, BlockPos p_270547_, ItemStack p_270354_, double rand) {
         int i = p_270603_.getValue(LEVEL);
-        float f = COMPOSTABLES.getFloat(p_270354_.getItem());
-        if ((i != 0 || !(f > 0.0F)) && !(p_270151_.getRandom().nextDouble() < f)) {
+        float f = getValue(p_270354_);
+        // Paper start - Add CompostItemEvent and EntityCompostItemEvent
+        boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f);
+        final io.papermc.paper.event.block.CompostItemEvent event;
+        if (p_270464_ == null) {
+            event = new io.papermc.paper.event.block.CompostItemEvent(org.bukkit.craftbukkit.block.CraftBlock.at(p_270151_, p_270547_), p_270354_.getBukkitStack(), willRaiseLevel);
+        } else {
+            event = new io.papermc.paper.event.entity.EntityCompostItemEvent(p_270464_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_270151_, p_270547_), p_270354_.getBukkitStack(), willRaiseLevel);
+        }
+        if (!event.callEvent()) { // check for cancellation of entity event (non entity event can't be cancelled cause of hoppers)
+            return null;
+        }
+        willRaiseLevel = event.willRaiseLevel();
+
+        if (!willRaiseLevel) {
+            // Paper end - Add CompostItemEvent and EntityCompostItemEvent
             return p_270603_;
         } else {
             int j = i + 1;
             BlockState blockstate = p_270603_.setValue(LEVEL, j);
+            // Paper start - move the EntityChangeBlockEvent here to avoid conflict later for the compost events
+            if (p_270464_ != null && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(p_270464_, p_270547_, blockstate)) {
+                return null;
+            }
+            // Paper end
             p_270151_.setBlock(p_270547_, blockstate, 3);
             p_270151_.gameEvent(GameEvent.BLOCK_CHANGE, p_270547_, GameEvent.Context.of(p_270464_, blockstate));
             if (j == 7) {
@@ -357,13 +_,14 @@
         if (i == 8) {
             return new ComposterBlock.OutputContainer(p_51956_, p_51957_, p_51958_, new ItemStack(Items.BONE_MEAL));
         } else {
-            return (WorldlyContainer)(i < 7 ? new ComposterBlock.InputContainer(p_51956_, p_51957_, p_51958_) : new ComposterBlock.EmptyContainer());
+            return (WorldlyContainer)(i < 7 ? new ComposterBlock.InputContainer(p_51956_, p_51957_, p_51958_) : new ComposterBlock.EmptyContainer(p_51957_, p_51958_)); // CraftBukkit - empty generatoraccess, blockposition
         }
     }
 
     public static class EmptyContainer extends SimpleContainer implements WorldlyContainer {
-        public EmptyContainer() {
+        public EmptyContainer(LevelAccessor levelAccessor, BlockPos blockPos) { // CraftBukkit
             super(0);
+            this.bukkitOwner = new org.bukkit.craftbukkit.inventory.CraftBlockInventoryHolder(levelAccessor, blockPos, this); // CraftBukkit
         }
 
         @Override
@@ -390,6 +_,7 @@
 
         public InputContainer(BlockState p_52022_, LevelAccessor p_52023_, BlockPos p_52024_) {
             super(1);
+            this.bukkitOwner = new org.bukkit.craftbukkit.inventory.CraftBlockInventoryHolder(p_52023_, p_52024_, this); // CraftBukkit
             this.state = p_52022_;
             this.level = p_52023_;
             this.pos = p_52024_;
@@ -407,7 +_,7 @@
 
         @Override
         public boolean canPlaceItemThroughFace(int p_52028_, ItemStack p_52029_, @Nullable Direction p_52030_) {
-            return !this.changed && p_52030_ == Direction.UP && ComposterBlock.COMPOSTABLES.containsKey(p_52029_.getItem());
+            return !this.changed && p_52030_ == Direction.UP && getValue(p_52029_) > 0f;
         }
 
         @Override
@@ -421,6 +_,11 @@
             if (!itemstack.isEmpty()) {
                 this.changed = true;
                 BlockState blockstate = ComposterBlock.addItem(null, this.state, this.level, this.pos, itemstack);
+                // Paper start - Add CompostItemEvent and EntityCompostItemEvent
+                if (blockstate == null) {
+                    return;
+                }
+                // Paper end - Add CompostItemEvent and EntityCompostItemEvent
                 this.level.levelEvent(1500, this.pos, blockstate != this.state ? 1 : 0);
                 this.removeItemNoUpdate(0);
             }
@@ -435,6 +_,7 @@
 
         public OutputContainer(BlockState p_52042_, LevelAccessor p_52043_, BlockPos p_52044_, ItemStack p_52045_) {
             super(p_52045_);
+            this.bukkitOwner = new org.bukkit.craftbukkit.inventory.CraftBlockInventoryHolder(p_52043_, p_52044_, this); // Paper
             this.state = p_52042_;
             this.level = p_52043_;
             this.pos = p_52044_;
@@ -465,5 +_,11 @@
             ComposterBlock.empty(null, this.state, this.level, this.pos);
             this.changed = true;
         }
+    }
+
+    public static float getValue(ItemStack item) {
+        var value = item.getItemHolder().getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.COMPOSTABLES);
+        if (value != null) return value.chance();
+        return -1f;
     }
 }
