--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -95,28 +_,68 @@
     @Override
     protected void randomTick(BlockState p_221139_, ServerLevel p_221140_, BlockPos p_221141_, RandomSource p_221142_) {
         int i = p_221139_.getValue(MOISTURE);
+        // Paper start - Configurable random tick rates for blocks
+        if (i > 0 && p_221140_.paperConfig().tickRates.wetFarmland != 1 && (p_221140_.paperConfig().tickRates.wetFarmland < 1 || (net.minecraft.server.MinecraftServer.currentTick + p_221141_.hashCode()) % p_221140_.paperConfig().tickRates.wetFarmland != 0)) { return; }
+        if (i == 0 && p_221140_.paperConfig().tickRates.dryFarmland != 1 && (p_221140_.paperConfig().tickRates.dryFarmland < 1 || (net.minecraft.server.MinecraftServer.currentTick + p_221141_.hashCode()) % p_221140_.paperConfig().tickRates.dryFarmland != 0)) { return; }
+        // Paper end - Configurable random tick rates for blocks
         if (!isNearWater(p_221140_, p_221141_) && !p_221140_.isRainingAt(p_221141_.above())) {
             if (i > 0) {
-                p_221140_.setBlock(p_221141_, p_221139_.setValue(MOISTURE, i - 1), 2);
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(p_221140_, p_221141_, p_221139_.setValue(MOISTURE, i - 1), 2); // CraftBukkit
             } else if (!shouldMaintainFarmland(p_221140_, p_221141_)) {
                 turnToDirt(null, p_221139_, p_221140_, p_221141_);
             }
         } else if (i < 7) {
-            p_221140_.setBlock(p_221141_, p_221139_.setValue(MOISTURE, 7), 2);
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(p_221140_, p_221141_, p_221139_.setValue(MOISTURE, 7), 2); // CraftBukkit
         }
     }
 
     @Override
     public void fallOn(Level p_153227_, BlockState p_153228_, BlockPos p_153229_, Entity p_153230_, double p_397639_) {
+        super.fallOn(p_153227_, p_153228_, p_153229_, p_153230_, p_397639_); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
+        if (com.mohistmc.youer.feature.ban.bans.BanEvents.banFarmlandTrample()) return; // Youer - after super.fallOn, so banning trample does not also cancel fall damage
         if (p_153227_ instanceof ServerLevel serverlevel
-            && p_153227_.random.nextFloat() < p_397639_ - 0.5
-            && p_153230_ instanceof LivingEntity
-            && (p_153230_ instanceof Player || serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))
-            && p_153230_.getBbWidth() * p_153230_.getBbWidth() * p_153230_.getBbHeight() > 0.512F) {
+            && net.neoforged.neoforge.common.CommonHooks.onFarmlandTrample(serverlevel, p_153229_, Blocks.DIRT.defaultBlockState(), p_397639_, p_153230_)) { // Forge: Move logic to Entity#canTrample
+            // CraftBukkit start - Interact soil
+            org.bukkit.event.Cancellable cancellable;
+            if (p_153230_ instanceof Player) {
+                cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent((Player) p_153230_, org.bukkit.event.block.Action.PHYSICAL, p_153229_, null, null, null);
+            } else {
+                cancellable = new org.bukkit.event.entity.EntityInteractEvent(p_153230_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_153227_, p_153229_));
+                p_153227_.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+            }
+
+            if (cancellable.isCancelled()) {
+                return;
+            }
+
+            if (p_153227_.purpurConfig.farmlandTramplingDisabled) return; // Purpur - Farmland trampling changes
+            if (p_153227_.purpurConfig.farmlandTramplingOnlyPlayers && !(p_153230_ instanceof Player)) return; // Purpur - Farmland trampling changes
+
+            // Purpur start - Ability to re-add farmland mechanics from Alpha
+            if (p_153227_.purpurConfig.farmlandAlpha) {
+                Block block = p_153227_.getBlockState(p_153229_.below()).getBlock();
+                if (block instanceof FenceBlock || block instanceof WallBlock) {
+                    return;
+                }
+            }
+            // Purpur end - Ability to re-add farmland mechanics from Alpha
+
+            // Purpur start - Farmland trampling changes
+            if (p_153227_.purpurConfig.farmlandTramplingFeatherFalling) {
+                net.minecraft.world.item.ItemStack bootsItem = ((net.minecraft.world.entity.LivingEntity) p_153230_).getItemBySlot(net.minecraft.world.entity.EquipmentSlot.FEET);
+
+                if (bootsItem != net.minecraft.world.item.ItemStack.EMPTY && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FEATHER_FALLING, bootsItem) >= (int) p_153230_.fallDistance) {
+                    return;
+                }
+            }
+            // Purpur end - Farmland trampling changes
+
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(p_153230_, p_153229_, Blocks.DIRT.defaultBlockState())) {
+                return;
+            }
+            // CraftBukkit end
             turnToDirt(p_153230_, p_153228_, p_153227_, p_153229_);
         }
-
-        super.fallOn(p_153227_, p_153228_, p_153229_, p_153230_, p_397639_);
     }
 
     public static void turnToDirt(@Nullable Entity p_270981_, BlockState p_270402_, Level p_270568_, BlockPos p_270551_) {
@@ -130,13 +_,16 @@
     }
 
     private static boolean isNearWater(LevelReader p_53259_, BlockPos p_53260_) {
+        BlockState state = p_53259_.getBlockState(p_53260_);
         for (BlockPos blockpos : BlockPos.betweenClosed(p_53260_.offset(-4, 0, -4), p_53260_.offset(4, 1, 4))) {
-            if (p_53259_.getFluidState(blockpos).is(FluidTags.WATER)) {
+            if (state.canBeHydrated(p_53259_, p_53260_, p_53259_.getFluidState(blockpos), blockpos)) {
                 return true;
             }
         }
 
-        return false;
+        return net.neoforged.neoforge.common.FarmlandWaterManager.hasBlockWaterTicket(p_53259_, p_53260_)
+            || ((ServerLevel) p_53259_).purpurConfig.farmlandGetsMoistFromBelow && p_53259_.getFluidState(p_53260_.relative(Direction.DOWN)).is(FluidTags.WATER); // Purpur - Allow soil to moisten from water directly under it
+        // Paper end - Perf: remove abstract block iteration
     }
 
     @Override
