--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -32,34 +_,54 @@
 
     @Override
     public BlockState getStateForPlacement(RandomSource p_374460_) {
-        return this.defaultBlockState().setValue(AGE, p_374460_.nextInt(25));
+        return this.defaultBlockState().setValue(AGE, getMaxGrowthAge() == 0 ? 0 : p_374460_.nextInt(getMaxGrowthAge())); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     @Override
     protected boolean isRandomlyTicking(BlockState p_53961_) {
-        return p_53961_.getValue(AGE) < 25;
+        return p_53961_.getValue(AGE) < getMaxGrowthAge(); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     @Override
     protected void randomTick(BlockState p_221350_, ServerLevel p_221351_, BlockPos p_221352_, RandomSource p_221353_) {
-        if (p_221350_.getValue(AGE) < 25 && p_221353_.nextDouble() < this.growPerTickProbability) {
+        // Spigot start
+        int modifier = 100;
+        if (this == Blocks.KELP) {
+            modifier = p_221351_.spigotConfig.kelpModifier;
+        } else if (this == Blocks.TWISTING_VINES) {
+            modifier = p_221351_.spigotConfig.twistingVinesModifier;
+        } else if (this == Blocks.WEEPING_VINES) {
+            modifier = p_221351_.spigotConfig.weepingVinesModifier;
+        } else if (this == Blocks.CAVE_VINES) {
+            modifier = p_221351_.spigotConfig.caveVinesModifier;
+        }
+        if (p_221350_.getValue(AGE) < 25 && net.neoforged.neoforge.common.CommonHooks.canCropGrow(p_221351_, p_221352_.relative(this.growthDirection), p_221350_, p_221353_.nextDouble() < ((modifier / 100.0D) * this.growPerTickProbability))) { // Spigot - SPIGOT-7159: Better modifier resolution
+            // Spigot end
             BlockPos blockpos = p_221352_.relative(this.growthDirection);
             if (this.canGrowInto(p_221351_.getBlockState(blockpos))) {
-                p_221351_.setBlockAndUpdate(blockpos, this.getGrowIntoState(p_221350_, p_221351_.random));
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(p_221351_, p_221352_, blockpos, this.getGrowIntoState(p_221350_, p_221351_.random, p_221351_), 3)) { // CraftBukkit // Paper - Fix Spigot growth modifiers
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(p_221351_, blockpos, p_221351_.getBlockState(blockpos));
+                }
             }
         }
     }
+
+    // Paper start - Fix Spigot growth modifiers
+    protected BlockState getGrowIntoState(BlockState state, RandomSource random, @javax.annotation.Nullable Level level) {
+        return this.getGrowIntoState(state, random);
+    }
+    // Paper end - Fix Spigot growth modifiers
 
     protected BlockState getGrowIntoState(BlockState p_221347_, RandomSource p_221348_) {
         return p_221347_.cycle(AGE);
     }
 
     public BlockState getMaxAgeState(BlockState p_187439_) {
-        return p_187439_.setValue(AGE, 25);
+        return p_187439_.setValue(AGE, getMaxGrowthAge()); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     public boolean isMaxAge(BlockState p_187441_) {
-        return p_187441_.getValue(AGE) == 25;
+        return p_187441_.getValue(AGE) >= getMaxGrowthAge(); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
     }
 
     protected BlockState updateBodyAfterConvertedFromHead(BlockState p_153329_, BlockState p_153330_) {
@@ -117,13 +_,13 @@
     @Override
     public void performBonemeal(ServerLevel p_221337_, RandomSource p_221338_, BlockPos p_221339_, BlockState p_221340_) {
         BlockPos blockpos = p_221339_.relative(this.growthDirection);
-        int i = Math.min(p_221340_.getValue(AGE) + 1, 25);
+        int i = Math.min(p_221340_.getValue(AGE) + 1, getMaxGrowthAge()); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
         int j = this.getBlocksToGrowWhenBonemealed(p_221338_);
 
         for (int k = 0; k < j && this.canGrowInto(p_221337_.getBlockState(blockpos)); k++) {
             p_221337_.setBlockAndUpdate(blockpos, p_221340_.setValue(AGE, i));
             blockpos = blockpos.relative(this.growthDirection);
-            i = Math.min(i + 1, 25);
+            i = Math.min(i + 1, getMaxGrowthAge()); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
         }
     }
 
@@ -135,4 +_,6 @@
     protected GrowingPlantHeadBlock getHeadBlock() {
         return this;
     }
+
+    public abstract int getMaxGrowthAge(); // Purpur - kelp, cave, weeping, and twisting configurable max growth age
 }
