--- a/net/minecraft/world/level/block/MangrovePropaguleBlock.java
+++ b/net/minecraft/world/level/block/MangrovePropaguleBlock.java
@@ -73,7 +_,12 @@
 
     @Override
     protected boolean canSurvive(BlockState p_221473_, LevelReader p_221474_, BlockPos p_221475_) {
-        return isHanging(p_221473_) ? p_221474_.getBlockState(p_221475_.above()).is(Blocks.MANGROVE_LEAVES) : super.canSurvive(p_221473_, p_221474_, p_221475_);
+        if (isHanging(p_221473_)) {
+            var soilDecision = p_221474_.getBlockState(p_221475_.above()).canSustainPlant(p_221474_, p_221475_.above(), Direction.DOWN, p_221473_);
+            if (!soilDecision.isDefault()) return soilDecision.isTrue();
+            return p_221474_.getBlockState(p_221475_.above()).is(Blocks.MANGROVE_LEAVES);
+        }
+        return super.canSurvive(p_221473_, p_221474_, p_221475_);
     }
 
     @Override
@@ -104,7 +_,7 @@
     @Override
     protected void randomTick(BlockState p_221488_, ServerLevel p_221489_, BlockPos p_221490_, RandomSource p_221491_) {
         if (!isHanging(p_221488_)) {
-            if (p_221491_.nextInt(7) == 0) {
+            if (p_221491_.nextFloat() < (p_221489_.spigotConfig.saplingModifier / (100.0F * 7))) { // Paper - Fix Spigot growth modifiers
                 this.advanceTree(p_221489_, p_221490_, p_221488_, p_221491_);
             }
         } else {
