--- a/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -60,6 +_,12 @@
 
     @Override
     protected void entityInside(BlockState p_154263_, Level p_154264_, BlockPos p_154265_, Entity p_154266_, InsideBlockEffectApplier p_405853_) {
+        // Paper start - Add EntityInsideBlockEvent
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_154266_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_154264_, p_154265_)).callEvent()) {
+            p_154266_.getEntityData().markDirty(net.minecraft.world.entity.Entity.DATA_TICKS_FROZEN); // prevent desync on cancelled event
+            return;
+        }
+        // Paper end - Add EntityInsideBlockEvent
         if (!(p_154266_ instanceof LivingEntity) || p_154266_.getInBlockState().is(this)) {
             p_154266_.makeStuckInBlock(p_154263_, new Vec3(0.9F, 1.5, 0.9F));
             if (p_154264_.isClientSide) {
@@ -85,8 +_,13 @@
             p_405326_ -> {
                 if (p_154264_ instanceof ServerLevel serverlevel
                     && p_405326_.isOnFire()
-                    && (serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) || p_405326_ instanceof Player)
+                        // CraftBukkit - move down
                     && p_405326_.mayInteract(serverlevel, blockpos)) {
+                    // CraftBukkit start
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(p_405326_, p_154265_, Blocks.AIR.defaultBlockState(), !(serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING, serverlevel.purpurConfig.powderSnowMobGriefingOverride) || p_405326_ instanceof Player))) { // Purpur - Add mobGriefing override to everything affected
+                        return;
+                    }
+                    // CraftBukkit end
                     p_154264_.destroyBlock(blockpos, false);
                 }
             }
@@ -138,7 +_,7 @@
         if (p_154256_.getType().is(EntityTypeTags.POWDER_SNOW_WALKABLE_MOBS)) {
             return true;
         } else {
-            return p_154256_ instanceof LivingEntity ? ((LivingEntity)p_154256_).getItemBySlot(EquipmentSlot.FEET).is(Items.LEATHER_BOOTS) : false;
+            return p_154256_ instanceof LivingEntity ? ((LivingEntity)p_154256_).getItemBySlot(EquipmentSlot.FEET).canWalkOnPowderedSnow((LivingEntity)p_154256_) : false;
         }
     }
 
