--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -19,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class SugarCaneBlock extends Block {
+public class SugarCaneBlock extends Block implements BonemealableBlock { // Purpur - bonemealable sugarcane
     public static final MapCodec<SugarCaneBlock> CODEC = simpleCodec(SugarCaneBlock::new);
     public static final IntegerProperty AGE = BlockStateProperties.AGE_15;
     private static final VoxelShape SHAPE = Block.column(12.0, 0.0, 16.0);
@@ -47,21 +_,26 @@
     }
 
     @Override
-    protected void randomTick(BlockState p_222548_, ServerLevel p_222549_, BlockPos p_222550_, RandomSource p_222551_) {
-        if (p_222549_.isEmptyBlock(p_222550_.above())) {
+    protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (level.isEmptyBlock(pos.above())) {
             int i = 1;
 
-            while (p_222549_.getBlockState(p_222550_.below(i)).is(this)) {
+            while (level.getBlockState(pos.below(i)).is(this)) {
                 i++;
             }
 
-            if (i < 3) {
-                int j = p_222548_.getValue(AGE);
-                if (j == 15) {
-                    p_222549_.setBlockAndUpdate(p_222550_.above(), this.defaultBlockState());
-                    p_222549_.setBlock(p_222550_, p_222548_.setValue(AGE, 0), 260);
-                } else {
-                    p_222549_.setBlock(p_222550_, p_222548_.setValue(AGE, j + 1), 260);
+            if (i < level.paperConfig().maxGrowthHeight.reeds) { // Paper - Configurable cactus/bamboo/reed growth height
+                int ageValue = state.getValue(AGE);
+                int modifier = level.spigotConfig.caneModifier; // Spigot - SPIGOT-7159: Better modifier resolution
+                if (net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, true)) {
+                if (ageValue >= 15 || (modifier != 100 && random.nextFloat() < (modifier / (100.0F * 16)))) { // Spigot - SPIGOT-7159: Better modifier resolution
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos.above(), this.defaultBlockState(), 3)) { // CraftBukkit
+                        net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos.above(), this.defaultBlockState());
+                    }
+                    level.setBlock(pos, state.setValue(AGE, 0), 260);
+                } else if (modifier == 100 || random.nextFloat() < (modifier / (100.0F * 16))) { // Spigot - SPIGOT-7159: Better modifier resolution
+                    level.setBlock(pos, state.setValue(AGE, ageValue + 1), 260);
+                }
                 }
             }
         }
@@ -91,13 +_,15 @@
         if (blockstate.is(this)) {
             return true;
         } else {
+            var soilDecision = blockstate.canSustainPlant(p_57176_, p_57177_.below(), Direction.UP, p_57175_);
+            if (!soilDecision.isDefault()) return soilDecision.isTrue();
             if (blockstate.is(BlockTags.DIRT) || blockstate.is(BlockTags.SAND)) {
                 BlockPos blockpos = p_57177_.below();
 
                 for (Direction direction : Direction.Plane.HORIZONTAL) {
                     BlockState blockstate1 = p_57176_.getBlockState(blockpos.relative(direction));
                     FluidState fluidstate = p_57176_.getFluidState(blockpos.relative(direction));
-                    if (fluidstate.is(FluidTags.WATER) || blockstate1.is(Blocks.FROSTED_ICE)) {
+                    if (p_57175_.canBeHydrated(p_57176_, p_57177_, fluidstate, blockpos.relative(direction)) || blockstate1.is(Blocks.FROSTED_ICE)) {
                         return true;
                     }
                 }
@@ -111,4 +_,34 @@
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> p_57186_) {
         p_57186_.add(AGE);
     }
+
+    // Purpur start - bonemealable sugarcane
+    @Override
+    public boolean isValidBonemealTarget(final LevelReader world, final BlockPos pos, final BlockState state) {
+        if (!((net.minecraft.world.level.Level) world).purpurConfig.sugarCanAffectedByBonemeal || !world.isEmptyBlock(pos.above())) return false;
+
+        int reedHeight = 0;
+        while (world.getBlockState(pos.below(reedHeight)).is(this)) {
+            reedHeight++;
+        }
+
+        return reedHeight < ((net.minecraft.world.level.Level) world).paperConfig().maxGrowthHeight.reeds;
+    }
+
+    @Override
+    public boolean isBonemealSuccess(net.minecraft.world.level.Level world, RandomSource random, BlockPos pos, BlockState state) {
+        return true;
+    }
+
+    @Override
+    public void performBonemeal(ServerLevel world, RandomSource random, BlockPos pos, BlockState state) {
+        int reedHeight = 0;
+        while (world.getBlockState(pos.below(reedHeight)).is(this)) {
+            reedHeight++;
+        }
+        for (int i = 0; i <= world.paperConfig().maxGrowthHeight.reeds - reedHeight; i++) {
+            world.setBlockAndUpdate(pos.above(i), state.setValue(SugarCaneBlock.AGE, 0));
+        }
+    }
+    // Purpur end - bonemealable sugarcane
 }
