--- a/net/minecraft/world/level/block/VegetationBlock.java
+++ b/net/minecraft/world/level/block/VegetationBlock.java
@@ -21,7 +_,7 @@
     protected abstract MapCodec<? extends VegetationBlock> codec();
 
     protected boolean mayPlaceOn(BlockState p_401274_, BlockGetter p_401010_, BlockPos p_401365_) {
-        return p_401274_.is(BlockTags.DIRT) || p_401274_.is(Blocks.FARMLAND);
+        return p_401274_.is(BlockTags.DIRT) || p_401274_.getBlock() instanceof net.minecraft.world.level.block.FarmBlock;
     }
 
     @Override
@@ -35,15 +_,24 @@
         BlockState p_401336_,
         RandomSource p_401169_
     ) {
-        return !p_401118_.canSurvive(p_401198_, p_401142_)
-            ? Blocks.AIR.defaultBlockState()
-            : super.updateShape(p_401118_, p_401198_, p_401107_, p_401142_, p_401236_, p_401082_, p_401336_, p_401169_);
+        // CraftBukkit start
+        if (!p_401118_.canSurvive(p_401198_, p_401142_)) {
+            // Suppress during worldgen
+            if (!(p_401198_ instanceof net.minecraft.server.level.ServerLevel serverLevel && serverLevel.hasPhysicsEvent) || !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPhysicsEvent(serverLevel, p_401142_).isCancelled()) { // Paper
+                return Blocks.AIR.defaultBlockState();
+            }
+        }
+        return super.updateShape(p_401118_, p_401198_, p_401107_, p_401142_, p_401236_, p_401082_, p_401336_, p_401169_);
+        // CraftBukkit end
     }
 
     @Override
     protected boolean canSurvive(BlockState p_401395_, LevelReader p_401031_, BlockPos p_401248_) {
         BlockPos blockpos = p_401248_.below();
-        return this.mayPlaceOn(p_401031_.getBlockState(blockpos), p_401031_, blockpos);
+        BlockState belowBlockState = p_401031_.getBlockState(blockpos);
+        var soilDecision = belowBlockState.canSustainPlant(p_401031_, blockpos, Direction.UP, p_401395_);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
+        return this.mayPlaceOn(belowBlockState, p_401031_, blockpos);
     }
 
     @Override
@@ -55,4 +_,24 @@
     protected boolean isPathfindable(BlockState p_401351_, PathComputationType p_401371_) {
         return p_401371_ == PathComputationType.AIR && !this.hasCollision ? true : super.isPathfindable(p_401351_, p_401371_);
     }
+
+    // Purpur start - Ability for hoe to replant crops
+    public void playerDestroyAndReplant(net.minecraft.world.level.Level world, net.minecraft.world.entity.player.Player player, BlockPos pos, BlockState state, @javax.annotation.Nullable net.minecraft.world.level.block.entity.BlockEntity blockEntity, net.minecraft.world.item.ItemStack itemInHand, net.minecraft.world.level.ItemLike itemToReplant) {
+        player.awardStat(net.minecraft.stats.Stats.BLOCK_MINED.get(this));
+        player.causeFoodExhaustion(0.005F, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.BLOCK_MINED);
+        java.util.List<net.minecraft.world.item.ItemStack> dropList = Block.getDrops(state, (net.minecraft.server.level.ServerLevel) world, pos, blockEntity, player, itemInHand);
+
+        boolean planted = false;
+        for (net.minecraft.world.item.ItemStack itemToDrop : dropList) {
+            if (!planted && itemToDrop.getItem() == itemToReplant) {
+                world.setBlock(pos, defaultBlockState(), 3);
+                itemToDrop.setCount(itemToDrop.getCount() - 1);
+                planted = true;
+            }
+            Block.popResource(world, pos, itemToDrop);
+        }
+
+        state.spawnAfterBreak((net.minecraft.server.level.ServerLevel) world, pos, itemInHand, true);
+    }
+    // Purpur end - Ability for hoe to replant crops
 }
