--- a/net/minecraft/world/level/block/WeepingVinesBlock.java
+++ b/net/minecraft/world/level/block/WeepingVinesBlock.java
@@ -34,4 +_,11 @@
     protected boolean canGrowInto(BlockState p_154971_) {
         return NetherVines.isValidGrowthState(p_154971_);
     }
+
+    // Purpur start - weeping vines configurable max growth age
+    @Override
+    public int getMaxGrowthAge() {
+        return org.purpurmc.purpur.PurpurConfig.weepingVinesMaxGrowthAge;
+    }
+    // Purpur end - weeping vines configurable max growth age
 }
