--- a/net/minecraft/world/damagesource/CombatTracker.java
+++ b/net/minecraft/world/damagesource/CombatTracker.java
@@ -29,15 +_,24 @@
     private int combatEndTime;
     public boolean inCombat;
     public boolean takingDamage;
+    public final io.papermc.paper.world.damagesource.PaperCombatTrackerWrapper paperCombatTracker; // Paper - Combat tracker API
 
     public CombatTracker(LivingEntity p_19285_) {
         this.mob = p_19285_;
+        this.paperCombatTracker = new io.papermc.paper.world.damagesource.PaperCombatTrackerWrapper(this); // Paper - Combat tracker API
     }
 
     public void recordDamage(DamageSource p_289533_, float p_289559_) {
         this.recheckStatus();
         FallLocation falllocation = FallLocation.getCurrentFallLocation(this.mob);
         CombatEntry combatentry = new CombatEntry(p_289533_, p_289559_, falllocation, (float)this.mob.fallDistance);
+    // Paper start - Combat tracker API
+        recordDamageAndCheckCombatState(combatentry);
+    }
+
+    public void recordDamageAndCheckCombatState(final CombatEntry combatentry) {
+        final DamageSource p_289533_ = combatentry.source();
+    // Paper end - Combat tracker API
         this.entries.add(combatentry);
         this.lastDamageTime = this.mob.tickCount;
         this.takingDamage = true;
@@ -55,7 +_,7 @@
 
     private Component getMessageForAssistedFall(Entity p_289547_, Component p_289532_, String p_289555_, String p_289548_) {
         ItemStack itemstack = p_289547_ instanceof LivingEntity livingentity ? livingentity.getMainHandItem() : ItemStack.EMPTY;
-        return !itemstack.isEmpty() && itemstack.has(DataComponents.CUSTOM_NAME)
+        return !itemstack.isEmpty() && (org.purpurmc.purpur.PurpurConfig.playerDeathsAlwaysShowItem || itemstack.has(DataComponents.CUSTOM_NAME)) // Purpur - always show item in player death messages
             ? Component.translatable(p_289555_, this.mob.getDisplayName(), p_289532_, itemstack.getDisplayName())
             : Component.translatable(p_289548_, this.mob.getDisplayName(), p_289532_);
     }
@@ -92,6 +_,11 @@
             DamageSource damagesource = combatentry.source();
             CombatEntry combatentry1 = this.getMostSignificantFall();
             DeathMessageType deathmessagetype = damagesource.type().deathMessageType();
+            // Neo: Implement IDeathMessageProvider#getDeathMessage
+            // Vanilla logic is replicated in IDeathMessageProvider.DEFAULT
+            if (true) {
+                return deathmessagetype.getMessageFunction().getDeathMessage(this.mob, combatentry, combatentry1);
+            }
             if (deathmessagetype == DeathMessageType.FALL_VARIANTS && combatentry1 != null) {
                 return this.getFallMessage(combatentry1, damagesource.getEntity());
             } else if (deathmessagetype == DeathMessageType.INTENTIONAL_GAME_DESIGN) {
@@ -99,6 +_,15 @@
                 Component component = ComponentUtils.wrapInSquareBrackets(Component.translatable(s + ".link")).withStyle(INTENTIONAL_GAME_DESIGN_STYLE);
                 return Component.translatable(s + ".message", this.mob.getDisplayName(), component);
             } else {
+                // Purpur start - Dont run with scissors!
+                if (damagesource.isScissors()) {
+                    return damagesource.getLocalizedDeathMessage(org.purpurmc.purpur.PurpurConfig.deathMsgRunWithScissors, this.mob);
+                // Purpur start - Stonecutter damage
+                } else if (damagesource.isStonecutter()) {
+                    return damagesource.getLocalizedDeathMessage(org.purpurmc.purpur.PurpurConfig.deathMsgStonecutter, this.mob);
+                // Purpur end - Stonecutter damage
+                }
+                // Purpur end - Dont run with scissors!
                 return damagesource.getLocalizedDeathMessage(this.mob);
             }
         }
@@ -147,6 +_,13 @@
     public void recheckStatus() {
         int i = this.inCombat ? 300 : 100;
         if (this.takingDamage && (!this.mob.isAlive() || this.mob.tickCount - this.lastDamageTime > i)) {
+    // Paper start - Combat tracker API
+            resetCombatState();
+        }
+    }
+
+    public void resetCombatState() {{
+    // Paper end - Combat tracker API
             boolean flag = this.inCombat;
             this.takingDamage = false;
             this.inCombat = false;
