--- a/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
@@ -17,15 +_,27 @@
         this.setPhase(EnderDragonPhase.HOVERING);
     }
 
-    public void setPhase(EnderDragonPhase<?> p_31417_) {
-        if (this.currentPhase == null || p_31417_ != this.currentPhase.getPhase()) {
+    public void setPhase(EnderDragonPhase<?> phase) {
+        if (this.currentPhase == null || phase != this.currentPhase.getPhase()) {
             if (this.currentPhase != null) {
                 this.currentPhase.end();
             }
 
-            this.currentPhase = this.getPhase((EnderDragonPhase<DragonPhaseInstance>)p_31417_);
+            // CraftBukkit start - Call EnderDragonChangePhaseEvent
+            org.bukkit.event.entity.EnderDragonChangePhaseEvent event = new org.bukkit.event.entity.EnderDragonChangePhaseEvent(
+                (org.bukkit.craftbukkit.entity.CraftEnderDragon) this.dragon.getBukkitEntity(),
+                (this.currentPhase == null) ? null : org.bukkit.craftbukkit.entity.CraftEnderDragon.getBukkitPhase(this.currentPhase.getPhase()),
+                org.bukkit.craftbukkit.entity.CraftEnderDragon.getBukkitPhase(phase)
+            );
+            if (!event.callEvent()) {
+                return;
+            }
+            phase = org.bukkit.craftbukkit.entity.CraftEnderDragon.getMinecraftPhase(event.getNewPhase());
+            // CraftBukkit end
+
+            this.currentPhase = this.getPhase((EnderDragonPhase<DragonPhaseInstance>)phase);
             if (!this.dragon.level().isClientSide) {
-                this.dragon.getEntityData().set(EnderDragon.DATA_PHASE, p_31417_.getId());
+                this.dragon.getEntityData().set(EnderDragon.DATA_PHASE, phase.getId());
             }
 
             LOGGER.debug("Dragon is now in phase {} on the {}", p_31417_, this.dragon.level().isClientSide ? "client" : "server");
