--- a/net/minecraft/util/random/WeightedList.java
+++ b/net/minecraft/util/random/WeightedList.java
@@ -13,14 +_,14 @@
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 
-public final class WeightedList<E> {
+public class WeightedList<E> { // Paper - non-final
     private static final int FLAT_THRESHOLD = 64;
     private final int totalWeight;
     private final List<Weighted<E>> items;
     @Nullable
     private final WeightedList.Selector<E> selector;
 
-    WeightedList(List<? extends Weighted<E>> p_393913_) {
+    protected WeightedList(List<? extends Weighted<E>> p_393913_) { // Paper - protected
         this.items = List.copyOf(p_393913_);
         this.totalWeight = WeightedRandom.getTotalWeight(p_393913_, Weighted::weight);
         if (this.totalWeight == 0) {
@@ -127,7 +_,7 @@
     }
 
     public static class Builder<E> {
-        private final ImmutableList.Builder<Weighted<E>> result = ImmutableList.builder();
+        protected final List<Weighted<E>> result = new java.util.ArrayList<>(); // Paper - protected
 
         public WeightedList.Builder<E> add(E p_394625_) {
             return this.add(p_394625_, 1);
@@ -138,8 +_,41 @@
             return this;
         }
 
+        public WeightedList.Builder<E> add(Weighted<E> value) {
+            this.result.add(value);
+            return this;
+        }
+
+        public WeightedList.Builder<E> addAll(WeightedList<E> values) {
+            return this.addAll(values.unwrap());
+        }
+
+        public WeightedList.Builder<E> addAll(java.util.Collection<Weighted<E>> values) {
+            this.result.addAll(values);
+            return this;
+        }
+
+        public WeightedList.Builder<E> remove(Weighted<E> value) {
+            this.result.remove(value);
+            return this;
+        }
+
+        public WeightedList.Builder<E> remove(E value) {
+            this.result.removeIf(weighted -> weighted.value().equals(value));
+            return this;
+        }
+
+        public WeightedList.Builder<E> removeIf(java.util.function.Predicate<Weighted<E>> filter) {
+            this.result.removeIf(filter);
+            return this;
+        }
+
+        public List<Weighted<E>> getList() {
+            return java.util.Collections.unmodifiableList(this.result);
+        }
+
         public WeightedList<E> build() {
-            return new WeightedList<>(this.result.build());
+            return new WeightedList<>(this.result);
         }
     }
 
