--- a/net/minecraft/network/protocol/game/ClientboundSetPlayerTeamPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetPlayerTeamPacket.java
@@ -31,6 +_,11 @@
     private final Collection<String> players;
     private final Optional<ClientboundSetPlayerTeamPacket.Parameters> parameters;
 
+    // Paper start - Multiple Entries with Scoreboards
+    public static ClientboundSetPlayerTeamPacket createMultiplePlayerPacket(PlayerTeam team, Collection<String> players, ClientboundSetPlayerTeamPacket.Action operation) {
+        return new ClientboundSetPlayerTeamPacket(team.getName(), operation == ClientboundSetPlayerTeamPacket.Action.ADD ? 3 : 4, Optional.empty(), players);
+    }
+    // Paper end - Multiple Entries with Scoreboards
     private ClientboundSetPlayerTeamPacket(
         String p_179318_, int p_179319_, Optional<ClientboundSetPlayerTeamPacket.Parameters> p_179320_, Collection<String> p_179321_
     ) {
@@ -200,7 +_,7 @@
             ComponentSerialization.TRUSTED_STREAM_CODEC.encode(p_323527_, this.displayName);
             p_323527_.writeByte(this.options);
             Team.Visibility.STREAM_CODEC.encode(p_323527_, this.nametagVisibility);
-            Team.CollisionRule.STREAM_CODEC.encode(p_323527_, this.collisionRule);
+            Team.CollisionRule.STREAM_CODEC.encode(p_323527_, !io.papermc.paper.configuration.GlobalConfiguration.get().collisions.enablePlayerCollisions ? PlayerTeam.CollisionRule.NEVER : this.collisionRule); // Paper - Configurable player collision
             p_323527_.writeEnum(this.color);
             ComponentSerialization.TRUSTED_STREAM_CODEC.encode(p_323527_, this.playerPrefix);
             ComponentSerialization.TRUSTED_STREAM_CODEC.encode(p_323527_, this.playerSuffix);
