--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -30,6 +_,7 @@
 
 public class Raids extends SavedData {
     private static final String RAID_FILE_ID = "raids";
+    public final java.util.Map<java.util.UUID, Integer> playerCooldowns = com.google.common.collect.Maps.newHashMap(); // Purpur - Raid cooldown setting
     public static final Codec<Raids> CODEC = RecordCodecBuilder.create(
         p_400930_ -> p_400930_.group(
                 Raids.RaidWithId.CODEC
@@ -58,6 +_,7 @@
     private Raids(List<Raids.RaidWithId> p_401252_, int p_401028_, int p_401314_) {
         for (Raids.RaidWithId raids$raidwithid : p_401252_) {
             this.raidMap.put(raids$raidwithid.id, raids$raidwithid.raid);
+            raids$raidwithid.raid.idOrNegativeOne = raids$raidwithid.id; // Paper - expose id of p_401252_ while method is kept around as deprecated for removal
         }
 
         this.nextId = p_401028_;
@@ -81,6 +_,17 @@
 
     public void tick(ServerLevel p_401204_) {
         this.tick++;
+        // Purpur start - Raid cooldown setting
+        if (p_401204_.purpurConfig.raidCooldownSeconds != 0 && this.tick % 20 == 0) {
+            com.google.common.collect.ImmutableMap.copyOf(playerCooldowns).forEach((uuid, i) -> {
+                if (i < 1) {
+                    playerCooldowns.remove(uuid);
+                } else {
+                    playerCooldowns.put(uuid, i - 1);
+                }
+            });
+        }
+        // Purpur end - Raid cooldown setting
         Iterator<Raid> iterator = this.raidMap.values().iterator();
 
         while (iterator.hasNext()) {
@@ -142,11 +_,25 @@
                     }
 
                     Raid raid = this.getOrCreateRaid(serverlevel, blockpos1);
-                    if (!raid.isStarted() && !this.raidMap.containsValue(raid)) {
-                        this.raidMap.put(this.getUniqueId(), raid);
-                    }
-
-                    if (!raid.isStarted() || raid.getRaidOmenLevel() < raid.getMaxRaidOmenLevel()) {
+                    // CraftBukkit - moved down
+                    // if (!raid.isStarted() && !this.raidMap.containsValue(raid)) {
+                    //     this.raidMap.put(this.getUniqueId(), raid);
+                    // }
+
+                    if (!raid.isStarted() || (raid.isInProgress() && raid.getRaidOmenLevel() < raid.getMaxRaidOmenLevel())) { // CraftBukkit - fixed a bug with raid: players could add up Bad Omen level even when the raid had finished
+                        if (serverlevel.purpurConfig.raidCooldownSeconds != 0 && playerCooldowns.containsKey(p_37964_.getUUID())) return null; // Purpur - Raid cooldown setting
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callRaidTriggerEvent(serverlevel, raid, p_37964_)) {
+                            p_37964_.removeEffect(net.minecraft.world.effect.MobEffects.RAID_OMEN);
+                            return null;
+                        }
+                        if (serverlevel.purpurConfig.raidCooldownSeconds != 0) playerCooldowns.put(p_37964_.getUUID(), serverlevel.purpurConfig.raidCooldownSeconds); // Purpur - Raid cooldown setting
+
+                        if (!raid.isStarted() && !this.raidMap.containsValue(raid)) {
+                            this.raidMap.put(this.getUniqueId(), raid);
+                            raid.idOrNegativeOne = this.nextId; // Paper - expose id of raids while method is kept around as deprecated for removal
+                        }
+                        // CraftBukkit end
                         raid.absorbRaidOmen(p_37964_);
                     }
 
