--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -48,8 +_,9 @@
     }
 
     private static void lookupPlayers(MinecraftServer p_11087_, Collection<String> p_11088_, ProfileLookupCallback p_11089_) {
-        String[] astring = p_11088_.stream().filter(p_11077_ -> !StringUtil.isNullOrEmpty(p_11077_)).toArray(String[]::new);
-        if (p_11087_.usesAuthentication()) {
+        String[] astring = p_11088_.stream().filter(name -> !StringUtil.isNullOrEmpty(name)).toArray(String[]::new);
+        if (p_11087_.usesAuthentication() ||
+            (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode())) { // Spigot: bungee = online mode, for now.  // Paper - Add setting for proxy online mode status
             p_11087_.getProfileRepository().findProfilesByNames(astring, p_11089_);
         } else {
             for (String s : astring) {
@@ -65,7 +_,7 @@
                 try {
                     userbanlist.load();
                 } catch (IOException ioexception1) {
-                    LOGGER.warn("Could not load existing file {}", userbanlist.getFile().getName(), ioexception1);
+                    LOGGER.warn("Could not load existing file {}", userbanlist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -120,7 +_,7 @@
                 try {
                     ipbanlist.load();
                 } catch (IOException ioexception1) {
-                    LOGGER.warn("Could not load existing file {}", ipbanlist.getFile().getName(), ioexception1);
+                    LOGGER.warn("Could not load existing file {}", ipbanlist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -156,7 +_,7 @@
                 try {
                     serveroplist.load();
                 } catch (IOException ioexception1) {
-                    LOGGER.warn("Could not load existing file {}", serveroplist.getFile().getName(), ioexception1);
+                    LOGGER.warn("Could not load existing file {}", serveroplist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -200,7 +_,7 @@
                 try {
                     userwhitelist.load();
                 } catch (IOException ioexception1) {
-                    LOGGER.warn("Could not load existing file {}", userwhitelist.getFile().getName(), ioexception1);
+                    LOGGER.warn("Could not load existing file {}", userwhitelist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -310,12 +_,43 @@
                         }
                     }
 
-                    private void movePlayerFile(File p_11168_, String p_11169_, String p_11170_) {
-                        File file5 = new File(file1, p_11169_ + ".dat");
-                        File file6 = new File(p_11168_, p_11170_ + ".dat");
-                        OldUsersConverter.ensureDirectoryExists(p_11168_);
-                        if (!file5.renameTo(file6)) {
-                            throw new OldUsersConverter.ConversionError("Could not convert file for " + p_11169_);
+                    private void movePlayerFile(File file3, String oldFileName, String newFileName) {
+                        File file4 = new File(file1, oldFileName + ".dat");
+                        File file5 = new File(file3, newFileName + ".dat");
+                        // CraftBukkit start - Use old file2 s to seed lastKnownName
+                        net.minecraft.nbt.CompoundTag root = null;
+
+                        try {
+                            root = net.minecraft.nbt.NbtIo.readCompressed(new java.io.FileInputStream(file4), net.minecraft.nbt.NbtAccounter.unlimitedHeap());
+                        } catch (Exception exception) {
+                            // Paper start
+                            io.papermc.paper.util.StacktraceDeobfuscator.INSTANCE.deobfuscateThrowable(exception);
+                            exception.printStackTrace();
+                            com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(exception);
+                            // Paper end
+                        }
+
+                        if (root != null) {
+                            if (!root.contains("bukkit")) {
+                                root.put("bukkit", new net.minecraft.nbt.CompoundTag());
+                            }
+                            net.minecraft.nbt.CompoundTag data = root.getCompoundOrEmpty("bukkit");
+                            data.putString("lastKnownName", oldFileName);
+
+                            try {
+                                net.minecraft.nbt.NbtIo.writeCompressed(root, new java.io.FileOutputStream(file3));
+                            } catch (Exception exception) {
+                                // Paper start
+                                io.papermc.paper.util.StacktraceDeobfuscator.INSTANCE.deobfuscateThrowable(exception);
+                                exception.printStackTrace();
+                                com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(exception);
+                                // Paper end
+                            }
+                       }
+                        // CraftBukkit end
+                        OldUsersConverter.ensureDirectoryExists(file3);
+                        if (!file4.renameTo(file5)) {
+                            throw new OldUsersConverter.ConversionError("Could not convert file2 for " + oldFileName);
                         }
                     }
 
