--- a/net/minecraft/world/item/BlockItem.java
+++ b/net/minecraft/world/item/BlockItem.java
@@ -57,6 +_,14 @@
                 return InteractionResult.FAIL;
             } else {
                 BlockState blockstate = this.getPlacementState(blockplacecontext);
+                // CraftBukkit start - special case for handling block placement with water lilies and snow buckets
+                org.bukkit.block.BlockState bukkitState = null;
+                if (this instanceof PlaceOnWaterBlockItem || this instanceof SolidBucketItem) {
+                    bukkitState = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(blockplacecontext.getLevel(), blockplacecontext.getClickedPos());
+                }
+                final org.bukkit.block.BlockState oldBukkitState = bukkitState != null ? bukkitState : org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(blockplacecontext.getLevel(), blockplacecontext.getClickedPos()); // Paper - Reset placed block on exception
+                // CraftBukkit end
+
                 if (blockstate == null) {
                     return InteractionResult.FAIL;
                 } else if (!this.placeBlock(blockplacecontext, blockstate)) {
@@ -77,27 +_,33 @@
                         }
                     }
 
-                    SoundType soundtype = blockstate1.getSoundType();
+                    SoundType soundtype = blockstate1.getSoundType(level, blockpos, p_40577_.getPlayer());
                     level.playSound(
                         player,
                         blockpos,
-                        this.getPlaceSound(blockstate1),
+                        this.getPlaceSound(blockstate1, level, blockpos, p_40577_.getPlayer()),
                         SoundSource.BLOCKS,
                         (soundtype.getVolume() + 1.0F) / 2.0F,
                         soundtype.getPitch() * 0.8F
                     );
                     level.gameEvent(GameEvent.BLOCK_PLACE, blockpos, GameEvent.Context.of(player, blockstate1));
                     itemstack.consume(1, player);
-                    return InteractionResult.SUCCESS;
+                    return InteractionResult.SUCCESS.configurePaper(e -> e.placedBlockAt(blockpos.immutable())); // Paper - track placed block position from block item
                 }
             }
         }
     }
 
+    @Deprecated //Forge: Use more sensitive version {@link BlockItem#getPlaceSound(BlockState, IBlockReader, BlockPos, Entity) }
     protected SoundEvent getPlaceSound(BlockState p_40588_) {
         return p_40588_.getSoundType().getPlaceSound();
     }
 
+    //Forge: Sensitive version of BlockItem#getPlaceSound
+    protected SoundEvent getPlaceSound(BlockState state, Level world, BlockPos pos, Player entity) {
+        return state.getSoundType(world, pos, entity).getPlaceSound();
+    }
+
     @Nullable
     public BlockPlaceContext updatePlacementContext(BlockPlaceContext p_40609_) {
         return p_40609_;
@@ -112,7 +_,16 @@
     }
 
     protected boolean updateCustomBlockEntityTag(BlockPos p_40597_, Level p_40598_, @Nullable Player p_40599_, ItemStack p_40600_, BlockState p_40601_) {
-        return updateCustomBlockEntityTag(p_40598_, p_40599_, p_40597_, p_40600_);
+        // Purpur start - Persistent BlockEntity Lore and DisplayName
+        boolean handled = updateCustomBlockEntityTag(p_40598_, p_40599_, p_40597_, p_40600_);
+        if (p_40598_.purpurConfig.persistentTileEntityLore) {
+            BlockEntity blockEntity1 = p_40598_.getBlockEntity(p_40597_);
+            if (blockEntity1 != null) {
+                blockEntity1.setPersistentLore(p_40600_.getOrDefault(DataComponents.LORE, net.minecraft.world.item.component.ItemLore.EMPTY));
+            }
+        }
+        return handled;
+        // Purpur end - Persistent BlockEntity Lore and DisplayName
     }
 
     @Nullable
@@ -137,8 +_,19 @@
 
     protected boolean canPlace(BlockPlaceContext p_40611_, BlockState p_40612_) {
         Player player = p_40611_.getPlayer();
-        return (!this.mustSurvive() || p_40612_.canSurvive(p_40611_.getLevel(), p_40611_.getClickedPos()))
-            && p_40611_.getLevel().isUnobstructed(p_40612_, p_40611_.getClickedPos(), CollisionContext.placementContext(player));
+        // CraftBukkit start
+        Level world = p_40611_.getLevel(); // Paper - Cancel hit for vanished players
+        boolean canBuild = (!this.mustSurvive() || p_40612_.canSurvive(world, p_40611_.getClickedPos())) && world.checkEntityCollision(p_40612_, player, CollisionContext.placementContext(player), p_40611_.getClickedPos(), true); // Paper - Cancel hit for vanished players
+        org.bukkit.entity.Player bukkitPlayer = (p_40611_.getPlayer() instanceof ServerPlayer) ? (org.bukkit.entity.Player) p_40611_.getPlayer().getBukkitEntity() : null;
+
+        org.bukkit.event.block.BlockCanBuildEvent event = new org.bukkit.event.block.BlockCanBuildEvent(
+            org.bukkit.craftbukkit.block.CraftBlock.at(world, p_40611_.getClickedPos()), bukkitPlayer,
+            org.bukkit.craftbukkit.block.data.CraftBlockData.fromData(p_40612_), canBuild, org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(p_40611_.getHand())
+        );
+        world.getCraftServer().getPluginManager().callEvent(event);
+
+        return event.isBuildable();
+        // CraftBukkit end
     }
 
     protected boolean mustSurvive() {
@@ -167,7 +_,7 @@
                         return false;
                     }
 
-                    if (!blockentitytype1.onlyOpCanSetNbt() || p_40584_ != null && p_40584_.canUseGameMasterBlocks()) {
+                    if (!blockentitytype1.onlyOpCanSetNbt() || p_40584_ != null && (p_40584_.canUseGameMasterBlocks() || (p_40584_.getAbilities().instabuild && p_40584_.getBukkitEntity().hasPermission("minecraft.nbt.place")))) { // Spigot - add permission
                         return customdata.loadInto(blockentity, p_40583_.registryAccess());
                     }
 
@@ -209,6 +_,7 @@
     public void onDestroyed(ItemEntity p_150700_) {
         ItemContainerContents itemcontainercontents = p_150700_.getItem().set(DataComponents.CONTAINER, ItemContainerContents.EMPTY);
         if (itemcontainercontents != null) {
+            if (p_150700_.level().purpurConfig.shulkerBoxItemDropContentsWhenDestroyed && this.getBlock() instanceof ShulkerBoxBlock) // Purpur - option to disable shulker box items from dropping contents when destroyed
             ItemUtils.onContainerDestroyed(p_150700_, itemcontainercontents.nonEmptyItemsCopy());
         }
     }
