--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -28,17 +_,24 @@
             return false;
         } else {
             ItemStack itemstack = player.getProjectile(p_40667_);
+            //  Purpur start - Infinity bow settings
+            if (level.purpurConfig.infinityWorksWithoutArrows && itemstack.isEmpty() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.INFINITY, p_40667_) > 0) {
+                itemstack = new ItemStack(Items.ARROW);
+            }
+            // Purpur end - Infinity bow settings
             if (itemstack.isEmpty()) {
                 return false;
             } else {
                 int i = this.getUseDuration(p_40667_, p_40669_) - p_40670_;
+                i = net.neoforged.neoforge.event.EventHooks.onArrowLoose(p_40667_, p_40668_, player, i, !itemstack.isEmpty());
+                if (i < 0) return false;
                 float f = getPowerForTime(i);
                 if (f < 0.1) {
                     return false;
                 } else {
                     List<ItemStack> list = draw(p_40667_, itemstack, player);
                     if (p_40668_ instanceof ServerLevel serverlevel && !list.isEmpty()) {
-                        this.shoot(serverlevel, player, player.getUsedItemHand(), p_40667_, list, f * 3.0F, 1.0F, f == 1.0F, null);
+                        this.shoot(serverlevel, player, player.getUsedItemHand(), p_40667_, list, f * 3.0F, 1.0F, f == 1.0F, null, f); // Paper - Pass draw strength
                     }
 
                     p_40668_.playSound(
@@ -89,6 +_,10 @@
     public InteractionResult use(Level p_40672_, Player p_40673_, InteractionHand p_40674_) {
         ItemStack itemstack = p_40673_.getItemInHand(p_40674_);
         boolean flag = !p_40673_.getProjectile(itemstack).isEmpty();
+
+        InteractionResult ret = net.neoforged.neoforge.event.EventHooks.onArrowNock(itemstack, p_40672_, p_40673_, p_40674_, flag);
+        if (ret != null) return ret;
+
         if (!p_40673_.hasInfiniteMaterials() && !flag) {
             return InteractionResult.FAIL;
         } else {
