--- a/net/minecraft/world/item/Items.java
+++ b/net/minecraft/world/item/Items.java
@@ -383,7 +_,7 @@
     public static final Item PURPUR_BLOCK = registerBlock(Blocks.PURPUR_BLOCK);
     public static final Item PURPUR_PILLAR = registerBlock(Blocks.PURPUR_PILLAR);
     public static final Item PURPUR_STAIRS = registerBlock(Blocks.PURPUR_STAIRS);
-    public static final Item SPAWNER = registerBlock(Blocks.SPAWNER);
+    public static final Item SPAWNER = registerBlock(Blocks.SPAWNER, org.purpurmc.purpur.item.SpawnerItem::new, new Item.Properties().rarity(Rarity.EPIC)); // Purpur - Silk touch spawners
     public static final Item CREAKING_HEART = registerBlock(Blocks.CREAKING_HEART);
     public static final Item CHEST = registerBlock(Blocks.CHEST, p_332711_ -> p_332711_.component(DataComponents.CONTAINER, ItemContainerContents.EMPTY));
     public static final Item CRAFTING_TABLE = registerBlock(Blocks.CRAFTING_TABLE);
@@ -2091,7 +_,7 @@
         "sweet_berries", createBlockItemWithCustomItemName(Blocks.SWEET_BERRY_BUSH), new Item.Properties().food(Foods.SWEET_BERRIES)
     );
     public static final Item GLOW_BERRIES = registerItem(
-        "glow_berries", createBlockItemWithCustomItemName(Blocks.CAVE_VINES), new Item.Properties().food(Foods.GLOW_BERRIES)
+        "glow_berries", p_403877_ -> new org.purpurmc.purpur.item.GlowBerryItem(Blocks.CAVE_VINES, p_403877_.useItemDescriptionPrefix()), new Item.Properties().food(Foods.GLOW_BERRIES) // Purpur - Eating glow berries adds glow effect
     );
     public static final Item CAMPFIRE = registerBlock(Blocks.CAMPFIRE, p_332745_ -> p_332745_.component(DataComponents.CONTAINER, ItemContainerContents.EMPTY));
     public static final Item SOUL_CAMPFIRE = registerBlock(
@@ -2292,11 +_,19 @@
     }
 
     public static Item registerBlock(Block p_252092_, Block... p_248886_) {
-        Item item = registerBlock(p_252092_);
-
-        for (Block block : p_248886_) {
-            Item.BY_BLOCK.put(block, item);
-        }
+        Item item = registerItem(
+                blockIdToItemId(p_252092_.builtInRegistryHolder().key()),
+                p -> new BlockItem(p_252092_, p) {
+                    @Override
+                    public void registerBlocks(java.util.Map<Block, Item> map, Item self) {
+                        super.registerBlocks(map, self);
+                        for (Block block : p_248886_) {
+                            map.put(block, self);
+                        }
+                    }
+                },
+                new Item.Properties().useBlockDescriptionPrefix()
+        );
 
         return item;
     }
