--- a/net/minecraft/world/item/MinecartItem.java
+++ b/net/minecraft/world/item/MinecartItem.java
@@ -30,11 +_,12 @@
         BlockPos blockpos = p_42943_.getClickedPos();
         BlockState blockstate = level.getBlockState(blockpos);
         if (!blockstate.is(BlockTags.RAILS)) {
-            return InteractionResult.FAIL;
-        } else {
+            if (!level.purpurConfig.minecartPlaceAnywhere) return InteractionResult.FAIL; // Purpur - Minecart settings and WASD controls
+            if (blockstate.isSolid()) blockpos = blockpos.relative(p_42943_.getClickedFace());
+        } // else { // Purpur - Minecart settings and WASD controls
             ItemStack itemstack = p_42943_.getItemInHand();
             RailShape railshape = blockstate.getBlock() instanceof BaseRailBlock
-                ? blockstate.getValue(((BaseRailBlock)blockstate.getBlock()).getShapeProperty())
+                    ? ((BaseRailBlock)blockstate.getBlock()).getRailDirection(blockstate, level, blockpos, null)
                 : RailShape.NORTH_SOUTH;
             double d0 = 0.0;
             if (railshape.isSlope()) {
@@ -57,7 +_,13 @@
                 }
 
                 if (level instanceof ServerLevel serverlevel) {
-                    serverlevel.addFreshEntity(abstractminecart);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPlaceEvent(p_42943_, abstractminecart).isCancelled()) {
+                        if (p_42943_.getPlayer() != null) p_42943_.getPlayer().containerMenu.forceHeldSlot(p_42943_.getHand()); // Paper - Fix inventory desync
+                        return InteractionResult.FAIL;
+                    }
+                    // CraftBukkit end
+                    if (!serverlevel.addFreshEntity(abstractminecart)) return InteractionResult.PASS; // CraftBukkit
                     serverlevel.gameEvent(
                         GameEvent.ENTITY_PLACE, blockpos, GameEvent.Context.of(p_42943_.getPlayer(), serverlevel.getBlockState(blockpos.below()))
                     );
@@ -66,6 +_,6 @@
                 itemstack.shrink(1);
                 return InteractionResult.SUCCESS;
             }
-        }
+        // } // Purpur - Minecart settings and WASD controls
     }
 }
