--- a/net/minecraft/world/item/SnowballItem.java
+++ b/net/minecraft/world/item/SnowballItem.java
@@ -21,24 +_,43 @@
     }
 
     @Override
-    public InteractionResult use(Level p_43142_, Player p_43143_, InteractionHand p_43144_) {
-        ItemStack itemstack = p_43143_.getItemInHand(p_43144_);
-        p_43142_.playSound(
-            null,
-            p_43143_.getX(),
-            p_43143_.getY(),
-            p_43143_.getZ(),
-            SoundEvents.SNOWBALL_THROW,
-            SoundSource.NEUTRAL,
-            0.5F,
-            0.4F / (p_43142_.getRandom().nextFloat() * 0.4F + 0.8F)
-        );
-        if (p_43142_ instanceof ServerLevel serverlevel) {
-            Projectile.spawnProjectileFromRotation(Snowball::new, serverlevel, itemstack, p_43143_, 0.0F, PROJECTILE_SHOOT_POWER, 1.0F);
+    public InteractionResult use(Level level, Player player, InteractionHand hand) {
+        ItemStack itemInHand = player.getItemInHand(hand);
+        // CraftBukkit start - moved down
+        if (level instanceof ServerLevel serverLevel) {
+            // Paper start - PlayerLaunchProjectileEvent
+            final Projectile.Delayed<Snowball> snowball = Projectile.spawnProjectileFromRotationDelayed(Snowball::new, serverLevel, itemInHand, player, 0.0F, SnowballItem.PROJECTILE_SHOOT_POWER, (float) serverLevel.purpurConfig.snowballProjectileOffset); // Purpur - Projectile offset config
+            com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent event = new com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemInHand), (org.bukkit.entity.Projectile) snowball.projectile().getBukkitEntity());
+            if (event.callEvent() && snowball.attemptSpawn()) {
+                player.awardStat(Stats.ITEM_USED.get(this));
+                if (event.shouldConsume()) {
+                    itemInHand.consume(1, player);
+                } else {
+                    player.containerMenu.forceHeldSlot(hand);
+                }
+                // Paper end - PlayerLaunchProjectileEvent
+
+                level.playSound(
+                    null,
+                    player.getX(),
+                    player.getY(),
+                    player.getZ(),
+                    SoundEvents.SNOWBALL_THROW,
+                    SoundSource.NEUTRAL,
+                    0.5F,
+                    0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F)
+                );
+                // Paper start - PlayerLaunchProjectileEvent - return fail
+            } else {
+                player.containerMenu.forceHeldSlot(hand);
+                return InteractionResult.FAIL;
+            }
+            // Paper end- PlayerLaunchProjectileEvent - return fail
+            // CraftBukkit end
         }
 
-        p_43143_.awardStat(Stats.ITEM_USED.get(this));
-        itemstack.consume(1, p_43143_);
+        // Paper - PlayerLaunchProjectileEvent - moved up
+        // itemInHand.consume(1, player); // CraftBukkit - moved up
         return InteractionResult.SUCCESS;
     }
 
