--- a/net/minecraft/world/item/WindChargeItem.java
+++ b/net/minecraft/world/item/WindChargeItem.java
@@ -26,34 +_,47 @@
     }
 
     @Override
-    public InteractionResult use(Level p_326306_, Player p_326042_, InteractionHand p_326470_) {
-        ItemStack itemstack = p_326042_.getItemInHand(p_326470_);
-        if (p_326306_ instanceof ServerLevel serverlevel) {
-            Projectile.spawnProjectileFromRotation(
-                (p_427163_, p_427164_, p_427165_) -> new WindCharge(
-                    p_326042_, p_326306_, p_326042_.position().x(), p_326042_.getEyePosition().y(), p_326042_.position().z()
-                ),
-                serverlevel,
-                itemstack,
-                p_326042_,
+    public InteractionResult use(Level level, Player player, InteractionHand hand) {
+        ItemStack itemInHand = player.getItemInHand(hand);
+        if (level instanceof ServerLevel serverLevel) {
+            final Projectile.Delayed<WindCharge> windCharge = Projectile.spawnProjectileFromRotationDelayed( // Paper - PlayerLaunchProjectileEvent
+                (level1, owner, spawnedFrom) -> new WindCharge(player, level, player.position().x(), player.getEyePosition().y(), player.position().z()),
+                serverLevel,
+                itemInHand,
+                player,
                 0.0F,
                 PROJECTILE_SHOOT_POWER,
                 1.0F
             );
+            // Paper start - PlayerLaunchProjectileEvent
+            com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent event = new com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemInHand), (org.bukkit.entity.Projectile) windCharge.projectile().getBukkitEntity());
+            if (!event.callEvent() || !windCharge.attemptSpawn()) {
+                player.containerMenu.forceHeldSlot(hand);
+                if (player instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+                    serverPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundCooldownPacket(player.getCooldowns().getCooldownGroup(itemInHand), 0)); // prevent visual desync of cooldown on the slot
+                }
+                return InteractionResult.FAIL;
+            }
+
+            player.awardStat(Stats.ITEM_USED.get(this));
+            if (event.shouldConsume()) itemInHand.consume(1, player);
+            else if (!player.hasInfiniteMaterials()) {
+                player.containerMenu.forceHeldSlot(hand);
+            }
+            // Paper end - PlayerLaunchProjectileEvent
         }
 
-        p_326306_.playSound(
+        level.playSound(
             null,
-            p_326042_.getX(),
-            p_326042_.getY(),
-            p_326042_.getZ(),
+            player.getX(),
+            player.getY(),
+            player.getZ(),
             SoundEvents.WIND_CHARGE_THROW,
             SoundSource.NEUTRAL,
             0.5F,
-            0.4F / (p_326306_.getRandom().nextFloat() * 0.4F + 0.8F)
+            0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F)
         );
-        p_326042_.awardStat(Stats.ITEM_USED.get(this));
-        itemstack.consume(1, p_326042_);
+        // Paper - PlayerLaunchProjectileEvent; moved up
         return InteractionResult.SUCCESS;
     }
 
