--- a/net/minecraft/world/effect/HungerMobEffect.java
+++ b/net/minecraft/world/effect/HungerMobEffect.java
@@ -12,7 +_,7 @@
     @Override
     public boolean applyEffectTick(ServerLevel p_376386_, LivingEntity p_296407_, int p_296356_) {
         if (p_296407_ instanceof Player player) {
-            player.causeFoodExhaustion(0.005F * (p_296356_ + 1));
+            player.causeFoodExhaustion(p_296407_.level().purpurConfig.humanHungerExhaustionAmount * (p_296356_ + 1), org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.HUNGER_EFFECT); // CraftBukkit - EntityExhaustionEvent // Purpur - Config MobEffect by world
         }
 
         return true;
