--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -45,20 +_,40 @@
     }
 
     public static List<ServerPlayer> addEffectToPlayersAround(
-        ServerLevel p_216947_, @Nullable Entity p_216948_, Vec3 p_216949_, double p_216950_, MobEffectInstance p_216951_, int p_216952_
+        ServerLevel level, @Nullable Entity source, Vec3 pos, double radius, MobEffectInstance effect, int duration
     ) {
-        Holder<MobEffect> holder = p_216951_.getEffect();
-        List<ServerPlayer> list = p_216947_.getPlayers(
-            p_267925_ -> p_267925_.gameMode.isSurvival()
-                && (p_216948_ == null || !p_216948_.isAlliedTo(p_267925_))
-                && p_216949_.closerThan(p_267925_.position(), p_216950_)
+        // CraftBukkit start
+        return MobEffectUtil.addEffectToPlayersAround(level, source, pos, radius, effect, duration, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.UNKNOWN);
+    }
+
+    public static List<ServerPlayer> addEffectToPlayersAround(ServerLevel level, @Nullable Entity source, Vec3 pos, double radius, MobEffectInstance effect, int duration, org.bukkit.event.entity.EntityPotionEffectEvent.Cause cause) {
+        // Paper start - Add ElderGuardianAppearanceEvent
+        return addEffectToPlayersAround(level, source, pos, radius, effect, duration, cause, null);
+    }
+
+    public static List<ServerPlayer> addEffectToPlayersAround(ServerLevel level, @Nullable Entity source, Vec3 pos, double radius, MobEffectInstance effect, int duration, org.bukkit.event.entity.EntityPotionEffectEvent.Cause cause, @Nullable java.util.function.Predicate<ServerPlayer> playerPredicate) {
+        // Paper end - Add ElderGuardianAppearanceEvent
+        // CraftBukkit end
+        Holder<MobEffect> effect1 = effect.getEffect();
+        List<ServerPlayer> players = level.getPlayers(
+            // Paper start - Add ElderGuardianAppearanceEvent
+            serverPlayer -> {
+            final boolean condition = serverPlayer.gameMode.isSurvival()
+                && (source == null || !source.isAlliedTo(serverPlayer))
+                && pos.closerThan(serverPlayer.position(), radius)
                 && (
-                    !p_267925_.hasEffect(holder)
-                        || p_267925_.getEffect(holder).getAmplifier() < p_216951_.getAmplifier()
-                        || p_267925_.getEffect(holder).endsWithin(p_216952_ - 1)
-                )
-        );
-        list.forEach(p_238232_ -> p_238232_.addEffect(new MobEffectInstance(p_216951_), p_216948_));
-        return list;
+                !serverPlayer.hasEffect(effect1)
+                    || serverPlayer.getEffect(effect1).getAmplifier() < effect.getAmplifier()
+                    || serverPlayer.getEffect(effect1).endsWithin(duration - 1)
+            );
+            if (condition) {
+                return playerPredicate == null || playerPredicate.test(serverPlayer); // Only test the player AFTER it is true
+            } else {
+                return false;
+            }
+        });
+        // Paper end - Add ElderGuardianAppearanceEvent
+        players.forEach(serverPlayer -> serverPlayer.addEffect(new MobEffectInstance(effect), source, cause)); // CraftBukkit
+        return players;
     }
 }
