--- a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
+++ b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
@@ -129,12 +_,12 @@
         .put(EntityType.IRON_GOLEM, IronGolem.createAttributes().build())
         .put(EntityType.LLAMA, Llama.createAttributes().build())
         .put(EntityType.MAGMA_CUBE, MagmaCube.createAttributes().build())
-        .put(EntityType.MOOSHROOM, Cow.createAttributes().build())
+        .put(EntityType.MOOSHROOM, net.minecraft.world.entity.animal.AbstractCow.createAttributes().build()) // Purpur - Cows naturally aggressive to players chance
         .put(EntityType.MULE, AbstractChestedHorse.createBaseChestedHorseAttributes().build())
         .put(EntityType.OCELOT, Ocelot.createAttributes().build())
         .put(EntityType.PANDA, Panda.createAttributes().build())
         .put(EntityType.PARROT, Parrot.createAttributes().build())
-        .put(EntityType.PHANTOM, Monster.createMonsterAttributes().build())
+        .put(EntityType.PHANTOM, net.minecraft.world.entity.monster.Phantom.createAttributes().build()) // Purpur - Ridables
         .put(EntityType.PIG, Pig.createAttributes().build())
         .put(EntityType.PIGLIN, Piglin.createAttributes().build())
         .put(EntityType.PIGLIN_BRUTE, PiglinBrute.createAttributes().build())
@@ -165,7 +_,7 @@
         .put(EntityType.VILLAGER, Villager.createAttributes().build())
         .put(EntityType.VINDICATOR, Vindicator.createAttributes().build())
         .put(EntityType.WARDEN, Warden.createAttributes().build())
-        .put(EntityType.WANDERING_TRADER, Mob.createMobAttributes().build())
+        .put(EntityType.WANDERING_TRADER, net.minecraft.world.entity.npc.WanderingTrader.createAttributes().build()) // Purpur - Villagers follow emerald blocks
         .put(EntityType.WITCH, Witch.createAttributes().build())
         .put(EntityType.WITHER, WitherBoss.createAttributes().build())
         .put(EntityType.WITHER_SKELETON, AbstractSkeleton.createAttributes().build())
@@ -178,11 +_,12 @@
         .build();
 
     public static AttributeSupplier getSupplier(EntityType<? extends LivingEntity> p_22298_) {
-        return SUPPLIERS.get(p_22298_);
+        AttributeSupplier supplier = net.neoforged.neoforge.common.CommonHooks.getAttributesView().get(p_22298_);
+        return supplier != null ? supplier : SUPPLIERS.get(p_22298_);
     }
 
     public static boolean hasSupplier(EntityType<?> p_22302_) {
-        return SUPPLIERS.containsKey(p_22302_);
+        return SUPPLIERS.containsKey(p_22302_) || net.neoforged.neoforge.common.CommonHooks.getAttributesView().containsKey(p_22302_);
     }
 
     public static void validate() {
