--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -62,10 +_,57 @@
         super(p_405549_, p_405858_);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.sheepRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.sheepRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.sheepControllable;
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.sheepMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.sheepScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Make entity breeding times configurable
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level().purpurConfig.sheepBreedingTicks;
+    }
+    // Purpur end - Make entity breeding times configurable
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.sheepTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.sheepAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
     @Override
     protected void registerGoals() {
         this.eatBlockGoal = new EatBlockGoal(this);
         this.goalSelector.addGoal(0, new FloatGoal(this));
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(1, new PanicGoal(this, 1.25));
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
         this.goalSelector.addGoal(3, new TemptGoal(this, 1.1, p_404879_ -> p_404879_.is(ItemTags.SHEEP_FOOD), false));
@@ -137,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player p_404982_, InteractionHand p_404782_) {
         ItemStack itemstack = p_404982_.getItemInHand(p_404782_);
-        if (itemstack.is(Items.SHEARS)) {
+        if (false && itemstack.is(Items.SHEARS)) { // Neo: Shear logic is handled by IShearable
             if (this.level() instanceof ServerLevel serverlevel && this.readyForShearing()) {
                 this.shear(serverlevel, SoundSource.PLAYERS, itemstack);
                 this.gameEvent(GameEvent.SHEAR, p_404982_);
@@ -153,14 +_,28 @@
 
     @Override
     public void shear(ServerLevel p_405075_, SoundSource p_405077_, ItemStack p_405709_) {
+        // Paper start - custom shear drops
+        this.shear(p_405075_, p_405077_, p_405709_, this.generateDefaultDrops(p_405075_, p_405709_));
+    }
+
+    @Override
+    public java.util.List<ItemStack> generateDefaultDrops(final ServerLevel serverLevel, final ItemStack p_405709_) {
+        final java.util.List<ItemStack> drops = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        this.dropFromShearingLootTable(serverLevel, BuiltInLootTables.SHEAR_SHEEP, p_405709_, (ignored, stack) -> {
+            for (int i = 0; i < stack.getCount(); ++i) drops.add(stack.copyWithCount(1));
+        });
+        return drops;
+    }
+
+    @Override
+    public void shear(ServerLevel p_405075_, SoundSource p_405077_, ItemStack p_405709_, java.util.List<ItemStack> drops) {
+        // Paper end - custom shear drops
         p_405075_.playSound(null, this, SoundEvents.SHEEP_SHEAR, p_405077_, 1.0F, 1.0F);
-        this.dropFromShearingLootTable(
-            p_405075_,
-            BuiltInLootTables.SHEAR_SHEEP,
-            p_405709_,
-            (p_405522_, p_405241_) -> {
-                for (int i = 0; i < p_405241_.getCount(); i++) {
-                    ItemEntity itementity = this.spawnAtLocation(p_405522_, p_405241_.copyWithCount(1), 1.0F);
+        drops.forEach(p_405241_ -> { // Paper - custom drops - loop in generated default drops
+            if (true) { // Paper - custom drops - loop in generated default drops
+                this.forceDrops = true; // CraftBukkit
+                ItemEntity itementity = this.spawnAtLocation(level, p_405241_, 1.0F); // Paper - custom drops - copy already done above
+                this.forceDrops = false; // CraftBukkit
                     if (itementity != null) {
                         itementity.setDeltaMovement(
                             itementity.getDeltaMovement()
@@ -197,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.SHEEP_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_405795_) {
+    public SoundEvent getHurtSound(DamageSource p_405795_) {
         return SoundEvents.SHEEP_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SHEEP_DEATH;
     }
 
@@ -279,6 +_,7 @@
 
     @Override
     public void ate() {
+        if (!new org.bukkit.event.entity.SheepRegrowWoolEvent((org.bukkit.entity.Sheep) this.getBukkitEntity()).callEvent()) return; // CraftBukkit
         super.ate();
         this.setSheared(false);
         if (this.isBaby()) {
