--- a/net/minecraft/world/entity/boss/EnderDragonPart.java
+++ b/net/minecraft/world/entity/boss/EnderDragonPart.java
@@ -15,18 +_,25 @@
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 
-public class EnderDragonPart extends Entity {
+public class EnderDragonPart extends net.neoforged.neoforge.entity.PartEntity<EnderDragon> {
     public final EnderDragon parentMob;
     public final String name;
     private final EntityDimensions size;
 
     public EnderDragonPart(EnderDragon p_31014_, String p_31015_, float p_31016_, float p_31017_) {
-        super(p_31014_.getType(), p_31014_.level());
+        super(p_31014_);
         this.size = EntityDimensions.scalable(p_31016_, p_31017_);
         this.refreshDimensions();
         this.parentMob = p_31014_;
         this.name = p_31015_;
     }
+
+    // Purpur start - Ridables
+    @Override
+    public net.minecraft.world.InteractionResult interact(net.minecraft.world.entity.player.Player player, net.minecraft.world.InteractionHand hand) {
+        return parentMob.isAlive() ? parentMob.tryRide(player, hand) : net.minecraft.world.InteractionResult.PASS;
+    }
+    // Purpur end - Ridables
 
     @Override
     protected void defineSynchedData(SynchedEntityData.Builder p_325943_) {
