--- a/net/minecraft/network/CompressionEncoder.java
+++ b/net/minecraft/network/CompressionEncoder.java
@@ -6,36 +_,71 @@
 import java.util.zip.Deflater;
 
 public class CompressionEncoder extends MessageToByteEncoder<ByteBuf> {
-    private final byte[] encodeBuf = new byte[8192];
+    @javax.annotation.Nullable private final byte[] encodeBuf; // Paper - Use Velocity cipher
+    @javax.annotation.Nullable // Paper - Use Velocity cipher
     private final Deflater deflater;
+    @javax.annotation.Nullable private final com.velocitypowered.natives.compression.VelocityCompressor compressor; // Paper - Use Velocity cipher
     private int threshold;
+    private static final boolean DISABLE_PACKET_DEBUG = Boolean.parseBoolean(System.getProperty("neoforge.disablePacketCompressionDebug", "false"));
+    private static final org.apache.logging.log4j.Logger LOGGER = org.apache.logging.log4j.LogManager.getLogger();
 
+    // Paper start - Use Velocity cipher
     public CompressionEncoder(int p_129448_) {
-        this.threshold = p_129448_;
-        this.deflater = new Deflater();
+        this(null, p_129448_);
+    }
+    public CompressionEncoder(@javax.annotation.Nullable com.velocitypowered.natives.compression.VelocityCompressor compressor, int threshold) {
+        this.threshold = threshold;
+        if (compressor == null) {
+            this.encodeBuf = new byte[8192];
+            this.deflater = new Deflater();
+        } else {
+            this.encodeBuf = null;
+            this.deflater = null;
+        }
+        this.compressor = compressor;
+        // Paper end - Use Velocity cipher
     }
 
-    protected void encode(ChannelHandlerContext p_129452_, ByteBuf p_129453_, ByteBuf p_129454_) {
-        int i = p_129453_.readableBytes();
+    @Override
+    protected void encode(ChannelHandlerContext context, ByteBuf encodingByteBuf, ByteBuf byteBuf) throws Exception { // Paper - Use Velocity cipher
+        int i = encodingByteBuf.readableBytes();
+        if (!DISABLE_PACKET_DEBUG && i > net.minecraft.network.CompressionDecoder.MAXIMUM_UNCOMPRESSED_LENGTH) {
+            encodingByteBuf.markReaderIndex();
+            LOGGER.error("Attempted to send packet over maximum protocol size: {} > {}\nData:\n{}", i, net.minecraft.network.CompressionDecoder.MAXIMUM_UNCOMPRESSED_LENGTH,
+                    net.neoforged.neoforge.logging.PacketDump.getContentDump(encodingByteBuf));
+            encodingByteBuf.resetReaderIndex();
+        }
         if (i > 8388608) {
             throw new IllegalArgumentException("Packet too big (is " + i + ", should be less than 8388608)");
         } else {
             if (i < this.threshold) {
-                VarInt.write(p_129454_, 0);
-                p_129454_.writeBytes(p_129453_);
+                VarInt.write(byteBuf, 0);
+                byteBuf.writeBytes(encodingByteBuf);
             } else {
-                byte[] abyte = new byte[i];
-                p_129453_.readBytes(abyte);
-                VarInt.write(p_129454_, abyte.length);
-                this.deflater.setInput(abyte, 0, i);
+                if (this.deflater != null) { // Paper - Use Velocity cipher
+                byte[] bytes = new byte[i];
+                encodingByteBuf.readBytes(bytes);
+                VarInt.write(byteBuf, bytes.length);
+                this.deflater.setInput(bytes, 0, i);
                 this.deflater.finish();
 
                 while (!this.deflater.finished()) {
-                    int j = this.deflater.deflate(this.encodeBuf);
-                    p_129454_.writeBytes(this.encodeBuf, 0, j);
+                    int i1 = this.deflater.deflate(this.encodeBuf);
+                    byteBuf.writeBytes(this.encodeBuf, 0, i1);
                 }
 
                 this.deflater.reset();
+                    // Paper start - Use Velocity cipher
+                    return;
+                }
+
+                VarInt.write(byteBuf, i);
+                final ByteBuf compatibleIn = com.velocitypowered.natives.util.MoreByteBufUtils.ensureCompatible(context.alloc(), this.compressor, encodingByteBuf);
+                try {
+                    this.compressor.deflate(compatibleIn, byteBuf);
+                } finally {
+                    compatibleIn.release();
+                }
             }
         }
     }
@@ -47,4 +_,31 @@
     public void setThreshold(int p_129450_) {
         this.threshold = p_129450_;
     }
+
+    // Paper start - Use Velocity cipher
+    @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) throws Exception {
+        if (this.compressor != null) {
+            // We allocate bytes to be compressed plus 1 byte. This covers two cases:
+            //
+            // - Compression
+            //    According to https://github.com/ebiggers/libdeflate/blob/master/libdeflate.h#L103,
+            //    if the data compresses well (and we do not have some pathological case) then the maximum
+            //    size the compressed size will ever be is the input size minus one.
+            // - Uncompressed
+            //    This is fairly obvious - we will then have one more than the uncompressed size.
+            final int initialBufferSize = msg.readableBytes() + 1;
+            return com.velocitypowered.natives.util.MoreByteBufUtils.preferredBuffer(ctx.alloc(), this.compressor, initialBufferSize);
+        }
+
+        return super.allocateBuffer(ctx, msg, preferDirect);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) {
+        if (this.compressor != null) {
+            this.compressor.close();
+        }
+    }
+    // Paper end - Use Velocity cipher
 }
