--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -34,6 +_,41 @@
     private static final int[][] CACHED_SLOTS = new int[54][];
     private static final int NO_COOLDOWN_TIME = -1;
     private NonNullList<ItemStack> items = NonNullList.withSize(5, ItemStack.EMPTY);
+    // Youer - fix: Container#setMaxStackSize(int) was never merged onto HopperBlockEntity - matches Paper's real implementation
+    // CraftBukkit start - add fields and methods
+    private int maxStack = MAX_STACK;
+    public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+
+    @Override
+    public void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        this.transaction.add(player);
+    }
+
+    @Override
+    public void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        this.transaction.remove(player);
+    }
+
+    @Override
+    public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return this.transaction;
+    }
+
+    @Override
+    public java.util.List<ItemStack> getContents() {
+        return this.items;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return this.maxStack;
+    }
+
+    @Override
+    public void setMaxStackSize(int size) {
+        this.maxStack = size;
+    }
+    // CraftBukkit end
     public int cooldownTime = -1;
     private long tickedGameTime;
     public Direction facing;
@@ -98,9 +_,53 @@
         p_155577_.tickedGameTime = p_155574_.getGameTime();
         if (!p_155577_.isOnCooldown()) {
             p_155577_.setCooldown(0);
-            tryMoveItems(p_155574_, p_155575_, p_155576_, p_155577_, () -> suckInItems(p_155574_, p_155577_));
-        }
-    }
+            // Spigot start
+            boolean result = tryMoveItems(p_155574_, p_155575_, p_155576_, p_155577_, () -> {
+                return suckInItems(p_155574_, p_155577_);
+            });
+            if (!result && p_155577_.level.spigotConfig.hopperCheck > 1) {
+                p_155577_.setCooldown(p_155577_.level.spigotConfig.hopperCheck);
+            }
+            // Spigot end
+        }
+    }
+
+    // Paper start - Perf: Optimize Hoppers
+    private static final int HOPPER_EMPTY = 0;
+    private static final int HOPPER_HAS_ITEMS = 1;
+    private static final int HOPPER_IS_FULL = 2;
+
+    private static int getFullState(final HopperBlockEntity hopper) {
+        hopper.unpackLootTable(null);
+
+        final List<ItemStack> hopperItems = hopper.items;
+
+        boolean empty = true;
+        boolean full = true;
+
+        for (int i = 0, len = hopperItems.size(); i < len; ++i) {
+            final ItemStack stack = hopperItems.get(i);
+            if (stack.isEmpty()) {
+                full = false;
+                continue;
+            }
+
+            if (!full) {
+                // can't be full
+                return HOPPER_HAS_ITEMS;
+            }
+
+            empty = false;
+
+            if (stack.getCount() != stack.getMaxStackSize()) {
+                // can't be full or empty
+                return HOPPER_HAS_ITEMS;
+            }
+        }
+
+        return empty ? HOPPER_EMPTY : (full ? HOPPER_IS_FULL : HOPPER_HAS_ITEMS);
+    }
+    // Paper end - Perf: Optimize Hoppers
 
     private static boolean tryMoveItems(Level p_155579_, BlockPos p_155580_, BlockState p_155581_, HopperBlockEntity p_155582_, BooleanSupplier p_155583_) {
         if (p_155579_.isClientSide) {
@@ -108,16 +_,17 @@
         } else {
             if (!p_155582_.isOnCooldown() && p_155581_.getValue(HopperBlock.ENABLED)) {
                 boolean flag = false;
-                if (!p_155582_.isEmpty()) {
+                final int fullState = getFullState(p_155582_); // Paper - Perf: Optimize Hoppers
+                if (fullState != HOPPER_EMPTY) { // Paper - Perf: Optimize Hoppers
                     flag = ejectItems(p_155579_, p_155580_, p_155582_);
                 }
 
-                if (!p_155582_.inventoryFull()) {
-                    flag |= p_155583_.getAsBoolean();
+                if (fullState != HOPPER_IS_FULL || flag) { // Paper - Perf: Optimize Hoppers
+                    flag |= p_155583_.getAsBoolean(); // Paper - note: this is not a validator, it's what adds/sucks in items
                 }
 
                 if (flag) {
-                    p_155582_.setCooldown(8);
+                    p_155582_.setCooldown(p_155579_.spigotConfig.hopperTransfer); // Spigot
                     setChanged(p_155579_, p_155580_, p_155581_);
                     return true;
                 }
@@ -137,34 +_,222 @@
         return true;
     }
 
+    // Paper start - Perf: Optimize Hoppers
+    public static boolean skipHopperEvents;
+    private static boolean skipPullModeEventFire;
+    private static boolean skipPushModeEventFire;
+
+    private static boolean hopperPush(final Level level, final Container destination, final Direction direction, final HopperBlockEntity hopper) {
+        skipPushModeEventFire = skipHopperEvents;
+        boolean foundItem = false;
+        for (int i = 0; i < hopper.getContainerSize(); ++i) {
+            final ItemStack item = hopper.getItem(i);
+            if (!item.isEmpty()) {
+                foundItem = true;
+                ItemStack origItemStack = item;
+                ItemStack movedItem = origItemStack;
+
+                final int originalItemCount = origItemStack.getCount();
+                final int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
+                origItemStack.setCount(movedItemCount);
+
+                // We only need to fire the event once to give protection plugins a chance to cancel this event
+                // Because nothing uses getItem, every event call should end up the same result.
+                if (!skipPushModeEventFire) {
+                    movedItem = callPushMoveEvent(destination, movedItem, hopper);
+                    if (movedItem == null) { // cancelled
+                        origItemStack.setCount(originalItemCount);
+                        return false;
+                    }
+                }
+
+                final ItemStack remainingItem = addItem(hopper, destination, movedItem, direction);
+                final int remainingItemCount = remainingItem.getCount();
+                if (remainingItemCount != movedItemCount) {
+                    origItemStack = origItemStack.copy(true);
+                    origItemStack.setCount(originalItemCount);
+                    if (!origItemStack.isEmpty()) {
+                        origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
+                    }
+                    hopper.setItem(i, origItemStack);
+                    destination.setChanged();
+                    return true;
+                }
+                origItemStack.setCount(originalItemCount);
+            }
+        }
+        if (foundItem && level.paperConfig().hopper.cooldownWhenFull) { // Inventory was full - cooldown
+            hopper.setCooldown(level.spigotConfig.hopperTransfer);
+        }
+        return false;
+    }
+
+    private static boolean hopperPull(final Level level, final Hopper hopper, final Container container, ItemStack origItemStack, final int i) {
+        ItemStack movedItem = origItemStack;
+        final int originalItemCount = origItemStack.getCount();
+        final int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
+        container.setChanged(); // original logic always marks source inv as changed even if no move happens.
+        movedItem.setCount(movedItemCount);
+
+        if (!skipPullModeEventFire) {
+            movedItem = callPullMoveEvent(hopper, container, movedItem);
+            if (movedItem == null) { // cancelled
+                origItemStack.setCount(originalItemCount);
+                // Drastically improve performance by returning true.
+                // No plugin could have relied on the behavior of false as the other call
+                // site for IMIE did not exhibit the same behavior
+                return true;
+            }
+        }
+
+        final ItemStack remainingItem = addItem(container, hopper, movedItem, null);
+        final int remainingItemCount = remainingItem.getCount();
+        if (remainingItemCount != movedItemCount) {
+            origItemStack = origItemStack.copy(true);
+            origItemStack.setCount(originalItemCount);
+            if (!origItemStack.isEmpty()) {
+                origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
+            }
+
+            ignoreBlockEntityUpdates = true;
+            container.setItem(i, origItemStack);
+            ignoreBlockEntityUpdates = false;
+            container.setChanged();
+            return true;
+        }
+        origItemStack.setCount(originalItemCount);
+
+        if (level.paperConfig().hopper.cooldownWhenFull) {
+            applyCooldown(hopper);
+        }
+
+        return false;
+    }
+
+    @Nullable
+    private static ItemStack callPushMoveEvent(Container destination, ItemStack itemStack, HopperBlockEntity hopper) {
+        final org.bukkit.inventory.Inventory destinationInventory = getInventory(destination);
+        final io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent event = new io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent(
+            hopper.getOwner(false).getInventory(),
+            org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack),
+            destinationInventory,
+            true
+        );
+        final boolean result = event.callEvent();
+        if (!event.calledGetItem && !event.calledSetItem) {
+            skipPushModeEventFire = true;
+        }
+        if (!result) {
+            applyCooldown(hopper);
+            return null;
+        }
+
+        if (event.calledSetItem) {
+            return org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+        } else {
+            return itemStack;
+        }
+    }
+
+    @Nullable
+    private static ItemStack callPullMoveEvent(final Hopper hopper, final Container container, final ItemStack itemstack) {
+        final org.bukkit.inventory.Inventory sourceInventory = getInventory(container);
+        final org.bukkit.inventory.Inventory destination = getInventory(hopper);
+
+        // Mirror is safe as no plugins ever use this item
+        final io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent event = new io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent(sourceInventory, org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), destination, false);
+        final boolean result = event.callEvent();
+        if (!event.calledGetItem && !event.calledSetItem) {
+            skipPullModeEventFire = true;
+        }
+        if (!result) {
+            applyCooldown(hopper);
+            return null;
+        }
+
+        if (event.calledSetItem) {
+            return org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+        } else {
+            return itemstack;
+        }
+    }
+
+    private static org.bukkit.inventory.Inventory getInventory(final Container container) {
+        final org.bukkit.inventory.Inventory sourceInventory;
+        if (container instanceof net.minecraft.world.CompoundContainer compoundContainer) {
+            // Have to special-case large chests as they work oddly
+            sourceInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest(compoundContainer);
+        } else if (container instanceof BlockEntity blockEntity) {
+            sourceInventory = blockEntity.getOwner(false).getInventory();
+        } else if (container.getOwner() != null) {
+            sourceInventory = container.getOwner().getInventory();
+        } else {
+            sourceInventory = new org.bukkit.craftbukkit.inventory.CraftInventory(container);
+        }
+        return sourceInventory;
+    }
+
+    private static void applyCooldown(final Hopper hopper) {
+        if (hopper instanceof HopperBlockEntity blockEntity && blockEntity.getLevel() != null) {
+            blockEntity.setCooldown(blockEntity.getLevel().spigotConfig.hopperTransfer);
+        }
+    }
+
+    private static boolean allMatch(Container container, Direction direction, java.util.function.BiPredicate<ItemStack, Integer> test) {
+        if (container instanceof WorldlyContainer) {
+            for (int slot : ((WorldlyContainer) container).getSlotsForFace(direction)) {
+                if (!test.test(container.getItem(slot), slot)) {
+                    return false;
+                }
+            }
+        } else {
+            int size = container.getContainerSize();
+            for (int slot = 0; slot < size; slot++) {
+                if (!test.test(container.getItem(slot), slot)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private static boolean anyMatch(Container container, Direction direction, java.util.function.BiPredicate<ItemStack, Integer> test) {
+        if (container instanceof WorldlyContainer) {
+            for (int slot : ((WorldlyContainer) container).getSlotsForFace(direction)) {
+                if (test.test(container.getItem(slot), slot)) {
+                    return true;
+                }
+            }
+        } else {
+            int size = container.getContainerSize();
+            for (int slot = 0; slot < size; slot++) {
+                if (test.test(container.getItem(slot), slot)) {
+                    return true;
+                }
+            }
+        }
+        return true;
+    }
+    private static final java.util.function.BiPredicate<ItemStack, Integer> STACK_SIZE_TEST = (itemStack, i) -> itemStack.getCount() >= itemStack.getMaxStackSize();
+    private static final java.util.function.BiPredicate<ItemStack, Integer> IS_EMPTY_TEST = (itemStack, i) -> itemStack.isEmpty();
+    // Paper end - Perf: Optimize Hoppers
+
     private static boolean ejectItems(Level p_155563_, BlockPos p_155564_, HopperBlockEntity p_326256_) {
-        Container container = getAttachedContainer(p_155563_, p_155564_, p_326256_);
-        if (container == null) {
+        var containerOrHandler = getContainerOrHandlerAt(p_155563_, p_155564_.relative(p_326256_.facing), p_326256_.facing.getOpposite());
+        if (containerOrHandler.isEmpty()) {
             return false;
-        } else {
+        } else if (containerOrHandler.container() != null) {
+            Container container = containerOrHandler.container();
             Direction direction = p_326256_.facing.getOpposite();
             if (isFullContainer(container, direction)) {
                 return false;
             } else {
-                for (int i = 0; i < p_326256_.getContainerSize(); i++) {
-                    ItemStack itemstack = p_326256_.getItem(i);
-                    if (!itemstack.isEmpty()) {
-                        int j = itemstack.getCount();
-                        ItemStack itemstack1 = addItem(p_326256_, container, p_326256_.removeItem(i, 1), direction);
-                        if (itemstack1.isEmpty()) {
-                            container.setChanged();
-                            return true;
-                        }
-
-                        itemstack.setCount(j);
-                        if (j == 1) {
-                            p_326256_.setItem(i, itemstack);
-                        }
-                    }
-                }
-
-                return false;
+                // Paper start - Perf: Optimize Hoppers
+                return hopperPush(p_155563_, container, direction, p_326256_);
+                // Paper end - Perf: Optimize Hoppers
             }
+        } else {
+            return net.neoforged.neoforge.items.VanillaInventoryCodeHooks.insertHook(p_326256_, containerOrHandler.itemHandler());
         }
     }
 
@@ -215,17 +_,21 @@
     public static boolean suckInItems(Level p_155553_, Hopper p_155554_) {
         BlockPos blockpos = BlockPos.containing(p_155554_.getLevelX(), p_155554_.getLevelY() + 1.0, p_155554_.getLevelZ());
         BlockState blockstate = p_155553_.getBlockState(blockpos);
-        Container container = getSourceContainer(p_155553_, p_155554_, blockpos, blockstate);
-        if (container != null) {
+        var containerOrHandler = getSourceContainerOrHandler(p_155553_, p_155554_, blockpos, blockstate);
+        if (containerOrHandler.container() != null) {
+            Container container = containerOrHandler.container();
             Direction direction = Direction.DOWN;
+            skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
 
             for (int i : getSlots(container, direction)) {
-                if (tryTakeInItemFromSlot(p_155554_, container, i, direction)) {
+                if (tryTakeInItemFromSlot(p_155554_, container, i, direction, p_155553_)) { // Spigot
                     return true;
                 }
             }
 
             return false;
+        } else if (containerOrHandler.itemHandler() != null) {
+            return net.neoforged.neoforge.items.VanillaInventoryCodeHooks.extractHook(p_155554_, containerOrHandler.itemHandler());
         } else {
             boolean flag = p_155554_.isGridAligned()
                 && blockstate.isCollisionShapeFullBlock(p_155553_, blockpos)
@@ -242,20 +_,12 @@
         }
     }
 
-    private static boolean tryTakeInItemFromSlot(Hopper p_59355_, Container p_59356_, int p_59357_, Direction p_59358_) {
+    private static boolean tryTakeInItemFromSlot(Hopper p_59355_, Container p_59356_, int p_59357_, Direction p_59358_, Level level) { // Spigot
         ItemStack itemstack = p_59356_.getItem(p_59357_);
         if (!itemstack.isEmpty() && canTakeItemFromContainer(p_59355_, p_59356_, itemstack, p_59357_, p_59358_)) {
-            int i = itemstack.getCount();
-            ItemStack itemstack1 = addItem(p_59356_, p_59355_, p_59356_.removeItem(p_59357_, 1), null);
-            if (itemstack1.isEmpty()) {
-                p_59356_.setChanged();
-                return true;
-            }
-
-            itemstack.setCount(i);
-            if (i == 1) {
-                p_59356_.setItem(p_59357_, itemstack);
-            }
+            // Paper start - Perf: Optimize Hoppers
+            return hopperPull(level, p_59355_, p_59356_, itemstack, p_59357_);
+            // Paper end - Perf: Optimize Hoppers
         }
 
         return false;
@@ -263,12 +_,22 @@
 
     public static boolean addItem(Container p_59332_, ItemEntity p_59333_) {
         boolean flag = false;
+        // CraftBukkit start
+        if (org.bukkit.event.inventory.InventoryPickupItemEvent.getHandlerList().getRegisteredListeners().length > 0) { // Paper - optimize hoppers
+        org.bukkit.event.inventory.InventoryPickupItemEvent event = new org.bukkit.event.inventory.InventoryPickupItemEvent(
+            getInventory(p_59332_), (org.bukkit.entity.Item) p_59333_.getBukkitEntity() // Paper - Perf: Optimize Hoppers; use getInventory() to avoid snapshot creation
+        );
+        if (!event.callEvent()) {
+            return false;
+        }
+        // CraftBukkit end
+        } // Paper - Perf: Optimize Hoppers
         ItemStack itemstack = p_59333_.getItem().copy();
         ItemStack itemstack1 = addItem(null, p_59332_, itemstack, null);
         if (itemstack1.isEmpty()) {
             flag = true;
             p_59333_.setItem(ItemStack.EMPTY);
-            p_59333_.discard();
+            p_59333_.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
         } else {
             p_59333_.setItem(itemstack1);
         }
@@ -312,11 +_,20 @@
             boolean flag = false;
             boolean flag1 = p_59322_.isEmpty();
             if (itemstack.isEmpty()) {
+                // Spigot start - SPIGOT-6693, SimpleContainer#setItem
+                ItemStack leftover = ItemStack.EMPTY; // Paper - Make hoppers respect inventory max stack size
+                if (!p_59323_.isEmpty() && p_59323_.getCount() > p_59322_.getMaxStackSize()) {
+                    leftover = p_59323_; // Paper - Make hoppers respect inventory max stack size
+                    p_59323_ = p_59323_.split(p_59322_.getMaxStackSize());
+                }
+                // Spigot end
+                ignoreBlockEntityUpdates = true; // Paper - Perf: Optimize Hoppers
                 p_59322_.setItem(p_59324_, p_59323_);
-                p_59323_ = ItemStack.EMPTY;
+                ignoreBlockEntityUpdates = false; // Paper - Perf: Optimize Hoppers
+                p_59323_ = leftover; // Paper - Make hoppers respect inventory max stack size
                 flag = true;
             } else if (canMergeItems(itemstack, p_59323_)) {
-                int i = p_59323_.getMaxStackSize() - itemstack.getCount();
+                int i = Math.min(p_59323_.getMaxStackSize(), p_59322_.getMaxStackSize()) - itemstack.getCount(); // Paper - Make hoppers respect inventory max stack size
                 int j = Math.min(p_59323_.getCount(), i);
                 p_59323_.shrink(j);
                 itemstack.grow(j);
@@ -330,7 +_,7 @@
                         k = 1;
                     }
 
-                    hopperblockentity1.setCooldown(8 - k);
+                    hopperblockentity1.setCooldown(hopperblockentity1.level.spigotConfig.hopperTransfer - k); // Spigot
                 }
 
                 p_59322_.setChanged();
@@ -340,14 +_,59 @@
         return p_59323_;
     }
 
+    // CraftBukkit start
+    @Nullable
+    private static Container runHopperInventorySearchEvent(
+        Container container,
+        org.bukkit.craftbukkit.block.CraftBlock hopper,
+        org.bukkit.craftbukkit.block.CraftBlock searchLocation,
+        org.bukkit.event.inventory.HopperInventorySearchEvent.ContainerType containerType
+    ) {
+        org.bukkit.event.inventory.HopperInventorySearchEvent event = new org.bukkit.event.inventory.HopperInventorySearchEvent(
+            (container != null) ? new org.bukkit.craftbukkit.inventory.CraftInventory(container) : null,
+            containerType,
+            hopper,
+            searchLocation
+        );
+        event.callEvent();
+        return (event.getInventory() != null) ? ((org.bukkit.craftbukkit.inventory.CraftInventory) event.getInventory()).getInventory() : null;
+    }
+    // CraftBukkit end
+
     @Nullable
     private static Container getAttachedContainer(Level p_155593_, BlockPos p_155594_, HopperBlockEntity p_326320_) {
-        return getContainerAt(p_155593_, p_155594_.relative(p_326320_.facing));
+        // Paper start
+        BlockPos searchPosition = p_155594_.relative(p_326320_.facing);
+        Container inventory = getContainerAt(p_155593_, searchPosition);
+        if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0) return inventory;
+
+        org.bukkit.craftbukkit.block.CraftBlock hopper = org.bukkit.craftbukkit.block.CraftBlock.at(p_155593_, p_155594_);
+        org.bukkit.craftbukkit.block.CraftBlock searchBlock = org.bukkit.craftbukkit.block.CraftBlock.at(p_155593_, searchPosition);
+        return HopperBlockEntity.runHopperInventorySearchEvent(
+            inventory,
+            hopper,
+            searchBlock,
+            org.bukkit.event.inventory.HopperInventorySearchEvent.ContainerType.DESTINATION
+        );
+        // Paper end
     }
 
     @Nullable
     private static Container getSourceContainer(Level p_155597_, Hopper p_155598_, BlockPos p_326315_, BlockState p_326093_) {
-        return getContainerAt(p_155597_, p_326315_, p_326093_, p_155598_.getLevelX(), p_155598_.getLevelY() + 1.0, p_155598_.getLevelZ());
+        // Paper start
+        final Container inventory = HopperBlockEntity.getContainerAt(p_155597_, p_326315_, p_326093_, p_155598_.getLevelX(), p_155598_.getLevelY() + 1.0D, p_155598_.getLevelZ());
+        if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0) return inventory;
+
+        final BlockPos hopperPos = BlockPos.containing(p_155598_.getLevelX(), p_155598_.getLevelY(), p_155598_.getLevelZ());
+        org.bukkit.craftbukkit.block.CraftBlock hopperBlock = org.bukkit.craftbukkit.block.CraftBlock.at(p_155597_, hopperPos);
+        org.bukkit.craftbukkit.block.CraftBlock containerBlock = org.bukkit.craftbukkit.block.CraftBlock.at(p_155597_, hopperPos.above());
+        return HopperBlockEntity.runHopperInventorySearchEvent(
+            inventory,
+            hopperBlock,
+            containerBlock,
+            org.bukkit.event.inventory.HopperInventorySearchEvent.ContainerType.SOURCE
+        );
+        // Paper end
     }
 
     public static List<ItemEntity> getItemsAtAndAbove(Level p_155590_, Hopper p_155591_) {
@@ -355,15 +_,21 @@
         return p_155590_.getEntitiesOfClass(ItemEntity.class, aabb, EntitySelector.ENTITY_STILL_ALIVE);
     }
 
+    /** @deprecated Use IItemHandler capability instead. To preserve Container-specific interactions, use {@link #getContainerOrHandlerAt} and handle both cases. */
+    @Deprecated
     @Nullable
     public static Container getContainerAt(Level p_59391_, BlockPos p_59392_) {
-        return getContainerAt(p_59391_, p_59392_, p_59391_.getBlockState(p_59392_), p_59392_.getX() + 0.5, p_59392_.getY() + 0.5, p_59392_.getZ() + 0.5);
+        return getContainerAt(p_59391_, p_59392_, p_59391_.getBlockState(p_59392_), p_59392_.getX() + 0.5, p_59392_.getY() + 0.5, p_59392_.getZ() + 0.5, true); // Paper - Optimize hoppers
     }
 
     @Nullable
     private static Container getContainerAt(Level p_59348_, BlockPos p_326114_, BlockState p_326445_, double p_59349_, double p_59350_, double p_59351_) {
+        return HopperBlockEntity.getContainerAt(p_59348_, p_326114_, p_326445_, p_59349_, p_59350_, p_59351_, false);
+    }
+    @Nullable
+    private static Container getContainerAt(Level p_59348_, BlockPos p_326114_, BlockState p_326445_, double p_59349_, double p_59350_, double p_59351_, final boolean optimizeEntities) {
         Container container = getBlockContainer(p_59348_, p_326114_, p_326445_);
-        if (container == null) {
+        if (container == null && (!optimizeEntities || !p_59348_.paperConfig().hopper.ignoreOccludingBlocks || !p_326445_.getBukkitMaterial().isOccluding())) { // Paper - Perf: Optimize Hoppers
             container = getEntityContainer(p_59348_, p_59349_, p_59350_, p_59351_);
         }
 
@@ -372,6 +_,7 @@
 
     @Nullable
     private static Container getBlockContainer(Level p_326127_, BlockPos p_326017_, BlockState p_326108_) {
+        if (!p_326127_.spigotConfig.hopperCanLoadChunks && !p_326127_.hasChunkAt(p_326017_)) return null; // Spigot
         Block block = p_326108_.getBlock();
         if (block instanceof WorldlyContainerHolder) {
             return ((WorldlyContainerHolder)block).getContainer(p_326108_, p_326127_, p_326017_);
@@ -388,16 +_,38 @@
 
     @Nullable
     private static Container getEntityContainer(Level p_326325_, double p_326012_, double p_326191_, double p_326289_) {
-        List<Entity> list = p_326325_.getEntities(
-            (Entity)null,
+        List<Entity> list = p_326325_.getEntitiesOfClass(
+            (Class) Container.class,
             new AABB(p_326012_ - 0.5, p_326191_ - 0.5, p_326289_ - 0.5, p_326012_ + 0.5, p_326191_ + 0.5, p_326289_ + 0.5),
-            EntitySelector.CONTAINER_ENTITY_SELECTOR
+            EntitySelector.CONTAINER_ENTITY_SELECTOR // Paper - Perf: Optimize hoppers
         );
         return !list.isEmpty() ? (Container)list.get(p_326325_.random.nextInt(list.size())) : null;
     }
 
+    private static net.neoforged.neoforge.items.ContainerOrHandler getSourceContainerOrHandler(Level p_155597_, Hopper p_155598_, BlockPos p_326315_, BlockState p_326093_) {
+        return getContainerOrHandlerAt(p_155597_, p_326315_, p_326093_, p_155598_.getLevelX(), p_155598_.getLevelY() + 1.0, p_155598_.getLevelZ(), Direction.DOWN);
+    }
+
+    public static net.neoforged.neoforge.items.ContainerOrHandler getContainerOrHandlerAt(Level level, BlockPos pos, @Nullable Direction side) {
+        return getContainerOrHandlerAt(
+                level, pos, level.getBlockState(pos), (double)pos.getX() + 0.5, (double)pos.getY() + 0.5, (double)pos.getZ() + 0.5, side
+        );
+    }
+
+    private static net.neoforged.neoforge.items.ContainerOrHandler getContainerOrHandlerAt(Level level, BlockPos pos, BlockState state, double x, double y, double z, @Nullable Direction side) {
+        Container container = getBlockContainer(level, pos, state);
+        if (container != null) {
+            return new net.neoforged.neoforge.items.ContainerOrHandler(container, null);
+        }
+        var blockItemHandler = level.getCapability(net.neoforged.neoforge.capabilities.Capabilities.ItemHandler.BLOCK, pos, state, null, side);
+        if (blockItemHandler != null) {
+            return new net.neoforged.neoforge.items.ContainerOrHandler(null, blockItemHandler);
+        }
+        return net.neoforged.neoforge.items.VanillaInventoryCodeHooks.getEntityContainerOrHandler(level, x, y, z, side);
+    }
+
     private static boolean canMergeItems(ItemStack p_59345_, ItemStack p_59346_) {
-        return p_59345_.getCount() <= p_59345_.getMaxStackSize() && ItemStack.isSameItemSameComponents(p_59345_, p_59346_);
+        return p_59345_.getCount() < p_59345_.getMaxStackSize() && ItemStack.isSameItemSameComponents(p_59345_, p_59346_); // Paper - Perf: Optimize Hoppers; used to return true for full itemstacks?!
     }
 
     @Override
@@ -453,5 +_,9 @@
     @Override
     protected AbstractContainerMenu createMenu(int p_59312_, Inventory p_59313_) {
         return new HopperMenu(p_59312_, p_59313_, this);
+    }
+
+    public long getLastUpdateTime() {
+        return this.tickedGameTime;
     }
 }
