--- a/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -86,6 +_,13 @@
 
     @Nullable
     public static ServerPlayer tryGetPlayer(@Nullable Entity p_222862_) {
+    // Paper start - check global player list where appropriate; ensure level is the same for sculk events
+        final ServerPlayer player = tryGetPlayer0(p_222862_);
+        return player != null && player.level() == p_222862_.level() ? player : null;
+    }
+    @Nullable
+    private static ServerPlayer tryGetPlayer0(@Nullable Entity p_222862_) {
+    // Paper end - check global player list where appropriate
         if (p_222862_ instanceof ServerPlayer serverplayer1) {
             return serverplayer1;
         } else if (p_222862_ != null && p_222862_.getControllingPassenger() instanceof ServerPlayer serverplayer) {
@@ -162,7 +_,7 @@
         return this.warningLevel < 4
             ? false
             : SpawnUtil.trySpawnMob(
-                    EntityType.WARDEN, EntitySpawnReason.TRIGGERED, p_222881_, this.getBlockPos(), 20, 5, 6, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER, false
+                    EntityType.WARDEN, EntitySpawnReason.TRIGGERED, p_222881_, this.getBlockPos(), 20, 5, 6, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER, false, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL, null // Paper - Entity#getEntitySpawnReason
                 )
                 .isPresent();
     }
