--- a/net/minecraft/world/level/block/entity/TestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TestBlockEntity.java
@@ -38,6 +_,7 @@
 
     @Override
     protected void loadAdditional(ValueInput p_422212_) {
+        super.loadAdditional(p_422212_); // Paper - load the PDC
         this.mode = p_422212_.read("mode", TestBlockMode.CODEC).orElse(TestBlockMode.FAIL);
         this.message = p_422212_.getStringOr("message", "");
         this.powered = p_422212_.getBooleanOr("powered", false);
