--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -93,7 +_,10 @@
     List<SerializableChunkData.SectionData> sectionData,
     List<CompoundTag> entities,
     List<CompoundTag> blockEntities,
-    CompoundTag structureData
+    CompoundTag structureData,
+    @Nullable CompoundTag attachmentData,
+    @Nullable ListTag auxLightData
+    , @Nullable net.minecraft.nbt.Tag persistentDataContainer // CraftBukkit - persistentDataContainer
 ) {
     private static final Codec<PalettedContainer<BlockState>> BLOCK_STATE_CODEC = PalettedContainer.codecRW(
         Block.BLOCK_STATE_REGISTRY, BlockState.CODEC, PalettedContainer.Strategy.SECTION_STATES, Blocks.AIR.defaultBlockState()
@@ -112,8 +_,62 @@
     public static final String BLOCK_LIGHT_TAG = "BlockLight";
     public static final String SKY_LIGHT_TAG = "SkyLight";
 
+    // Paper start - guard against serializing mismatching coordinates
+    // TODO Note: This needs to be re-checked each update
+    public static ChunkPos getChunkCoordinate(final CompoundTag chunkData) {
+        final int dataVersion = ChunkStorage.getVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompoundOrEmpty("Level");
+            return new ChunkPos(levelData.getIntOr("xPos", 0), levelData.getIntOr("zPos", 0));
+        } else {
+            return new ChunkPos(chunkData.getIntOr("xPos", 0), chunkData.getIntOr("zPos", 0));
+        }
+    }
+    // Paper end - guard against serializing mismatching coordinates
+
+    // Paper start - Attempt to recalculate regionfile header if it is corrupt
+    // TODO: Check on update
+    public static long getLastWorldSaveTime(final CompoundTag chunkData) {
+        final int dataVersion = ChunkStorage.getVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompoundOrEmpty("Level");
+            return levelData.getLongOr("LastUpdate", 0L);
+        } else {
+            return chunkData.getLongOr("LastUpdate", 0L);
+        }
+    }
+    // Paper end - Attempt to recalculate regionfile header if it is corrupt
+
+    /**
+     * @deprecated Neo: use constructor with additional data instead
+     */
+    @Deprecated
+    SerializableChunkData(
+            Registry<Biome> biomeRegistry,
+            ChunkPos chunkPos,
+            int minSectionY,
+            long lastUpdateTime,
+            long inhabitedTime,
+            ChunkStatus chunkStatus,
+            @Nullable BlendingData.Packed blendingData,
+            @Nullable BelowZeroRetrogen belowZeroRetrogen,
+            UpgradeData upgradeData,
+            @Nullable long[] carvingMask,
+            Map<Heightmap.Types, long[]> heightmaps,
+            ChunkAccess.PackedTicks packedTicks,
+            ShortList[] postProcessingSections,
+            boolean lightCorrect,
+            List<SerializableChunkData.SectionData> sectionData,
+            List<CompoundTag> entities,
+            List<CompoundTag> blockEntities,
+            CompoundTag structureData
+    ) {
+        this(biomeRegistry, chunkPos, minSectionY, lastUpdateTime, inhabitedTime, chunkStatus, blendingData, belowZeroRetrogen, upgradeData, carvingMask, heightmaps, packedTicks, postProcessingSections, lightCorrect, sectionData, entities, blockEntities, structureData, null, null);
+    }
+
     @Nullable
     public static SerializableChunkData parse(LevelHeightAccessor p_361938_, RegistryAccess p_365010_, CompoundTag p_362040_) {
+        net.minecraft.server.level.ServerLevel serverLevel = (net.minecraft.server.level.ServerLevel) p_361938_; // Paper - Anti-Xray This is is seemingly only called from ChunkMap, where, we have a server level. We'll fight this later if needed.
         if (p_362040_.getString("Status").isEmpty()) {
             return null;
         } else {
@@ -155,7 +_,7 @@
             ListTag listtag2 = p_362040_.getListOrEmpty("sections");
             List<SerializableChunkData.SectionData> list2 = new ArrayList<>(listtag2.size());
             Registry<Biome> registry = p_365010_.lookupOrThrow(Registries.BIOME);
-            Codec<PalettedContainerRO<Holder<Biome>>> codec = makeBiomeCodec(registry);
+            Codec<PalettedContainer<Holder<Biome>>> codec = makeBiomeCodecRW(registry); // CraftBukkit - read/write
 
             for (int i1 = 0; i1 < listtag2.size(); i1++) {
                 Optional<CompoundTag> optional = listtag2.getCompound(i1);
@@ -164,18 +_,20 @@
                     int j1 = compoundtag.getByteOr("Y", (byte)0);
                     LevelChunkSection levelchunksection;
                     if (j1 >= p_361938_.getMinSectionY() && j1 <= p_361938_.getMaxSectionY()) {
+                        final BlockState[] presetBlockStates = serverLevel.chunkPacketBlockController.getPresetBlockStates(serverLevel, chunkpos, j1); // Paper - Anti-Xray - Add preset block states
+                        final Codec<PalettedContainer<BlockState>> blockStateCodec = presetBlockStates == null ? BLOCK_STATE_CODEC : PalettedContainer.codecRW(Block.BLOCK_STATE_REGISTRY, BlockState.CODEC, PalettedContainer.Strategy.SECTION_STATES, Blocks.AIR.defaultBlockState(), presetBlockStates); // Paper - Anti-Xray
                         PalettedContainer<BlockState> palettedcontainer = compoundtag.getCompound("block_states")
                             .map(
-                                p_409528_ -> BLOCK_STATE_CODEC.parse(NbtOps.INSTANCE, p_409528_)
+                                p_409528_ -> blockStateCodec.parse(NbtOps.INSTANCE, p_409528_) // Paper - Anti-Xray
                                     .promotePartial(p_361842_ -> logErrors(chunkpos, j1, p_361842_))
                                     .getOrThrow(SerializableChunkData.ChunkReadException::new)
                             )
                             .orElseGet(
                                 () -> new PalettedContainer<>(
-                                    Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState(), PalettedContainer.Strategy.SECTION_STATES
+                                    Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState(), PalettedContainer.Strategy.SECTION_STATES, presetBlockStates // Paper - Anti-Xray - Add preset biomes
                                 )
                             );
-                        PalettedContainerRO<Holder<Biome>> palettedcontainerro = compoundtag.getCompound("biomes")
+                        PalettedContainer<Holder<Biome>> palettedcontainerro = compoundtag.getCompound("biomes") // CraftBukkit - read/write
                             .map(
                                 p_409533_ -> codec.parse(NbtOps.INSTANCE, p_409533_)
                                     .promotePartial(p_361282_ -> logErrors(chunkpos, j1, p_361282_))
@@ -183,7 +_,7 @@
                             )
                             .orElseGet(
                                 () -> new PalettedContainer<>(
-                                    registry.asHolderIdMap(), registry.getOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES
+                                    registry.asHolderIdMap(), registry.getOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES, null // Paper - Anti-Xray - Add preset biomes
                                 )
                             );
                         levelchunksection = new LevelChunkSection(palettedcontainer, palettedcontainerro);
@@ -197,6 +_,10 @@
                 }
             }
 
+            CompoundTag attachmentData = p_362040_.getCompound(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY).orElse(null);
+            ListTag auxLightData = p_362040_.getList(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY).orElse(null);
+            net.minecraft.nbt.Tag chunkPersistentDataContainer = p_362040_.get("ChunkBukkitValues"); // CraftBukkit - persistentDataContainer
+
             return new SerializableChunkData(
                 registry,
                 chunkpos,
@@ -215,7 +_,10 @@
                 list2,
                 list3,
                 list4,
-                compoundtag1
+                compoundtag1,
+                attachmentData,
+                auxLightData
+                , chunkPersistentDataContainer // CraftBukkit - persistentDataContainer
             );
         }
     }
@@ -275,6 +_,9 @@
                 postLoadChunk(p_360452_, this.entities, this.blockEntities),
                 BlendingData.unpack(this.blendingData)
             );
+            if (this.auxLightData != null) {
+                ((LevelChunk) chunkaccess).getAuxLightManager(chunkPos).deserializeNBT(this.auxLightData);
+            }
         } else {
             ProtoChunkTicks<Block> protochunkticks = ProtoChunkTicks.load(this.packedTicks.blocks());
             ProtoChunkTicks<Fluid> protochunkticks1 = ProtoChunkTicks.load(this.packedTicks.fluids());
@@ -293,6 +_,12 @@
             }
         }
 
+        // CraftBukkit start - load chunk persistent data from nbt - SPIGOT-6814: Already load PDC here to account for 1.17 to 1.18 chunk upgrading.
+        if (this.persistentDataContainer instanceof CompoundTag compoundTag) {
+            chunkaccess.persistentDataContainer.putAll(compoundTag);
+        }
+        // CraftBukkit end
+
         chunkaccess.setLightCorrect(this.lightCorrect);
         EnumSet<Heightmap.Types> enumset = EnumSet.noneOf(Heightmap.Types.class);
 
@@ -313,6 +_,10 @@
             chunkaccess.addPackedPostProcess(this.postProcessingSections[j], j);
         }
 
+        if (this.attachmentData != null) {
+            chunkaccess.readAttachmentsFromNBT(p_360452_.registryAccess(), this.attachmentData);
+        }
+
         if (chunktype == ChunkType.LEVELCHUNK) {
             return new ImposterProtoChunk((LevelChunk)chunkaccess, false);
         } else {
@@ -344,6 +_,12 @@
         );
     }
 
+    // CraftBukkit start - read/write
+    private static Codec<PalettedContainer<Holder<Biome>>> makeBiomeCodecRW(Registry<Biome> biomeRegistry) {
+        return PalettedContainer.codecRW(biomeRegistry.asHolderIdMap(), biomeRegistry.holderByNameCodec(), PalettedContainer.Strategy.SECTION_BIOMES, biomeRegistry.getOrThrow(Biomes.PLAINS), null); // Paper - Anti-Xray - Add preset biomes
+    }
+    // CraftBukkit end
+
     public static SerializableChunkData copyOf(ServerLevel p_365319_, ChunkAccess p_362284_) {
         if (!p_362284_.canBeSerialized()) {
             throw new IllegalArgumentException("Chunk can't be serialized: " + p_362284_);
@@ -402,6 +_,23 @@
             CompoundTag compoundtag1 = packStructureData(
                 StructurePieceSerializationContext.fromLevel(p_365319_), chunkpos, p_362284_.getAllStarts(), p_362284_.getAllReferences()
             );
+
+            CompoundTag attachmentData = null;
+            try {
+                attachmentData = p_362284_.writeAttachmentsToNBT(p_365319_.registryAccess());
+            } catch (Exception exception) {
+                LOGGER.error("Failed to write chunk attachments. An attachment has likely thrown an exception trying to write state. It will not persist. Report this to the mod author", exception);
+            }
+            ListTag auxLightData = null;
+            if (p_362284_ instanceof LevelChunk levelChunk) {
+                auxLightData = levelChunk.getAuxLightManager(chunkpos).serializeNBT();
+            }
+            // CraftBukkit start - store chunk persistent data in nbt
+            CompoundTag persistentDataContainer = null;
+            if (!p_362284_.persistentDataContainer.isEmpty()) { // SPIGOT-6814: Always save PDC to account for 1.17 to 1.18 chunk upgrading.
+                persistentDataContainer = p_362284_.persistentDataContainer.toTagCompound();
+            }
+            // CraftBukkit end
             return new SerializableChunkData(
                 p_365319_.registryAccess().lookupOrThrow(Registries.BIOME),
                 chunkpos,
@@ -420,7 +_,10 @@
                 list,
                 list2,
                 list1,
-                compoundtag1
+                compoundtag1,
+                attachmentData,
+                auxLightData
+                , persistentDataContainer // CraftBukkit - persistentDataContainer
             );
         }
     }
@@ -430,7 +_,7 @@
         compoundtag.putInt("xPos", this.chunkPos.x);
         compoundtag.putInt("yPos", this.minSectionY);
         compoundtag.putInt("zPos", this.chunkPos.z);
-        compoundtag.putLong("LastUpdate", this.lastUpdateTime);
+        compoundtag.putLong("LastUpdate", this.lastUpdateTime); // Paper - Diff on change
         compoundtag.putLong("InhabitedTime", this.inhabitedTime);
         compoundtag.putString("Status", BuiltInRegistries.CHUNK_STATUS.getKey(this.chunkStatus).toString());
         compoundtag.storeNullable("blending_data", BlendingData.Packed.CODEC, this.blendingData);
@@ -487,6 +_,19 @@
         this.heightmaps.forEach((p_362472_, p_363515_) -> compoundtag2.put(p_362472_.getSerializationKey(), new LongArrayTag(p_363515_)));
         compoundtag.put("Heightmaps", compoundtag2);
         compoundtag.put("structures", this.structureData);
+        // CraftBukkit start - store chunk persistent data in nbt
+        if (this.persistentDataContainer != null) { // SPIGOT-6814: Always save PDC to account for 1.17 to 1.18 chunk upgrading.
+            compoundtag.put("ChunkBukkitValues", this.persistentDataContainer);
+        }
+        // CraftBukkit end
+
+        if (attachmentData != null) {
+            compoundtag.put(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY, attachmentData);
+        }
+        if (auxLightData != null) {
+            compoundtag.put(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, auxLightData);
+        }
+
         return compoundtag;
     }
 
@@ -568,6 +_,12 @@
             } else {
                 StructureStart structurestart = StructureStart.loadStaticStart(p_360982_, compoundtag.getCompoundOrEmpty(s), p_362695_);
                 if (structurestart != null) {
+                    // CraftBukkit start - load persistent data for structure start
+                    net.minecraft.nbt.Tag persistentBase = compoundtag.getCompoundOrEmpty(s).get("StructureBukkitValues");
+                    if (persistentBase instanceof CompoundTag compoundTag) {
+                        structurestart.persistentDataContainer.putAll(compoundTag);
+                    }
+                    // CraftBukkit end
                     map.put(structure, structurestart);
                 }
             }
