--- a/net/minecraft/world/entity/monster/Bogged.java
+++ b/net/minecraft/world/entity/monster/Bogged.java
@@ -43,6 +_,31 @@
         super(p_326920_, p_326946_);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.boggedRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.boggedRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.boggedControllable;
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.boggedMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.boggedScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
     @Override
     protected void defineSynchedData(SynchedEntityData.Builder p_331269_) {
         super.defineSynchedData(p_331269_);
@@ -72,7 +_,7 @@
     @Override
     protected InteractionResult mobInteract(Player p_330736_, InteractionHand p_331786_) {
         ItemStack itemstack = p_330736_.getItemInHand(p_331786_);
-        if (itemstack.is(Items.SHEARS) && this.readyForShearing()) {
+        if (false && itemstack.is(Items.SHEARS) && this.readyForShearing()) { // Neo: Shear logic is handled by IShearable
             if (this.level() instanceof ServerLevel serverlevel) {
                 this.shear(serverlevel, SoundSource.PLAYERS, itemstack);
                 this.gameEvent(GameEvent.SHEAR, p_330736_);
@@ -86,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.BOGGED_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_326909_) {
+    public SoundEvent getHurtSound(DamageSource p_326909_) {
         return SoundEvents.BOGGED_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.BOGGED_DEATH;
     }
 
@@ -127,15 +_,33 @@
 
     @Override
     public void shear(ServerLevel p_376748_, SoundSource p_331493_, ItemStack p_373125_) {
+        // Paper start - custom shear drops
+        this.shear(p_376748_, p_331493_, p_373125_, this.generateDefaultDrops(p_376748_, p_373125_));
+    }
+
+    @Override
+    public java.util.List<ItemStack> generateDefaultDrops(final ServerLevel serverLevel, final ItemStack p_373125_) {
+        final java.util.List<ItemStack> drops = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        this.dropFromShearingLootTable(serverLevel, BuiltInLootTables.BOGGED_SHEAR, p_373125_, (ignored, stack) -> {
+            drops.add(stack);
+        });
+        return drops;
+    }
+
+    @Override
+    public void shear(ServerLevel p_376748_, SoundSource p_331493_, ItemStack p_373125_, java.util.List<ItemStack> drops) {
+        // Paper end - custom shear drops
         p_376748_.playSound(null, this, SoundEvents.BOGGED_SHEAR, p_331493_, 1.0F, 1.0F);
-        this.spawnShearedMushrooms(p_376748_, p_373125_);
+        this.spawnShearedMushrooms(p_376748_, p_373125_, drops); // Paper - custom shear drops
         this.setSheared(true);
     }
 
-    private void spawnShearedMushrooms(ServerLevel p_376846_, ItemStack p_372900_) {
-        this.dropFromShearingLootTable(
-            p_376846_, BuiltInLootTables.BOGGED_SHEAR, p_372900_, (p_427097_, p_427098_) -> this.spawnAtLocation(p_427097_, p_427098_, this.getBbHeight())
-        );
+    // Paper start - custom shear drops
+    private void spawnShearedMushrooms(ServerLevel level, ItemStack stack, java.util.List<ItemStack> drops) {
+        this.forceDrops = true; // Paper - Add missing forceDrop toggles
+        drops.forEach(drop -> this.spawnAtLocation(level, drop, this.getBbHeight()));
+        this.forceDrops = false; // Paper - Add missing forceDrop toggles
+        // Paper end - custom shear drops
     }
 
     @Override
