--- a/net/minecraft/world/entity/monster/Phantom.java
+++ b/net/minecraft/world/entity/monster/Phantom.java
@@ -49,13 +_,110 @@
     @Nullable
     public BlockPos anchorPoint;
     Phantom.AttackPhase attackPhase = Phantom.AttackPhase.CIRCLE;
+    Vec3 crystalPosition; // Purpur - Phantoms attracted to crystals and crystals shoot phantoms
+    // Paper start
+    @Nullable
+    public java.util.UUID spawningEntity;
+    //public boolean shouldBurnInDay = true; // Purpur - API for any mob to burn daylight
+    // Paper end
+    private static final net.minecraft.world.item.crafting.Ingredient TORCH = net.minecraft.world.item.crafting.Ingredient.of(net.minecraft.world.item.Items.TORCH, net.minecraft.world.item.Items.SOUL_TORCH); // Purpur - Phantoms burn in light
 
     public Phantom(EntityType<? extends Phantom> p_33101_, Level p_33102_) {
         super(p_33101_, p_33102_);
         this.xpReward = 5;
         this.moveControl = new Phantom.PhantomMoveControl(this);
         this.lookControl = new Phantom.PhantomLookControl(this);
-    }
+        this.setShouldBurnInDay(true); // Purpur - API for any mob to burn daylight
+    }
+
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.phantomRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.phantomRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.phantomControllable;
+    }
+
+    @Override
+    public double getMaxY() {
+        return level().purpurConfig.phantomMaxY;
+    }
+
+    public static net.minecraft.world.entity.ai.attributes.AttributeSupplier.Builder createAttributes() {
+        return Monster.createMonsterAttributes().add(Attributes.FLYING_SPEED, 3.0D);
+    }
+
+    @Override
+    public boolean onSpacebar() {
+        if (getRider() != null && getRider().getBukkitEntity().hasPermission("allow.special.phantom")) {
+            shoot();
+        }
+        return false;
+    }
+
+    public boolean shoot() {
+        org.bukkit.Location loc = ((org.bukkit.entity.LivingEntity) getBukkitEntity()).getEyeLocation();
+        loc.setPitch(-loc.getPitch());
+        org.bukkit.util.Vector target = loc.getDirection().normalize().multiply(100).add(loc.toVector());
+
+        org.purpurmc.purpur.entity.projectile.PhantomFlames flames = new org.purpurmc.purpur.entity.projectile.PhantomFlames(level(), this);
+        flames.canGrief = level().purpurConfig.phantomAllowGriefing;
+        flames.shoot(target.getX() - getX(), target.getY() - getY(), target.getZ() - getZ(), 1.0F, 5.0F);
+        level().addFreshEntity(flames);
+        return true;
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Phantoms attracted to crystals and crystals shoot phantoms
+    @Override
+    protected void dropFromLootTable(ServerLevel world, DamageSource damageSource, boolean causedByPlayer) {
+        boolean dropped = false;
+        if (lastHurtByPlayer == null && damageSource.getEntity() instanceof net.minecraft.world.entity.boss.enderdragon.EndCrystal) {
+            if (random.nextInt(5) < 1) {
+                dropped = spawnAtLocation(world, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.PHANTOM_MEMBRANE)) != null;
+            }
+        }
+        if (!dropped) {
+            super.dropFromLootTable(world, damageSource, causedByPlayer);
+        }
+    }
+
+    public boolean isCirclingCrystal() {
+        return crystalPosition != null;
+    }
+    // Purpur end - Phantoms attracted to crystals and crystals shoot phantoms
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.phantomTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    //private boolean shouldBurnInDay = true; // Purpur - moved to LivingEntity; keep methods for ABI compatibility - API for any mob to burn daylight
+    // Purpur start - API for any mob to burn daylight
+    public boolean shouldBurnInDay() {
+        boolean burnFromDaylight = this.shouldBurnInDay && this.level().purpurConfig.phantomBurnInDaylight;
+        boolean burnFromLightSource = this.level().purpurConfig.phantomBurnInLight > 0 && this.level().getMaxLocalRawBrightness(blockPosition()) >= this.level().purpurConfig.phantomBurnInLight;
+        return burnFromDaylight || burnFromLightSource;
+    }
+    public void setShouldBurnInDay(boolean shouldBurnInDay) { this.shouldBurnInDay = shouldBurnInDay; }
+    // Purpur end - API for any mob to burn daylight
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.phantomAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
 
     @Override
     public boolean isFlapping() {
@@ -69,9 +_,17 @@
 
     @Override
     protected void registerGoals() {
-        this.goalSelector.addGoal(1, new Phantom.PhantomAttackStrategyGoal());
-        this.goalSelector.addGoal(2, new Phantom.PhantomSweepAttackGoal());
-        this.goalSelector.addGoal(3, new Phantom.PhantomCircleAroundAnchorGoal());
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
+        // Purpur start - Phantoms attracted to crystals and crystals shoot phantoms
+        if (level().purpurConfig.phantomOrbitCrystalRadius > 0) {
+            this.goalSelector.addGoal(1, new PhantomFindCrystalGoal(this));
+            this.goalSelector.addGoal(2, new PhantomOrbitCrystalGoal(this));
+        }
+        this.goalSelector.addGoal(3, new Phantom.PhantomAttackStrategyGoal());
+        this.goalSelector.addGoal(4, new Phantom.PhantomSweepAttackGoal());
+        this.goalSelector.addGoal(5, new Phantom.PhantomCircleAroundAnchorGoal());
+        // Purpur end - Phantoms attracted to crystals and crystals shoot phantoms
+        this.targetSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.targetSelector.addGoal(1, new Phantom.PhantomAttackPlayerTargetGoal());
     }
 
@@ -87,7 +_,11 @@
 
     private void updatePhantomSizeInfo() {
         this.refreshDimensions();
-        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(6 + this.getPhantomSize());
+        if (level().purpurConfig.phantomFlamesOnSwoop && attackPhase == AttackPhase.SWOOP) shoot(); // Purpur - Ridables - Phantom flames on swoop
+        // Purpur start - Configurable entity base attributes
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(getFromCache(() -> this.level().purpurConfig.phantomMaxHealth, () -> this.level().purpurConfig.phantomMaxHealthCache, () -> 20.0D));
+        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(getFromCache(() -> this.level().purpurConfig.phantomAttackDamage, () -> this.level().purpurConfig.phantomAttackDamageCache, () -> (double) (6 + this.getPhantomSize())));
+        // Purpur end - Configurable entity base attributes
     }
 
     public int getPhantomSize() {
@@ -112,6 +_,23 @@
         return true;
     }
 
+    // Purpur start - Configurable entity base attributes
+    private double getFromCache(java.util.function.Supplier<String> equation, java.util.function.Supplier<java.util.Map<Integer, Double>> cache, java.util.function.Supplier<Double> defaultValue) {
+        int size = getPhantomSize();
+        Double value = cache.get().get(size);
+        if (value == null) {
+            try {
+                value = ((Number) scriptEngine.eval("let size = " + size + "; " + equation.get())).doubleValue();
+            } catch (javax.script.ScriptException e) {
+                e.printStackTrace();
+                value = defaultValue.get();
+            }
+            cache.get().put(size, value);
+        }
+        return value;
+    }
+    // Purpur end - Configurable entity base attributes
+
     @Override
     public void tick() {
         super.tick();
@@ -143,10 +_,7 @@
 
     @Override
     public void aiStep() {
-        if (this.isAlive() && this.isSunBurnTick()) {
-            this.igniteForSeconds(8.0F);
-        }
-
+        // Purpur - implemented in LivingEntity; moved down to shouldBurnInDay() - API for any mob to burn daylight
         super.aiStep();
     }
 
@@ -160,17 +_,30 @@
     }
 
     @Override
-    public void travel(Vec3 p_416093_) {
-        this.travelFlying(p_416093_, 0.2F);
+    public void travel(Vec3 travelVector) {
+        this.travelFlying(travelVector, 0.2F);
+        // Purpur start - Ridables
+        if (this.getRider() != null && this.isControllable() && !this.onGround) {
+            float speed = (float) this.getAttributeValue(Attributes.FLYING_SPEED);
+            this.setSpeed(speed);
+            Vec3 mot = this.getDeltaMovement();
+            this.move(net.minecraft.world.entity.MoverType.SELF, mot.multiply(speed, speed, speed));
+            this.setDeltaMovement(mot.scale(0.9D));
+        }
+        // Purpur end - Ridables
     }
 
     @Override
     public SpawnGroupData finalizeSpawn(
-        ServerLevelAccessor p_33126_, DifficultyInstance p_33127_, EntitySpawnReason p_364084_, @Nullable SpawnGroupData p_33129_
+        ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData spawnGroupData
     ) {
         this.anchorPoint = this.blockPosition().above(5);
-        this.setPhantomSize(0);
-        return super.finalizeSpawn(p_33126_, p_33127_, p_364084_, p_33129_);
+        // Purpur start - Configurable phantom size
+        int min = level.getLevel().purpurConfig.phantomMinSize;
+        int max = level.getLevel().purpurConfig.phantomMaxSize;
+        this.setPhantomSize(min == max ? min : level.getRandom().nextInt(max + 1 - min) + min);
+        // Purpur end - Configurable phantom size
+        return super.finalizeSpawn(level, difficulty, spawnReason, spawnGroupData);
     }
 
     @Override
@@ -178,6 +_,10 @@
         super.readAdditionalSaveData(p_421906_);
         this.anchorPoint = p_421906_.read("anchor_pos", BlockPos.CODEC).orElse(null);
         this.setPhantomSize(p_421906_.getIntOr("size", 0));
+        // Paper start
+        this.spawningEntity = p_421906_.read("Paper.SpawningEntity", net.minecraft.core.UUIDUtil.CODEC).orElse(null);
+        //this.shouldBurnInDay = p_421906_.getBooleanOr("Paper.ShouldBurnInDay", true); // Purpur - implemented in LivingEntity - API for any mob to burn daylight
+        // Paper end
     }
 
     @Override
@@ -185,6 +_,10 @@
         super.addAdditionalSaveData(p_422420_);
         p_422420_.storeNullable("anchor_pos", BlockPos.CODEC, this.anchorPoint);
         p_422420_.putInt("size", this.getPhantomSize());
+        // Paper start
+        p_422420_.storeNullable("Paper.SpawningEntity", net.minecraft.core.UUIDUtil.CODEC, this.spawningEntity);
+        //p_422420_.putBoolean("Paper.ShouldBurnInDay", this.shouldBurnInDay); // Purpur - implemented in LivingEntity - API for any mob to burn daylight
+        // Paper end
     }
 
     @Override
@@ -203,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_33152_) {
+    public SoundEvent getHurtSound(DamageSource p_33152_) { // Youer - fix: LivingEntity.getHurtSound is public, can't narrow to protected in override
         return SoundEvents.PHANTOM_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() { // Youer - fix: LivingEntity.getDeathSound is public, can't narrow to protected in override
         return SoundEvents.PHANTOM_DEATH;
     }
 
@@ -250,14 +_,19 @@
             } else {
                 this.nextScanTick = reducedTickDelay(60);
                 ServerLevel serverlevel = getServerLevel(Phantom.this.level());
-                List<Player> list = serverlevel.getNearbyPlayers(this.attackTargeting, Phantom.this, Phantom.this.getBoundingBox().inflate(16.0, 64.0, 16.0));
+                List<Player> list = serverlevel.getNearbyPlayers(
+                    this.attackTargeting, Phantom.this, Phantom.this.getBoundingBox().inflate(16.0, 64.0, 16.0)
+                );
+                if (level().purpurConfig.phantomIgnorePlayersWithTorch) list.removeIf(human -> TORCH.test(human.getItemInHand(net.minecraft.world.InteractionHand.MAIN_HAND)) || TORCH.test(human.getItemInHand(net.minecraft.world.InteractionHand.OFF_HAND))); // Purpur - Phantoms burn in light
                 if (!list.isEmpty()) {
                     list.sort(Comparator.<Player, Double>comparing(Entity::getY).reversed());
 
                     for (Player player : list) {
                         if (Phantom.this.canAttack(serverlevel, player, TargetingConditions.DEFAULT)) {
-                            Phantom.this.setTarget(player);
+                            if (!level().paperConfig().entities.behavior.phantomsOnlyAttackInsomniacs || EntitySelector.IS_INSOMNIAC.test(player)) { // Paper - Add phantom creative and insomniac controls
+                            Phantom.this.setTarget(player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER); // CraftBukkit - reason
                             return true;
+                            } // Paper - Add phantom creative and insomniac controls
                         }
                     }
                 }
@@ -399,60 +_,195 @@
         }
     }
 
-    static class PhantomLookControl extends LookControl {
+    static class PhantomLookControl extends org.purpurmc.purpur.controller.LookControllerWASD { // Purpur - Ridables
         public PhantomLookControl(Mob p_33235_) {
             super(p_33235_);
         }
 
+        // Purpur start - Ridables
+        public void purpurTick(Player rider) {
+            setYawPitch(rider.getYRot(), -rider.xRotO * 0.75F);
+        }
+        // Purpur end - Ridables
+
+        @Override
+        public void vanillaTick() { // Purpur - Ridables
+        }
+    }
+
+    // Purpur start - Phantoms attracted to crystals and crystals shoot phantoms
+    class PhantomFindCrystalGoal extends Goal {
+        private final Phantom phantom;
+        private net.minecraft.world.entity.boss.enderdragon.EndCrystal crystal;
+        private Comparator<net.minecraft.world.entity.boss.enderdragon.EndCrystal> comparator;
+
+        PhantomFindCrystalGoal(Phantom phantom) {
+            this.phantom = phantom;
+            this.comparator = Comparator.comparingDouble(phantom::distanceToSqr);
+            this.setFlags(EnumSet.of(Flag.LOOK));
+        }
+
+        @Override
+        public boolean canUse() {
+            double range = maxTargetRange();
+            List<net.minecraft.world.entity.boss.enderdragon.EndCrystal> crystals = level().getEntitiesOfClass(net.minecraft.world.entity.boss.enderdragon.EndCrystal.class, phantom.getBoundingBox().inflate(range));
+            if (crystals.isEmpty()) {
+                return false;
+            }
+            crystals.sort(comparator);
+            crystal = crystals.get(0);
+            if (phantom.distanceToSqr(crystal) > range * range) {
+                crystal = null;
+                return false;
+            }
+            return true;
+        }
+
+        @Override
+        public boolean canContinueToUse() {
+            if (crystal == null || !crystal.isAlive()) {
+                return false;
+            }
+            double range = maxTargetRange();
+            return phantom.distanceToSqr(crystal) <= (range * range) * 2;
+        }
+
+        @Override
+        public void start() {
+            phantom.crystalPosition = new Vec3(crystal.getX(), crystal.getY() + (phantom.random.nextInt(10) + 10), crystal.getZ());
+        }
+
+        @Override
+        public void stop() {
+            crystal = null;
+            phantom.crystalPosition = null;
+            super.stop();
+        }
+
+        private double maxTargetRange() {
+            return phantom.level().purpurConfig.phantomOrbitCrystalRadius;
+        }
+    }
+
+    class PhantomOrbitCrystalGoal extends Goal {
+        private final Phantom phantom;
+        private float offset;
+        private float radius;
+        private float verticalChange;
+        private float direction;
+
+        PhantomOrbitCrystalGoal(Phantom phantom) {
+            this.phantom = phantom;
+            this.setFlags(EnumSet.of(Flag.MOVE));
+        }
+
+        @Override
+        public boolean canUse() {
+            return phantom.isCirclingCrystal();
+        }
+
+        @Override
+        public void start() {
+            this.radius = 5.0F + phantom.random.nextFloat() * 10.0F;
+            this.verticalChange = -4.0F + phantom.random.nextFloat() * 9.0F;
+            this.direction = phantom.random.nextBoolean() ? 1.0F : -1.0F;
+            updateOffset();
+        }
+
         @Override
         public void tick() {
+            if (phantom.random.nextInt(350) == 0) {
+                this.verticalChange = -4.0F + phantom.random.nextFloat() * 9.0F;
+            }
+            if (phantom.random.nextInt(250) == 0) {
+                ++this.radius;
+                if (this.radius > 15.0F) {
+                    this.radius = 5.0F;
+                    this.direction = -this.direction;
+                }
+            }
+            if (phantom.random.nextInt(450) == 0) {
+                this.offset = phantom.random.nextFloat() * 2.0F * 3.1415927F;
+                updateOffset();
+            }
+            if (phantom.moveTargetPoint.distanceToSqr(phantom.getX(), phantom.getY(), phantom.getZ()) < 4.0D) {
+                updateOffset();
+            }
+            if (phantom.moveTargetPoint.y < phantom.getY() && !phantom.level().isEmptyBlock(phantom.blockPosition().below(1))) { // Youer - fix: BlockPos(Vec3i) ctor cannot take an Entity; use its blockPosition()
+                this.verticalChange = Math.max(1.0F, this.verticalChange);
+                updateOffset();
+            }
+            if (phantom.moveTargetPoint.y > phantom.getY() && !phantom.level().isEmptyBlock(phantom.blockPosition().above(1))) { // Youer - fix: BlockPos(Vec3i) ctor cannot take an Entity; use its blockPosition()
+                this.verticalChange = Math.min(-1.0F, this.verticalChange);
+                updateOffset();
+            }
+        }
+
+        private void updateOffset() {
+            this.offset += this.direction * 15.0F * 0.017453292F;
+            phantom.moveTargetPoint = phantom.crystalPosition.add(
+                    this.radius * Mth.cos(this.offset),
+                    -4.0F + this.verticalChange,
+                    this.radius * Mth.sin(this.offset));
         }
     }
+    // Purpur end - Phantoms attracted to crystals and crystals shoot phantoms
 
-    class PhantomMoveControl extends MoveControl {
+    class PhantomMoveControl extends org.purpurmc.purpur.controller.FlyingMoveControllerWASD { // Purpur - Ridables
         private float speed = 0.1F;
 
         public PhantomMoveControl(Mob p_33241_) {
             super(p_33241_);
         }
 
+        // Purpur start - Ridables
+        public void purpurTick(Player rider) {
+            if (!Phantom.this.onGround) {
+                // phantom is always in motion when flying
+                // TODO - FIX THIS
+                // rider.setForward(1.0F);
+            }
+            super.purpurTick(rider);
+        }
+        // Purpur end - Ridables
+
         @Override
-        public void tick() {
+        public void vanillaTick() { // Purpur - Ridables
             if (Phantom.this.horizontalCollision) {
                 Phantom.this.setYRot(Phantom.this.getYRot() + 180.0F);
                 this.speed = 0.1F;
             }
 
-            double d0 = Phantom.this.moveTargetPoint.x - Phantom.this.getX();
+            double d = Phantom.this.moveTargetPoint.x - Phantom.this.getX();
             double d1 = Phantom.this.moveTargetPoint.y - Phantom.this.getY();
             double d2 = Phantom.this.moveTargetPoint.z - Phantom.this.getZ();
-            double d3 = Math.sqrt(d0 * d0 + d2 * d2);
-            if (Math.abs(d3) > 1.0E-5F) {
-                double d4 = 1.0 - Math.abs(d1 * 0.7F) / d3;
-                d0 *= d4;
-                d2 *= d4;
-                d3 = Math.sqrt(d0 * d0 + d2 * d2);
-                double d5 = Math.sqrt(d0 * d0 + d2 * d2 + d1 * d1);
-                float f = Phantom.this.getYRot();
-                float f1 = (float)Mth.atan2(d2, d0);
-                float f2 = Mth.wrapDegrees(Phantom.this.getYRot() + 90.0F);
-                float f3 = Mth.wrapDegrees(f1 * (180.0F / (float)Math.PI));
-                Phantom.this.setYRot(Mth.approachDegrees(f2, f3, 4.0F) - 90.0F);
+            double squareRoot = Math.sqrt(d * d + d2 * d2);
+            if (Math.abs(squareRoot) > 1.0E-5F) {
+                double d3 = 1.0 - Math.abs(d1 * 0.7F) / squareRoot;
+                d *= d3;
+                d2 *= d3;
+                squareRoot = Math.sqrt(d * d + d2 * d2);
+                double squareRoot1 = Math.sqrt(d * d + d2 * d2 + d1 * d1);
+                float yRot = Phantom.this.getYRot();
+                float f = (float)Mth.atan2(d2, d);
+                float f1 = Mth.wrapDegrees(Phantom.this.getYRot() + 90.0F);
+                float f2 = Mth.wrapDegrees(f * (180.0F / (float)Math.PI));
+                Phantom.this.setYRot(Mth.approachDegrees(f1, f2, 4.0F) - 90.0F);
                 Phantom.this.yBodyRot = Phantom.this.getYRot();
-                if (Mth.degreesDifferenceAbs(f, Phantom.this.getYRot()) < 3.0F) {
+                if (Mth.degreesDifferenceAbs(yRot, Phantom.this.getYRot()) < 3.0F) {
                     this.speed = Mth.approach(this.speed, 1.8F, 0.005F * (1.8F / this.speed));
                 } else {
                     this.speed = Mth.approach(this.speed, 0.2F, 0.025F);
                 }
 
-                float f4 = (float)(-(Mth.atan2(-d1, d3) * 180.0F / (float)Math.PI));
-                Phantom.this.setXRot(f4);
-                float f5 = Phantom.this.getYRot() + 90.0F;
-                double d6 = this.speed * Mth.cos(f5 * (float) (Math.PI / 180.0)) * Math.abs(d0 / d5);
-                double d7 = this.speed * Mth.sin(f5 * (float) (Math.PI / 180.0)) * Math.abs(d2 / d5);
-                double d8 = this.speed * Mth.sin(f4 * (float) (Math.PI / 180.0)) * Math.abs(d1 / d5);
-                Vec3 vec3 = Phantom.this.getDeltaMovement();
-                Phantom.this.setDeltaMovement(vec3.add(new Vec3(d6, d8, d7).subtract(vec3).scale(0.2)));
+                float f3 = (float)(-(Mth.atan2(-d1, squareRoot) * 180.0F / (float)Math.PI));
+                Phantom.this.setXRot(f3);
+                float f4 = Phantom.this.getYRot() + 90.0F;
+                double d4 = this.speed * Mth.cos(f4 * (float) (Math.PI / 180.0)) * Math.abs(d / squareRoot1);
+                double d5 = this.speed * Mth.sin(f4 * (float) (Math.PI / 180.0)) * Math.abs(d2 / squareRoot1);
+                double d6 = this.speed * Mth.sin(f3 * (float) (Math.PI / 180.0)) * Math.abs(d1 / squareRoot1);
+                Vec3 deltaMovement = Phantom.this.getDeltaMovement();
+                Phantom.this.setDeltaMovement(deltaMovement.add(new Vec3(d4, d6, d5).subtract(deltaMovement).scale(0.2)));
             }
         }
     }
@@ -484,6 +_,12 @@
                 return false;
             } else if (!livingentity.isAlive()) {
                 return false;
+            // Purpur start - Phantoms burn in light
+            } else if (level().purpurConfig.phantomBurnInLight > 0 && level().getLightEmission(Phantom.this.blockPosition()) >= level().purpurConfig.phantomBurnInLight) { // Youer - fix: BlockPos(Vec3i) ctor cannot take an Entity; use its blockPosition()
+                return false;
+            } else if (level().purpurConfig.phantomIgnorePlayersWithTorch && (TORCH.test(livingentity.getItemInHand(net.minecraft.world.InteractionHand.MAIN_HAND)) || TORCH.test(livingentity.getItemInHand(net.minecraft.world.InteractionHand.OFF_HAND)))) {
+                return false;
+            // Purpur end - Phantoms burn in light
             } else if (livingentity instanceof Player player && (livingentity.isSpectator() || player.isCreative())) {
                 return false;
             } else if (!this.canUse()) {
