--- a/net/minecraft/world/entity/monster/Ravager.java
+++ b/net/minecraft/world/entity/monster/Ravager.java
@@ -145,7 +_,7 @@
 
             if (this.level() instanceof ServerLevel serverlevel
                 && this.horizontalCollision
-                && serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                && net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverlevel, this)) {
                 boolean flag = false;
                 AABB aabb = this.getBoundingBox().inflate(0.2);
 
@@ -155,12 +_,19 @@
                     BlockState blockstate = serverlevel.getBlockState(blockpos);
                     Block block = blockstate.getBlock();
                     if (block instanceof LeavesBlock) {
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, blockstate.getFluidState().createLegacyBlock())) { // Paper - fix wrong block state
+                            continue;
+                        }
+                        // CraftBukkit end
                         flag = serverlevel.destroyBlock(blockpos, true, this) || flag;
                     }
                 }
 
                 if (!flag && this.onGround()) {
+                    if (new com.destroystokyo.paper.event.entity.EntityJumpEvent(getBukkitLivingEntity()).callEvent()) { // Paper - Entity Jump API
                     this.jumpFromGround();
+                    } else { this.setJumping(false); } // Paper - Entity Jump API; setJumping(false) stops a potential loop
                 }
             }
 
@@ -223,7 +_,7 @@
 
     private void roar() {
         if (this.isAlive() && this.level() instanceof ServerLevel serverlevel) {
-            Predicate<Entity> predicate = serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+            Predicate<Entity> predicate = net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverlevel, this)
                 ? ROAR_TARGET_WITH_GRIEFING
                 : ROAR_TARGET_WITHOUT_GRIEFING;
 
@@ -252,7 +_,7 @@
         double d0 = p_33340_.getX() - this.getX();
         double d1 = p_33340_.getZ() - this.getZ();
         double d2 = Math.max(d0 * d0 + d1 * d1, 0.001);
-        p_33340_.push(d0 / d2 * 4.0, 0.2, d1 / d2 * 4.0);
+        p_33340_.push(d0 / d2 * 4.0, 0.2, d1 / d2 * 4.0, this); // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
     }
 
     @Override
