--- a/net/minecraft/world/entity/monster/Slime.java
+++ b/net/minecraft/world/entity/monster/Slime.java
@@ -58,6 +_,8 @@
     public float squish;
     public float oSquish;
     private boolean wasOnGround = false;
+    private boolean canWander = true; // Paper - Slime pathfinder events
+    protected boolean actualJump; // Purpur - Ridables
 
     public Slime(EntityType<? extends Slime> p_33588_, Level p_33589_) {
         super(p_33588_, p_33589_);
@@ -65,12 +_,95 @@
         this.moveControl = new Slime.SlimeMoveControl(this);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.slimeRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.slimeRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.slimeControllable;
+    }
+
+    @Override
+    public float getJumpPower() {
+        float height = super.getJumpPower();
+        return getRider() != null && this.isControllable() && actualJump ? height * 1.5F : height;
+    }
+
+    @Override
+    public boolean onSpacebar() {
+        if (onGround && getRider() != null && this.isControllable()) {
+            actualJump = true;
+            if (getRider().getForwardMot() == 0 || getRider().getStrafeMot() == 0) {
+                jumpFromGround(); // jump() here if not moving
+            }
+        }
+        return true; // do not jump() in wasd controller, let vanilla controller handle
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    protected String getMaxHealthEquation() {
+        return level().purpurConfig.slimeMaxHealth;
+    }
+
+    protected String getAttackDamageEquation() {
+        return level().purpurConfig.slimeAttackDamage;
+    }
+
+    protected java.util.Map<Integer, Double> getMaxHealthCache() {
+        return level().purpurConfig.slimeMaxHealthCache;
+    }
+
+    protected java.util.Map<Integer, Double> getAttackDamageCache() {
+        return level().purpurConfig.slimeAttackDamageCache;
+    }
+
+    protected double getFromCache(java.util.function.Supplier<String> equation, java.util.function.Supplier<java.util.Map<Integer, Double>> cache, java.util.function.Supplier<Double> defaultValue) {
+        int size = getSize();
+        Double value = cache.get().get(size);
+        if (value == null) {
+            try {
+                value = ((Number) scriptEngine.eval("let size = " + size + "; " + equation.get())).doubleValue();
+            } catch (javax.script.ScriptException e) {
+                e.printStackTrace();
+                value = defaultValue.get();
+            }
+            cache.get().put(size, value);
+        }
+        return value;
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.slimeTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.slimeAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
     @Override
     protected void registerGoals() {
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(1, new Slime.SlimeFloatGoal(this));
         this.goalSelector.addGoal(2, new Slime.SlimeAttackGoal(this));
         this.goalSelector.addGoal(3, new Slime.SlimeRandomDirectionGoal(this));
         this.goalSelector.addGoal(5, new Slime.SlimeKeepOnJumpingGoal(this));
+        this.targetSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.targetSelector
             .addGoal(
                 1,
@@ -98,9 +_,9 @@
         this.entityData.set(ID_SIZE, i);
         this.reapplyPosition();
         this.refreshDimensions();
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(i * i);
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(getFromCache(this::getMaxHealthEquation, this::getMaxHealthCache, () -> (double) (p_33594_ * p_33594_))); // Purpur - Configurable entity base attributes
         this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(0.2F + 0.1F * i);
-        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(i);
+        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(getFromCache(this::getAttackDamageEquation, this::getAttackDamageCache, () -> (double) i)); // Purpur - Configurable entity base attributes
         if (p_33595_) {
             this.setHealth(this.getMaxHealth());
         }
@@ -117,6 +_,7 @@
         super.addAdditionalSaveData(p_422317_);
         p_422317_.putInt("Size", this.getSize() - 1);
         p_422317_.putBoolean("wasOnGround", this.wasOnGround);
+        p_422317_.putBoolean("Paper.canWander", this.canWander); // Paper
     }
 
     @Override
@@ -124,6 +_,7 @@
         this.setSize(p_421591_.getIntOr("Size", 0) + 1, false);
         super.readAdditionalSaveData(p_421591_);
         this.wasOnGround = p_421591_.getBooleanOr("wasOnGround", false);
+        this.canWander = p_421591_.getBooleanOr("Paper.canWander", true); // Paper
     }
 
     public boolean isTiny() {
@@ -148,6 +_,8 @@
             float f = this.getDimensions(this.getPose()).width() * 2.0F;
             float f1 = f / 2.0F;
 
+            // Forge: Don't spawn particles if it's handled by the implementation itself
+            if (!spawnCustomParticles())
             for (int i = 0; i < f * 16.0F; i++) {
                 float f2 = this.random.nextFloat() * (float) (Math.PI * 2);
                 float f3 = this.random.nextFloat() * 0.5F + 0.5F;
@@ -203,28 +_,57 @@
     }
 
     @Override
-    public void remove(Entity.RemovalReason p_149847_) {
-        int i = this.getSize();
-        if (!this.level().isClientSide && i > 1 && this.isDeadOrDying()) {
-            float f = this.getDimensions(this.getPose()).width();
-            float f1 = f / 2.0F;
-            int j = i / 2;
-            int k = 2 + this.random.nextInt(3);
-            PlayerTeam playerteam = this.getTeam();
-
-            for (int l = 0; l < k; l++) {
-                float f2 = (l % 2 - 0.5F) * f1;
-                float f3 = (l / 2 - 0.5F) * f1;
-                this.convertTo(
-                    this.getType(), new ConversionParams(ConversionType.SPLIT_ON_DEATH, false, false, playerteam), EntitySpawnReason.TRIGGERED, p_427109_ -> {
-                        p_427109_.setSize(j, true);
-                        p_427109_.snapTo(this.getX() + f2, this.getY() + 0.5, this.getZ() + f3, this.random.nextFloat() * 360.0F, 0.0F);
-                    }
-                );
-            }
+    public void remove(Entity.RemovalReason reason, @Nullable org.bukkit.event.entity.EntityRemoveEvent.Cause eventCause) { // CraftBukkit - add Bukkit remove cause
+        int size = this.getSize();
+        if (!this.level().isClientSide && size > 1 && this.isDeadOrDying()) {
+            float width = this.getDimensions(this.getPose()).width();
+            float f = width / 2.0F;
+            int i = size / 2;
+            int i1 = 2 + this.random.nextInt(3);
+            PlayerTeam team = this.getTeam();
+            // CraftBukkit start
+            org.bukkit.event.entity.SlimeSplitEvent event = new org.bukkit.event.entity.SlimeSplitEvent((org.bukkit.entity.Slime) this.getBukkitEntity(), i1);
+            if (event.callEvent() && event.getCount() > 0) {
+                i1 = event.getCount();
+            } else {
+                super.remove(reason, eventCause); // CraftBukkit - add Bukkit remove cause
+                return;
+            }
+
+            java.util.List<LivingEntity> slimes = new java.util.ArrayList<>(i1);
+            var children = new java.util.ArrayList<Mob>(); // Neo: Children list for MobSplitEvent.
+            // CraftBukkit end
+
+            this.preventConversionSpawns = true;
+            for (int i2 = 0; i2 < i1; i2++) {
+                float f1 = (i2 % 2 - 0.5F) * f;
+                float f2 = (i2 / 2 - 0.5F) * f;
+                Slime converted = this.convertTo(this.getType(), new ConversionParams(ConversionType.SPLIT_ON_DEATH, false, false, team), EntitySpawnReason.TRIGGERED, (mob) -> { // CraftBukkit
+                    mob.setSize(i, true);
+                    mob.snapTo(this.getX() + f1, this.getY() + 0.5, this.getZ() + f2, this.random.nextFloat() * 360.0F, 0.0F);
+                // CraftBukkit start
+                }, null, null);
+                if (converted != null) {
+                    slimes.add(converted);
+                    children.add(converted); // Neo: Record the slime until after event firing.
+                }
+                // CraftBukkit end
+            }
+            this.preventConversionSpawns = false;
+            // CraftBukkit start
+            if (!slimes.isEmpty() && org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTransformEvent(this, slimes, org.bukkit.event.entity.EntityTransformEvent.TransformReason.SPLIT).isCancelled()) { // check for empty converted entities or cancel event
+                super.remove(reason, eventCause); // add Bukkit remove cause
+                return;
+            }
+            if (!net.neoforged.neoforge.event.EventHooks.onMobSplit(this, children).isCanceled()) {
+                for (LivingEntity living : slimes) {
+                    this.level().addFreshEntity(living, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SLIME_SPLIT);
+                }
+            }
+            // CraftBukkit end
         }
 
-        super.remove(p_149847_);
+        super.remove(reason, eventCause); // CraftBukkit - add Bukkit remove cause
     }
 
     @Override
@@ -280,30 +_,37 @@
     }
 
     public static boolean checkSlimeSpawnRules(
-        EntityType<Slime> p_219113_, LevelAccessor p_219114_, EntitySpawnReason p_363569_, BlockPos p_219116_, RandomSource p_219117_
+        EntityType<Slime> entityType, LevelAccessor level, EntitySpawnReason spawnReason, BlockPos pos, RandomSource random
     ) {
-        if (p_219114_.getDifficulty() != Difficulty.PEACEFUL) {
-            if (EntitySpawnReason.isSpawner(p_363569_)) {
-                return checkMobSpawnRules(p_219113_, p_219114_, p_363569_, p_219116_, p_219117_);
-            }
-
-            if (p_219114_.getBiome(p_219116_).is(BiomeTags.ALLOWS_SURFACE_SLIME_SPAWNS)
-                && p_219116_.getY() > 50
-                && p_219116_.getY() < 70
-                && p_219117_.nextFloat() < 0.5F
-                && p_219117_.nextFloat() < p_219114_.getMoonBrightness()
-                && p_219114_.getMaxLocalRawBrightness(p_219116_) <= p_219117_.nextInt(8)) {
-                return checkMobSpawnRules(p_219113_, p_219114_, p_363569_, p_219116_, p_219117_);
-            }
-
-            if (!(p_219114_ instanceof WorldGenLevel)) {
+        if (level.getDifficulty() != Difficulty.PEACEFUL) {
+            if (EntitySpawnReason.isSpawner(spawnReason)) {
+                return checkMobSpawnRules(entityType, level, spawnReason, pos, random);
+            }
+
+            // Paper start - Replace rules for Height in Swamp Biomes
+            final double maxHeightSwamp = level.getMinecraftWorld().paperConfig().entities.spawning.slimeSpawnHeight.surfaceBiome.maximum;
+            final double minHeightSwamp = level.getMinecraftWorld().paperConfig().entities.spawning.slimeSpawnHeight.surfaceBiome.minimum;
+            // Paper end - Replace rules for Height in Swamp Biomes
+            if (level.getBiome(pos).is(BiomeTags.ALLOWS_SURFACE_SLIME_SPAWNS)
+                && pos.getY() > minHeightSwamp // Paper - Replace rules for Height in Swamp Biomes
+                && pos.getY() < maxHeightSwamp // Paper - Replace rules for Height in Swamp Biomes
+                && random.nextFloat() < 0.5F
+                && random.nextFloat() < level.getMoonBrightness()
+                && level.getMaxLocalRawBrightness(pos) <= random.nextInt(8)) {
+                return checkMobSpawnRules(entityType, level, spawnReason, pos, random);
+            }
+
+            if (!(level instanceof WorldGenLevel)) {
                 return false;
             }
 
-            ChunkPos chunkpos = new ChunkPos(p_219116_);
-            boolean flag = WorldgenRandom.seedSlimeChunk(chunkpos.x, chunkpos.z, ((WorldGenLevel)p_219114_).getSeed(), 987234911L).nextInt(10) == 0;
-            if (p_219117_.nextInt(10) == 0 && flag && p_219116_.getY() < 40) {
-                return checkMobSpawnRules(p_219113_, p_219114_, p_363569_, p_219116_, p_219117_);
+            ChunkPos chunkPos = new ChunkPos(pos);
+            boolean flag = level.getMinecraftWorld().paperConfig().entities.spawning.allChunksAreSlimeChunks || WorldgenRandom.seedSlimeChunk(chunkPos.x, chunkPos.z, ((WorldGenLevel) level).getSeed(), level.getMinecraftWorld().spigotConfig.slimeSeed).nextInt(10) == 0; // Paper
+                // Paper start - Replace rules for Height in Slime Chunks
+                final double maxHeightSlimeChunk = level.getMinecraftWorld().paperConfig().entities.spawning.slimeSpawnHeight.slimeChunk.maximum;
+                if (random.nextInt(10) == 0 && flag && pos.getY() < maxHeightSlimeChunk) {
+                // Paper end - Replace rules for Height in Slime Chunks
+                return checkMobSpawnRules(entityType, level, spawnReason, pos, random);
             }
         }
 
@@ -329,6 +_,8 @@
         Vec3 vec3 = this.getDeltaMovement();
         this.setDeltaMovement(vec3.x, this.getJumpPower(), vec3.z);
         this.hasImpulse = true;
+        net.neoforged.neoforge.common.CommonHooks.onLivingJump(this);
+        this.actualJump = false; // Purpur - Ridables
     }
 
     @Nullable
@@ -361,6 +_,22 @@
         return super.getDefaultDimensions(p_316359_).scale(this.getSize());
     }
 
+    /**
+     * Called when the slime spawns particles on landing, see onUpdate.
+     * Return true to prevent the spawning of the default particles.
+     */
+    protected boolean spawnCustomParticles() { return false; }
+
+    // Paper start - Slime pathfinder events
+    public boolean canWander() {
+        return this.canWander;
+    }
+
+    public void setWander(boolean canWander) {
+        this.canWander = canWander;
+    }
+    // Paper end - Slime pathfinder events
+
     static class SlimeAttackGoal extends Goal {
         private final Slime slime;
         private int growTiredTimer;
@@ -373,11 +_,16 @@
         @Override
         public boolean canUse() {
             LivingEntity livingentity = this.slime.getTarget();
-            if (livingentity == null) {
-                return false;
-            } else {
-                return !this.slime.canAttack(livingentity) ? false : this.slime.getMoveControl() instanceof Slime.SlimeMoveControl;
-            }
+
+            // Paper start - Slime pathfinder events
+            if (livingentity == null || !livingentity.isAlive()) {
+                return false;
+            }
+            if (!this.slime.canAttack(livingentity)) {
+                return false;
+            }
+            return this.slime.getMoveControl() instanceof Slime.SlimeMoveControl && this.slime.canWander && new com.destroystokyo.paper.event.entity.SlimeTargetLivingEntityEvent((org.bukkit.entity.Slime) this.slime.getBukkitEntity(), (org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity()).callEvent();
+            // Paper end - Slime pathfinder events
         }
 
         @Override
@@ -389,11 +_,16 @@
         @Override
         public boolean canContinueToUse() {
             LivingEntity livingentity = this.slime.getTarget();
-            if (livingentity == null) {
-                return false;
-            } else {
-                return !this.slime.canAttack(livingentity) ? false : --this.growTiredTimer > 0;
-            }
+
+            // Paper start - Slime pathfinder events
+            if (livingentity == null || !livingentity.isAlive()) {
+                return false;
+            }
+            if (!this.slime.canAttack(livingentity)) {
+                return false;
+            }
+            return --this.growTiredTimer > 0 && this.slime.canWander && new com.destroystokyo.paper.event.entity.SlimeTargetLivingEntityEvent((org.bukkit.entity.Slime) this.slime.getBukkitEntity(), (org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity()).callEvent();
+            // Paper end - Slime pathfinder events
         }
 
         @Override
@@ -412,6 +_,13 @@
                 slime$slimemovecontrol.setDirection(this.slime.getYRot(), this.slime.isDealsDamage());
             }
         }
+
+        // Paper start - Slime pathfinder events; clear timer and target when goal resets
+        public void stop() {
+            this.growTiredTimer = 0;
+            this.slime.setTarget(null);
+        }
+        // Paper end - Slime pathfinder events
     }
 
     static class SlimeFloatGoal extends Goal {
@@ -425,7 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return (this.slime.isInWater() || this.slime.isInLava()) && this.slime.getMoveControl() instanceof Slime.SlimeMoveControl;
+            return (this.slime.isInWater() || this.slime.isInLava()) && this.slime.getMoveControl() instanceof Slime.SlimeMoveControl && this.slime.canWander && new com.destroystokyo.paper.event.entity.SlimeSwimEvent((org.bukkit.entity.Slime) this.slime.getBukkitEntity()).callEvent(); // Paper - Slime pathfinder events
         }
 
         @Override
@@ -455,7 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return !this.slime.isPassenger();
+            return !this.slime.isPassenger() && this.slime.canWander && new com.destroystokyo.paper.event.entity.SlimeWanderEvent((org.bukkit.entity.Slime) this.slime.getBukkitEntity()).callEvent(); // Paper - Slime pathfinder events
         }
 
         @Override
@@ -466,7 +_,7 @@
         }
     }
 
-    static class SlimeMoveControl extends MoveControl {
+    static class SlimeMoveControl extends org.purpurmc.purpur.controller.MoveControllerWASD { // Purpur - Ridables
         private float yRot;
         private int jumpDelay;
         private final Slime slime;
@@ -484,21 +_,33 @@
         }
 
         public void setWantedMovement(double p_33671_) {
-            this.speedModifier = p_33671_;
+            this.setSpeedModifier(p_33671_); // Purpur - Ridables
             this.operation = MoveControl.Operation.MOVE_TO;
         }
 
         @Override
         public void tick() {
+            // Purpur start - Ridables
+            if (slime.getRider() != null && slime.isControllable()) {
+                purpurTick(slime.getRider());
+                if (slime.getForwardMot() != 0 || slime.getStrafeMot() != 0) {
+                    if (jumpDelay > 10) {
+                        jumpDelay = 6;
+                    }
+                } else {
+                    jumpDelay = 20;
+                }
+            } else {
+            // Purpur end - Ridables
             this.mob.setYRot(this.rotlerp(this.mob.getYRot(), this.yRot, 90.0F));
             this.mob.yHeadRot = this.mob.getYRot();
             this.mob.yBodyRot = this.mob.getYRot();
-            if (this.operation != MoveControl.Operation.MOVE_TO) {
+            } if ((slime.getRider() == null || !slime.isControllable()) && this.operation != MoveControl.Operation.MOVE_TO) { // Purpur - Ridables
                 this.mob.setZza(0.0F);
             } else {
                 this.operation = MoveControl.Operation.WAIT;
                 if (this.mob.onGround()) {
-                    this.mob.setSpeed((float)(this.speedModifier * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED)));
+                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (slime.getRider() != null && slime.isControllable() && (slime.getRider().getForwardMot() != 0 || slime.getRider().getStrafeMot() != 0) ? 2.0D : 1.0D))); // Purpur - Ridables
                     if (this.jumpDelay-- <= 0) {
                         this.jumpDelay = this.slime.getJumpDelay();
                         if (this.isAggressive) {
@@ -515,7 +_,7 @@
                         this.mob.setSpeed(0.0F);
                     }
                 } else {
-                    this.mob.setSpeed((float)(this.speedModifier * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED)));
+                    this.mob.setSpeed((float)(this.getSpeedModifier() * this.mob.getAttributeValue(Attributes.MOVEMENT_SPEED) * (slime.getRider() != null && slime.isControllable() && (slime.getRider().getForwardMot() != 0 || slime.getRider().getStrafeMot() != 0) ? 2.0D : 1.0D))); // Purpur - Ridables
                 }
             }
         }
@@ -533,7 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return this.slime.getTarget() == null
+            return this.slime.getTarget() == null && this.slime.canWander // Paper - Slime pathfinder events
                 && (this.slime.onGround() || this.slime.isInWater() || this.slime.isInLava() || this.slime.hasEffect(MobEffects.LEVITATION))
                 && this.slime.getMoveControl() instanceof Slime.SlimeMoveControl;
         }
@@ -543,10 +_,15 @@
             if (--this.nextRandomizeTime <= 0) {
                 this.nextRandomizeTime = this.adjustedTickDelay(40 + this.slime.getRandom().nextInt(60));
                 this.chosenDegrees = this.slime.getRandom().nextInt(360);
+                // Paper start - Slime pathfinder events
+                com.destroystokyo.paper.event.entity.SlimeChangeDirectionEvent event = new com.destroystokyo.paper.event.entity.SlimeChangeDirectionEvent((org.bukkit.entity.Slime) this.slime.getBukkitEntity(), this.chosenDegrees);
+                if (!this.slime.canWander || !event.callEvent()) return;
+                this.chosenDegrees = event.getNewYaw();
+                // Paper end - Slime pathfinder events
             }
 
-            if (this.slime.getMoveControl() instanceof Slime.SlimeMoveControl slime$slimemovecontrol) {
-                slime$slimemovecontrol.setDirection(this.chosenDegrees, false);
+            if (this.slime.getMoveControl() instanceof Slime.SlimeMoveControl slimeMoveControl) {
+                slimeMoveControl.setDirection(this.chosenDegrees, false);
             }
         }
     }
