--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -76,6 +_,66 @@
         super(p_34368_, p_34369_);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.zombieVillagerRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.zombieVillagerRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.zombieVillagerControllable;
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.zombieVillagerMaxHealth);
+    }
+
+    @Override
+    protected void randomizeReinforcementsChance() {
+        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.random.nextDouble() * this.level().purpurConfig.zombieVillagerSpawnReinforcements);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Configurable jockey options
+    @Override
+    public boolean jockeyOnlyBaby() {
+        return level().purpurConfig.zombieVillagerJockeyOnlyBaby;
+    }
+
+    @Override
+    public double jockeyChance() {
+        return level().purpurConfig.zombieVillagerJockeyChance;
+    }
+
+    @Override
+    public boolean jockeyTryExistingChickens() {
+        return level().purpurConfig.zombieVillagerJockeyTryExistingChickens;
+    }
+    // Purpur end - Configurable jockey options
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.zombieVillagerTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.zombieVillagerAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
     @Override
     protected void defineSynchedData(SynchedEntityData.Builder p_326075_) {
         super.defineSynchedData(p_326075_);
@@ -117,7 +_,7 @@
         if (!this.level().isClientSide && this.isAlive() && this.isConverting()) {
             int i = this.getConversionProgress();
             this.villagerConversionTime -= i;
-            if (this.villagerConversionTime <= 0) {
+            if (this.villagerConversionTime <= 0 && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.VILLAGER, (timer) -> this.villagerConversionTime = timer)) {
                 this.finishConversion((ServerLevel)this.level());
             }
         }
@@ -129,10 +_,10 @@
     public InteractionResult mobInteract(Player p_34394_, InteractionHand p_34395_) {
         ItemStack itemstack = p_34394_.getItemInHand(p_34395_);
         if (itemstack.is(Items.GOLDEN_APPLE)) {
-            if (this.hasEffect(MobEffects.WEAKNESS)) {
+            if (this.hasEffect(MobEffects.WEAKNESS) && level().purpurConfig.zombieVillagerCureEnabled) { // Purpur - Add option to disable zombie villagers cure
                 itemstack.consume(1, p_34394_);
                 if (!this.level().isClientSide) {
-                    this.startConverting(p_34394_.getUUID(), this.random.nextInt(2401) + 3600);
+                    this.startConverting(p_34394_.getUUID(), this.random.nextInt(level().purpurConfig.zombieVillagerCuringTimeMax - level().purpurConfig.zombieVillagerCuringTimeMin + 1) + level().purpurConfig.zombieVillagerCuringTimeMin); // Purpur - Customizable Zombie Villager curing times
                 }
 
                 return InteractionResult.SUCCESS_SERVER;
@@ -159,12 +_,20 @@
     }
 
     public void startConverting(@Nullable UUID p_34384_, int p_34385_) {
-        this.conversionStarter = p_34384_;
-        this.villagerConversionTime = p_34385_;
+        // Paper start - missing entity behaviour api - converting without entity event
+        this.startConverting(p_34384_, p_34385_, true);
+    }
+
+    public void startConverting(@Nullable UUID conversionStarter, int villagerConversionTime, boolean broadcastEntityEvent) {
+        // Paper end - missing entity behaviour api - converting without entity event
+        this.conversionStarter = conversionStarter;
+        this.villagerConversionTime = villagerConversionTime;
         this.getEntityData().set(DATA_CONVERTING_ID, true);
-        this.removeEffect(MobEffects.WEAKNESS);
-        this.addEffect(new MobEffectInstance(MobEffects.STRENGTH, p_34385_, Math.min(this.level().getDifficulty().getId() - 1, 0)));
-        this.level().broadcastEntityEvent(this, (byte)16);
+        // CraftBukkit start
+        this.removeEffect(MobEffects.WEAKNESS, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
+        this.addEffect(new MobEffectInstance(MobEffects.STRENGTH, villagerConversionTime, Math.min(this.level().getDifficulty().getId() - 1, 0)), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
+        // CraftBukkit end
+        if (broadcastEntityEvent) this.level().broadcastEntityEvent(this, (byte)16); // Paper - missing entity behaviour api - converting without entity event
     }
 
     @Override
@@ -188,13 +_,13 @@
         }
     }
 
-    private void finishConversion(ServerLevel p_34399_) {
-        this.convertTo(
+    private void finishConversion(ServerLevel level) {
+        Villager converted = this.convertTo( // CraftBukkit
             EntityType.VILLAGER,
             ConversionParams.single(this, false, false),
             p_375894_ -> {
                 for (EquipmentSlot equipmentslot : this.dropPreservedEquipment(
-                    p_34399_, p_351901_ -> !EnchantmentHelper.has(p_351901_, EnchantmentEffectComponents.PREVENT_ARMOR_CHANGE)
+                    level, p_351901_ -> !EnchantmentHelper.has(p_351901_, EnchantmentEffectComponents.PREVENT_ARMOR_CHANGE)
                 )) {
                     SlotAccess slotaccess = p_375894_.getSlot(equipmentslot.getIndex() + 300);
                     slotaccess.set(this.getItemBySlot(equipmentslot));
@@ -210,22 +_,29 @@
                 }
 
                 p_375894_.setVillagerXp(this.villagerXp);
-                p_375894_.finalizeSpawn(p_34399_, p_34399_.getCurrentDifficultyAt(p_375894_.blockPosition()), EntitySpawnReason.CONVERSION, null);
-                p_375894_.refreshBrain(p_34399_);
+                p_375894_.finalizeSpawn(level, level.getCurrentDifficultyAt(p_375894_.blockPosition()), EntitySpawnReason.CONVERSION, null);
+                p_375894_.refreshBrain(level);
                 if (this.conversionStarter != null) {
-                    Player player = p_34399_.getPlayerByUUID(this.conversionStarter);
-                    if (player instanceof ServerPlayer) {
-                        CriteriaTriggers.CURED_ZOMBIE_VILLAGER.trigger((ServerPlayer)player, this, p_375894_);
-                        p_34399_.onReputationEvent(ReputationEventType.ZOMBIE_VILLAGER_CURED, player, p_375894_);
+                    Player playerByUuid = level.getPlayerByUUID(this.conversionStarter);
+                    if (playerByUuid instanceof ServerPlayer) {
+                        CriteriaTriggers.CURED_ZOMBIE_VILLAGER.trigger((ServerPlayer)playerByUuid, this, p_375894_);
+                        level.onReputationEvent(ReputationEventType.ZOMBIE_VILLAGER_CURED, playerByUuid, p_375894_);
                     }
                 }
 
-                p_375894_.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0));
+                p_375894_.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION); // CraftBukkit
                 if (!this.isSilent()) {
-                    p_34399_.levelEvent(null, 1027, this.blockPosition(), 0);
+                    level.levelEvent(null, 1027, this.blockPosition(), 0);
                 }
-            }
+
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_375894_);
+                // CraftBukkit start
+            }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.CURED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CURED // CraftBukkit
         );
+        if (converted == null) {
+            ((org.bukkit.entity.ZombieVillager) this.getBukkitEntity()).setConversionTime(-1); // SPIGOT-5208: End conversion to stop event spam
+        }
+        // CraftBukkit end
     }
 
     @VisibleForTesting
