--- a/net/minecraft/network/chat/contents/TranslatableContents.java
+++ b/net/minecraft/network/chat/contents/TranslatableContents.java
@@ -81,6 +_,13 @@
         this.key = p_265775_;
         this.fallback = p_265204_;
         this.args = p_265752_;
+        if (!net.neoforged.fml.loading.FMLEnvironment.production) {
+            for (Object arg : this.args) {
+                if (!(arg instanceof Component) && !isAllowedPrimitiveArgument(arg)) {
+                    throw new IllegalArgumentException("TranslatableContents' arguments must be either a Component, Number, Boolean, or a String. Was given " + arg + " for " + this.key);
+                }
+            }
+        }
     }
 
     @Override
@@ -92,6 +_,13 @@
         Language language = Language.getInstance();
         if (language != this.decomposedWith) {
             this.decomposedWith = language;
+
+            Component langComponent = language.getComponent(this.key);
+            if (langComponent != null) {
+                this.decomposedParts = ImmutableList.of(langComponent);
+                return;
+            }
+
             String s = this.fallback != null ? language.getOrDefault(this.key, this.fallback) : language.getOrDefault(this.key);
 
             try {
@@ -170,6 +_,12 @@
     public <T> Optional<T> visit(FormattedText.StyledContentConsumer<T> p_237521_, Style p_237522_) {
         this.decompose();
 
+        if (!net.neoforged.neoforge.common.util.InsertingContents.pushTranslation(this)) {
+            // Reference cycle.
+            return Optional.empty();
+        }
+
+        try {
         for (FormattedText formattedtext : this.decomposedParts) {
             Optional<T> optional = formattedtext.visit(p_237521_, p_237522_);
             if (optional.isPresent()) {
@@ -178,21 +_,64 @@
         }
 
         return Optional.empty();
+        } finally {
+            net.neoforged.neoforge.common.util.InsertingContents.popTranslation();
+        }
     }
 
     @Override
-    public <T> Optional<T> visit(FormattedText.ContentConsumer<T> p_237519_) {
+    public <T> Optional<T> visit(FormattedText.ContentConsumer<T> contentConsumer) {
+        // Paper start - Count visited parts
+        try {
+            return this.visit(new TranslatableContentConsumer<>(contentConsumer));
+        } catch (IllegalArgumentException ignored) {
+            return contentConsumer.accept("...");
+        }
+    }
+
+    private <T> Optional<T> visit(TranslatableContentConsumer<T> contentConsumer) {
+        // Paper end - Count visited parts
         this.decompose();
 
-        for (FormattedText formattedtext : this.decomposedParts) {
-            Optional<T> optional = formattedtext.visit(p_237519_);
+        if (!net.neoforged.neoforge.common.util.InsertingContents.pushTranslation(this)) {
+            // Reference cycle.
+            return Optional.empty();
+        }
+
+        try {
+        for (FormattedText formattedText : this.decomposedParts) {
+            Optional<T> optional = formattedText.visit(contentConsumer);
             if (optional.isPresent()) {
                 return optional;
             }
         }
 
         return Optional.empty();
-    }
+        } finally {
+            net.neoforged.neoforge.common.util.InsertingContents.popTranslation();
+        }
+    }
+
+    // Paper start - Count visited parts
+    private static final class TranslatableContentConsumer<T> implements FormattedText.ContentConsumer<T> {
+        private static final IllegalArgumentException NESTED_TOO_LONG = new IllegalArgumentException("Too long");
+
+        private final FormattedText.ContentConsumer<T> visitor;
+        private int visited;
+
+        private TranslatableContentConsumer(final FormattedText.ContentConsumer<T> visitor) {
+            this.visitor = visitor;
+        }
+
+        @Override
+        public Optional<T> accept(final String asString) {
+            if (this.visited++ > 32) {
+                throw NESTED_TOO_LONG;
+            }
+            return this.visitor.accept(asString);
+        }
+    }
+    // Paper end - Count visited parts
 
     @Override
     public MutableComponent resolve(@Nullable CommandSourceStack p_237512_, @Nullable Entity p_237513_, int p_237514_) throws CommandSyntaxException {
