--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -77,6 +_,13 @@
     @Nullable
     @VisibleForDebug
     private NaturalSpawner.SpawnState lastSpawnState;
+    // Paper start
+    public final ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<net.minecraft.world.level.chunk.LevelChunk> fullChunks = new ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<>();
+    public int getFullChunksCount() {
+        return this.fullChunks.size();
+    }
+    long chunkFutureAwaitCounter;
+    // Paper end
 
     public ServerChunkCache(
         ServerLevel p_214982_,
@@ -127,6 +_,64 @@
         this.clearCache();
     }
 
+    // CraftBukkit start - properly implement isChunkLoaded
+    public boolean isChunkLoaded(int chunkX, int chunkZ) {
+        ChunkHolder chunk = this.chunkMap.getUpdatingChunkIfPresent(ChunkPos.asLong(chunkX, chunkZ));
+        if (chunk == null) {
+            return false;
+        }
+        return chunk.getFullChunkNow() != null;
+    }
+    // CraftBukkit end
+    // Paper start
+    public void addLoadedChunk(LevelChunk chunk) {
+        this.fullChunks.put(chunk.coordinateKey, chunk);
+    }
+
+    public void removeLoadedChunk(LevelChunk chunk) {
+        this.fullChunks.remove(chunk.coordinateKey);
+    }
+
+    @Nullable
+    public ChunkAccess getChunkAtImmediately(int x, int z) {
+        ChunkHolder holder = this.chunkMap.getVisibleChunkIfPresent(ChunkPos.asLong(x, z));
+        if (holder == null) {
+            return null;
+        }
+
+        return holder.getLatestChunk();
+    }
+
+    public void addTicketAtLevel(TicketType ticketType, ChunkPos chunkPos, int ticketLevel) {
+        this.ticketStorage.addTicket(new Ticket(ticketType, ticketLevel), chunkPos);
+    }
+
+    public void removeTicketAtLevel(TicketType ticketType, ChunkPos chunkPos, int ticketLevel) {
+        this.ticketStorage.removeTicket(new Ticket(ticketType, ticketLevel), chunkPos);
+    }
+
+    // "real" get chunk if loaded
+    // Note: Partially copied from the getChunkAt method below
+    @Nullable
+    public LevelChunk getChunkAtIfCachedImmediately(int x, int z) {
+        long k = ChunkPos.asLong(x, z);
+
+        // Note: Bypass cache since we need to check ticket level, and to make this MT-Safe
+
+        ChunkHolder playerChunk = this.getVisibleChunkIfPresent(k);
+        if (playerChunk == null) {
+            return null;
+        }
+
+        return playerChunk.getFullChunkNowUnchecked();
+    }
+
+    @Nullable
+    public LevelChunk getChunkAtIfLoadedImmediately(int x, int z) {
+        return this.fullChunks.get(ChunkPos.asLong(x, z));
+    }
+    // Paper end
+
     public ThreadedLevelLightEngine getLightEngine() {
         return this.lightEngine;
     }
@@ -158,6 +_,12 @@
         if (Thread.currentThread() != this.mainThread) {
             return CompletableFuture.<ChunkAccess>supplyAsync(() -> this.getChunk(p_8360_, p_8361_, p_330876_, p_8363_), this.mainThreadProcessor).join();
         } else {
+            // Paper start - Perf: Optimise getChunkAt calls for loaded chunks
+            LevelChunk ifLoaded = this.getChunkAtIfCachedImmediately(p_8360_, p_8361_);
+            if (ifLoaded != null) {
+                return ifLoaded;
+            }
+            // Paper end - Perf: Optimise getChunkAt calls for loaded chunks
             ProfilerFiller profilerfiller = Profiler.get();
             profilerfiller.incrementCounter("getChunk");
             long i = ChunkPos.asLong(p_8360_, p_8361_);
@@ -165,15 +_,20 @@
             for (int j = 0; j < 4; j++) {
                 if (i == this.lastChunkPos[j] && p_330876_ == this.lastChunkStatus[j]) {
                     ChunkAccess chunkaccess = this.lastChunk[j];
-                    if (chunkaccess != null || !p_8363_) {
+                    if (chunkaccess != null) { // CraftBukkit - the chunk can become accessible in the meantime TODO for non-null chunks it might also make sense to check that the chunk's state hasn't changed in the meantime
                         return chunkaccess;
                     }
                 }
             }
 
+            ChunkHolder chunkholder = this.getVisibleChunkIfPresent(i);
+            if (chunkholder != null && chunkholder.currentlyLoading != null)
+                return chunkholder.currentlyLoading; // Neo: If the requested chunk is loading, bypass the future chain to prevent a deadlock.
+
             profilerfiller.incrementCounter("getChunkCacheMiss");
             CompletableFuture<ChunkResult<ChunkAccess>> completablefuture = this.getChunkFutureMainThread(p_8360_, p_8361_, p_330876_, p_8363_);
             this.mainThreadProcessor.managedBlock(completablefuture::isDone);
+            // com.destroystokyo.paper.io.SyncLoadFinder.logSyncLoad(this.level, x, z); // Paper - Add debug for sync chunk loads
             ChunkResult<ChunkAccess> chunkresult = completablefuture.join();
             ChunkAccess chunkaccess1 = chunkresult.orElse(null);
             if (chunkaccess1 == null && p_8363_) {
@@ -205,6 +_,7 @@
             if (chunkholder == null) {
                 return null;
             } else {
+                if (chunkholder.currentlyLoading != null) return chunkholder.currentlyLoading; // Forge: If the requested chunk is loading, bypass the future chain to prevent a deadlock.
                 ChunkAccess chunkaccess1 = chunkholder.getChunkIfPresent(ChunkStatus.FULL);
                 if (chunkaccess1 != null) {
                     this.storeInCache(i, chunkaccess1, ChunkStatus.FULL);
@@ -245,7 +_,15 @@
         long i = chunkpos.toLong();
         int j = ChunkLevel.byStatus(p_331599_);
         ChunkHolder chunkholder = this.getVisibleChunkIfPresent(i);
-        if (p_8460_) {
+        // CraftBukkit start - don't add new ticket for currently unloading chunk
+        boolean currentlyUnloading = false;
+        if (chunkholder != null) {
+            FullChunkStatus oldChunkState = ChunkLevel.fullStatus(chunkholder.oldTicketLevel);
+            FullChunkStatus currentChunkState = ChunkLevel.fullStatus(chunkholder.getTicketLevel());
+            currentlyUnloading = (oldChunkState.isOrAfter(FullChunkStatus.FULL) && !currentChunkState.isOrAfter(FullChunkStatus.FULL));
+        }
+        if (p_8460_ && !currentlyUnloading) {
+            // CraftBukkit end
             this.addTicket(new Ticket(TicketType.UNKNOWN, j), chunkpos);
             if (this.chunkAbsent(chunkholder, j)) {
                 ProfilerFiller profilerfiller = Profiler.get();
@@ -265,7 +_,7 @@
     }
 
     private boolean chunkAbsent(@Nullable ChunkHolder p_8417_, int p_8418_) {
-        return p_8417_ == null || p_8417_.getTicketLevel() > p_8418_;
+        return p_8417_ == null || p_8417_.oldTicketLevel > p_8418_; // CraftBukkit using oldTicketLevel for isLoaded checks
     }
 
     @Override
@@ -319,17 +_,39 @@
 
     @Override
     public void close() throws IOException {
-        this.save(true);
+        // CraftBukkit start
+        this.close(true);
+    }
+
+    public void close(boolean save) throws IOException {
+        if (save) {
+            this.save(true);
+        }
+        // CraftBukkit end
         this.dataStorage.close();
         this.lightEngine.close();
         this.chunkMap.close();
     }
 
+    // CraftBukkit start - modelled on below
+    public void purgeUnload() {
+        ProfilerFiller gameprofilerfiller = Profiler.get();
+
+        gameprofilerfiller.push("purge");
+        this.ticketStorage.purgeStaleTickets(this.chunkMap);
+        this.runDistanceManagerUpdates();
+        gameprofilerfiller.popPush("unload");
+        this.chunkMap.tick(() -> true);
+        gameprofilerfiller.pop();
+        this.clearCache();
+    }
+    // CraftBukkit end
+
     @Override
     public void tick(BooleanSupplier p_201913_, boolean p_201914_) {
         ProfilerFiller profilerfiller = Profiler.get();
         profilerfiller.push("purge");
-        if (this.level.tickRateManager().runsNormally() || !p_201914_) {
+        if (this.level.tickRateManager().runsNormally() || !p_201914_ || this.level.spigotConfig.unloadFrozenChunks) { // Spigot
             this.ticketStorage.purgeStaleTickets(this.chunkMap);
         }
 
@@ -391,7 +_,7 @@
         List<MobCategory> list;
         if (flag && (this.spawnEnemies || this.spawnFriendlies)) {
             boolean flag1 = this.level.getLevelData().getGameTime() % 400L == 0L;
-            list = NaturalSpawner.getFilteredSpawningCategories(naturalspawner$spawnstate, this.spawnFriendlies, this.spawnEnemies, flag1);
+            list = NaturalSpawner.getFilteredSpawningCategories(naturalspawner$spawnstate, this.spawnFriendlies, this.spawnEnemies, flag1, this.level);
         } else {
             list = List.of();
         }
@@ -545,8 +_,13 @@
 
     @Override
     public void setSpawnSettings(boolean p_8425_) {
+        // CraftBukkit start
+        this.setSpawnSettings(p_8425_, this.spawnFriendlies);
+    }
+    public void setSpawnSettings(boolean p_8425_, boolean spawnFriendlies) {
         this.spawnEnemies = p_8425_;
-        this.spawnFriendlies = this.spawnFriendlies;
+        this.spawnFriendlies = spawnFriendlies;
+        // CraftBukkit end
     }
 
     public String getChunkDebugData(ChunkPos p_8449_) {
@@ -581,7 +_,7 @@
         }
     }
 
-    public final class MainThreadExecutor extends BlockableEventLoop<Runnable> {
+    final class MainThreadExecutor extends BlockableEventLoop<Runnable> {
         MainThreadExecutor(Level p_8494_) {
             super("Chunk source main thread executor for " + p_8494_.dimension().location());
         }
@@ -619,12 +_,18 @@
 
         @Override
         public boolean pollTask() {
+            try { // CraftBukkit - process pending Chunk loadCallback() and unloadCallback() after each run task
             if (ServerChunkCache.this.runDistanceManagerUpdates()) {
                 return true;
             } else {
                 ServerChunkCache.this.lightEngine.tryScheduleUpdate();
                 return super.pollTask();
             }
+            // CraftBukkit start - process pending Chunk loadCallback() and unloadCallback() after each run task
+            } finally {
+                ServerChunkCache.this.chunkMap.callbackExecutor.run();
+            }
+            // CraftBukkit end - process pending Chunk loadCallback() and unloadCallback() after each run task
         }
     }
 }
