--- a/net/minecraft/world/phys/shapes/Shapes.java
+++ b/net/minecraft/world/phys/shapes/Shapes.java
@@ -298,9 +_,22 @@
     }
 
     @VisibleForTesting
-    protected static IndexMerger createIndexMerger(int p_83059_, DoubleList p_83060_, DoubleList p_83061_, boolean p_83062_, boolean p_83063_) {
+    private static IndexMerger createIndexMerger(int p_83059_, DoubleList p_83060_, DoubleList p_83061_, boolean p_83062_, boolean p_83063_) { // Paper - private
+        // Paper start - fast track the most common scenario
+        // doublelist is usually a DoubleArrayList with Infinite head/tails that falls to the final else clause
+        // This is actually the most common path, so jump to it straight away
+        if (p_83060_.getDouble(0) == Double.NEGATIVE_INFINITY && p_83060_.getDouble(p_83060_.size() - 1) == Double.POSITIVE_INFINITY) {
+            return new IndirectMerger(p_83060_, p_83061_, p_83062_, p_83063_);
+        }
+        // Split out rest to hopefully inline the above
+        return lessCommonMerge(p_83059_, p_83060_, p_83061_, p_83062_, p_83063_);
+    }
+
+    private static IndexMerger lessCommonMerge(int p_83059_, DoubleList p_83060_, DoubleList p_83061_, boolean p_83062_, boolean p_83063_) {
+        // Paper end - fast track the most common scenario
         int i = p_83060_.size() - 1;
         int j = p_83061_.size() - 1;
+        // Paper note - Rewrite below as optimized order if instead of nasty ternary
         if (p_83060_ instanceof CubePointRange && p_83061_ instanceof CubePointRange) {
             long k = lcm(i, j);
             if (p_83059_ * k <= 256L) {
@@ -308,14 +_,21 @@
             }
         }
 
-        if (p_83060_.getDouble(i) < p_83061_.getDouble(0) - 1.0E-7) {
+        // Paper start - Identical happens more often than Disjoint
+        if (i == j && Objects.equals(p_83060_, p_83061_)) {
+            if (p_83060_ instanceof IdenticalMerger) {
+                return (IndexMerger) p_83060_;
+            } else if (p_83061_ instanceof IdenticalMerger) {
+                return (IndexMerger) p_83061_;
+            }
+            return new IdenticalMerger(p_83060_);
+        } else if (p_83060_.getDouble(i) < p_83061_.getDouble(0) - 1.0E-7) {
+            // Paper end - Identical happens more often than Disjoint
             return new NonOverlappingMerger(p_83060_, p_83061_, false);
         } else if (p_83061_.getDouble(j) < p_83060_.getDouble(0) - 1.0E-7) {
             return new NonOverlappingMerger(p_83061_, p_83060_, true);
         } else {
-            return (IndexMerger)(i == j && Objects.equals(p_83060_, p_83061_)
-                ? new IdenticalMerger(p_83060_)
-                : new IndirectMerger(p_83060_, p_83061_, p_83062_, p_83063_));
+            return new IndirectMerger(p_83060_, p_83061_, p_83062_, p_83063_); // Paper - Identical happens more often than Disjoint
         }
     }
 
