--- a/net/minecraft/server/commands/RideCommand.java
+++ b/net/minecraft/server/commands/RideCommand.java
@@ -63,7 +_,7 @@
         Entity entity = p_265711_.getVehicle();
         if (entity != null) {
             throw ERROR_ALREADY_RIDING.create(p_265711_.getDisplayName(), entity.getDisplayName());
-        } else if (p_265339_.getType() == EntityType.PLAYER) {
+        } else if (p_265339_.getType() == EntityType.PLAYER && !io.papermc.paper.configuration.GlobalConfiguration.get().commands.rideCommandAllowPlayerAsVehicle) { // Paper - allow player as vehicle
             throw ERROR_MOUNTING_PLAYER.create();
         } else if (p_265711_.getSelfAndPassengers().anyMatch(p_265501_ -> p_265501_ == p_265339_)) {
             throw ERROR_MOUNTING_LOOP.create();
